{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c524d897-1e6a-460f-97c2-ed9d05ca0f3e",
   "metadata": {},
   "source": [
    "## CRISPR screen analysis\n",
    "\n",
    "The __Clustered Regularly Interspaced Short Palindromic Repeats (CRISPR) system__ is a bacterial immune system that has been modified for genome engineering. CRISPR consists of two components: a guide RNA (gRNA) and a non-specific CRISPR-associated endonuclease (Cas9). The gRNA is a short synthetic RNA composed of a scaffold sequence necessary for Cas9-binding (trRNA) and ~20 nucleotide spacer or targeting sequence which defines the genomic target to be modified (crRNA). sgRNA (single guide RNA) is a single RNA molecule that contains both the custom-designed short crRNA sequence fused to the scaffold tracrRNA sequence. Cas9 induces double-stranded breaks (DSB) within the target DNA. The resulting DSB is then repaired by either error-prone Non-Homologous End Joining (NHEJ) pathway or less efficient but high-fidelity Homology Directed Repair (HDR) pathway. The NHEJ pathway is the most active repair mechanism and it leads to small nucleotide insertions or deletions (indels) at the DSB site. This results in in-frame amino acid deletions, insertions or frameshift mutations leading to premature stop codons within the open reading frame (ORF) of the targeted gene. The end result is a loss-of-function mutation within the targeted gene.\n",
    "\n",
    "The ease of generating gRNAs makes CRISPR one of the most scalable genome editing technologies. Genome-wide screens enable systematic targeting of thousands of genes, with one gene targeted per cell, to identify genes driving phenotypes, such as cell survival, drug resistance or sensitivity. \n",
    "\n",
    "Below, a __CRISPR perturbation screen analysis to identify genes whose knockout increases cancer cells sensitivity to a drug__. \n",
    "\n",
    "CRISPR screen data is analysed using MAGeCK and standard read quality tools. CRISPR screen reads are assessed for quality using __fastp__ sequencing tool. The detection of enriched guides can be performed using MAGeCK. Downstream analysis include visualisations, such as volcano plot, and pathway analysis with tools like fgsea."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb25cb8c-13b7-426d-bbf4-c3e7e28b1450",
   "metadata": {},
   "source": [
    "### Data\n",
    "\n",
    "3 samples from human esophageal cancer cell line (OACM5.1): a baseline sample taken at time zero (T0-Control), a sample treated with drug for 8 days (T8-APR-246) and a control sample treated with vehicle for 8 days (T8-Vehicle) [10.1101/2020.11.29.398867](10.1101/2020.11.29.398867).\n",
    "\n",
    "APR-246 is a mutant-p53 targeted therapeutic (targeting antioxidant pathways) under clinical investigation in acute myeloid leukemia (AML)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3b45c5f-96c2-4a04-bdd2-21775be8a3e8",
   "metadata": {},
   "source": [
    "### Processing reads \n",
    "\n",
    "#### QC and adapter trimming\n",
    "\n",
    "Use __fastp__ - fast all-in-one preprocessing tool for fastq files which performs quality control, adapter trimming, quality filtering, per-read quality pruning, and provides the filtering results in various formats (HTML, JSON)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "45294892-c73f-4404-af33-6e62782602da",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5e97c34e-09b4-4ebc-b897-bee25b41642b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Illumina Universal Adapter</th>\n",
       "      <th>AGATCGGAAGAG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Illumina Small RNA 3' Adapter</td>\n",
       "      <td>TGGAATTCTCGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Illumina Small RNA 5' Adapter</td>\n",
       "      <td>GATCGTCGGACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Nextera Transposase Sequence</td>\n",
       "      <td>CTGTCTCTTATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CRISPR lentiGuide Puro 5prime</td>\n",
       "      <td>TGTGGAAAGGAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CRISPR lentiGuide Puro 3prime</td>\n",
       "      <td>GTTTTAGAGCTA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Illumina Universal Adapter  AGATCGGAAGAG\n",
       "0  Illumina Small RNA 3' Adapter  TGGAATTCTCGG\n",
       "1  Illumina Small RNA 5' Adapter  GATCGTCGGACT\n",
       "2   Nextera Transposase Sequence  CTGTCTCTTATA\n",
       "3  CRISPR lentiGuide Puro 5prime  TGTGGAAAGGAC\n",
       "4  CRISPR lentiGuide Puro 3prime  GTTTTAGAGCTA"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = ['T0-Control', 'T8-APR-246', 'T8-Vehicle']\n",
    "\n",
    "adapters = pd.read_csv(\"adapter_list.tsv\", delimiter='\\t')\n",
    "adapters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a1ea789-68f2-460a-91cc-84de51754a8b",
   "metadata": {},
   "source": [
    "The 5' adapter sequence, TGTGGAAAGGAC, from the Brunello library vector used with this dataset (lentiGuide-Puro) will be trimmed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "67ccb4e9-bedd-45dd-9d78-dd9894b102bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read1 before filtering:\n",
      "total reads: 213111\n",
      "total bases: 15983325\n",
      "Q20 bases: 14506936(90.7629%)\n",
      "Q30 bases: 13902387(86.9806%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 1488\n",
      "total bases: 111505\n",
      "Q20 bases: 94780(85.0007%)\n",
      "Q30 bases: 89011(79.8269%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 1488\n",
      "reads failed due to low quality: 6858\n",
      "reads failed due to too many N: 0\n",
      "reads failed due to too short: 204765\n",
      "reads with adapter trimmed: 211347\n",
      "bases trimmed due to adapters: 14849002\n",
      "\n",
      "Duplication rate (may be overestimated since this is SE data): 8.47774%\n",
      "\n",
      "JSON report: qc/T0-Control.json\n",
      "HTML report: qc/T0-Control.html\n",
      "\n",
      "fastp -i T0-Control.fastq.gz -o qc/T0-Control.fastq.gz --adapter_sequence=TGTGGAAAGGAC -w 4 -h qc/T0-Control.html -j qc/T0-Control.json \n",
      "fastp v0.23.2, time used: 2 seconds\n",
      "Read1 before filtering:\n",
      "total reads: 187963\n",
      "total bases: 14097225\n",
      "Q20 bases: 12812053(90.8835%)\n",
      "Q30 bases: 12284537(87.1415%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 1202\n",
      "total bases: 90092\n",
      "Q20 bases: 77306(85.8078%)\n",
      "Q30 bases: 72890(80.9062%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 1202\n",
      "reads failed due to low quality: 5864\n",
      "reads failed due to too many N: 0\n",
      "reads failed due to too short: 180897\n",
      "reads with adapter trimmed: 186534\n",
      "bases trimmed due to adapters: 13104835\n",
      "\n",
      "Duplication rate (may be overestimated since this is SE data): 8.76555%\n",
      "\n",
      "JSON report: qc/T8-APR-246.json\n",
      "HTML report: qc/T8-APR-246.html\n",
      "\n",
      "fastp -i T8-APR-246.fastq.gz -o qc/T8-APR-246.fastq.gz --adapter_sequence=TGTGGAAAGGAC -w 4 -h qc/T8-APR-246.html -j qc/T8-APR-246.json \n",
      "fastp v0.23.2, time used: 1 seconds\n",
      "Read1 before filtering:\n",
      "total reads: 184346\n",
      "total bases: 13825950\n",
      "Q20 bases: 12557012(90.8221%)\n",
      "Q30 bases: 12036916(87.0603%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 1221\n",
      "total bases: 91308\n",
      "Q20 bases: 77593(84.9794%)\n",
      "Q30 bases: 72988(79.936%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 1221\n",
      "reads failed due to low quality: 5896\n",
      "reads failed due to too many N: 0\n",
      "reads failed due to too short: 177229\n",
      "reads with adapter trimmed: 182823\n",
      "bases trimmed due to adapters: 12843341\n",
      "\n",
      "Duplication rate (may be overestimated since this is SE data): 8.6853%\n",
      "\n",
      "JSON report: qc/T8-Vehicle.json\n",
      "HTML report: qc/T8-Vehicle.html\n",
      "\n",
      "fastp -i T8-Vehicle.fastq.gz -o qc/T8-Vehicle.fastq.gz --adapter_sequence=TGTGGAAAGGAC -w 4 -h qc/T8-Vehicle.html -j qc/T8-Vehicle.json \n",
      "fastp v0.23.2, time used: 1 seconds\n"
     ]
    }
   ],
   "source": [
    "for sample in samples:\n",
    "    ! fastp -i {sample}.fastq.gz -o qc/{sample}.fastq.gz --adapter_sequence=TGTGGAAAGGAC -w 4 -h qc/{sample}.html -j qc/{sample}.json"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccf965ca-2c9c-46cf-af2c-77438653af67",
   "metadata": {},
   "source": [
    "#### Counting\n",
    "\n",
    "Use __[MAGeCK](10.1186/s13059-014-0554-4)__ (Model-based Analysis of Genome-wide CRISPR/Cas9 Knockout) method for prioritizing single-guide RNAs, genes and pathways in genome-scale CRISPR/Cas9 knockout screens. MAGeCK identifies both positively and negatively selected genes simultaneously, and reports robust results across different experimental conditions. \n",
    "\n",
    "To count the number of guides for each gene, [__MAGeCK__](https://sourceforge.net/p/mageck/wiki/usage/) __`count`__ needs a library file containing list of sgRNA names, the sequences and target genes, specifying which guide sequence belongs to which gene. The guides here are from the [Brunello](10.1038/nbt.3437) library which contains 77,441 sgRNAs, an average of 4 sgRNAs per gene, and 1000 non-targeting control sgRNAs. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bcc1d844-efd5-4ca4-95e3-286af7a029db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ID_1</td>\n",
       "      <td>CATCTTCTTTCACCTGAACG</td>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ID_2</td>\n",
       "      <td>CTCCGGGGAGAACTCCGGCG</td>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ID_3</td>\n",
       "      <td>TCTCCATGGTGCATCAGCAC</td>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ID_4</td>\n",
       "      <td>TGGAAGTCCACTCCACTCAG</td>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ID_5</td>\n",
       "      <td>ACTGCATCTGTGCAAACGGG</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ID_6</td>\n",
       "      <td>ATGTCTCATGAACTACCCTG</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ID_7</td>\n",
       "      <td>TGAAATGAAACTTCACACTG</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ID_8</td>\n",
       "      <td>TTACTCATATAGGATCCCAA</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ID_9</td>\n",
       "      <td>CGGAAGACACAAGGCACCTG</td>\n",
       "      <td>NAT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ID_10</td>\n",
       "      <td>GAACCTTAACATCCATTGTG</td>\n",
       "      <td>NAT1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0                     1     2\n",
       "0   ID_1  CATCTTCTTTCACCTGAACG  A1BG\n",
       "1   ID_2  CTCCGGGGAGAACTCCGGCG  A1BG\n",
       "2   ID_3  TCTCCATGGTGCATCAGCAC  A1BG\n",
       "3   ID_4  TGGAAGTCCACTCCACTCAG  A1BG\n",
       "4   ID_5  ACTGCATCTGTGCAAACGGG   A2M\n",
       "5   ID_6  ATGTCTCATGAACTACCCTG   A2M\n",
       "6   ID_7  TGAAATGAAACTTCACACTG   A2M\n",
       "7   ID_8  TTACTCATATAGGATCCCAA   A2M\n",
       "8   ID_9  CGGAAGACACAAGGCACCTG  NAT1\n",
       "9  ID_10  GAACCTTAACATCCATTGTG  NAT1"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sgRNA_library = pd.read_csv(\"counting/sgRNA_library.tsv\", delimiter=\"\\t\", header=None)\n",
    "sgRNA_library.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6b1ec049-5ff0-478a-a85c-22422c1fe684",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Parameters: /usr/local/bin/mageck count -l counting/sgRNA_library.tsv --fastq qc/T0-Control.fastq.gz qc/T8-APR-246.fastq.gz qc/T8-Vehicle.fastq.gz --sample-label T0-Control,T8-APR-246,T8-Vehicle -n output --pdf-report --keep-tmp --norm-method median \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Welcome to MAGeCK v0.5.9.5. Command: count \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Loading 77441 predefined sgRNAs. \n",
      "WARNING @ Thu, 21 Dec 2023 07:39:30: There are 0 sgRNAs with duplicated sequences. \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Parsing FASTQ file qc/T0-Control.fastq.gz... \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Determining the trim-5 length of FASTQ file qc/T0-Control.fastq.gz... \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Possible gRNA lengths:20 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Processing 0M reads ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Read length:75,63,40,61,54,62 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Total tested reads: 1488, mapped: 1016(0.6827956989247311) \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: --trim-5 test data: (trim_length reads fraction) \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 23\t150\t0.14763779527559054 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 25\t138\t0.13582677165354332 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 24\t129\t0.12696850393700787 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 22\t115\t0.11318897637795275 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 27\t97\t0.09547244094488189 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 29\t91\t0.08956692913385826 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 28\t89\t0.08759842519685039 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 21\t75\t0.07381889763779527 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 26\t51\t0.05019685039370079 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 30\t35\t0.0344488188976378 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 31\t16\t0.015748031496062992 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 20\t16\t0.015748031496062992 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 18\t6\t0.005905511811023622 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 19\t3\t0.002952755905511811 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 16\t2\t0.001968503937007874 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 17\t2\t0.001968503937007874 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 14\t1\t0.000984251968503937 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Auto determination of trim5 results: 23,25,24,22,27,29,28 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Possible gRNA lengths:20 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Processing 0M reads .. \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Total: 1488. \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Mapped: 809. \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Parsing FASTQ file qc/T8-APR-246.fastq.gz... \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Determining the trim-5 length of FASTQ file qc/T8-APR-246.fastq.gz... \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Possible gRNA lengths:20 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Processing 0M reads ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Read length:75,28,64 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Total tested reads: 1202, mapped: 799(0.6647254575707154) \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: --trim-5 test data: (trim_length reads fraction) \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 23\t103\t0.12891113892365458 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 24\t100\t0.1251564455569462 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 27\t98\t0.12265331664580725 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 25\t97\t0.1214017521902378 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 22\t88\t0.11013767209011265 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 28\t83\t0.10387984981226533 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 29\t79\t0.09887359198998748 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 21\t76\t0.0951188986232791 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 26\t32\t0.04005006257822278 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 30\t13\t0.016270337922403004 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 18\t7\t0.008760951188986232 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 20\t7\t0.008760951188986232 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 31\t7\t0.008760951188986232 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 19\t5\t0.006257822277847309 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 17\t3\t0.0037546933667083854 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 14\t1\t0.0012515644555694619 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Auto determination of trim5 results: 23,24,27,25,22,28,29 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Possible gRNA lengths:20 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Processing 0M reads .. \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Total: 1202. \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Mapped: 648. \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Parsing FASTQ file qc/T8-Vehicle.fastq.gz... \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Determining the trim-5 length of FASTQ file qc/T8-Vehicle.fastq.gz... \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Possible gRNA lengths:20 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Processing 0M reads ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Read length:75,58,20,63,37,56,30,61 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Total tested reads: 1221, mapped: 799(0.6543816543816544) \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: --trim-5 test data: (trim_length reads fraction) \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 24\t104\t0.13016270337922403 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 25\t96\t0.12015018773466833 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 27\t92\t0.11514392991239049 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 29\t91\t0.11389236545682102 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 22\t90\t0.11264080100125157 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 28\t86\t0.10763454317897372 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 23\t85\t0.10638297872340426 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 21\t83\t0.10387984981226533 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 26\t35\t0.04380475594493116 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 30\t16\t0.02002503128911139 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 20\t7\t0.008760951188986232 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 31\t5\t0.006257822277847309 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 14\t4\t0.0050062578222778474 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 17\t2\t0.0025031289111389237 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 19\t2\t0.0025031289111389237 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: 18\t1\t0.0012515644555694619 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Auto determination of trim5 results: 24,25,27,29,22,28,23 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Possible gRNA lengths:20 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Processing 0M reads .. \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Total: 1221. \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Mapped: 644. \n",
      "WARNING @ Thu, 21 Dec 2023 07:39:30: Sample 0 has zero median count, so median normalization is not possible. Switch to total read count normalization. \n",
      "WARNING @ Thu, 21 Dec 2023 07:39:30: Sample 0 has too many zero-count sgRNAs (0.6113813229571985), and median normalization is unstable. Switch to total read count normalization. \n",
      "WARNING @ Thu, 21 Dec 2023 07:39:30: Sample 1 has zero median count, so median normalization is not possible. Switch to total read count normalization. \n",
      "WARNING @ Thu, 21 Dec 2023 07:39:30: Sample 1 has too many zero-count sgRNAs (0.6857976653696498), and median normalization is unstable. Switch to total read count normalization. \n",
      "WARNING @ Thu, 21 Dec 2023 07:39:30: Sample 2 has zero median count, so median normalization is not possible. Switch to total read count normalization. \n",
      "WARNING @ Thu, 21 Dec 2023 07:39:30: Sample 2 has too many zero-count sgRNAs (0.6896887159533074), and median normalization is unstable. Switch to total read count normalization. \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Final size factor: 0.865677791512155 1.080761316872428 1.0874741200828157 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Summary of file qc/T0-Control.fastq.gz: \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: label\tT0-Control \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: reads\t1488 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: mappedreads\t809 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: totalsgrnas\t77441 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: zerosgrnas\t76642 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: giniindex\t0.9897692979128031 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Summary of file qc/T8-APR-246.fastq.gz: \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: label\tT8-APR-246 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: reads\t1202 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: mappedreads\t648 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: totalsgrnas\t77441 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: zerosgrnas\t76795 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: giniindex\t0.9916860045784042 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Summary of file qc/T8-Vehicle.fastq.gz: \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: label\tT8-Vehicle \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: reads\t1221 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: mappedreads\t644 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: totalsgrnas\t77441 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: zerosgrnas\t76803 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: giniindex\t0.991818927504039 \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Loading Rnw template file: /usr/local/lib/python3.8/site-packages/mageck/count_report.Rmd. \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Loading Rnw template file: /usr/local/lib/python3.8/site-packages/mageck/fastq_template.Rnw. \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:30: Running command: cd ./; Rscript output_countsummary.R \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33: Command message: \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:   Writing to file output_countsummary.tex \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:   Processing code chunks with options ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:    1 : keep.source term verbatim (label = funcdef, output_countsummary.Rnw:28) \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:    2 : keep.source term tex (label = tab1, output_countsummary.Rnw:174) \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:    3 : keep.source term tex (label = tab2, output_countsummary.Rnw:192) \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:    4 : keep.source term tex (label = tab3, output_countsummary.Rnw:206) \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:    5 : keep.source term verbatim pdf  (output_countsummary.Rnw:239) \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:    6 : keep.source term verbatim pdf  (output_countsummary.Rnw:246) \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:    7 : keep.source term verbatim pdf  (output_countsummary.Rnw:256) \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:    8 : keep.source term verbatim pdf  (output_countsummary.Rnw:260) \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:    9 : keep.source term verbatim pdf  (output_countsummary.Rnw:269) \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:    \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:   You can now run (pdf)latex on ‘output_countsummary.tex’ \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:   Error in texi2dvi(\"output_countsummary.tex\", pdf = TRUE) :  \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:     pdflatex is not available \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:   In addition: Warning message: \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:   In texi2dvi(\"output_countsummary.tex\", pdf = TRUE) : \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:     texi2dvi script/program not available, using emulation \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:   Execution halted \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33:    \n",
      "INFO  @ Thu, 21 Dec 2023 07:39:33: End command message. \n"
     ]
    }
   ],
   "source": [
    "! mageck count -l \"counting/sgRNA_library.tsv\" \\\n",
    "--fastq \"qc/T0-Control.fastq.gz\" \"qc/T8-APR-246.fastq.gz\" \"qc/T8-Vehicle.fastq.gz\" \\\n",
    "--sample-label \"T0-Control\",\"T8-APR-246\",\"T8-Vehicle\" -n output --pdf-report --keep-tmp   --norm-method median "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5d6363d3-6bb5-4709-b3ca-7ff0058820cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! gs -dBATCH -dNOPAUSE -q -dPDFSETTINGS=/prepress -sDEVICE=pdfwrite -sOutputFile=merged.pdf *.pdf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4eae9cc-2645-4105-a160-1388ff34a591",
   "metadata": {},
   "source": [
    "Inspect the MAGeCK count files (sgRNA counts, counts summary and plots pdf) previously generated for the full dataset (as mageck count above was performed on 1% of the full dataset)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed29b191-dc14-4179-92bd-3c199eb4a4b3",
   "metadata": {},
   "source": [
    "__sgRNA counts__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "da31a202-841f-4813-862a-525d5246cd05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgRNA</th>\n",
       "      <th>Gene</th>\n",
       "      <th>T8-APR-246</th>\n",
       "      <th>T8-Vehicle</th>\n",
       "      <th>T0-Control</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ID_30030</td>\n",
       "      <td>DUSP14</td>\n",
       "      <td>172</td>\n",
       "      <td>195</td>\n",
       "      <td>205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ID_21657</td>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>84</td>\n",
       "      <td>55</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ID_69143</td>\n",
       "      <td>CTAGE6</td>\n",
       "      <td>118</td>\n",
       "      <td>232</td>\n",
       "      <td>257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ID_70279</td>\n",
       "      <td>TBC1D10C</td>\n",
       "      <td>272</td>\n",
       "      <td>232</td>\n",
       "      <td>308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ID_21824</td>\n",
       "      <td>PIP5K1B</td>\n",
       "      <td>352</td>\n",
       "      <td>388</td>\n",
       "      <td>395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ID_77019</td>\n",
       "      <td>controlguide578</td>\n",
       "      <td>445</td>\n",
       "      <td>365</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ID_49661</td>\n",
       "      <td>GALNT12</td>\n",
       "      <td>243</td>\n",
       "      <td>213</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ID_23709</td>\n",
       "      <td>MAP7</td>\n",
       "      <td>273</td>\n",
       "      <td>181</td>\n",
       "      <td>242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ID_27699</td>\n",
       "      <td>APBB3</td>\n",
       "      <td>96</td>\n",
       "      <td>114</td>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ID_41971</td>\n",
       "      <td>EXD2</td>\n",
       "      <td>172</td>\n",
       "      <td>121</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sgRNA             Gene  T8-APR-246  T8-Vehicle  T0-Control\n",
       "0  ID_30030           DUSP14         172         195         205\n",
       "1  ID_21657        HIST1H2BL          84          55          85\n",
       "2  ID_69143           CTAGE6         118         232         257\n",
       "3  ID_70279         TBC1D10C         272         232         308\n",
       "4  ID_21824          PIP5K1B         352         388         395\n",
       "5  ID_77019  controlguide578         445         365         293\n",
       "6  ID_49661          GALNT12         243         213         224\n",
       "7  ID_23709             MAP7         273         181         242\n",
       "8  ID_27699            APBB3          96         114         159\n",
       "9  ID_41971             EXD2         172         121         152"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sgRNA_counts = pd.read_csv(\"counting/full_mageck_sgrna_counts.tsv\", delimiter=\"\\t\")\n",
    "sgRNA_counts.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82bd1c90-be17-46c5-afb4-1132034c9dba",
   "metadata": {},
   "source": [
    "__Counts summary__\n",
    "\n",
    "Counts summary file summarizes QC measurements of the fastq files, and evaluates the degree of negative selection in known essential genes - only if option --day0-label is provided, MAGeck will run pathway analysis for each sample, and use GSEA metrics (columns 9-13 in table below) to evaluate the quality of the samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4a04361f-09af-41b7-81f1-482edf3909f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File</th>\n",
       "      <th>Label</th>\n",
       "      <th>Reads</th>\n",
       "      <th>Mapped</th>\n",
       "      <th>Percentage</th>\n",
       "      <th>TotalsgRNAs</th>\n",
       "      <th>Zerocounts</th>\n",
       "      <th>GiniIndex</th>\n",
       "      <th>NegSelQC</th>\n",
       "      <th>NegSelQCPval</th>\n",
       "      <th>NegSelQCPvalPermutation</th>\n",
       "      <th>NegSelQCPvalPermutationFDR</th>\n",
       "      <th>NegSelQCGene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T8-APR-246.fq.gz</td>\n",
       "      <td>T8-APR-246</td>\n",
       "      <td>18165059</td>\n",
       "      <td>14665554</td>\n",
       "      <td>0.8073</td>\n",
       "      <td>77441</td>\n",
       "      <td>2102</td>\n",
       "      <td>0.13620</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T8-Vehicle.fq.gz</td>\n",
       "      <td>T8-Vehicle</td>\n",
       "      <td>17855968</td>\n",
       "      <td>15019692</td>\n",
       "      <td>0.8412</td>\n",
       "      <td>77441</td>\n",
       "      <td>1659</td>\n",
       "      <td>0.12750</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T0-Control.fq.gz</td>\n",
       "      <td>T0-Control</td>\n",
       "      <td>20646404</td>\n",
       "      <td>17272052</td>\n",
       "      <td>0.8366</td>\n",
       "      <td>77441</td>\n",
       "      <td>535</td>\n",
       "      <td>0.09077</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               File       Label     Reads    Mapped  Percentage  TotalsgRNAs  \\\n",
       "0  T8-APR-246.fq.gz  T8-APR-246  18165059  14665554      0.8073        77441   \n",
       "1  T8-Vehicle.fq.gz  T8-Vehicle  17855968  15019692      0.8412        77441   \n",
       "2  T0-Control.fq.gz  T0-Control  20646404  17272052      0.8366        77441   \n",
       "\n",
       "   Zerocounts  GiniIndex  NegSelQC  NegSelQCPval  NegSelQCPvalPermutation  \\\n",
       "0        2102    0.13620         0             1                        1   \n",
       "1        1659    0.12750         0             1                        1   \n",
       "2         535    0.09077         0             1                        1   \n",
       "\n",
       "   NegSelQCPvalPermutationFDR  NegSelQCGene  \n",
       "0                           1           0.0  \n",
       "1                           1           0.0  \n",
       "2                           1           0.0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_summary = pd.read_csv(\"counting/full_mageck_count_summary.tsv\", delimiter=\"\\t\")\n",
    "counts_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "055946f9-6ae2-4c5d-986c-aab5ca3ff62c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['File', 'Label', 'Reads', 'Mapped', 'Percentage', 'TotalsgRNAs',\n",
       "       'Zerocounts', 'GiniIndex', 'NegSelQC', 'NegSelQCPval',\n",
       "       'NegSelQCPvalPermutation', 'NegSelQCPvalPermutationFDR',\n",
       "       'NegSelQCGene'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_summary.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2a825618-7722-4441-bc8d-a47b98438571",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Column_name</th>\n",
       "      <th>Content</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>File</td>\n",
       "      <td>The fastq file used</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Label</td>\n",
       "      <td>The label of that fastq file assigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Reads</td>\n",
       "      <td>Total number reads in the fastq file (Recommen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mapped</td>\n",
       "      <td>Total number of reads that can be mapped to li...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Percentage</td>\n",
       "      <td>Mapped percentage, calculated as Mapped/Reads ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TotalsgRNAs</td>\n",
       "      <td>Total number of sgRNAs in the library</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Zerocounts</td>\n",
       "      <td>Total number of missing sgRNAs (Recommended: n...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GiniIndex</td>\n",
       "      <td>The Gini Index of the read count distribution....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NegSelQC</td>\n",
       "      <td>The enrichment score (ES) of GSEA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NegSelQCPval</td>\n",
       "      <td>The p-value of the GSEA analysis (Recommended:...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NegSelQCPvalPermutation</td>\n",
       "      <td>The permutation p-value</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NegSelQCPvalPermutationFDR</td>\n",
       "      <td>The FDR of the permutation p-value</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NegSelQCGene</td>\n",
       "      <td>The number of essential genes found in the lib...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Column_name  \\\n",
       "0                         File   \n",
       "1                        Label   \n",
       "2                        Reads   \n",
       "3                       Mapped   \n",
       "4                   Percentage   \n",
       "5                  TotalsgRNAs   \n",
       "6                   Zerocounts   \n",
       "7                    GiniIndex   \n",
       "8                     NegSelQC   \n",
       "9                 NegSelQCPval   \n",
       "10     NegSelQCPvalPermutation   \n",
       "11  NegSelQCPvalPermutationFDR   \n",
       "12                NegSelQCGene   \n",
       "\n",
       "                                              Content  \n",
       "0                                 The fastq file used  \n",
       "1               The label of that fastq file assigned  \n",
       "2   Total number reads in the fastq file (Recommen...  \n",
       "3   Total number of reads that can be mapped to li...  \n",
       "4   Mapped percentage, calculated as Mapped/Reads ...  \n",
       "5               Total number of sgRNAs in the library  \n",
       "6   Total number of missing sgRNAs (Recommended: n...  \n",
       "7   The Gini Index of the read count distribution....  \n",
       "8                   The enrichment score (ES) of GSEA  \n",
       "9   The p-value of the GSEA analysis (Recommended:...  \n",
       "10                            The permutation p-value  \n",
       "11                 The FDR of the permutation p-value  \n",
       "12  The number of essential genes found in the lib...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_summary_info= {'Column_name': counts_summary.columns, \n",
    "                     'Content': ['The fastq file used', \n",
    "                                 'The label of that fastq file assigned', \n",
    "                                 'Total number reads in the fastq file (Recommended: 100-300 times the number of sgRNAs)', \n",
    "                                 'Total number of reads that can be mapped to library',\n",
    "                                 'Mapped percentage, calculated as Mapped/Reads (Recommended: at least 60%)', \n",
    "                                 'Total number of sgRNAs in the library',\n",
    "                                 'Total number of missing sgRNAs (Recommended: no more than 1%)',\n",
    "                                 'The Gini Index of the read count distribution. A smaller value indicates more eveness of the count distribution (Recommended: around 0.1 for plasmid or initial state samples, and around 0.2-0.3 for negative selection samples)',\n",
    "                                 'The enrichment score (ES) of GSEA',\n",
    "                                 'The p-value of the GSEA analysis (Recommended: smaller than 1e-10)',\n",
    "                                 'The permutation p-value',\n",
    "                                 'The FDR of the permutation p-value',\n",
    "                                 'The number of essential genes found in the library that are evaluated for GSEA analysis'\n",
    "                                ]}\n",
    "pd.DataFrame.from_dict(counts_summary_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47a11006-0045-4de7-b793-90cbea074b43",
   "metadata": {},
   "source": [
    "Data quality is good for the 3 samples. \n",
    "\n",
    "- There are enough sequenced reads: for T0 control sample there are 17,272,052 reads mapped to guides, 77,441 guides so we there are ~220 reads per guide (17,272,052/77,441), for T8-APR 14,665,554 reads mapped to guides/77441 guides ~ 190 reads per guide, and for T8-Vehicle 15019692/77441 ~ 190 reads/guide. A minimum of 100 reads per guide, preferably 300, is recommended. OK\n",
    "- The mapped percentage is good:  >80% mapped for all 3 samples (percentage column). recommended at least 60%. OK\n",
    "- The sgRNA zero count is good: T0-Control has 0.69% (535/77441 * 100) sgRNAs that have no reads mapped, which is good. The T8 samples are just slightly high at 2.1% (1659/77441 * 100) and 2.7% (2102/77441 * 100). Recommended no more than 1%. \n",
    "- The Gini Index is good: Gini index (a measure of inequality) is used in CRISPR analysis to assess if sgRNAs are present in equal amounts. In positive selection experiments, where only some sgRNAs dominate, the index can be high. However, in plasmid library, in early time points, or negative selection experiments, the remaining sgRNAs that haven’t been negatively selected are expected to display a fairly even distribution. A high Gini index in the latter types of sample can indicate CRISPR oligonucleotide synthesis unevenness, low viral transfection efficiency, and overselection, respectively. The Gini Index is 0.09 for T0-Control (initial state) which is good. The T8 samples are higher at 0.12 and 0.13 but good for a negative selection experiment. Recommended: around 0.1 for plasmid or initial state samples, and around 0.2-0.3 for negative selection samples\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5c23315-76fc-48b1-8ad0-6f8a1ce65a89",
   "metadata": {},
   "source": [
    "__Plots PDF__ \n",
    "\n",
    "[MAGeCK count report](./counting/full_mageck_count_report.pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e78366f3-6550-4cb4-857d-242d55041184",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"600\"\n",
       "            height=\"300\"\n",
       "            src=\"counting/full_mageck_count_report.pdf\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7fc17ff88910>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import IFrame\n",
    "IFrame(\"counting/full_mageck_count_report.pdf\", width=600, height=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ac391f9-8809-4c41-ae59-1d973ce22bbc",
   "metadata": {},
   "source": [
    "- The __boxplots__ show largely similiar distributions of counts for the 3 samples. The greater length of the box and between whiskers in the T8 samples compared to the control reflect a bit more variability of counts in those samples, more guides with low and high counts in T8 compared to the T0.\n",
    "- The __distribution of read counts plot__ shows how many guides there are for each count (the frequency sums to the total number of guides 77,441). Similar to the boxplots, the wider distribution for T8 compared to T0 shows that those samples have more guides with low and high counts. The peak for T0 appears lower because it has more points (bins) in the plot.\n",
    "- The __PCA plot__ shows that the samples from the different conditions separate well. The T8 samples are more similar to each other on PC1 axis than to T0. If the dataset would contain more samples, PCA plots could be used to check for clustering of replicates, batch effect or outliers.\n",
    "- The __hierarchical clustering plot__ also shows that the T8 samples are more similar to each other than to T0."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2572e975-43fb-4933-bc67-0430654e4d10",
   "metadata": {},
   "source": [
    "### Testing\n",
    "\n",
    "Use __MAGeCK__ __`test`__ to identify essential genes. Essential means positively or negatively selected sgRNAs and genes. CRISPR positive or negative selection screens can be performed. \n",
    "\n",
    "With a positive selection screen, most cells die after the treatment (selection) and it is of interest to identify genes whose sgRNAs increase and dominate, indicating loss of those genes helps cells survive that treatment. This can help identify genes essential for drug resistance. \n",
    "\n",
    "With a negative selection screen, most cells survive after the treatment, and it is of interest to identify genes whose sgRNAs decrease (drop out) compared to a control (e.g. vehicle), indicating those genes are needed for the cells to survive with that treatment. This can help identify genes essential for drug sensitivity. \n",
    "\n",
    "Regardless of the type of screen performed (positive or negative), MAGeCK can identify both positively and negatively selected genes in the screen by the following algorithm: \n",
    "- Raw read counts corresponding to single-guided RNAs (sgRNAs) from different experiments are first normalized using median normalization and mean-variance modeling is used to capture the relationship of mean and variance in replicates.\n",
    "- The statistical significance of each sgRNA is calculated using the learned mean-variance model.\n",
    "- Essential genes (both positively and negatively selected) are then identified by looking for genes whose sgRNAs are ranked consistently higher (by significance) using robust rank aggregation (RRA).\n",
    "\n",
    "__The dataset is from a negative selection screen and the aim is to identify genes whose knockout increases the cancer cells sensitivity to the drug__."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bff2f165-e6ea-402f-97c4-15da2825b74d",
   "metadata": {},
   "source": [
    "#### Two conditions\n",
    "\n",
    "__Compare the drug treatment (T8-APR-246) to the vehicle control (T8-Vehicle) using MAGeCK test which employs a robust ranking aggregation (RRA) algorithm__."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a2147640-fa6c-4e6b-8340-cb6a74738d73",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  @ Thu, 21 Dec 2023 07:40:33: Parameters: /usr/local/bin/mageck test -k counting/full_mageck_sgrna_counts.tsv -t 0 -c 1 -n output_test --normcounts-to-file --pdf-report --norm-method median --adjust-method fdr --sort-criteria neg --remove-zero both --gene-lfc-method median \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:33: Welcome to MAGeCK v0.5.9.5. Command: test \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:33: Loading count table from counting/full_mageck_sgrna_counts.tsv  \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:33: Processing 1 lines.. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:33: Loaded 77441 records. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:33: Loading Rnw template file: /usr/local/lib/python3.8/site-packages/mageck/test_report.Rmd. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:33: Loading R template file: /usr/local/lib/python3.8/site-packages/mageck/plot_template.RTemplate. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:33: Loading R template file: /usr/local/lib/python3.8/site-packages/mageck/plot_template_indvgene.RTemplate. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:33: Loading Rnw template file: /usr/local/lib/python3.8/site-packages/mageck/plot_template.Rnw. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:33: Treatment samples:0 \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:33: Treatment sample index:0 \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:33: Control samples:1 \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:33: Control sample index:1 \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:33: Final size factor: 0.9852674473812402 1.0333261386046741 \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:33: Writing normalized read counts to output_test.normalized.txt \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:34: Detecting outliers from variance estimation.. Avg read count:[158.19520942086456], Var: [23877.925254507176] \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:34: Skipping 273 sgRNAs from variance calculation because of their extreme high counts (> 4* STD (mean counts) ). \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:35: Before RRA, 1318 sgRNAs are removed with zero counts in both group(s). \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:35: Use qnorm to reversely calculate sgRNA scores ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:38: Running command: RRA -i output_test.plow.txt -o output_test.gene.low.txt -p 0.34833099063357986 --skip-gene NA --skip-gene na  \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46: Command message: \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   Welcome to RRA v 0.5.9. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   Skipping gene NA for permutation ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   Skipping gene na for permutation ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   Reading input file... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   Summary: 76123 sgRNAs, 20103 genes, 1 lists; skipped sgRNAs:0 \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   Computing lo-values for each group... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   Computing false discovery rate... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   Permuting genes with 1 sgRNAs... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   Permuting genes with 2 sgRNAs... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   Permuting genes with 3 sgRNAs... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   Permuting genes with 4 sgRNAs... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   Permuting genes with 7 sgRNAs... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   Permuting genes with 8 sgRNAs... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   Number of genes under FDR adjustment: 20103 \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   Saving to output file... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   RRA completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   Groups deletion complete. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:   Lists deletion complete. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46:    \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46: End command message. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:46: Running command: RRA -i output_test.phigh.txt -o output_test.gene.high.txt -p 0.35997004847417996 --skip-gene NA --skip-gene na  \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55: Command message: \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   Welcome to RRA v 0.5.9. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   Skipping gene NA for permutation ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   Skipping gene na for permutation ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   Reading input file... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   Summary: 76123 sgRNAs, 20103 genes, 1 lists; skipped sgRNAs:0 \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   Computing lo-values for each group... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   Computing false discovery rate... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   Permuting genes with 1 sgRNAs... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   Permuting genes with 2 sgRNAs... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   Permuting genes with 3 sgRNAs... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   Permuting genes with 4 sgRNAs... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   Permuting genes with 7 sgRNAs... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   Permuting genes with 8 sgRNAs... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   Number of genes under FDR adjustment: 20103 \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   Saving to output file... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   RRA completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   Groups deletion complete. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:   Lists deletion complete. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55:    \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55: End command message. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:55: Loading top 10 genes from output_test.gene.low.txt: ESD,MTHFD1L,SHMT2,FLI1,TACC3,ACYP1,SYNRG,KIAA1468,ZMYND11,B3GALTL \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:56: Loading top 10 genes from output_test.gene.high.txt: ATP5E,ATP5A1,PDE12,NDUFA13,OXSM,NDUFA1,MDH2,NDUFA9,NUBPL,SLC25A19 \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:56: Running command: rm output_test.plow.txt \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:56: Running command: rm output_test.phigh.txt \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:56: Running command: rm output_test.gene.low.txt \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:56: Running command: rm output_test.gene.high.txt \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:56: Running command: cd ./; Rscript output_test.R \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58: Command message: \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:   null device  \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:             1  \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:   Writing to file output_test_summary.tex \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:   Processing code chunks with options ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:    1 : keep.source term verbatim (label = funcdef, output_test_summary.Rnw:27) \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:    2 : keep.source term tex (label = tab1, output_test_summary.Rnw:37) \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:    3 : keep.source term verbatim (output_test_summary.Rnw:77) \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:    4 : keep.source term verbatim pdf  (output_test_summary.Rnw:83) \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:    5 : keep.source term verbatim pdf  (output_test_summary.Rnw:201) \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:    6 : keep.source term verbatim pdf  (output_test_summary.Rnw:345) \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:    7 : keep.source term verbatim pdf  (output_test_summary.Rnw:489) \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:    8 : keep.source term verbatim (output_test_summary.Rnw:567) \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:    9 : keep.source term verbatim pdf  (output_test_summary.Rnw:573) \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:   10 : keep.source term verbatim pdf  (output_test_summary.Rnw:691) \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:   11 : keep.source term verbatim pdf  (output_test_summary.Rnw:835) \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:   12 : keep.source term verbatim pdf  (output_test_summary.Rnw:979) \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:    \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:   You can now run (pdf)latex on ‘output_test_summary.tex’ \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:   Error in texi2dvi(\"output_test_summary.tex\", pdf = TRUE) :  \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:     pdflatex is not available \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:   In addition: Warning message: \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:   In texi2dvi(\"output_test_summary.tex\", pdf = TRUE) : \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:     texi2dvi script/program not available, using emulation \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:   Execution halted \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:    \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58: End command message. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58: Running command: cd ./; rm -rf output_test_summary-*.pdf \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58: Command message: \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:    \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58: End command message. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58: Running command: cd ./; rm -rf output_test_summary.aux \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58: Command message: \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:    \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58: End command message. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58: Running command: cd ./; rm -rf output_test_summary.tex \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58: Command message: \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:    \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58: End command message. \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58: Running command: cd ./; rm -rf output_test_summary.toc \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58: Command message: \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58:    \n",
      "INFO  @ Thu, 21 Dec 2023 07:40:58: End command message. \n"
     ]
    }
   ],
   "source": [
    "! mageck test -k \"counting/full_mageck_sgrna_counts.tsv\" -t '0' -c '1' -n output_test --normcounts-to-file \\\n",
    "--pdf-report --norm-method median --adjust-method fdr --sort-criteria neg --remove-zero both --gene-lfc-method median"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a06d221-f566-4ae6-82d7-e1f9912d8952",
   "metadata": {},
   "source": [
    "Inspect the MAGeCK test files (sgRNA summary, gene summary and PDF report). "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e06b5017-d674-49f9-91ad-893e517c1bff",
   "metadata": {},
   "source": [
    "__sgRNA summary__ \n",
    "\n",
    "Used to check how the individual guides for the genes of interest performed. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2f1f7cf7-9d35-4613-98fb-d51a8d3fb473",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(77441, 15)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sgRNA_summary = pd.read_table('testing/output_test.sgrna_summary.txt')\n",
    "sgRNA_summary.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bd4bb720-ad35-41db-9a3a-19d2cfa78dbe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgrna</th>\n",
       "      <th>Gene</th>\n",
       "      <th>control_count</th>\n",
       "      <th>treatment_count</th>\n",
       "      <th>control_mean</th>\n",
       "      <th>treat_mean</th>\n",
       "      <th>LFC</th>\n",
       "      <th>control_var</th>\n",
       "      <th>adj_var</th>\n",
       "      <th>score</th>\n",
       "      <th>p.low</th>\n",
       "      <th>p.high</th>\n",
       "      <th>p.twosided</th>\n",
       "      <th>FDR</th>\n",
       "      <th>high_in_treatment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ID_73563</td>\n",
       "      <td>LOC554223</td>\n",
       "      <td>677.86000</td>\n",
       "      <td>5228.600</td>\n",
       "      <td>677.86000</td>\n",
       "      <td>5228.600</td>\n",
       "      <td>2.9455</td>\n",
       "      <td>10355000.0</td>\n",
       "      <td>6690.3000</td>\n",
       "      <td>55.637</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ID_24663</td>\n",
       "      <td>NRXN1</td>\n",
       "      <td>126.11000</td>\n",
       "      <td>1733.900</td>\n",
       "      <td>126.11000</td>\n",
       "      <td>1733.900</td>\n",
       "      <td>3.7707</td>\n",
       "      <td>1292500.0</td>\n",
       "      <td>883.5100</td>\n",
       "      <td>54.091</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ID_28115</td>\n",
       "      <td>ACAA2</td>\n",
       "      <td>175.38000</td>\n",
       "      <td>2000.500</td>\n",
       "      <td>175.38000</td>\n",
       "      <td>2000.500</td>\n",
       "      <td>3.5044</td>\n",
       "      <td>1665600.0</td>\n",
       "      <td>1312.3000</td>\n",
       "      <td>50.382</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ID_38719</td>\n",
       "      <td>AMOTL2</td>\n",
       "      <td>44.33700</td>\n",
       "      <td>829.760</td>\n",
       "      <td>44.33700</td>\n",
       "      <td>829.760</td>\n",
       "      <td>4.1957</td>\n",
       "      <td>308450.0</td>\n",
       "      <td>253.3000</td>\n",
       "      <td>49.350</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ID_33954</td>\n",
       "      <td>RPL36</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>67.166</td>\n",
       "      <td>0.98527</td>\n",
       "      <td>67.166</td>\n",
       "      <td>5.1017</td>\n",
       "      <td>2255.6</td>\n",
       "      <td>2.9063</td>\n",
       "      <td>38.821</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ID_24933</td>\n",
       "      <td>EIF4E2</td>\n",
       "      <td>0.98527</td>\n",
       "      <td>66.133</td>\n",
       "      <td>0.98527</td>\n",
       "      <td>66.133</td>\n",
       "      <td>5.0796</td>\n",
       "      <td>2122.1</td>\n",
       "      <td>2.9063</td>\n",
       "      <td>38.215</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.121381e-319</td>\n",
       "      <td>2.242910e-319</td>\n",
       "      <td>2.894800e-315</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ID_30112</td>\n",
       "      <td>PRSS23</td>\n",
       "      <td>389.18000</td>\n",
       "      <td>2237.200</td>\n",
       "      <td>389.18000</td>\n",
       "      <td>2237.200</td>\n",
       "      <td>2.5201</td>\n",
       "      <td>1707500.0</td>\n",
       "      <td>3424.3000</td>\n",
       "      <td>31.580</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.517300e-219</td>\n",
       "      <td>7.034600e-219</td>\n",
       "      <td>7.782400e-215</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ID_4797</td>\n",
       "      <td>DLD</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>51.666</td>\n",
       "      <td>0.98527</td>\n",
       "      <td>51.666</td>\n",
       "      <td>4.7295</td>\n",
       "      <td>1334.7</td>\n",
       "      <td>2.9063</td>\n",
       "      <td>29.729</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.273400e-194</td>\n",
       "      <td>4.546900e-194</td>\n",
       "      <td>4.401400e-190</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ID_75745</td>\n",
       "      <td>PET117</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>50.633</td>\n",
       "      <td>0.98527</td>\n",
       "      <td>50.633</td>\n",
       "      <td>4.7009</td>\n",
       "      <td>1281.8</td>\n",
       "      <td>2.9063</td>\n",
       "      <td>29.123</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.293100e-186</td>\n",
       "      <td>2.586200e-186</td>\n",
       "      <td>2.225300e-182</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ID_32417</td>\n",
       "      <td>NCAPD3</td>\n",
       "      <td>0.98527</td>\n",
       "      <td>48.566</td>\n",
       "      <td>0.98527</td>\n",
       "      <td>48.566</td>\n",
       "      <td>4.6420</td>\n",
       "      <td>1132.0</td>\n",
       "      <td>2.9063</td>\n",
       "      <td>27.910</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.391300e-171</td>\n",
       "      <td>2.782700e-171</td>\n",
       "      <td>2.154900e-167</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sgrna       Gene  control_count  treatment_count  control_mean  \\\n",
       "0  ID_73563  LOC554223      677.86000         5228.600     677.86000   \n",
       "1  ID_24663      NRXN1      126.11000         1733.900     126.11000   \n",
       "2  ID_28115      ACAA2      175.38000         2000.500     175.38000   \n",
       "3  ID_38719     AMOTL2       44.33700          829.760      44.33700   \n",
       "4  ID_33954      RPL36        0.00000           67.166       0.98527   \n",
       "5  ID_24933     EIF4E2        0.98527           66.133       0.98527   \n",
       "6  ID_30112     PRSS23      389.18000         2237.200     389.18000   \n",
       "7   ID_4797        DLD        0.00000           51.666       0.98527   \n",
       "8  ID_75745     PET117        0.00000           50.633       0.98527   \n",
       "9  ID_32417     NCAPD3        0.98527           48.566       0.98527   \n",
       "\n",
       "   treat_mean     LFC  control_var    adj_var   score  p.low         p.high  \\\n",
       "0    5228.600  2.9455   10355000.0  6690.3000  55.637    1.0   0.000000e+00   \n",
       "1    1733.900  3.7707    1292500.0   883.5100  54.091    1.0   0.000000e+00   \n",
       "2    2000.500  3.5044    1665600.0  1312.3000  50.382    1.0   0.000000e+00   \n",
       "3     829.760  4.1957     308450.0   253.3000  49.350    1.0   0.000000e+00   \n",
       "4      67.166  5.1017       2255.6     2.9063  38.821    1.0   0.000000e+00   \n",
       "5      66.133  5.0796       2122.1     2.9063  38.215    1.0  1.121381e-319   \n",
       "6    2237.200  2.5201    1707500.0  3424.3000  31.580    1.0  3.517300e-219   \n",
       "7      51.666  4.7295       1334.7     2.9063  29.729    1.0  2.273400e-194   \n",
       "8      50.633  4.7009       1281.8     2.9063  29.123    1.0  1.293100e-186   \n",
       "9      48.566  4.6420       1132.0     2.9063  27.910    1.0  1.391300e-171   \n",
       "\n",
       "      p.twosided            FDR  high_in_treatment  \n",
       "0   0.000000e+00   0.000000e+00               True  \n",
       "1   0.000000e+00   0.000000e+00               True  \n",
       "2   0.000000e+00   0.000000e+00               True  \n",
       "3   0.000000e+00   0.000000e+00               True  \n",
       "4   0.000000e+00   0.000000e+00               True  \n",
       "5  2.242910e-319  2.894800e-315               True  \n",
       "6  7.034600e-219  7.782400e-215               True  \n",
       "7  4.546900e-194  4.401400e-190               True  \n",
       "8  2.586200e-186  2.225300e-182               True  \n",
       "9  2.782700e-171  2.154900e-167               True  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sgRNA_summary.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "721e56da-1119-43f3-807d-350b2d753322",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sgrna', 'Gene', 'control_count', 'treatment_count', 'control_mean',\n",
       "       'treat_mean', 'LFC', 'control_var', 'adj_var', 'score', 'p.low',\n",
       "       'p.high', 'p.twosided', 'FDR', 'high_in_treatment'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sgRNA_summary.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b1a49020-aade-49f4-bd04-328eaa1f2c0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Column_name</th>\n",
       "      <th>Content</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sgrna</td>\n",
       "      <td>sgRNA ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gene</td>\n",
       "      <td>The targeting gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>control_count</td>\n",
       "      <td>Normalized read counts in control samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>treatment_count</td>\n",
       "      <td>Normalized read counts in treatment samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>control_mean</td>\n",
       "      <td>Mean read counts in control samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>treat_mean</td>\n",
       "      <td>Mean read counts in treatment samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>LFC</td>\n",
       "      <td>The log2 fold change of sgRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>control_var</td>\n",
       "      <td>The raw variance in control samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>adj_var</td>\n",
       "      <td>The adjusted variance in control samples</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>score</td>\n",
       "      <td>The score of this sgRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>p.low</td>\n",
       "      <td>p-value (lower tail)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>p.high</td>\n",
       "      <td>p-value (higher tail</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>p.twosided</td>\n",
       "      <td>p-value (two-sided)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>FDR</td>\n",
       "      <td>False discovery rate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>high_in_treatment</td>\n",
       "      <td>Whether the abundance is higher in treatment s...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Column_name                                            Content\n",
       "0               sgrna                                           sgRNA ID\n",
       "1                Gene                                 The targeting gene\n",
       "2       control_count          Normalized read counts in control samples\n",
       "3     treatment_count        Normalized read counts in treatment samples\n",
       "4        control_mean                Mean read counts in control samples\n",
       "5          treat_mean              Mean read counts in treatment samples\n",
       "6                 LFC                      The log2 fold change of sgRNA\n",
       "7         control_var                The raw variance in control samples\n",
       "8             adj_var           The adjusted variance in control samples\n",
       "9               score                            The score of this sgRNA\n",
       "10              p.low                               p-value (lower tail)\n",
       "11             p.high                               p-value (higher tail\n",
       "12         p.twosided                                p-value (two-sided)\n",
       "13                FDR                               False discovery rate\n",
       "14  high_in_treatment  Whether the abundance is higher in treatment s..."
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sgRNA_summary_info= {'Column_name': sgRNA_summary.columns, \n",
    "                     'Content': ['sgRNA ID', \n",
    "                                 'The targeting gene', \n",
    "                                 'Normalized read counts in control samples', \n",
    "                                 'Normalized read counts in treatment samples',\n",
    "                                 'Mean read counts in control samples', \n",
    "                                 'Mean read counts in treatment samples',\n",
    "                                 'The log2 fold change of sgRNA',\n",
    "                                 'The raw variance in control samples',\n",
    "                                 'The adjusted variance in control samples',\n",
    "                                 'The score of this sgRNA',\n",
    "                                 'p-value (lower tail)',\n",
    "                                 'p-value (higher tail',\n",
    "                                 'p-value (two-sided)',\n",
    "                                 'False discovery rate',\n",
    "                                 'Whether the abundance is higher in treatment samples'\n",
    "                                ]}\n",
    "pd.DataFrame.from_dict(sgRNA_summary_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aadb73cd-3cf0-4361-a036-7e6df5f153f6",
   "metadata": {},
   "source": [
    "__Gene summary__ \n",
    "\n",
    "The gene summary file contains a row for each gene targeted by sgRNAs ranked by the p.neg field (by default). The values are for both negative and positive selection. There are >20,000 genes in the file for this dataset from a negative selection screen so we negative values are of most interest. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "551eec83-2d2d-482d-b93f-f5db17ba50ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20103, 14)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_summary = pd.read_table('testing/output_test.gene_summary.txt')\n",
    "gene_summary.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "34c535d8-bcbe-4239-a520-df13c42498e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>num</th>\n",
       "      <th>neg|score</th>\n",
       "      <th>neg|p-value</th>\n",
       "      <th>neg|fdr</th>\n",
       "      <th>neg|rank</th>\n",
       "      <th>neg|goodsgrna</th>\n",
       "      <th>neg|lfc</th>\n",
       "      <th>pos|score</th>\n",
       "      <th>pos|p-value</th>\n",
       "      <th>pos|fdr</th>\n",
       "      <th>pos|rank</th>\n",
       "      <th>pos|goodsgrna</th>\n",
       "      <th>pos|lfc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ESD</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.324257</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>-1.92830</td>\n",
       "      <td>0.995120</td>\n",
       "      <td>0.99512</td>\n",
       "      <td>0.999775</td>\n",
       "      <td>20004</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.92830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MTHFD1L</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.324257</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.89778</td>\n",
       "      <td>0.999990</td>\n",
       "      <td>0.99999</td>\n",
       "      <td>0.999991</td>\n",
       "      <td>20103</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.89778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SHMT2</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>0.454208</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>-1.13280</td>\n",
       "      <td>0.999980</td>\n",
       "      <td>0.99998</td>\n",
       "      <td>0.999991</td>\n",
       "      <td>20102</td>\n",
       "      <td>0</td>\n",
       "      <td>-1.13280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FLI1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.454208</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.08872</td>\n",
       "      <td>0.337410</td>\n",
       "      <td>0.48212</td>\n",
       "      <td>0.999775</td>\n",
       "      <td>9461</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.08872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TACC3</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>0.000121</td>\n",
       "      <td>0.486139</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>-0.75335</td>\n",
       "      <td>0.075059</td>\n",
       "      <td>0.16968</td>\n",
       "      <td>0.998666</td>\n",
       "      <td>3241</td>\n",
       "      <td>2</td>\n",
       "      <td>-0.75335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACYP1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000079</td>\n",
       "      <td>0.000298</td>\n",
       "      <td>0.767027</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>-0.13267</td>\n",
       "      <td>0.413280</td>\n",
       "      <td>0.55952</td>\n",
       "      <td>0.999775</td>\n",
       "      <td>10986</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.13267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SYNRG</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000098</td>\n",
       "      <td>0.000369</td>\n",
       "      <td>0.767027</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>-1.25300</td>\n",
       "      <td>0.244650</td>\n",
       "      <td>0.38823</td>\n",
       "      <td>0.999603</td>\n",
       "      <td>7600</td>\n",
       "      <td>1</td>\n",
       "      <td>-1.25300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>KIAA1468</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000110</td>\n",
       "      <td>0.000413</td>\n",
       "      <td>0.767027</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.70376</td>\n",
       "      <td>0.999890</td>\n",
       "      <td>0.99990</td>\n",
       "      <td>0.999991</td>\n",
       "      <td>20101</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.70376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ZMYND11</td>\n",
       "      <td>3</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.000361</td>\n",
       "      <td>0.767027</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>-0.78281</td>\n",
       "      <td>0.999880</td>\n",
       "      <td>0.99988</td>\n",
       "      <td>0.999991</td>\n",
       "      <td>20100</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.78281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>B3GALTL</td>\n",
       "      <td>4</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.000498</td>\n",
       "      <td>0.767027</td>\n",
       "      <td>10</td>\n",
       "      <td>3</td>\n",
       "      <td>-0.27607</td>\n",
       "      <td>0.964200</td>\n",
       "      <td>0.96411</td>\n",
       "      <td>0.999775</td>\n",
       "      <td>19342</td>\n",
       "      <td>0</td>\n",
       "      <td>-0.27607</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         id  num  neg|score  neg|p-value   neg|fdr  neg|rank  neg|goodsgrna  \\\n",
       "0       ESD    4   0.000007     0.000019  0.324257         1              4   \n",
       "1   MTHFD1L    4   0.000010     0.000032  0.324257         2              4   \n",
       "2     SHMT2    4   0.000021     0.000076  0.454208         3              4   \n",
       "3      FLI1    4   0.000026     0.000090  0.454208         4              1   \n",
       "4     TACC3    4   0.000035     0.000121  0.486139         5              2   \n",
       "5     ACYP1    4   0.000079     0.000298  0.767027         6              2   \n",
       "6     SYNRG    4   0.000098     0.000369  0.767027         7              3   \n",
       "7  KIAA1468    4   0.000110     0.000413  0.767027         8              4   \n",
       "8   ZMYND11    3   0.000118     0.000361  0.767027         9              3   \n",
       "9   B3GALTL    4   0.000131     0.000498  0.767027        10              3   \n",
       "\n",
       "   neg|lfc  pos|score  pos|p-value   pos|fdr  pos|rank  pos|goodsgrna  pos|lfc  \n",
       "0 -1.92830   0.995120      0.99512  0.999775     20004              0 -1.92830  \n",
       "1 -0.89778   0.999990      0.99999  0.999991     20103              0 -0.89778  \n",
       "2 -1.13280   0.999980      0.99998  0.999991     20102              0 -1.13280  \n",
       "3 -0.08872   0.337410      0.48212  0.999775      9461              1 -0.08872  \n",
       "4 -0.75335   0.075059      0.16968  0.998666      3241              2 -0.75335  \n",
       "5 -0.13267   0.413280      0.55952  0.999775     10986              1 -0.13267  \n",
       "6 -1.25300   0.244650      0.38823  0.999603      7600              1 -1.25300  \n",
       "7 -0.70376   0.999890      0.99990  0.999991     20101              0 -0.70376  \n",
       "8 -0.78281   0.999880      0.99988  0.999991     20100              0 -0.78281  \n",
       "9 -0.27607   0.964200      0.96411  0.999775     19342              0 -0.27607  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_summary.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0512fa54-acc4-4e00-91f0-675c9952d72d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['id', 'num', 'neg|score', 'neg|p-value', 'neg|fdr', 'neg|rank',\n",
       "       'neg|goodsgrna', 'neg|lfc', 'pos|score', 'pos|p-value', 'pos|fdr',\n",
       "       'pos|rank', 'pos|goodsgrna', 'pos|lfc'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_summary.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9bca150b-2fef-4ccc-a183-3d0ce4c7d9f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Column_name</th>\n",
       "      <th>Content</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>id</td>\n",
       "      <td>Gene ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>num</td>\n",
       "      <td>The number of targeting sgRNAs for each gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>neg|score</td>\n",
       "      <td>The RRA lo value of this gene in negative sele...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>neg|p-value</td>\n",
       "      <td>The raw p-value (using permutation) of this ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>neg|fdr</td>\n",
       "      <td>The false discovery rate for this gene in nega...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>neg|rank</td>\n",
       "      <td>The ranking of this gene in negative selection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>neg|goodsgrna</td>\n",
       "      <td>The number of good sgRNA - those whose ranking...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>neg|lfc</td>\n",
       "      <td>The log2 fold change of this gene in negative ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>pos|score</td>\n",
       "      <td>The RRA lo value of this gene in positive sele...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>pos|p-value</td>\n",
       "      <td>The raw p-value (using permutation) of this ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>pos|fdr</td>\n",
       "      <td>The false discovery rate for this gene in posi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>pos|rank</td>\n",
       "      <td>The ranking of this gene in positive selection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>pos|goodsgrna</td>\n",
       "      <td>The number of good sgRNA - those whose ranking...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>pos|lfc</td>\n",
       "      <td>The log2 fold change of this gene in positive ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Column_name                                            Content\n",
       "0              id                                            Gene ID\n",
       "1             num       The number of targeting sgRNAs for each gene\n",
       "2       neg|score  The RRA lo value of this gene in negative sele...\n",
       "3     neg|p-value  The raw p-value (using permutation) of this ge...\n",
       "4         neg|fdr  The false discovery rate for this gene in nega...\n",
       "5        neg|rank     The ranking of this gene in negative selection\n",
       "6   neg|goodsgrna  The number of good sgRNA - those whose ranking...\n",
       "7         neg|lfc  The log2 fold change of this gene in negative ...\n",
       "8       pos|score  The RRA lo value of this gene in positive sele...\n",
       "9     pos|p-value  The raw p-value (using permutation) of this ge...\n",
       "10        pos|fdr  The false discovery rate for this gene in posi...\n",
       "11       pos|rank     The ranking of this gene in positive selection\n",
       "12  pos|goodsgrna  The number of good sgRNA - those whose ranking...\n",
       "13        pos|lfc  The log2 fold change of this gene in positive ..."
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_summary_info = {'Column_name': gene_summary.columns, \n",
    "                     'Content': ['Gene ID', \n",
    "                                 'The number of targeting sgRNAs for each gene', \n",
    "                                 'The RRA lo value of this gene in negative selection', \n",
    "                                 'The raw p-value (using permutation) of this gene in negative selection',\n",
    "                                 'The false discovery rate for this gene in negative selection', \n",
    "                                 'The ranking of this gene in negative selection',\n",
    "                                 'The number of good sgRNA - those whose ranking is below the alpha cutoff (determined by the -gene-test-fdr-threshold option, in negative selection',\n",
    "                                 'The log2 fold change of this gene in negative selection. Gene lfc calculation is controlled by the -gene-lfc-method option',\n",
    "                                 'The RRA lo value of this gene in positive selection', \n",
    "                                 'The raw p-value (using permutation) of this gene in positive selection',\n",
    "                                 'The false discovery rate for this gene in positive selection', \n",
    "                                 'The ranking of this gene in positive selection',\n",
    "                                 'The number of good sgRNA - those whose ranking is below the alpha cutoff (determined by the -gene-test-fdr-threshold option, in positive selection',\n",
    "                                 'The log2 fold change of this gene in positive selection'\n",
    "                                 ]}\n",
    "pd.DataFrame.from_dict(gene_summary_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74400505-46d9-4fa6-9ecb-45b3b07e3263",
   "metadata": {},
   "source": [
    "__PDF report__\n",
    "\n",
    "[MAGeck test report](./testing/output_test.pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "da5eaa1d-0479-4c4f-ac96-10d9c72276c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"600\"\n",
       "            height=\"300\"\n",
       "            src=\"testing/output_test.pdf\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7fc1a477f290>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import IFrame\n",
    "IFrame(\"testing/output_test.pdf\", width=600, height=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eefe535a-7313-4656-8ad2-395f48472a6d",
   "metadata": {},
   "source": [
    "The PDF shows plots of the top 10 negatively and positively selected genes. \n",
    "- The distribution plots show the top genes ranked by RRA scores or p values found in the gene summary file. The top 3 negatively selected genes: ESD, MTHFD1L and SHMT2 have similar RRA/p-value scores and are invloved in the glutathione (GSH) pathway and mitochondrial metabolism in determining APR-246 efficacy/sensitivity. APR-246 triggers GSH depletion in cancer cells, and its sensitivity is increased through loss of key enzymes in mitochondrial one-carbon metabolism.\n",
    "- The plots with the sgRNA counts for the top 10 genes show whether the counts of all the sgRNAs for the top genes are changing similarly. The values are the normalized counts for each sgRNA from the sgRNA summary file. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be6340b2-bd88-4b12-8d22-11c3b40f5be4",
   "metadata": {},
   "source": [
    "#### Visualisation\n",
    "\n",
    "In addition to the visualisations automatically generated by MAGeCK in the PDF, the output can be further visualised with a volcano plot to plot the magnitude of change for drug treatment versus vehicle control (lfc) versus significance (p-value). As the gene summary file contains two columns for lfc and p-value, one for negative selection and one for positive, these two columns are combined into one for each gene using __awk__. If the `neg|p-value` is smaller than the `pos|p-value` the gene is negatively selected. If the `neg|p-value` is larger than the `pos|p-value` the gene is positively selected. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "448692b4-bf69-4e90-950e-6ce0e3c6c6e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "! $(which awk) -v FS='\t' \\\n",
    "      -v OFS='\t' \\\n",
    "      -f scripts/awk.script testing/output_test.gene_summary.txt > testing/output_test.reformatted.gene_summary.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "4dc7f625-8dbb-48db-970f-e465773df639",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Print new header for first line\n",
      "NR == 1 { print \"gene\", \"pval\", \"fdr\", \"lfc\" }\n",
      "\n",
      "# Only process lines after first\n",
      "NR > 1 {\n",
      "    # check if neg pval (column 4) is less than pos pval (column 10)\n",
      "    if ($4 < $10){\n",
      "       # if it is, print negative selection values\n",
      "        print $1, $4, $5, $8\n",
      "    } else {\n",
      "       # if it's not, print positive selection values\n",
      "        print $1, $10, $11, $14\n",
      "    }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "! cat scripts/awk.script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "1b22a9a1-d1d1-4e36-8db5-4a66ff0be2cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>pval</th>\n",
       "      <th>fdr</th>\n",
       "      <th>lfc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ESD</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.324257</td>\n",
       "      <td>-1.92830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MTHFD1L</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.324257</td>\n",
       "      <td>-0.89778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SHMT2</td>\n",
       "      <td>0.000076</td>\n",
       "      <td>0.454208</td>\n",
       "      <td>-1.13280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FLI1</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.454208</td>\n",
       "      <td>-0.08872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TACC3</td>\n",
       "      <td>0.000121</td>\n",
       "      <td>0.486139</td>\n",
       "      <td>-0.75335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACYP1</td>\n",
       "      <td>0.000298</td>\n",
       "      <td>0.767027</td>\n",
       "      <td>-0.13267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SYNRG</td>\n",
       "      <td>0.000369</td>\n",
       "      <td>0.767027</td>\n",
       "      <td>-1.25300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>KIAA1468</td>\n",
       "      <td>0.000413</td>\n",
       "      <td>0.767027</td>\n",
       "      <td>-0.70376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ZMYND11</td>\n",
       "      <td>0.000361</td>\n",
       "      <td>0.767027</td>\n",
       "      <td>-0.78281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>B3GALTL</td>\n",
       "      <td>0.000498</td>\n",
       "      <td>0.767027</td>\n",
       "      <td>-0.27607</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene      pval       fdr      lfc\n",
       "0       ESD  0.000019  0.324257 -1.92830\n",
       "1   MTHFD1L  0.000032  0.324257 -0.89778\n",
       "2     SHMT2  0.000076  0.454208 -1.13280\n",
       "3      FLI1  0.000090  0.454208 -0.08872\n",
       "4     TACC3  0.000121  0.486139 -0.75335\n",
       "5     ACYP1  0.000298  0.767027 -0.13267\n",
       "6     SYNRG  0.000369  0.767027 -1.25300\n",
       "7  KIAA1468  0.000413  0.767027 -0.70376\n",
       "8   ZMYND11  0.000361  0.767027 -0.78281\n",
       "9   B3GALTL  0.000498  0.767027 -0.27607"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_summary_trf = pd.read_table('testing/output_test.reformatted.gene_summary.txt')\n",
    "gene_summary_trf.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "31ae255f-fc14-44a7-92e7-efe98fc16a04",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hnull device \n",
      "          1 \n",
      "\u001b[?25h[1] \"Create volcano_plot.pdf....... DONE.\"\n",
      "\u001b[?25h\u001b[?25h"
     ]
    }
   ],
   "source": [
    "! Rscript scripts/volcanoplot.rscript --input testing/output_test.reformatted.gene_summary.txt \\\n",
    "--fdr_col 3 --pval_col 2 --lfc_col 4 --label_col 1 --signif_thresh 0.05 --lfc_thresh 0.0 --topn 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "3dab681d-8cd3-4a2f-b41a-a53b6612de68",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Load packages -----------------------------------------------------------\n",
      "\n",
      "suppressPackageStartupMessages({\n",
      "    library(dplyr)\n",
      "    library(ggplot2)\n",
      "    library(ggrepel)\n",
      "    library(optparse)    \n",
      "})\n",
      "\n",
      "# Fetch arguments  --------------------------------------------------------\n",
      "\n",
      "option_list = list(\n",
      "  make_option(c(\"-i\", \"--input\"), action=\"store\", default=NA, type='character',\n",
      "              help=\"input gene summary file\"),\n",
      "  make_option(c(\"-a\", \"--fdr_col\"), action=\"store\", default=NA, type='integer',\n",
      "              help=\"FDR (adjusted p-value) column number\"),\n",
      "  make_option(c(\"-p\", \"--pval_col\"), action=\"store\", default=NA, type='integer',\n",
      "              help=\"pvalue column number\"),\n",
      "  make_option(c(\"-f\", \"--lfc_col\"), action=\"store\", default=NA, type='integer',\n",
      "              help=\"logfc column number\"),   \n",
      "  make_option(c(\"-l\", \"--label_col\"), action=\"store\", default=NA, type='integer',\n",
      "              help=\"labels column number\"),  \n",
      "  make_option(c(\"-s\", \"--signif_thresh\"), action=\"store\", default=NA, type='double',\n",
      "              help=\"significance threshold\"),\n",
      "  make_option(c(\"-c\", \"--lfc_thresh\"), action=\"store\", default=NA, type='double',\n",
      "              help=\"log fold change threshold\"),\n",
      "  make_option(c(\"-t\", \"--topn\"), action=\"store\", default=NA, type='integer',\n",
      "              help=\"number of top most significant genes to label\")       \n",
      ")\n",
      "opt = parse_args(OptionParser(option_list=option_list))\n",
      "\n",
      "# Read input  -------------------------------------------------------------\n",
      "\n",
      "results <- read.delim(opt$input, header = TRUE)\n",
      "\n",
      "# Format data  ------------------------------------------------------------\n",
      "\n",
      "# Create columns from the column numbers specified\n",
      "results <- results %>% mutate(fdr = .[[opt$fdr_col]],\n",
      "                              pvalue = .[[opt$pval_col]],\n",
      "                              logfc = .[[opt$lfc_col]],\n",
      "                              labels = .[[opt$label_col]])\n",
      "\n",
      "# Get names for legend\n",
      "down <- unlist(strsplit('Down,Not Sig,Up', split = \",\"))[1]\n",
      "notsig <- unlist(strsplit('Down,Not Sig,Up', split = \",\"))[2]\n",
      "up <- unlist(strsplit('Down,Not Sig,Up', split = \",\"))[3]\n",
      "\n",
      "# Set colours\n",
      "colours <- setNames(c(\"cornflowerblue\", \"grey\", \"firebrick\"), c(down, notsig, up))\n",
      "\n",
      "# Create significant (sig) column\n",
      "results <- mutate(results, sig = case_when(\n",
      "                                fdr < opt$signif_thresh & logfc > opt$lfc_thresh ~ up,\n",
      "                                fdr < opt$signif_thresh & logfc < -opt$lfc_thresh ~ down,\n",
      "                                TRUE ~ notsig))\n",
      "\n",
      "\n",
      "# Specify genes to label --------------------------------------------------\n",
      "\n",
      "# Get top genes by P value\n",
      "top <- slice_min(results, order_by = pvalue, n = opt$topn)\n",
      "\n",
      "# Extract into vector\n",
      "toplabels <- pull(top, labels)\n",
      "\n",
      "# Label just the top genes in results table\n",
      "results <- mutate(results, labels = ifelse(labels %in% toplabels, labels, \"\"))\n",
      "\n",
      "\n",
      "# Create plot -------------------------------------------------------------\n",
      "\n",
      "# Open file to save plot as PDF\n",
      "pdf(\"volcano_plot.pdf\")\n",
      "\n",
      "# Set up base plot\n",
      "p <- ggplot(data = results, aes(x = logfc, y = -log10(pvalue))) +\n",
      "    geom_point(aes(colour = sig)) +\n",
      "    scale_color_manual(values = colours) +\n",
      "    theme(panel.grid.major = element_blank(),\n",
      "        panel.grid.minor = element_blank(),\n",
      "        panel.background = element_blank(),\n",
      "        axis.line = element_line(colour = \"black\"),\n",
      "        legend.key = element_blank())\n",
      "\n",
      "# Add gene labels\n",
      "p <- p + geom_text_repel(data = filter(results, labels != \"\"), aes(label = labels),\n",
      "                         min.segment.length = 0,\n",
      "                         max.overlaps = Inf,\n",
      "                         show.legend = FALSE)\n",
      "\n",
      "# Set legend title\n",
      "p <- p + theme(legend.title = element_blank())\n",
      "\n",
      "# Print plot\n",
      "print(p)\n",
      "\n",
      "# Close PDF graphics device\n",
      "dev.off()\n",
      "\n",
      "print(\"Create volcano_plot.pdf....... DONE.\")\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! cat scripts/volcanoplot.rscript"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a022819c-5c5f-4b8c-8190-78cd83be9081",
   "metadata": {},
   "source": [
    "[MAGeCK test volcano plot](./testing/volcano_plot.pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "42e55315-744f-4fe8-8e0f-a376f1c5e71c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"600\"\n",
       "            height=\"300\"\n",
       "            src=\"testing/volcano_plot.pdf\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7fc195c72690>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import IFrame\n",
    "IFrame(\"testing/volcano_plot.pdf\", width=600, height=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56f18164-dac1-443f-a56d-4cd6fb249279",
   "metadata": {},
   "source": [
    "#### Pathway analysis\n",
    "\n",
    "Use [__fgsea__](https://bioconductor.org/packages/devel/bioc/manuals/fgsea/man/fgsea.pdf) to perform pathway analysis on the results and identify pathways that are changing with the treatment. fgsea needs a ranked list of genes - the list from mageck gene summary result file ranked by `neg|score` - the RRA score column for negative results, and a pathways file in GMT format - the Hallmark pathways from MSigDB."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "fdf0c021-8b3b-413f-bfe8-6aaae4711e65",
   "metadata": {},
   "outputs": [],
   "source": [
    "! cut -f 1,3 testing/output_test.gene_summary.txt > testing/output_test.cut.gene_summary.txt    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "58bfccb1-7eca-458d-976a-d1603cf06255",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>neg|score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ESD</td>\n",
       "      <td>0.000007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MTHFD1L</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SHMT2</td>\n",
       "      <td>0.000021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FLI1</td>\n",
       "      <td>0.000026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TACC3</td>\n",
       "      <td>0.000035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACYP1</td>\n",
       "      <td>0.000079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SYNRG</td>\n",
       "      <td>0.000098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>KIAA1468</td>\n",
       "      <td>0.000110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ZMYND11</td>\n",
       "      <td>0.000118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>B3GALTL</td>\n",
       "      <td>0.000131</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         id  neg|score\n",
       "0       ESD   0.000007\n",
       "1   MTHFD1L   0.000010\n",
       "2     SHMT2   0.000021\n",
       "3      FLI1   0.000026\n",
       "4     TACC3   0.000035\n",
       "5     ACYP1   0.000079\n",
       "6     SYNRG   0.000098\n",
       "7  KIAA1468   0.000110\n",
       "8   ZMYND11   0.000118\n",
       "9   B3GALTL   0.000131"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_table('testing/output_test.cut.gene_summary.txt', delimiter='\\t').head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d4a21973-a0c3-48c2-9753-985104a63f86",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "package ‘fgsea’ was built under R version 4.3.2 \n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hWarning messages:\n",
      "1: In preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam,  :\n",
      "  There are ties in the preranked stats (14.51% of the list).\n",
      "The order of those tied genes will be arbitrary, which may produce unexpected results.\n",
      "2: In preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam,  :\n",
      "  All values in the stats vector are greater than zero and scoreType is \"std\", maybe you should switch to scoreType = \"pos\".\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h[1] \"Create fgsea_plots.pdf...... DOne.\"\n",
      "\u001b[?25h\u001b[?25h\u001b[?25h"
     ]
    }
   ],
   "source": [
    "! Rscript scripts/fgsea.R --rnk_file testing/output_test.cut.gene_summary.txt \\\n",
    "--header True --sets_file fgsea/hallmark.pathways.all.v7.4.symbols.gmt --gmt True \\\n",
    "--min_size 15 --max_size 500 --n_perm 1000 --out_tab fgsea/fgsea_pathways.txt \\\n",
    "--plot_opt True --top_num 10 --rda_opt True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a01c480f-c709-4f25-a515-db5db488e451",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "suppressPackageStartupMessages({\n",
      "  library(fgsea)\n",
      "  library(ggplot2)\n",
      "  library(optparse)\n",
      "})\n",
      "\n",
      "option_list <- list(\n",
      "  make_option(c(\"-r\", \"--rnk_file\"), type=\"character\", help=\"Path to ranked genes file\"),\n",
      "  make_option(c(\"-h\", \"--header\"), type=\"logical\", help = \"Does ranked genes file have a header\"),\n",
      "  make_option(c(\"-s\", \"--sets_file\"), type=\"character\", help = \"Path to gene sets file\"),\n",
      "  make_option(c(\"-g\", \"--gmt\"), type=\"logical\", help = \"Is the sets file in GMT format\"),\n",
      "  make_option(c(\"-o\",\"--out_tab\"), type=\"character\", help=\"Path to output file\"),\n",
      "  make_option(c(\"-m\", \"--min_size\"), type=\"integer\", help=\"Minimal size of a gene set to test. All pathways below the threshold are excluded.\"),\n",
      "  make_option(c(\"-x\", \"--max_size\"), type=\"integer\", help=\"Maximal size of a gene set to test. All pathways above the threshold are excluded.\"),\n",
      "  make_option(c(\"-n\", \"--n_perm\"), type=\"integer\", help=\"Number of permutations to do. Minimial possible nominal p-value is about 1/nperm\"),\n",
      "  make_option(c(\"-d\", \"--rda_opt\"), type=\"logical\", help=\"Output RData file\"),\n",
      "  make_option(c(\"-p\", \"--plot_opt\"), type=\"logical\", help=\"Output plot\"),\n",
      "  make_option(c(\"-t\", \"--top_num\"), type=\"integer\", help=\"Top number of pathways to plot\")\n",
      ")\n",
      "\n",
      "parser <- OptionParser(usage = \"%prog [options] file\", option_list=option_list, add_help_option=FALSE)\n",
      "args = parse_args(parser)\n",
      "\n",
      "rnk_file = args$rnk_file\n",
      "if (args$header) {\n",
      "  header = TRUE\n",
      "} else {\n",
      "  header = FALSE\n",
      "}\n",
      "sets_file = args$sets_file\n",
      "gmt = args$gmt\n",
      "out_tab = args$out_tab\n",
      "min_size = args$min_size\n",
      "max_size = args$max_size\n",
      "n_perm = args$n_perm\n",
      "rda_opt = args$rda_opt\n",
      "plot_opt = args$plot_opt\n",
      "top_num = args$top_num\n",
      "\n",
      "## using the steps from the fgsea vignette\n",
      "rankTab <- read.table(rnk_file, header=header, colClasses = c(\"character\", \"numeric\"))\n",
      "\n",
      "ranks <-rankTab[,2]\n",
      "names(ranks) <- rankTab[,1]\n",
      "\n",
      "if (gmt) {\n",
      "  pathways <- gmtPathways(sets_file)\n",
      "} else {\n",
      "  pathways <- load(sets_file)\n",
      "  pathways <- get(pathways)\n",
      "}\n",
      "\n",
      "# set seed for reproducibility https://github.com/ctlab/fgsea/issues/12\n",
      "set.seed(42)\n",
      "fgseaRes <- fgsea(pathways, ranks, minSize=min_size, maxSize=max_size)\n",
      "fgseaRes <- fgseaRes[order(pval), ]\n",
      "# convert leadingEdge column from list to character to output\n",
      "fgseaRes$leadingEdge <- sapply(fgseaRes$leadingEdge, toString)\n",
      "\n",
      "write.table(fgseaRes, out_tab, sep=\"\\t\", row.names=FALSE, quote=FALSE)\n",
      "\n",
      "if (plot_opt) {\n",
      "  pdf(\"fgsea_plots.pdf\", width=8)\n",
      "\n",
      "  topPathways <- head(fgseaRes, n=top_num)\n",
      "  topPathways <- topPathways$pathway\n",
      "\n",
      "  ## make summary table plot for top pathways\n",
      "  p <- plotGseaTable(pathways[topPathways], ranks, fgseaRes, gseaParam = 0.5,\n",
      "  colwidths = c(5.3,3,0.7, 0.9, 0.9))\n",
      "  print(p)\n",
      "\n",
      "  # make enrichment plots for top pathways\n",
      "  for (i in topPathways) {\n",
      "    p <- plotEnrichment(pathways[[i]], ranks) + labs(title=i)\n",
      "    print(p)\n",
      "  }\n",
      "\n",
      "  dev.off()\n",
      "  \n",
      "  print(\"Create fgsea_plots.pdf...... DOne.\")\n",
      "}\n",
      "\n",
      "## output RData file\n",
      "if (rda_opt) {\n",
      "    save.image(file = \"fgsea_analysis.RData\")\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "! cat scripts/fgsea.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "4eebd1e9-0042-4984-a6b5-3696a10ba7ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway</th>\n",
       "      <th>pval</th>\n",
       "      <th>padj</th>\n",
       "      <th>log2err</th>\n",
       "      <th>ES</th>\n",
       "      <th>NES</th>\n",
       "      <th>size</th>\n",
       "      <th>leadingEdge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HALLMARK_OXIDATIVE_PHOSPHORYLATION</td>\n",
       "      <td>0.002719</td>\n",
       "      <td>0.078159</td>\n",
       "      <td>0.431708</td>\n",
       "      <td>0.422450</td>\n",
       "      <td>1.217541</td>\n",
       "      <td>184</td>\n",
       "      <td>CS, PDHB, NDUFS7, NDUFB7, COX7B, NDUFA9, MRPS1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HALLMARK_UV_RESPONSE_DN</td>\n",
       "      <td>0.003126</td>\n",
       "      <td>0.078159</td>\n",
       "      <td>0.431708</td>\n",
       "      <td>0.430468</td>\n",
       "      <td>1.226789</td>\n",
       "      <td>138</td>\n",
       "      <td>ABCC1, GCNT1, ADD3, SIPA1L1, NFKB1, SERPINE1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HALLMARK_PEROXISOME</td>\n",
       "      <td>0.016071</td>\n",
       "      <td>0.267852</td>\n",
       "      <td>0.352488</td>\n",
       "      <td>0.431720</td>\n",
       "      <td>1.210982</td>\n",
       "      <td>104</td>\n",
       "      <td>CRABP2, SLC25A19, FADS1, PEX11A, CADM1, IDH1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HALLMARK_GLYCOLYSIS</td>\n",
       "      <td>0.034965</td>\n",
       "      <td>0.437063</td>\n",
       "      <td>0.241340</td>\n",
       "      <td>0.395600</td>\n",
       "      <td>1.142261</td>\n",
       "      <td>197</td>\n",
       "      <td>ARPP19, GAL3ST1, P4HA2, HDLBP, B3GAT1, VCAN, I...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HALLMARK_FATTY_ACID_METABOLISM</td>\n",
       "      <td>0.044955</td>\n",
       "      <td>0.449550</td>\n",
       "      <td>0.211400</td>\n",
       "      <td>0.400885</td>\n",
       "      <td>1.147461</td>\n",
       "      <td>155</td>\n",
       "      <td>PDHB, PPARA, IDH1, HSP90AA1, HSD17B4, FABP1, P...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY</td>\n",
       "      <td>0.079920</td>\n",
       "      <td>0.585129</td>\n",
       "      <td>0.155242</td>\n",
       "      <td>0.449017</td>\n",
       "      <td>1.201711</td>\n",
       "      <td>47</td>\n",
       "      <td>ABCC1, MPO, NDUFS2, CDKN2D, PRDX6, STK25, HMOX...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HALLMARK_APICAL_SURFACE</td>\n",
       "      <td>0.081918</td>\n",
       "      <td>0.585129</td>\n",
       "      <td>0.153159</td>\n",
       "      <td>0.456963</td>\n",
       "      <td>1.217509</td>\n",
       "      <td>43</td>\n",
       "      <td>LYN, ADAM10, FLOT2, CROCC, APP, NTNG1, LYPD3, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HALLMARK_BILE_ACID_METABOLISM</td>\n",
       "      <td>0.105894</td>\n",
       "      <td>0.599401</td>\n",
       "      <td>0.132846</td>\n",
       "      <td>0.399053</td>\n",
       "      <td>1.125537</td>\n",
       "      <td>112</td>\n",
       "      <td>FADS1, PEX11A, AGXT, IDH1, SLC27A5, GNPAT, HSD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HALLMARK_IL2_STAT5_SIGNALING</td>\n",
       "      <td>0.107892</td>\n",
       "      <td>0.599401</td>\n",
       "      <td>0.131458</td>\n",
       "      <td>0.381330</td>\n",
       "      <td>1.100535</td>\n",
       "      <td>194</td>\n",
       "      <td>MXD1, GPR83, NDRG1, WLS, TGM2, CST7, SWAP70, B...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HALLMARK_ADIPOGENESIS</td>\n",
       "      <td>0.147852</td>\n",
       "      <td>0.699301</td>\n",
       "      <td>0.109684</td>\n",
       "      <td>0.375415</td>\n",
       "      <td>1.083243</td>\n",
       "      <td>192</td>\n",
       "      <td>CS, ANGPT1, NDUFB7, COX7B, GPHN, NKIRAS1, COQ5...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    pathway      pval      padj   log2err  \\\n",
       "0        HALLMARK_OXIDATIVE_PHOSPHORYLATION  0.002719  0.078159  0.431708   \n",
       "1                   HALLMARK_UV_RESPONSE_DN  0.003126  0.078159  0.431708   \n",
       "2                       HALLMARK_PEROXISOME  0.016071  0.267852  0.352488   \n",
       "3                       HALLMARK_GLYCOLYSIS  0.034965  0.437063  0.241340   \n",
       "4            HALLMARK_FATTY_ACID_METABOLISM  0.044955  0.449550  0.211400   \n",
       "5  HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY  0.079920  0.585129  0.155242   \n",
       "6                   HALLMARK_APICAL_SURFACE  0.081918  0.585129  0.153159   \n",
       "7             HALLMARK_BILE_ACID_METABOLISM  0.105894  0.599401  0.132846   \n",
       "8              HALLMARK_IL2_STAT5_SIGNALING  0.107892  0.599401  0.131458   \n",
       "9                     HALLMARK_ADIPOGENESIS  0.147852  0.699301  0.109684   \n",
       "\n",
       "         ES       NES  size                                        leadingEdge  \n",
       "0  0.422450  1.217541   184  CS, PDHB, NDUFS7, NDUFB7, COX7B, NDUFA9, MRPS1...  \n",
       "1  0.430468  1.226789   138  ABCC1, GCNT1, ADD3, SIPA1L1, NFKB1, SERPINE1, ...  \n",
       "2  0.431720  1.210982   104  CRABP2, SLC25A19, FADS1, PEX11A, CADM1, IDH1, ...  \n",
       "3  0.395600  1.142261   197  ARPP19, GAL3ST1, P4HA2, HDLBP, B3GAT1, VCAN, I...  \n",
       "4  0.400885  1.147461   155  PDHB, PPARA, IDH1, HSP90AA1, HSD17B4, FABP1, P...  \n",
       "5  0.449017  1.201711    47  ABCC1, MPO, NDUFS2, CDKN2D, PRDX6, STK25, HMOX...  \n",
       "6  0.456963  1.217509    43  LYN, ADAM10, FLOT2, CROCC, APP, NTNG1, LYPD3, ...  \n",
       "7  0.399053  1.125537   112  FADS1, PEX11A, AGXT, IDH1, SLC27A5, GNPAT, HSD...  \n",
       "8  0.381330  1.100535   194  MXD1, GPR83, NDRG1, WLS, TGM2, CST7, SWAP70, B...  \n",
       "9  0.375415  1.083243   192  CS, ANGPT1, NDUFB7, COX7B, GPHN, NKIRAS1, COQ5...  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_table('fgsea/fgsea_pathways.txt', delimiter='\\t').head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc9bc253-5bed-46ed-8063-dac37970b4f7",
   "metadata": {},
   "source": [
    "[FGSEA plots](./fgsea/fgsea_plots.pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "40a5d5f4-0424-4faa-8d76-edd99f3427a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"600\"\n",
       "            height=\"300\"\n",
       "            src=\"fgsea/fgsea_plots.pdf\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7fc17d71e3d0>"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import IFrame\n",
    "IFrame(\"fgsea/fgsea_plots.pdf\", width=600, height=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4cbdeb3-2a21-4e45-9244-4a43181e6839",
   "metadata": {},
   "source": [
    "#### Multiple conditions\n",
    "\n",
    "Use __MAGeck mle__ to compare multiple conditions: the drug treatment (T8-APR-246) to T0-COntrol and the vehicle (T8-Vehicle) to T0-Control.\n",
    "\n",
    "MAGeCK mle uses a maximum likelihood estimation (MLE) algorithm. It outputs a single value (beta score) per gene instead of a score for both negative and positive selection. A negative beta score indicates negative selection and a positive indicates positive selection. MAGeCK mle can also be used for comparing 2 conditions instead of MAGeCK test (RRA algorithm) but it is slower.\n",
    "\n",
    "A MAGeCK mle run requires the MAGeCK sgRNA counts and a design matrix. \n",
    "\n",
    "Design matrix for 3 conditions: baseline (T0), Vehicle and APR. The 1st row indicates that T0 is baseline sample, the 2nd row indicates Vehicle to T0 comparison, and the 3rd row indicates APR to T0 comparison.\n",
    "\n",
    "For MAGeCK mle the design matrix:\n",
    "\n",
    "- must include a header line of condition labels\n",
    "- the first column is the sample labels that must match labels in read count file\n",
    "- the second column must be a “baseline” column that sets all values to “1”\n",
    "- the value in the design matrix is either “0” or “1”\n",
    "- must have at least one sample of “initial state” (e.g., time 0) that has only one “1” in the corresponding row. That only “1” must be in the baseline column\n",
    "- the initial state (baseline) sample must be the 1st sample listed, under the header row."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "483ccc67-74ad-401b-83f7-32b64f11c773",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Samples</th>\n",
       "      <th>baseline</th>\n",
       "      <th>Vehicle</th>\n",
       "      <th>APR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T0-Control</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T8-Vehicle</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T8-APR-246</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Samples  baseline  Vehicle  APR\n",
       "0  T0-Control         1        0    0\n",
       "1  T8-Vehicle         1        1    0\n",
       "2  T8-APR-246         1        0    1"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "design_matrix = pd.read_csv(\"mle/full_mageck_mle_design_matrix.tsv\", delimiter=\"\\t\")\n",
    "design_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "dd502bfb-cd9b-4e3d-b6e8-08269dffa6b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  @ Thu, 21 Dec 2023 07:44:27: Parameters: /usr/local/bin/mageck mle -k counting/full_mageck_sgrna_counts.tsv -d mle/full_mageck_mle_design_matrix.tsv -n output_mle --norm-method median --threads 4 \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:27: Cannot parse design matrix as a string; try to parse it as a file name ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:27: Design matrix: \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:27: [[1. 0. 0.] \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:27:  [1. 1. 0.] \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:27:  [1. 0. 1.]] \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:27: Beta labels:baseline,Vehicle,APR \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:27: Included samples:T0-Control,T8-Vehicle,T8-APR-246 \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:27: Loaded samples:T0-Control;T8-Vehicle;T8-APR-246 \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:27: Sample index: 2;1;0 \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:27: Loaded 20112 genes. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:28: Final size factor: 0.8688780183350758 1.0565381424478189 1.1107589343544333 \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:28: size factor: 1.150909539541784,0.946487362664608,0.9002853536183296 \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 0 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 0: total 1 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 1 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 1: total 0 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 2 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 2: total 0 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 3 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 3: total 0 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 0 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 1 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 2 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 3 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: All threads completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Modeling the mean and variance ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Run the algorithm for the second time ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 0 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 0: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 1 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 1: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 2 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 2: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 3 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 3: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 0: Calculating PIP5K1B (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 1: Calculating controlguide578 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 2: Calculating GALNT12 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:29: Thread 3: Calculating MAP7 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:33: Thread 0: Calculating TRIP11 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:33: Thread 2: Calculating PASD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:33: Thread 1: Calculating EFHD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:34: Thread 3: Calculating WI2-2373I1.2 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:38: Thread 0: Calculating TMEM181 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:38: Thread 2: Calculating CLCF1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:38: Thread 3: Calculating HINT1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:38: Thread 1: Calculating FAM133B (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:42: Thread 0: Calculating ADGRL2 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:43: Thread 2: Calculating TJP3 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:43: Thread 3: Calculating ALOX12 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:43: Thread 1: Calculating controlguide609 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:47: Thread 0: Calculating PZP (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:47: Thread 2: Calculating EIF3M (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:47: Thread 3: Calculating CHRM5 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:47: Thread 1: Calculating DAZ2 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:51: Thread 0: Calculating TXN2 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:52: Thread 0 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:52: Thread 1: Calculating CCT7 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:52: Thread 3: Calculating NBPF15 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:52: Thread 2: Calculating PLK1 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:52: Thread 1 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:52: Thread 2 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:52: Thread 3 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:52: All threads completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:56: Permuting groups of gene with 1 sgRNAs per gene. Group progress: 1/4 \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:56: Start permuting 2 rounds ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:56: Collecting 77441 sgRNAs from 20112 genes. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:58: Permuting round 0 ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:58: Thread 0 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:58: Thread 0: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:58: Thread 1 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:58: Thread 1: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:58: Thread 2 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:58: Thread 2: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:58: Thread 3 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:58: Thread 3: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:58: Thread 0: Calculating PIP5K1B (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:58: Thread 1: Calculating controlguide578 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:58: Thread 2: Calculating GALNT12 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:44:58: Thread 3: Calculating MAP7 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:01: Thread 2: Calculating PASD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:02: Thread 0: Calculating TRIP11 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:02: Thread 1: Calculating EFHD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:02: Thread 3: Calculating WI2-2373I1.2 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:04: Thread 2: Calculating CLCF1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:04: Thread 1: Calculating FAM133B (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:04: Thread 0: Calculating TMEM181 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:05: Thread 3: Calculating HINT1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:07: Thread 2: Calculating TJP3 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:07: Thread 1: Calculating controlguide609 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:07: Thread 0: Calculating ADGRL2 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:07: Thread 3: Calculating ALOX12 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:10: Thread 1: Calculating DAZ2 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:10: Thread 0: Calculating PZP (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:10: Thread 2: Calculating EIF3M (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:10: Thread 3: Calculating CHRM5 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:13: Thread 1: Calculating CCT7 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:13: Thread 0: Calculating TXN2 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:13: Thread 2: Calculating PLK1 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:13: Thread 3: Calculating NBPF15 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:13: Thread 0 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:13: Thread 1 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:14: Thread 2 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:14: Thread 3 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:14: All threads completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:18: Permuting round 1 ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:18: Thread 0 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:18: Thread 0: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:18: Thread 1 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:18: Thread 1: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:18: Thread 2 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:18: Thread 2: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:18: Thread 3 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:18: Thread 3: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:18: Thread 0: Calculating PIP5K1B (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:18: Thread 1: Calculating controlguide578 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:18: Thread 2: Calculating GALNT12 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:18: Thread 3: Calculating MAP7 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:21: Thread 1: Calculating EFHD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:21: Thread 2: Calculating PASD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:22: Thread 0: Calculating TRIP11 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:22: Thread 3: Calculating WI2-2373I1.2 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:25: Thread 2: Calculating CLCF1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:25: Thread 1: Calculating FAM133B (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:25: Thread 0: Calculating TMEM181 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:25: Thread 3: Calculating HINT1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:28: Thread 2: Calculating TJP3 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:28: Thread 1: Calculating controlguide609 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:28: Thread 0: Calculating ADGRL2 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:28: Thread 3: Calculating ALOX12 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:31: Thread 2: Calculating EIF3M (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:31: Thread 1: Calculating DAZ2 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:31: Thread 0: Calculating PZP (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:31: Thread 3: Calculating CHRM5 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:33: Thread 2: Calculating PLK1 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:34: Thread 1: Calculating CCT7 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:34: Thread 0: Calculating TXN2 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:34: Thread 3: Calculating NBPF15 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:34: Thread 0 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:34: Thread 1 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:34: Thread 2 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:34: Thread 3 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:34: All threads completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:38: Assigning p values... \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:40: Permuting groups of gene with 2 sgRNAs per gene. Group progress: 2/4 \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:40: Start permuting 2 rounds ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:41: Collecting 77441 sgRNAs from 20112 genes. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:42: Permuting round 0 ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:43: Thread 0 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:43: Thread 0: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:43: Thread 1 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:43: Thread 1: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:43: Thread 2 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:43: Thread 2: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:43: Thread 3 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:43: Thread 0: Calculating PIP5K1B (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:43: Thread 3: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:43: Thread 1: Calculating controlguide578 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:43: Thread 2: Calculating GALNT12 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:43: Thread 3: Calculating MAP7 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:48: Thread 3: Calculating WI2-2373I1.2 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:48: Thread 2: Calculating PASD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:48: Thread 0: Calculating TRIP11 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:48: Thread 1: Calculating EFHD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:52: Thread 3: Calculating HINT1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:52: Thread 2: Calculating CLCF1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:52: Thread 1: Calculating FAM133B (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:52: Thread 0: Calculating TMEM181 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:57: Thread 3: Calculating ALOX12 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:57: Thread 2: Calculating TJP3 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:57: Thread 1: Calculating controlguide609 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:45:57: Thread 0: Calculating ADGRL2 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:02: Thread 3: Calculating CHRM5 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:02: Thread 2: Calculating EIF3M (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:02: Thread 0: Calculating PZP (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:06: Thread 1: Calculating DAZ2 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:07: Thread 3: Calculating NBPF15 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:07: Thread 0: Calculating TXN2 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:07: Thread 0 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:08: Thread 2: Calculating PLK1 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:10: Thread 1: Calculating CCT7 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:10: Thread 1 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:10: Thread 2 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:10: Thread 3 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:10: All threads completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:15: Permuting round 1 ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:15: Thread 0 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:15: Thread 0: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:15: Thread 1 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:15: Thread 1: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:15: Thread 2 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:15: Thread 2: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:15: Thread 0: Calculating PIP5K1B (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:15: Thread 3 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:15: Thread 3: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:15: Thread 1: Calculating controlguide578 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:15: Thread 2: Calculating GALNT12 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:16: Thread 3: Calculating MAP7 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:20: Thread 1: Calculating EFHD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:21: Thread 2: Calculating PASD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:21: Thread 3: Calculating WI2-2373I1.2 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:21: Thread 0: Calculating TRIP11 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:25: Thread 1: Calculating FAM133B (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:25: Thread 3: Calculating HINT1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:25: Thread 0: Calculating TMEM181 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:26: Thread 2: Calculating CLCF1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:29: Thread 1: Calculating controlguide609 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:30: Thread 3: Calculating ALOX12 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:30: Thread 0: Calculating ADGRL2 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:31: Thread 2: Calculating TJP3 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:34: Thread 1: Calculating DAZ2 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:34: Thread 3: Calculating CHRM5 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:34: Thread 0: Calculating PZP (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:36: Thread 2: Calculating EIF3M (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:39: Thread 1: Calculating CCT7 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:39: Thread 3: Calculating NBPF15 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:39: Thread 0: Calculating TXN2 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:39: Thread 0 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:39: Thread 1 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:40: Thread 2: Calculating PLK1 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:40: Thread 2 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:40: Thread 3 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:40: All threads completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:45: Assigning p values... \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:45: Permuting groups of gene with 4 sgRNAs per gene. Group progress: 3/4 \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:45: Start permuting 2 rounds ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:46: Collecting 77441 sgRNAs from 20112 genes. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:47: Permuting round 0 ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:47: Thread 0 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:47: Thread 0: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:47: Thread 1 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:47: Thread 1: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:47: Thread 2 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:47: Thread 2: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:47: Thread 0: Calculating PIP5K1B (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:47: Thread 3 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:47: Thread 3: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:47: Thread 1: Calculating controlguide578 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:47: Thread 2: Calculating GALNT12 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:47: Thread 3: Calculating MAP7 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:52: Thread 0: Calculating TRIP11 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:52: Thread 1: Calculating EFHD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:52: Thread 2: Calculating PASD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:53: Thread 3: Calculating WI2-2373I1.2 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:57: Thread 1: Calculating FAM133B (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:58: Thread 0: Calculating TMEM181 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:58: Thread 3: Calculating HINT1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:46:58: Thread 2: Calculating CLCF1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:02: Thread 1: Calculating controlguide609 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:02: Thread 0: Calculating ADGRL2 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:03: Thread 3: Calculating ALOX12 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:03: Thread 2: Calculating TJP3 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:07: Thread 0: Calculating PZP (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:07: Thread 1: Calculating DAZ2 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:08: Thread 3: Calculating CHRM5 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:08: Thread 2: Calculating EIF3M (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:12: Thread 0: Calculating TXN2 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:13: Thread 1: Calculating CCT7 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:13: Thread 0 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:13: Thread 1 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:13: Thread 2: Calculating PLK1 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:13: Thread 3: Calculating NBPF15 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:13: Thread 2 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:13: Thread 3 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:13: All threads completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:18: Permuting round 1 ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:18: Thread 0 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:18: Thread 0: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:18: Thread 1 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:18: Thread 1: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:18: Thread 2 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:18: Thread 2: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:18: Thread 3 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:18: Thread 3: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:18: Thread 0: Calculating PIP5K1B (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:18: Thread 1: Calculating controlguide578 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:18: Thread 2: Calculating GALNT12 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:18: Thread 3: Calculating MAP7 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:23: Thread 1: Calculating EFHD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:23: Thread 2: Calculating PASD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:23: Thread 3: Calculating WI2-2373I1.2 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:24: Thread 0: Calculating TRIP11 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:29: Thread 1: Calculating FAM133B (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:29: Thread 2: Calculating CLCF1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:29: Thread 3: Calculating HINT1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:29: Thread 0: Calculating TMEM181 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:34: Thread 3: Calculating ALOX12 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:34: Thread 1: Calculating controlguide609 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:34: Thread 2: Calculating TJP3 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:34: Thread 0: Calculating ADGRL2 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:39: Thread 2: Calculating EIF3M (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:39: Thread 1: Calculating DAZ2 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:39: Thread 0: Calculating PZP (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:39: Thread 3: Calculating CHRM5 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:44: Thread 2: Calculating PLK1 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:44: Thread 0: Calculating TXN2 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:44: Thread 1: Calculating CCT7 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:44: Thread 0 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:44: Thread 3: Calculating NBPF15 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:45: Thread 1 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:45: Thread 2 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:45: Thread 3 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:45: All threads completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:47:49: Assigning p values... \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:28: Permuting groups of gene with 8 sgRNAs per gene. Group progress: 4/4 \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:28: Start permuting 2 rounds ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:29: Collecting 77441 sgRNAs from 20112 genes. \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:30: Permuting round 0 ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:31: Thread 0 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:31: Thread 0: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:31: Thread 1 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:31: Thread 1: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:31: Thread 2 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:31: Thread 2: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:31: Thread 0: Calculating PIP5K1B (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:31: Thread 3 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:31: Thread 3: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:31: Thread 1: Calculating controlguide578 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:31: Thread 2: Calculating GALNT12 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:31: Thread 3: Calculating MAP7 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:37: Thread 2: Calculating PASD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:37: Thread 3: Calculating WI2-2373I1.2 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:37: Thread 0: Calculating TRIP11 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:37: Thread 1: Calculating EFHD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:43: Thread 0: Calculating TMEM181 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:43: Thread 1: Calculating FAM133B (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:43: Thread 2: Calculating CLCF1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:43: Thread 3: Calculating HINT1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:48: Thread 0: Calculating ADGRL2 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:48: Thread 1: Calculating controlguide609 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:49: Thread 2: Calculating TJP3 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:49: Thread 3: Calculating ALOX12 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:54: Thread 0: Calculating PZP (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:54: Thread 1: Calculating DAZ2 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:54: Thread 2: Calculating EIF3M (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:48:55: Thread 3: Calculating CHRM5 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:00: Thread 0: Calculating TXN2 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:00: Thread 1: Calculating CCT7 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:00: Thread 0 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:00: Thread 2: Calculating PLK1 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:00: Thread 1 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:00: Thread 2 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:00: Thread 3: Calculating NBPF15 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:01: Thread 3 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:01: All threads completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:07: Permuting round 1 ... \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:07: Thread 0 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:07: Thread 0: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:07: Thread 1 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:07: Thread 1: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:07: Thread 2 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:07: Thread 2: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:07: Thread 0: Calculating PIP5K1B (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:07: Thread 3 started. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:07: Thread 3: total 5028 instances. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:07: Thread 1: Calculating controlguide578 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:07: Thread 2: Calculating GALNT12 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:07: Thread 3: Calculating MAP7 (1) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:13: Thread 2: Calculating PASD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:13: Thread 1: Calculating EFHD1 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:13: Thread 0: Calculating TRIP11 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:14: Thread 3: Calculating WI2-2373I1.2 (1001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:19: Thread 1: Calculating FAM133B (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:19: Thread 2: Calculating CLCF1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:19: Thread 0: Calculating TMEM181 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:19: Thread 3: Calculating HINT1 (2001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:25: Thread 2: Calculating TJP3 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:25: Thread 1: Calculating controlguide609 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:25: Thread 0: Calculating ADGRL2 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:25: Thread 3: Calculating ALOX12 (3001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:30: Thread 2: Calculating EIF3M (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:30: Thread 0: Calculating PZP (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:30: Thread 1: Calculating DAZ2 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:31: Thread 3: Calculating CHRM5 (4001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:36: Thread 2: Calculating PLK1 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:36: Thread 1: Calculating CCT7 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:36: Thread 0: Calculating TXN2 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:37: Thread 0 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:37: Thread 1 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:37: Thread 2 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:37: Thread 3: Calculating NBPF15 (5001) ...  \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:37: Thread 3 completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:37: All threads completed. \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:43: Assigning p values... \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:45: Writing gene results to output_mle.gene_summary.txt \n",
      "INFO  @ Thu, 21 Dec 2023 07:49:45: Writing sgRNA results to output_mle.sgrna_summary.txt \n"
     ]
    }
   ],
   "source": [
    "! mageck mle -k counting/full_mageck_sgrna_counts.tsv -d mle/full_mageck_mle_design_matrix.tsv \\\n",
    "-n output_mle --norm-method median --threads 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "7ea3780c-74da-4ec3-9c5f-ca2f7e4d3474",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(77441, 3)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sgRNA_summary_mle = pd.read_table('mle/output_mle.sgrna_summary.txt')\n",
    "sgRNA_summary_mle.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "a2392524-a595-4eec-aadf-ef4462573cf1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>sgRNA</th>\n",
       "      <th>eff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DUSP14</td>\n",
       "      <td>ID_30030</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DUSP14</td>\n",
       "      <td>ID_30029</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DUSP14</td>\n",
       "      <td>ID_30028</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DUSP14</td>\n",
       "      <td>ID_30027</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>ID_21657</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>ID_21660</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>ID_21658</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>ID_21659</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CTAGE6</td>\n",
       "      <td>ID_69143</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CTAGE6</td>\n",
       "      <td>ID_69142</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene     sgRNA  eff\n",
       "0     DUSP14  ID_30030    1\n",
       "1     DUSP14  ID_30029    1\n",
       "2     DUSP14  ID_30028    1\n",
       "3     DUSP14  ID_30027    1\n",
       "4  HIST1H2BL  ID_21657    1\n",
       "5  HIST1H2BL  ID_21660    1\n",
       "6  HIST1H2BL  ID_21658    1\n",
       "7  HIST1H2BL  ID_21659    1\n",
       "8     CTAGE6  ID_69143    1\n",
       "9     CTAGE6  ID_69142    1"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sgRNA_summary_mle.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "b897058f-c18a-4961-adda-6011648a9e20",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20112, 14)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_summary_mle = pd.read_table('mle/output_mle.gene_summary.txt')\n",
    "gene_summary_mle.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e6cf212e-7a8c-41b0-81ab-b0393a7834a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>sgRNA</th>\n",
       "      <th>Vehicle|beta</th>\n",
       "      <th>Vehicle|z</th>\n",
       "      <th>Vehicle|p-value</th>\n",
       "      <th>Vehicle|fdr</th>\n",
       "      <th>Vehicle|wald-p-value</th>\n",
       "      <th>Vehicle|wald-fdr</th>\n",
       "      <th>APR|beta</th>\n",
       "      <th>APR|z</th>\n",
       "      <th>APR|p-value</th>\n",
       "      <th>APR|fdr</th>\n",
       "      <th>APR|wald-p-value</th>\n",
       "      <th>APR|wald-fdr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DUSP14</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.046593</td>\n",
       "      <td>-0.406180</td>\n",
       "      <td>0.81056</td>\n",
       "      <td>0.99079</td>\n",
       "      <td>0.68461</td>\n",
       "      <td>0.96053</td>\n",
       "      <td>0.071544</td>\n",
       "      <td>0.62381</td>\n",
       "      <td>0.78187</td>\n",
       "      <td>0.98870</td>\n",
       "      <td>0.53275</td>\n",
       "      <td>0.84452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HIST1H2BL</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.387780</td>\n",
       "      <td>-1.370100</td>\n",
       "      <td>0.09825</td>\n",
       "      <td>0.64533</td>\n",
       "      <td>0.17065</td>\n",
       "      <td>0.95209</td>\n",
       "      <td>-0.066854</td>\n",
       "      <td>-0.23705</td>\n",
       "      <td>0.73523</td>\n",
       "      <td>0.98462</td>\n",
       "      <td>0.81262</td>\n",
       "      <td>0.94193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTAGE6</td>\n",
       "      <td>4</td>\n",
       "      <td>0.070307</td>\n",
       "      <td>0.389960</td>\n",
       "      <td>0.70744</td>\n",
       "      <td>0.98061</td>\n",
       "      <td>0.69656</td>\n",
       "      <td>0.96074</td>\n",
       "      <td>-0.107730</td>\n",
       "      <td>-0.59718</td>\n",
       "      <td>0.60800</td>\n",
       "      <td>0.97167</td>\n",
       "      <td>0.55039</td>\n",
       "      <td>0.85088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TBC1D10C</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.004010</td>\n",
       "      <td>-0.053744</td>\n",
       "      <td>0.98086</td>\n",
       "      <td>0.99824</td>\n",
       "      <td>0.95714</td>\n",
       "      <td>0.99509</td>\n",
       "      <td>-0.045206</td>\n",
       "      <td>-0.60519</td>\n",
       "      <td>0.80842</td>\n",
       "      <td>0.99177</td>\n",
       "      <td>0.54505</td>\n",
       "      <td>0.84882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PIP5K1B</td>\n",
       "      <td>4</td>\n",
       "      <td>0.190520</td>\n",
       "      <td>1.373800</td>\n",
       "      <td>0.29410</td>\n",
       "      <td>0.87983</td>\n",
       "      <td>0.16950</td>\n",
       "      <td>0.95209</td>\n",
       "      <td>0.179410</td>\n",
       "      <td>1.29340</td>\n",
       "      <td>0.43541</td>\n",
       "      <td>0.93838</td>\n",
       "      <td>0.19587</td>\n",
       "      <td>0.75718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>controlguide578</td>\n",
       "      <td>1</td>\n",
       "      <td>0.415150</td>\n",
       "      <td>0.391540</td>\n",
       "      <td>0.22817</td>\n",
       "      <td>0.83798</td>\n",
       "      <td>0.69540</td>\n",
       "      <td>0.96074</td>\n",
       "      <td>0.662850</td>\n",
       "      <td>0.62516</td>\n",
       "      <td>0.11182</td>\n",
       "      <td>0.67497</td>\n",
       "      <td>0.53187</td>\n",
       "      <td>0.84402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GALNT12</td>\n",
       "      <td>4</td>\n",
       "      <td>0.111720</td>\n",
       "      <td>0.582340</td>\n",
       "      <td>0.54196</td>\n",
       "      <td>0.95750</td>\n",
       "      <td>0.56034</td>\n",
       "      <td>0.95300</td>\n",
       "      <td>0.212830</td>\n",
       "      <td>1.10950</td>\n",
       "      <td>0.34596</td>\n",
       "      <td>0.90939</td>\n",
       "      <td>0.26723</td>\n",
       "      <td>0.76828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MAP7</td>\n",
       "      <td>4</td>\n",
       "      <td>0.246180</td>\n",
       "      <td>1.049700</td>\n",
       "      <td>0.17363</td>\n",
       "      <td>0.78962</td>\n",
       "      <td>0.29388</td>\n",
       "      <td>0.95209</td>\n",
       "      <td>0.342050</td>\n",
       "      <td>1.45860</td>\n",
       "      <td>0.12197</td>\n",
       "      <td>0.68885</td>\n",
       "      <td>0.14468</td>\n",
       "      <td>0.75718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>APBB3</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.043763</td>\n",
       "      <td>-0.338840</td>\n",
       "      <td>0.82244</td>\n",
       "      <td>0.99079</td>\n",
       "      <td>0.73473</td>\n",
       "      <td>0.96644</td>\n",
       "      <td>-0.106320</td>\n",
       "      <td>-0.82249</td>\n",
       "      <td>0.61088</td>\n",
       "      <td>0.97209</td>\n",
       "      <td>0.41080</td>\n",
       "      <td>0.81037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EXD2</td>\n",
       "      <td>4</td>\n",
       "      <td>0.151600</td>\n",
       "      <td>0.782790</td>\n",
       "      <td>0.40528</td>\n",
       "      <td>0.92816</td>\n",
       "      <td>0.43375</td>\n",
       "      <td>0.95209</td>\n",
       "      <td>0.225530</td>\n",
       "      <td>1.16470</td>\n",
       "      <td>0.31678</td>\n",
       "      <td>0.89732</td>\n",
       "      <td>0.24414</td>\n",
       "      <td>0.76466</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Gene  sgRNA  Vehicle|beta  Vehicle|z  Vehicle|p-value  \\\n",
       "0           DUSP14      4     -0.046593  -0.406180          0.81056   \n",
       "1        HIST1H2BL      4     -0.387780  -1.370100          0.09825   \n",
       "2           CTAGE6      4      0.070307   0.389960          0.70744   \n",
       "3         TBC1D10C      4     -0.004010  -0.053744          0.98086   \n",
       "4          PIP5K1B      4      0.190520   1.373800          0.29410   \n",
       "5  controlguide578      1      0.415150   0.391540          0.22817   \n",
       "6          GALNT12      4      0.111720   0.582340          0.54196   \n",
       "7             MAP7      4      0.246180   1.049700          0.17363   \n",
       "8            APBB3      4     -0.043763  -0.338840          0.82244   \n",
       "9             EXD2      4      0.151600   0.782790          0.40528   \n",
       "\n",
       "   Vehicle|fdr  Vehicle|wald-p-value  Vehicle|wald-fdr  APR|beta    APR|z  \\\n",
       "0      0.99079               0.68461           0.96053  0.071544  0.62381   \n",
       "1      0.64533               0.17065           0.95209 -0.066854 -0.23705   \n",
       "2      0.98061               0.69656           0.96074 -0.107730 -0.59718   \n",
       "3      0.99824               0.95714           0.99509 -0.045206 -0.60519   \n",
       "4      0.87983               0.16950           0.95209  0.179410  1.29340   \n",
       "5      0.83798               0.69540           0.96074  0.662850  0.62516   \n",
       "6      0.95750               0.56034           0.95300  0.212830  1.10950   \n",
       "7      0.78962               0.29388           0.95209  0.342050  1.45860   \n",
       "8      0.99079               0.73473           0.96644 -0.106320 -0.82249   \n",
       "9      0.92816               0.43375           0.95209  0.225530  1.16470   \n",
       "\n",
       "   APR|p-value  APR|fdr  APR|wald-p-value  APR|wald-fdr  \n",
       "0      0.78187  0.98870           0.53275       0.84452  \n",
       "1      0.73523  0.98462           0.81262       0.94193  \n",
       "2      0.60800  0.97167           0.55039       0.85088  \n",
       "3      0.80842  0.99177           0.54505       0.84882  \n",
       "4      0.43541  0.93838           0.19587       0.75718  \n",
       "5      0.11182  0.67497           0.53187       0.84402  \n",
       "6      0.34596  0.90939           0.26723       0.76828  \n",
       "7      0.12197  0.68885           0.14468       0.75718  \n",
       "8      0.61088  0.97209           0.41080       0.81037  \n",
       "9      0.31678  0.89732           0.24414       0.76466  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_summary_mle.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "847a362d-1e32-44d9-894a-46c085dbbc0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Gene', 'sgRNA', 'Vehicle|beta', 'Vehicle|z', 'Vehicle|p-value',\n",
       "       'Vehicle|fdr', 'Vehicle|wald-p-value', 'Vehicle|wald-fdr', 'APR|beta',\n",
       "       'APR|z', 'APR|p-value', 'APR|fdr', 'APR|wald-p-value', 'APR|wald-fdr'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_summary_mle.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "42f95d10-ec27-45cd-b686-bd81827a1455",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Column_name</th>\n",
       "      <th>Content</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sgRNA</td>\n",
       "      <td>The number of targeting sgRNAs for each gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Vehicle|beta</td>\n",
       "      <td>The beta score of this gene in Vehicle condition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Vehicle|z</td>\n",
       "      <td>The raw p-value (using permutation) of this ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Vehicle|p-value</td>\n",
       "      <td>The false discovery rate for this gene in Vehi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Vehicle|fdr</td>\n",
       "      <td>The z-score associated with Wald test in Vehic...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Vehicle|wald-p-value</td>\n",
       "      <td>The p-value using Wald test in the Vehicle sel...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Vehicle|wald-fdr</td>\n",
       "      <td>The false discovery rate of the Wald test in t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>APR|beta</td>\n",
       "      <td>The beta score of this gene in APR condition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>APR|z</td>\n",
       "      <td>The raw p-value (using permutation) of this ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>APR|p-value</td>\n",
       "      <td>The false discovery rate for this gene in APR ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>APR|fdr</td>\n",
       "      <td>The z-score associated with Wald test in APR s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>APR|wald-p-value</td>\n",
       "      <td>The p-value using Wald test in the APR selection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>APR|wald-fdr</td>\n",
       "      <td>The false discovery rate of the Wald test in t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Column_name                                            Content\n",
       "0                   Gene                                            Gene ID\n",
       "1                  sgRNA       The number of targeting sgRNAs for each gene\n",
       "2           Vehicle|beta   The beta score of this gene in Vehicle condition\n",
       "3              Vehicle|z  The raw p-value (using permutation) of this ge...\n",
       "4        Vehicle|p-value  The false discovery rate for this gene in Vehi...\n",
       "5            Vehicle|fdr  The z-score associated with Wald test in Vehic...\n",
       "6   Vehicle|wald-p-value  The p-value using Wald test in the Vehicle sel...\n",
       "7       Vehicle|wald-fdr  The false discovery rate of the Wald test in t...\n",
       "8               APR|beta       The beta score of this gene in APR condition\n",
       "9                  APR|z  The raw p-value (using permutation) of this ge...\n",
       "10           APR|p-value  The false discovery rate for this gene in APR ...\n",
       "11               APR|fdr  The z-score associated with Wald test in APR s...\n",
       "12      APR|wald-p-value   The p-value using Wald test in the APR selection\n",
       "13          APR|wald-fdr  The false discovery rate of the Wald test in t..."
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_summary_mle_info = {'Column_name': gene_summary_mle.columns, \n",
    "                     'Content': ['Gene ID', \n",
    "                                 'The number of targeting sgRNAs for each gene', \n",
    "                                 'The beta score of this gene in Vehicle condition', \n",
    "                                 'The raw p-value (using permutation) of this gene in Vehicle condition',\n",
    "                                 'The false discovery rate for this gene in Vehicle selection', \n",
    "                                 'The z-score associated with Wald test in Vehicle selection',\n",
    "                                 'The p-value using Wald test in the Vehicle selection',\n",
    "                                 'The false discovery rate of the Wald test in the Vehicle selection',\n",
    "                                 'The beta score of this gene in APR condition', \n",
    "                                 'The raw p-value (using permutation) of this gene in APR condition',\n",
    "                                 'The false discovery rate for this gene in APR selection', \n",
    "                                 'The z-score associated with Wald test in APR selection',\n",
    "                                 'The p-value using Wald test in the APR selection',\n",
    "                                 'The false discovery rate of the Wald test in the APR selection'\n",
    "                                 ]}\n",
    "pd.DataFrame.from_dict(gene_summary_mle_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ca4c5e6-7b2e-4bbd-87dc-a74c45758cdc",
   "metadata": {},
   "source": [
    "#### Visualize mle results\n",
    "\n",
    "Create volcano plot from mle gene summary to visualize the result for APT vs T0 using `Gene` column for label option, `APR|beta` for log fold change, `APR|wald-p-value` for p-value, and `APR|wald-fdr` for fdr option."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "55346e56-dabd-4e23-944b-1d512a91f771",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hnull device \n",
      "          1 \n",
      "\u001b[?25h[1] \"Create volcano_plot.pdf....... DONE.\"\n",
      "\u001b[?25h\u001b[?25h"
     ]
    }
   ],
   "source": [
    "! Rscript scripts/volcanoplot.rscript --input mle/output_mle.gene_summary.txt \\\n",
    "--fdr_col 14 --pval_col 13 --lfc_col 9 --label_col 1 --signif_thresh 0.05 --lfc_thresh 0.0 --topn 10"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5f61cfa-bd35-4d02-85db-ba799ecbb985",
   "metadata": {},
   "source": [
    "[MAGeCK mle APT vs T0 volcano plot](./mle/volcano_plot_APR-T0.pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "6c8f597f-3db9-4b67-a53e-2a091bc3ad7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"600\"\n",
       "            height=\"300\"\n",
       "            src=\"mle/volcano_plot_APR-T0.pdf\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7fc17d72fd10>"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import IFrame\n",
    "IFrame(\"mle/volcano_plot_APR-T0.pdf\", width=600, height=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d90c754-f1ac-4919-a62a-bba7db43cc26",
   "metadata": {},
   "source": [
    "Create volcano plot from mle gene summary to visualize the result for Vehicle vs T0 using `Gene` column for label option, `Vehicle|beta` for log fold change, `Vehicle|wald-p-value` for p-value, and `Vehicle|wald-fdr` for fdr option."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "f1bf5267-48a8-4823-97bd-c94841328fd6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25h\u001b[?25hnull device \n",
      "          1 \n",
      "\u001b[?25h[1] \"Create volcano_plot.pdf....... DONE.\"\n",
      "\u001b[?25h\u001b[?25h"
     ]
    }
   ],
   "source": [
    "! Rscript scripts/volcanoplot.rscript --input mle/output_mle.gene_summary.txt \\\n",
    "--fdr_col 8 --pval_col 7 --lfc_col 3 --label_col 1 --signif_thresh 0.05 --lfc_thresh 0.0 --topn 10"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81c51f01-7682-4ce2-a59c-9f605aff9236",
   "metadata": {},
   "source": [
    "[MAGeCK mle Vehicle vs T0 volcano plot](./mle/volcano_plot_Vehicle-T0.pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "eddbc902-0171-402c-9030-b49557b12fe2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"600\"\n",
       "            height=\"300\"\n",
       "            src=\"mle/volcano_plot_Vehicle-T0.pdf\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7fc17d723b10>"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import IFrame\n",
    "IFrame(\"mle/volcano_plot_Vehicle-T0.pdf\", width=600, height=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2b0c024-fd56-4739-a001-c9f96bf47a21",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
