---
title: "Spatial Transcriptomics Data Analysis with Seurat"
author: "Lavinia Carabet"
output:
  html_notebook: default
---

```{r, message=FALSE, warning=FALSE}
library(Seurat)    # single-cell, spatial transcriptomics data analysis
library(dplyr)     # data manipulation
library(ggplot2)   # visualization
library(patchwork) # combine separate ggplots into the same graphic
library(ggpubr)    # ggplot2 based publication ready plots
library(Rfast2)    # fast statistical and utility functions for data analysis
library(hdf5r)     # interface to the HDF5 binary data format 
                   # for storing and managing large amounts of data
```

## Data Loading

Read the 10X Genomics Visium spatial gene expression dataset of mouse brain serial section 2 (Sagittal-Posterior) collected by Space Ranger 1.1.0.

The data encompasses:

* gene expression matrix (V1_Mouse_Brain_Sagittal_Posterior_Section_2_filtered_feature_bc_matrix.h5),

* spatial imaging data (V1_Mouse_Brain_Sagittal_Posterior_Section_2_spatial.tar.gz)

```{r, message=FALSE, warning=FALSE}
data.dir <- './data/'
files <- list.files(data.dir)

feature_matrix_name <-files[grepl("feature", files)]
feature_matrix_name

spatial_folder_name <- paste0(data.dir, files[grepl("spatial", files)])
spatial_folder_name
untar(spatial_folder_name, exdir = data.dir)
untar(spatial_folder_name,list = TRUE)
```
### Load 10x Genomics Visium Spatial Experiment into a Seurat object

```{r, message=FALSE, warning=FALSE}
spatial_brain_data <- Seurat::Load10X_Spatial(
  data.dir = data.dir,            # directory containing the H5 file specified by filename and the image data in a subdirectory called spatial
  filename = feature_matrix_name, # name of H5 file containing the feature barcode matrix
  assay = "Spatial",              # name of the initial assay
  slice = "slice1",               # name for the stored image of the tissue slice
  filter.matrix = TRUE,           # Only keep spots that have been determined to be over tissue
  to.upper = FALSE
)

spatial_brain_data
```
### Explore the Seurat object

```{r, message=FALSE, warning=FALSE}
# number of features and samples (spots) in the active assay
dim(spatial_brain_data)

nrow(spatial_brain_data)           # number of features
head(rownames(spatial_brain_data)) # features

ncol(spatial_brain_data)           # number of spots
head(colnames(spatial_brain_data)) # spots

# specific assays
names(spatial_brain_data)
# pulling specific Assay
spatial_brain_data[['Spatial']]

SeuratObject::GetAssayData(spatial_brain_data)[1:10, 7:10]
```
```{r, message=FALSE, warning=FALSE}
# sample-level metadata data frame
class(spatial_brain_data[[]])
colnames(spatial_brain_data[[]])

head(spatial_brain_data@meta.data)
head(spatial_brain_data[[c('nCount_Spatial', 'nFeature_Spatial')]])

spatial_brain_data@assays
spatial_brain_data$Spatial@counts[1:10, 1:3]
spatial_brain_data$Spatial@meta.features[1:10,]

spatial_brain_data@images
```
```{r, message=FALSE, warning=FALSE}
str(spatial_brain_data)
```

## Data Preprocessing 

### 1. Quality Control (QC)

QC metrics

* `nFeature_Spatial` - remove low numbered unique genes detected in each sample.

* `nCount_Spatial` - remove high numbered total molecules detected in each sample. 

* `percent_mt` - filter out low-quality samples having high percent of mitochondrial mapped/contaminated read counts


```{r}
# compute the percentage of transcripts that map to mitochondrial genes
spatial_brain_data[["percent.mt"]] <- PercentageFeatureSet(spatial_brain_data, pattern = "^mt-")
head(spatial_brain_data[[]])

# visualize QC metrics as a violin plot
VlnPlot(
  spatial_brain_data, features = c('nFeature_Spatial','nCount_Spatial', 'percent.mt'), 
  pt.size = 0.1, ncol = 3) &
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())

# Jointly (rather than separately) consider the QC metrics when filtering
plot1 <- FeatureScatter(
  spatial_brain_data, feature1 = "nCount_Spatial", feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(
  spatial_brain_data, feature1 = "nCount_Spatial", feature2 = "nFeature_Spatial") +
  NoLegend()
plot1 + plot2
```
```{r, message=FALSE, warning=FALSE}
# plot data on tissue sample
SpatialFeaturePlot(
  spatial_brain_data, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt")) &
  theme(legend.position = "bottom", legend.text = element_text(size=6)) 
```
Filter data 

```{r}
spatial_brain_subset <- subset(
  spatial_brain_data, 
  subset = nFeature_Spatial < 8000 & nFeature_Spatial > 1000 & 
    nCount_Spatial < 50000 & percent.mt < 30)

print(paste("Filter out", ncol(spatial_brain_data) - ncol(spatial_brain_subset), 
            "samples because of the outlier QC metrics, with", ncol(spatial_brain_subset),
            "samples left."))
```
```{r, message=FALSE, warning=FALSE}
# plot filtered data on tissue sample
SpatialFeaturePlot(
  spatial_brain_subset, features = c("nFeature_Spatial", "nCount_Spatial", "percent.mt")) &
  theme(legend.position = "bottom", legend.text = element_text(size=6)) 
```
### 2. Normalization using SCTransform

* account for technical bias while preserving true biological differences

* uses regularized negative binomial regression to normalize UMI count data

```{r, message=FALSE, warning=FALSE}
library(printr)
help(SCTransform) 
```

```{r}
spatial_brain_norm <- SCTransform(spatial_brain_subset, assay = "Spatial", 
                                  return.only.var.genes = TRUE, variable.features.n = 3000, 
                                  verbose = TRUE)

names(spatial_brain_norm)
```
```{r}
spatial_brain_norm@assays$SCT
# `counts` slot of the SCT assay stores the corrected UMI counts
dim(spatial_brain_norm@assays$SCT@counts)
spatial_brain_norm@assays$SCT@counts[1:5, 1:5]
# `data` slot stores the log-normalized version of the corrected UMI counts
dim(spatial_brain_norm@assays$SCT@data)
spatial_brain_norm@assays$SCT@data[1:5, 1:5]
# 'scale.data` slot stores the normalized and scaled values (pearson residuals) of the 3000 variable genes
# used in downstream PCA analysis
dim(spatial_brain_norm@assays$SCT@scale.data)
spatial_brain_norm@assays$SCT@scale.data[1:5, 1:5]
#spatial_brain_norm@assays$SCT@var.features
```
## Analysis (PCA, cluster annotation, UMAP, DE)

```{r}
# PCA dimensionality reduction
spatial_brain_sobj <- RunPCA(spatial_brain_norm, assay = "SCT", npcs = 50)
ElbowPlot(spatial_brain_sobj, ndim = 50, reduction = "pca")

# clustering
spatial_brain_sobj <- FindNeighbors(spatial_brain_sobj, reduction = "pca", dims = 1:30)
spatial_brain_sobj <- FindClusters(spatial_brain_sobj)


# UMAP Uniform Manifold Approximation and Projection (UMAP) nonlinear dimensional reduction
spatial_brain_sobj <- RunUMAP(spatial_brain_sobj,
                              reduction = "pca", dims = 1:30)

plot3 <- DimPlot(spatial_brain_sobj, reduction = "umap", label = TRUE) + NoLegend()
plot4 <- SpatialDimPlot(spatial_brain_sobj, label = TRUE, label.size = 3) + NoLegend()
plot3 + plot4
```

```{r}
# plot individual clusters
SpatialDimPlot(spatial_brain_sobj, 
               cells.highlight = CellsByIdentities(spatial_brain_sobj), 
               facet.highlight = TRUE,
               ncol = 5)
```
```{r}
spatial_brain_sobj@reductions
```
```{r}
# identity class of each sample
table(spatial_brain_sobj@active.ident)
```

```{r, message=FALSE, warning=FALSE}
# find all markers of cluster 1
cluster1_markers <- FindMarkers(spatial_brain_sobj, ident.1 = 1, min.pct = 0.25)
head(cluster1_markers, n = 5)
```

```{r}
VlnPlot(spatial_brain_sobj, features = c("Fth1", "Plp1", "Mbp"))
```
```{r, warning=FALSE}
SpatialFeaturePlot(object = spatial_brain_sobj, 
                   features = c("Fth1", "Plp1", "Mbp"), 
                   alpha = c(0.1, 1), ncol = 3)
```
```{r, message=FALSE, warning=FALSE}
# find all markers of cluster 11
cluster11_markers <- FindMarkers(spatial_brain_sobj, ident.1 = 11, min.pct = 0.25)
head(cluster11_markers, n = 10)
```

```{r}
VlnPlot(spatial_brain_sobj, features = c("Gm13944", "Hes3", "Nrk"))
```

```{r}
SpatialFeaturePlot(object = spatial_brain_sobj, 
                   features = rownames(cluster11_markers)[1:3], 
                   alpha = c(0.1, 1), ncol = 3)
```
```{r}
# find all markers of cluster 10
cluster10_markers <- FindMarkers(spatial_brain_sobj, ident.1 = 10, min.pct = 0.25)
head(cluster10_markers, n = 5)

VlnPlot(spatial_brain_sobj, features = c("Esr1", "Trh", "Ccn3"))

SpatialFeaturePlot(object = spatial_brain_sobj, 
                   features = rownames(cluster10_markers)[1:3], 
                   alpha = c(0.1, 1), ncol = 3)
```
## Identify spatially variable genes

### Identify molecular features that correlate with spatial location within tissue by differential expression based on spatially distinct clusters
```{r}
# differential expression between clusters
de_markers <- FindMarkers(spatial_brain_sobj, ident.1 = 0, ident.2 = 1, min.pct = 0.25)
head(de_markers, n = 5)

SpatialFeaturePlot(object = spatial_brain_sobj, 
                   features = rownames(de_markers)[1:3], 
                   alpha = c(0.1, 1), ncol = 3)
```

```{r}
# differential expression between clusters
de12_markers <- FindMarkers(spatial_brain_sobj, ident.1 = 1, ident.2 = 2, min.pct = 0.5)
head(de12_markers, n = 5)

SpatialFeaturePlot(object = spatial_brain_sobj, 
                   features = rownames(de12_markers)[1:3], 
                   alpha = c(0.1, 1), ncol = 3)
```
## Identify spatially variable genes

### Moran's I statistic - a spatial autocorrelation metric measuring the correlation of gene expression values between local observed values and the average of neighboring values

```{r}
brain_moransi <- FindSpatiallyVariableFeatures(
  spatial_brain_sobj, assay = "SCT", 
  features = VariableFeatures(spatial_brain_sobj)[1:10],
    selection.method = "moransi") 
```

```{r}
moransi_output_df <- brain_moransi@assays$SCT@meta.features %>%
  na.exclude
head(moransi_output_df[order(moransi_output_df$MoransI_observed, decreasing = T), ])
```
```{r}
svf <- SpatiallyVariableFeatures(brain_moransi, selection.method = "moransi")
```

```{r}
SpatiallyVariableFeatures_workaround <- function(object, assay="SCT", selection.method = "moransi") {
  # Work around function to replace SeuratObject::SpatiallyVariableFeatures function.
  # return ranked list of Spatially Variable Features
  
  # Check if object is a Seurat object
  if (!inherits(object, "Seurat")) {
    stop("object must be a Seurat object")
  }

  # Check if assay is a valid assay
  if (!assay %in% names(object@assays)) {
    stop("assay must be a valid assay")
  }
  
  # Extract meta.features from the specified object and assay
  data <- object@assays[[assay]]@meta.features
  
  # Select columns starting with the provided col_prefix
  sel_cols <- grep(paste0("^", selection.method), colnames(data), value = TRUE)

  # Filter rows where "moransi.spatially.variable" is TRUE
  filtered_data <- data[data[[paste0(selection.method, ".spatially.variable")]], sel_cols]

  # Sort filtered data by "moransi.spatially.variable.rank" column in ascending order
  sorted_data <- filtered_data[order(filtered_data[[paste0(selection.method, ".spatially.variable.rank")]]), ]

  # Return row names of the sorted data frame
  rownames(sorted_data)
}
```

```{r, warning=FALSE}
top_features_moransi <- head(SpatiallyVariableFeatures_workaround(brain_moransi, selection.method = "moransi"), 3)

SpatialFeaturePlot(brain_moransi, 
                   features = top_features_moransi, ncol = 3, alpha = c(0.1, 1)) +
  plot_annotation(
    title = "Top 3 genes with the largest Moran's I",
    subtitle = "among 10 top variable genes")
```

## Identify spatially variable genes

### marker-variogram - a spatial autocorrelation metric measuring the variability between data points as a function of distance

```{r, warning=FALSE}
brain_variogram <- FindSpatiallyVariableFeatures(
  spatial_brain_sobj, assay = "SCT", 
  features = VariableFeatures(spatial_brain_sobj)[1:10],
    selection.method = "markvariogram")  

variogram_output_df <- brain_variogram@assays$SCT@meta.features %>%
  na.exclude # there are NA rows b/c we only calculated the variogram for 10 genes

head(variogram_output_df[order(variogram_output_df$r.metric.5), ])
```

```{r, warning=FALSE}
top_features_variogram <- head(
  SpatiallyVariableFeatures_workaround(brain_variogram, 
                            selection.method = "markvariogram"), 3)
SpatialFeaturePlot(brain_variogram, 
                   features = top_features_variogram, ncol = 3, alpha = c(0.1, 1)) + 
  plot_annotation(
  title = "3 genes with the top spatially variable rank (by mark-variogram)",
  subtitle = "among 10 top variable genes")
```

```{r}
sessionInfo()
```





