<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Integrative analysis between scRNA-seq and scATAC-seq multiome datasets</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="scRNA-seq_scATAC-seq_integration_files/libs/clipboard/clipboard.min.js"></script>
<script src="scRNA-seq_scATAC-seq_integration_files/libs/quarto-html/quarto.js"></script>
<script src="scRNA-seq_scATAC-seq_integration_files/libs/quarto-html/popper.min.js"></script>
<script src="scRNA-seq_scATAC-seq_integration_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="scRNA-seq_scATAC-seq_integration_files/libs/quarto-html/anchor.min.js"></script>
<link href="scRNA-seq_scATAC-seq_integration_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="scRNA-seq_scATAC-seq_integration_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="scRNA-seq_scATAC-seq_integration_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="scRNA-seq_scATAC-seq_integration_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="scRNA-seq_scATAC-seq_integration_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Integrative analysis between scRNA-seq and scATAC-seq multiome datasets</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="multiome-data" class="level2">
<h2 class="anchored" data-anchor-id="multiome-data">Multiome Data</h2>
<p>Publicly available ~12,000 human PBMC ‘multiome’ dataset from 10x Genomics: scRNA-seq and scATAC-seq profiles simultaneously collected in the same cells - ground truth to assess accuracy of integration. <a href="https://www.cell.com/cell/fulltext/S0092-8674(19)30559-8">Integrate the scRNA-seq and scATAC-seq data</a> - here, as if it originates from two different experiments.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Installed datasets ──────────────────────────────── SeuratData v0.2.2.9001 ──</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>✔ pbmcMultiome 0.1.4                    </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>────────────────────────────────────── Key ─────────────────────────────────────</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>✔ Dataset loaded successfully
❯ Dataset built with a newer version of Seurat than installed
❓ Unknown version of Seurat installed</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># install the dataset </span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">InstallData</span>(<span class="st">'pbmcMultiome'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The following packages are already installed and will not be
reinstalled: pbmcMultiome</code></pre>
</div>
</div>
<section id="load-libraries" class="level3">
<h3 class="anchored" data-anchor-id="load-libraries">Load Libraries</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat) <span class="co"># QC, analysis, ad exploration of scRNA-seq data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: SeuratObject</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: sp</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'SeuratObject'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:base':

    intersect</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Signac) <span class="co"># Analysis of single-cell chromatin datasets, including scATAC-seq, and multimodal datasets</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(EnsDb.Hsapiens.v86)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: ensembldb</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BiocGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:SeuratObject':

    intersect</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,
    table, tapply, union, unique, unsplit, which.max, which.min</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomicRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: stats4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:utils':

    findMatches</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: IRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'IRanges'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:sp':

    %over%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:grDevices':

    windows</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomeInfoDb</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomicFeatures</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: AnnotationDbi</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Biobase</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: AnnotationFilter</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'ensembldb'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:stats':

    filter</code></pre>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-individual-data-modalities" class="level3">
<h3 class="anchored" data-anchor-id="load-individual-data-modalities">Load Individual Data Modalities</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load scRNA-seq data</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>pbmc.rna <span class="ot">&lt;-</span> <span class="fu">LoadData</span>(<span class="st">"pbmcMultiome"</span>, <span class="st">"pbmc.rna"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Validating object structure</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Updating object slots</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Ensuring keys are in the proper structure
Ensuring keys are in the proper structure</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Ensuring feature names don't have underscores or pipes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Updating slots in RNA</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Validating object structure for Assay 'RNA'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Object representation is consistent with the most current Seurat version</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Assay RNA changing from Assay to Assay5</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>pbmc.rna[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">as</span>(pbmc.rna[[<span class="st">"RNA"</span>]], <span class="at">Class =</span> <span class="st">"Assay5"</span>)</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="co"># load scATAC-seq data</span></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>pbmc.atac <span class="ot">&lt;-</span> <span class="fu">LoadData</span>(<span class="st">"pbmcMultiome"</span>, <span class="st">"pbmc.atac"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Validating object structure</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Updating object slots</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Ensuring keys are in the proper structure
Ensuring keys are in the proper structure</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Ensuring feature names don't have underscores or pipes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Updating slots in ATAC</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Validating object structure for ChromatinAssay 'ATAC'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Object representation is consistent with the most current Seurat version</code></pre>
</div>
</div>
</section>
<section id="qc" class="level3">
<h3 class="anchored" data-anchor-id="qc">QC</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>pbmc.rna <span class="ot">&lt;-</span> <span class="fu">subset</span>(pbmc.rna, seurat_annotations <span class="sc">!=</span> <span class="st">"filtered"</span>)</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>pbmc.atac <span class="ot">&lt;-</span> <span class="fu">subset</span>(pbmc.atac, seurat_annotations <span class="sc">!=</span> <span class="st">"filtered"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>pbmc.rna</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
36601 features across 10412 samples within 1 assay 
Active assay: RNA (36601 features, 0 variable features)
 2 layers present: counts, data</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>pbmc.atac </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
108377 features across 10412 samples within 1 assay 
Active assay: ATAC (108377 features, 0 variable features)
 2 layers present: counts, data</code></pre>
</div>
</div>
</section>
<section id="individual-modality-analysis" class="level3">
<h3 class="anchored" data-anchor-id="individual-modality-analysis">Individual Modality Analysis</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>pbmc.rna <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(pbmc.rna)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts</code></pre>
</div>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>pbmc.rna <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(pbmc.rna)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding variable features for layer counts</code></pre>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>pbmc.rna <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(pbmc.rna)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering and scaling data matrix</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>pbmc.rna <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(pbmc.rna)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>PC_ 1 
Positive:  PLXDC2, SLC8A1, LRMDA, FCN1, TYMP, MCTP1, JAK2, RBM47, IRAK3, NAMPT 
       DMXL2, TBXAS1, ZEB2, LYN, LRRK2, SAT1, GAB2, CYBB, TNFAIP2, CSF3R 
       HCK, TLR2, CLEC7A, LYST, VCAN, DENND1A, FGD4, CD36, GRK3, FAM49A 
Negative:  CD247, IL32, CAMK4, IL7R, LTB, LEF1, BACH2, INPP4B, BCL2, SYNE2 
       THEMIS, TRBC2, RORA, TXK, ANK3, CD69, TRBC1, MLLT3, APBA2, ACTG1 
       RASGRF2, NELL2, LINC01934, TAFA1, SAMD3, PCAT1, PFN1, NCALD, CTSW, KCNQ5 
PC_ 2 
Positive:  CD247, IL32, DPYD, IL7R, INPP4B, AOAH, CAMK4, PDE3B, THEMIS, TXK 
       RORA, LEF1, SLCO3A1, NEAT1, FNDC3B, ARHGAP26, ANXA1, SRGN, ADGRE5, SAMD3 
       SYNE2, NCALD, TRBC1, CTSW, PLCB1, APBA2, CCL5, S100A4, TGFBR3, PRF1 
Negative:  BANK1, PAX5, MS4A1, FCRL1, NIBAN3, AFF3, IGHM, EBF1, LINC00926, OSBPL10 
       RALGPS2, CD79A, CD22, COBLL1, BLK, BLNK, AP002075.1, ADAM28, IGHD, COL19A1 
       BCL11A, CD79B, PLEKHG1, GNG7, DENND5B, WDFY4, TCF4, AC120193.1, RUBCNL, SPIB 
PC_ 3 
Positive:  BACH2, LEF1, PDE3B, CAMK4, IL7R, LTB, ANK3, FHIT, INPP4B, NELL2 
       BCL2, RASGRF2, CSGALNACT1, SLC2A3, MLLT3, AC139720.1, PLCL1, PRKN, PAX5, FCRL1 
       LINC00926, TSHZ2, MS4A1, PTPRK, AL589693.1, CD79A, EBF1, ST6GALNAC3, BANK1, COL19A1 
Negative:  GZMB, NKG7, GNLY, PRF1, KLRD1, GZMA, CST7, CLIC3, CCL5, FGFBP2 
       SPON2, ADGRG1, KLRF1, GZMH, CCL4, TGFBR3, LINC00996, LINC01478, PDGFD, PTGDR 
       C1orf21, FCRL6, CLEC4C, SLAMF7, BNC2, CTSW, IL2RB, EPHB1, LILRA4, HOPX 
PC_ 4 
Positive:  NKG7, GNLY, PRF1, KLRD1, GZMA, CST7, CCL5, FGFBP2, MCTP2, ADGRG1 
       GZMH, CCL4, KLRF1, FCGR3A, SPON2, TGFBR3, PDGFD, PAX5, FCRL1, MS4A1 
       FCRL6, PTGDR, BNC2, MTSS1, CX3CR1, LINC00926, C1orf21, IL2RB, PPP2R2B, HOPX 
Negative:  LINC01478, CUX2, EPHB1, CLEC4C, LILRA4, COL26A1, AC023590.1, PTPRS, SCN9A, LINC01374 
       COL24A1, LINC00996, NRP1, PHEX, FAM160A1, TNFRSF21, PLXNA4, PACSIN1, SCAMP5, SLC35F3 
       DNASE1L3, P3H2, LRRC26, PLD4, ITM2C, SCN1A-AS1, PLEKHD1, SERPINF1, SCT, PTCRA 
PC_ 5 
Positive:  CDKN1C, IFITM3, HES4, CSF1R, SMIM25, FCGR3A, MS4A7, LST1, AIF1, LRRC25 
       AC104809.2, MAFB, CALHM6, CFD, AC020651.2, SERPINA1, GPBAR1, SIGLEC10, CST3, HMOX1 
       CTSL, TCF7L2, FMNL2, HLA-DPA1, CKB, MTSS1, COTL1, IFI30, CCDC26, SPRED1 
Negative:  PLCB1, VCAN, ARHGAP24, CSF3R, DYSF, LINC00937, FNDC3B, GLT1D1, VCAN-AS1, AC020916.1 
       PDE4D, PADI4, ARHGAP26, CREB5, TREM1, MCTP2, CD36, MEGF9, KIF13A, MICAL2 
       FCAR, ACSL1, CLMN, JUN, ANXA1, MIR646HG, RAB11FIP1, CRISPLD2, GNLY, ZBTB16 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>pbmc.rna <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(pbmc.rna, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
This message will be shown once per session</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>19:28:11 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>19:28:11 Read 10412 rows and found 30 numeric columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>19:28:11 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>19:28:11 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[----|----|----|----|----|----|----|----|----|----|</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**************************************************|
19:28:13 Writing NN index file to temp file C:\Users\lac_2019\AppData\Local\Temp\RtmpkT1Nyl\file766c56d87575
19:28:13 Searching Annoy index using 1 thread, search_k = 3000
19:28:16 Annoy recall = 100%
19:28:16 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30
19:28:18 Initializing from normalized Laplacian + noise (using RSpectra)
19:28:19 Commencing optimization for 200 epochs, with 450622 positive edges
19:28:29 Optimization finished</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>annotations <span class="ot">&lt;-</span> <span class="fu">GetGRangesFromEnsDb</span>(<span class="at">ensdb =</span> EnsDb.Hsapiens.v86)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)

Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="fu">seqlevelsStyle</span>(annotations) <span class="ot">&lt;-</span> <span class="st">"UCSC"</span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a><span class="fu">genome</span>(annotations) <span class="ot">&lt;-</span> <span class="st">"hg38"</span></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a><span class="fu">Annotation</span>(pbmc.atac) <span class="ot">&lt;-</span> annotations</span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>pbmc.atac <span class="ot">&lt;-</span> <span class="fu">RunTFIDF</span>(pbmc.atac)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Performing TF-IDF normalization</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in RunTFIDF.default(object = GetAssayData(object = object, slot =
"counts"), : Some features contain 0 total counts</code></pre>
</div>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>pbmc.atac <span class="ot">&lt;-</span> <span class="fu">FindTopFeatures</span>(pbmc.atac, <span class="at">min.cutoff =</span> <span class="st">'q0'</span>)</span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>pbmc.atac <span class="ot">&lt;-</span> <span class="fu">RunSVD</span>(pbmc.atac)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Running SVD</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Scaling cell embeddings</code></pre>
</div>
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>pbmc.atac <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(pbmc.atac, <span class="at">reduction =</span> <span class="st">'lsi'</span>, <span class="at">dims =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">30</span>, </span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">reduction.name =</span> <span class="st">"umap.atac"</span>, <span class="at">reduction.key =</span> <span class="st">"atacUMAP_"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>19:31:45 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>19:31:45 Read 10412 rows and found 29 numeric columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>19:31:45 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>19:31:45 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[----|----|----|----|----|----|----|----|----|----|</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**************************************************|
19:31:46 Writing NN index file to temp file C:\Users\lac_2019\AppData\Local\Temp\RtmpkT1Nyl\file766c3638610d
19:31:47 Searching Annoy index using 1 thread, search_k = 3000
19:31:49 Annoy recall = 100%
19:31:50 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30
19:31:52 Initializing from normalized Laplacian + noise (using RSpectra)
19:31:52 Commencing optimization for 200 epochs, with 414208 positive edges
19:32:03 Optimization finished</code></pre>
</div>
</div>
<p>plot the results from both modalities. Cells have been previously annotated based on transcriptomic state. We will predict annotations for the scATAC-seq cells.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(pbmc.rna, <span class="at">group.by =</span> <span class="st">'seurat_annotations'</span>, <span class="at">label =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> </span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"RNA"</span>)</span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(pbmc.atac, <span class="at">group.by =</span> <span class="st">'orig.ident'</span>, <span class="at">label =</span> <span class="cn">FALSE</span>) <span class="sc">+</span> </span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"ATAC"</span>)</span>
<span id="cb102-5"><a href="#cb102-5" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="scRNA-seq_scATAC-seq_integration_files/figure-html/viz1-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>plot <span class="ot">&lt;-</span> (p1 <span class="sc">+</span> p2) <span class="sc">&amp;</span></span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"UMAP 1"</span>) <span class="sc">&amp;</span> <span class="fu">ylab</span>(<span class="st">"UMAP 2"</span>) <span class="sc">&amp;</span> </span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>))</span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="at">filename =</span> <span class="st">"./output/images/rnaseq_atacseq_integration.jpg"</span>, </span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">height =</span> <span class="dv">7</span>, <span class="at">width =</span> <span class="dv">12</span>, <span class="at">plot =</span> plot, <span class="at">quality =</span> <span class="dv">50</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="integrative-analysis-between-scrna-seq-and-scatac-seq-datasets" class="level2">
<h2 class="anchored" data-anchor-id="integrative-analysis-between-scrna-seq-and-scatac-seq-datasets">Integrative analysis between scRNA-seq and scATAC-seq datasets</h2>
<p>Integrative analysis between scRNA-seq and scATAC-seq datasets focuses primarily on label transfer across data modalities. Optionally, it includes co-embedding of scRNA-seq and scATAC-seq datasets and visualization of co-embedded single cells on the same plot.</p>
<section id="identify-anchors-between-scrna-seq-and-scatac-seq-datasets" class="level3">
<h3 class="anchored" data-anchor-id="identify-anchors-between-scrna-seq-and-scatac-seq-datasets">Identify anchors between scRNA-seq and scATAC-seq datasets</h3>
<p>To identify ‘anchors’ between scRNA-seq and scATAC-seq experiments, we first generate a rough estimate of the transcriptional activity of each gene by quantifying ATAC-seq counts in the 2 kb-upstream region and gene body, using the <code>GeneActivity()</code> function in the Signac package. The ensuing gene activity scores from the scATAC-seq data are then used as input for canonical correlation analysis, along with the gene expression quantifications from scRNA-seq. This quantification is performed for all genes identified as being highly variable from the scRNA-seq dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="co"># quantify gene activity</span></span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>gene.activities <span class="ot">&lt;-</span> <span class="fu">GeneActivity</span>(pbmc.atac, <span class="at">features =</span> <span class="fu">VariableFeatures</span>(pbmc.rna))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Extracting gene coordinates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Extracting reads overlapping genomic regions</code></pre>
</div>
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add gene activities as a new assay</span></span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>pbmc.atac[[<span class="st">"ACTIVITY"</span>]] <span class="ot">&lt;-</span> <span class="fu">CreateAssayObject</span>(<span class="at">counts =</span> gene.activities)</span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a><span class="co"># normalize gene activities</span></span>
<span id="cb107-5"><a href="#cb107-5" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(pbmc.atac) <span class="ot">&lt;-</span> <span class="st">"ACTIVITY"</span></span>
<span id="cb107-6"><a href="#cb107-6" aria-hidden="true" tabindex="-1"></a>pbmc.atac <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(pbmc.atac)</span>
<span id="cb107-7"><a href="#cb107-7" aria-hidden="true" tabindex="-1"></a>pbmc.atac <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(pbmc.atac, <span class="at">features =</span> <span class="fu">rownames</span>(pbmc.atac))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering and scaling data matrix</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>pbmc.rna</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
36601 features across 10412 samples within 1 assay 
Active assay: RNA (36601 features, 2000 variable features)
 3 layers present: counts, data, scale.data
 2 dimensional reductions calculated: pca, umap</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>pbmc.atac</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
109663 features across 10412 samples within 2 assays 
Active assay: ACTIVITY (1286 features, 0 variable features)
 3 layers present: counts, data, scale.data
 1 other assay present: ATAC
 2 dimensional reductions calculated: lsi, umap.atac</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify anchors </span></span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a>transfer.anchors <span class="ot">&lt;-</span> <span class="fu">FindTransferAnchors</span>(</span>
<span id="cb113-3"><a href="#cb113-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">reference =</span> pbmc.rna,</span>
<span id="cb113-4"><a href="#cb113-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">query =</span> pbmc.atac,</span>
<span id="cb113-5"><a href="#cb113-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">features =</span> <span class="fu">VariableFeatures</span>(<span class="at">object =</span> pbmc.rna),</span>
<span id="cb113-6"><a href="#cb113-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">reference.assay =</span> <span class="st">'RNA'</span>,</span>
<span id="cb113-7"><a href="#cb113-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">query.assay =</span> <span class="st">'ACTIVITY'</span>,</span>
<span id="cb113-8"><a href="#cb113-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">reduction =</span> <span class="st">'cca'</span></span>
<span id="cb113-9"><a href="#cb113-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: 714 features of the features specified were not present in both the reference query assays. 
Continuing with remaining 1286 features.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Running CCA</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Merging objects</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding neighborhoods</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding anchors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>    Found 23956 anchors</code></pre>
</div>
</div>
</section>
<section id="annotate-scatac-seq-cells-via-label-transfer" class="level3">
<h3 class="anchored" data-anchor-id="annotate-scatac-seq-cells-via-label-transfer">Annotate scATAC-seq cells via label transfer</h3>
<p>After identifying anchors, annotations are transferred from the scRNA-seq dataset onto the scATAC-seq cells. The annotations are stored in the <code>seurat_annotations</code> field, and are provided as input to the <code>refdata</code> parameter. The output will contain a matrix with predictions and confidence scores for each ATAC-seq cell.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a>celltype.predictions <span class="ot">&lt;-</span> <span class="fu">TransferData</span>(</span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">anchorset =</span> transfer.anchors,</span>
<span id="cb120-3"><a href="#cb120-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">refdata =</span> pbmc.rna<span class="sc">$</span>seurat_annotations,</span>
<span id="cb120-4"><a href="#cb120-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">weight.reduction =</span> pbmc.atac[[<span class="st">'lsi'</span>]],</span>
<span id="cb120-5"><a href="#cb120-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">dims =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">30</span></span>
<span id="cb120-6"><a href="#cb120-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vector weights</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Predicting cell labels</code></pre>
</div>
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>pbmc.atac <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(pbmc.atac, <span class="at">metadata =</span> celltype.predictions)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>After performing transfer, the ATAC-seq cells have predicted annotations (transferred from the scRNA-seq dataset) stored in the <code>predicted.id</code> field. Since these cells were measured with a multiome kit, a ground-truth annotation is available to be used for evaluation. The plots below show that the predicted and actual annotations are extremely similar.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a>pbmc.atac<span class="sc">$</span>annotation_correct <span class="ot">&lt;-</span> pbmc.atac<span class="sc">$</span>predicted.id <span class="sc">==</span></span>
<span id="cb125-2"><a href="#cb125-2" aria-hidden="true" tabindex="-1"></a>  pbmc.atac<span class="sc">$</span>seurat_annotations</span>
<span id="cb125-3"><a href="#cb125-3" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(pbmc.atac, <span class="at">group.by =</span> <span class="st">'predicted.id'</span>, <span class="at">label =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb125-4"><a href="#cb125-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Predicted annotation"</span>)</span>
<span id="cb125-5"><a href="#cb125-5" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(pbmc.atac, <span class="at">group.by =</span> <span class="st">'seurat_annotations'</span>, <span class="at">label =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb125-6"><a href="#cb125-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"Ground-truth annotation"</span>)</span>
<span id="cb125-7"><a href="#cb125-7" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">|</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="scRNA-seq_scATAC-seq_integration_files/figure-html/viz.label.accuracy-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The annotation for an scATAC-seq profile is correctly predicted via scRNA-seq integration ~90% of the time.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(pbmc.atac<span class="sc">$</span>annotation_correct)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
FALSE  TRUE 
 1105  9307 </code></pre>
</div>
</div>
<p>In addition, the `prediction.score.max` field quantifies the uncertainty associated with the predicted annotations. Cells that are correctly annotated are typically associated with high prediction scores (&gt;90%), while cells that are incorrectly annotated are associated with sharply lower prediction scores (&lt;50%). Incorrect assignments also tend to reflect closely related cell types (i.e.&nbsp;Intermediate vs.&nbsp;Naive B cells).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>predictions <span class="ot">&lt;-</span> <span class="fu">table</span>(pbmc.atac<span class="sc">$</span>seurat_annotations, pbmc.atac<span class="sc">$</span>predicted.id)</span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a>predictions <span class="ot">&lt;-</span> predictions <span class="sc">/</span> <span class="fu">rowSums</span>(predictions) <span class="co"># normalize for number of cells in each cell type</span></span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a>predictions <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(predictions)</span>
<span id="cb128-4"><a href="#cb128-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(predictions)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Var1      Var2        Freq
1 CD14 Mono CD14 Mono 0.938833570
2 CD16 Mono CD14 Mono 0.293774319
3 CD4 Naive CD14 Mono 0.001409443
4   CD4 TCM CD14 Mono 0.000870322
5   CD4 TEM CD14 Mono 0.000000000
6 CD8 Naive CD14 Mono 0.001418440</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(predictions, <span class="fu">aes</span>(Var1, Var2, <span class="at">fill =</span> Freq)) <span class="sc">+</span></span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tile</span>() <span class="sc">+</span></span>
<span id="cb130-3"><a href="#cb130-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_gradient</span>(<span class="at">name =</span> <span class="st">"Fraction of cells"</span>, <span class="at">low =</span> <span class="st">"#ffffc8"</span>, <span class="at">high =</span> <span class="st">"#7d0025"</span>) <span class="sc">+</span></span>
<span id="cb130-4"><a href="#cb130-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Cell type annotation (RNA)"</span>) <span class="sc">+</span></span>
<span id="cb130-5"><a href="#cb130-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Predicted cell type label (ATAC)"</span>) <span class="sc">+</span></span>
<span id="cb130-6"><a href="#cb130-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_cowplot</span>() <span class="sc">+</span></span>
<span id="cb130-7"><a href="#cb130-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust=</span><span class="dv">1</span>)) </span>
<span id="cb130-8"><a href="#cb130-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-9"><a href="#cb130-9" aria-hidden="true" tabindex="-1"></a>correct <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">which</span>(pbmc.atac<span class="sc">$</span>seurat_annotations <span class="sc">==</span> pbmc.atac<span class="sc">$</span>predicted.id))</span>
<span id="cb130-10"><a href="#cb130-10" aria-hidden="true" tabindex="-1"></a>incorrect <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">which</span>(pbmc.atac<span class="sc">$</span>seurat_annotations <span class="sc">!=</span> pbmc.atac<span class="sc">$</span>predicted.id))</span>
<span id="cb130-11"><a href="#cb130-11" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(pbmc.atac, <span class="at">vars =</span> <span class="fu">c</span>(<span class="st">"prediction.score.max"</span>, <span class="st">"annotation_correct"</span>)) </span>
<span id="cb130-12"><a href="#cb130-12" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(data, <span class="fu">aes</span>(prediction.score.max, <span class="at">fill =</span> annotation_correct, <span class="at">colour =</span> annotation_correct)) <span class="sc">+</span> </span>
<span id="cb130-13"><a href="#cb130-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span> <span class="fu">theme_cowplot</span>() <span class="sc">+</span> </span>
<span id="cb130-14"><a href="#cb130-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_discrete</span>(<span class="at">name =</span> <span class="st">"Annotation Correct"</span>, <span class="at">labels =</span> <span class="fu">c</span>(<span class="fu">paste0</span>(<span class="st">"FALSE (n = "</span>, incorrect, <span class="st">")"</span>), <span class="fu">paste0</span>(<span class="st">"TRUE (n = "</span>, correct, <span class="st">")"</span>))) <span class="sc">+</span> </span>
<span id="cb130-15"><a href="#cb130-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_discrete</span>(<span class="at">name =</span> <span class="st">"Annotation Correct"</span>, <span class="at">labels =</span> <span class="fu">c</span>(<span class="fu">paste0</span>(<span class="st">"FALSE (n = "</span>, incorrect, <span class="st">")"</span>), <span class="fu">paste0</span>(<span class="st">"TRUE (n = "</span>, correct, <span class="st">")"</span>))) <span class="sc">+</span></span>
<span id="cb130-16"><a href="#cb130-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Prediction Score"</span>)</span>
<span id="cb130-17"><a href="#cb130-17" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="scRNA-seq_scATAC-seq_integration_files/figure-html/score.viz-1.png" class="img-fluid" width="1152"></p>
</div>
</div>
</section>
<section id="co-embed-scrna-seq-and-scatac-seq-datasets" class="level3">
<h3 class="anchored" data-anchor-id="co-embed-scrna-seq-and-scatac-seq-datasets">Co-embed scRNA-seq and scATAC-seq datasets</h3>
<p>To perform co-embedding, RNA expression is first <code>imputed</code> into the scATAC-seq cells based on the previously computed anchors, and then the datasets are merged.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="co"># restrict the imputation to variable genes from scRNA-seq, but could impute the</span></span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a><span class="co"># full transcriptome if wanted</span></span>
<span id="cb131-3"><a href="#cb131-3" aria-hidden="true" tabindex="-1"></a>genes.use <span class="ot">&lt;-</span> <span class="fu">VariableFeatures</span>(pbmc.rna)</span>
<span id="cb131-4"><a href="#cb131-4" aria-hidden="true" tabindex="-1"></a>refdata <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(pbmc.rna, <span class="at">assay =</span> <span class="st">"RNA"</span>, <span class="at">slot =</span> <span class="st">"data"</span>)[genes.use, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The `slot` argument of `GetAssayData()` is deprecated as of SeuratObject 5.0.0.
ℹ Please use the `layer` argument instead.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="co"># refdata (input) contains a scRNA-seq expression matrix for the scRNA-seq cells  </span></span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a><span class="co"># imputation (output) contains an imputed scRNA-seq matrix for each of the ATAC cells</span></span>
<span id="cb133-3"><a href="#cb133-3" aria-hidden="true" tabindex="-1"></a>imputation <span class="ot">&lt;-</span> <span class="fu">TransferData</span>(<span class="at">anchorset =</span> transfer.anchors, </span>
<span id="cb133-4"><a href="#cb133-4" aria-hidden="true" tabindex="-1"></a>                           <span class="at">refdata =</span> refdata, </span>
<span id="cb133-5"><a href="#cb133-5" aria-hidden="true" tabindex="-1"></a>                           <span class="at">weight.reduction =</span> pbmc.atac[[<span class="st">"lsi"</span>]], </span>
<span id="cb133-6"><a href="#cb133-6" aria-hidden="true" tabindex="-1"></a>                           <span class="at">dims =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vector weights</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Transfering 2000 features onto reference data</code></pre>
</div>
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>pbmc.atac[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> imputation</span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb137-3"><a href="#cb137-3" aria-hidden="true" tabindex="-1"></a>coembed <span class="ot">&lt;-</span> <span class="fu">merge</span>(<span class="at">x =</span> pbmc.rna, <span class="at">y =</span> pbmc.atac)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Some cell names are duplicated across objects provided. Renaming to
enforce unique cell names.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run PCA and UMAP on the combined object </span></span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a>coembed <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(coembed, <span class="at">features =</span> genes.use, <span class="at">do.scale =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering data matrix</code></pre>
</div>
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a>coembed <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(coembed, <span class="at">features =</span> genes.use, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb141-2"><a href="#cb141-2" aria-hidden="true" tabindex="-1"></a>coembed <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(coembed, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>19:41:51 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>19:41:51 Read 20824 rows and found 30 numeric columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>19:41:51 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>19:41:51 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[----|----|----|----|----|----|----|----|----|----|</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**************************************************|
19:41:57 Writing NN index file to temp file C:\Users\lac_2019\AppData\Local\Temp\RtmpkT1Nyl\file766c37dd6d55
19:41:59 Searching Annoy index using 1 thread, search_k = 3000
19:42:11 Annoy recall = 100%
19:42:12 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30
19:42:15 Initializing from normalized Laplacian + noise (using RSpectra)
19:42:16 Commencing optimization for 200 epochs, with 990646 positive edges
19:42:37 Optimization finished</code></pre>
</div>
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="co"># visualize the co-embedding of both datasets</span></span>
<span id="cb153-2"><a href="#cb153-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(coembed, <span class="at">group.by =</span> <span class="fu">c</span>(<span class="st">"orig.ident"</span>,<span class="st">"seurat_annotations"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="scRNA-seq_scATAC-seq_integration_files/figure-html/coembed-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb154"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb154-1"><a href="#cb154-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(refdata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
  ..@ i       : int [1:1976785] 22 39 46 47 51 58 59 67 77 81 ...
  ..@ p       : int [1:10413] 0 209 347 521 666 747 919 1022 1319 1428 ...
  ..@ Dim     : int [1:2] 2000 10412
  ..@ Dimnames:List of 2
  .. ..$ : chr [1:2000] "JCHAIN" "PTGDS" "IGKC" "IGHA1" ...
  .. ..$ : chr [1:10412] "AAACAGCCAAGGAATC-1" "AAACAGCCAATCCCTT-1" "AAACAGCCAATGCGCT-1" "AAACAGCCACCAACCG-1" ...
  ..@ x       : num [1:1976785] 0.785 1.22 0.785 0.785 0.785 ...
  ..@ factors : list()</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb156"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb156-1"><a href="#cb156-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(imputation)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Formal class 'Assay' [package "SeuratObject"] with 8 slots
  ..@ counts       : num[0 , 0 ] 
  ..@ data         :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
  .. .. ..@ i       : int [1:9629093] 0 1 2 3 5 6 10 11 12 13 ...
  .. .. ..@ p       : int [1:10413] 0 799 1659 2573 3383 4222 5070 5938 7021 7887 ...
  .. .. ..@ Dim     : int [1:2] 2000 10412
  .. .. ..@ Dimnames:List of 2
  .. .. .. ..$ : chr [1:2000] "JCHAIN" "PTGDS" "IGKC" "IGHA1" ...
  .. .. .. ..$ : chr [1:10412] "AAACAGCCAAGGAATC-1" "AAACAGCCAATCCCTT-1" "AAACAGCCAATGCGCT-1" "AAACAGCCACCAACCG-1" ...
  .. .. ..@ x       : num [1:9629093] 0.0508 0.0704 0.2323 0.0636 0.097 ...
  .. .. ..@ factors : list()
  ..@ scale.data   : num[0 , 0 ] 
  ..@ assay.orig   : NULL
  ..@ var.features : logi(0) 
  ..@ meta.features:'data.frame':   2000 obs. of  0 variables
  ..@ misc         :List of 1
  .. ..$ calcN: logi TRUE
  ..@ key          : chr "id_"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb158"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb158-1"><a href="#cb158-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(coembed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Formal class 'Seurat' [package "SeuratObject"] with 13 slots
  ..@ assays      :List of 3
  .. ..$ RNA     :Formal class 'Assay5' [package "SeuratObject"] with 8 slots
  .. .. .. ..@ layers    :List of 5
  .. .. .. .. ..$ counts.RNA    :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
  .. .. .. .. .. .. ..@ i       : int [1:21013327] 17 48 53 59 86 109 113 137 168 170 ...
  .. .. .. .. .. .. ..@ p       : int [1:10413] 0 3308 5204 8108 10390 11743 14804 16495 19523 21306 ...
  .. .. .. .. .. .. ..@ Dim     : int [1:2] 36601 10412
  .. .. .. .. .. .. ..@ Dimnames:List of 2
  .. .. .. .. .. .. .. ..$ : NULL
  .. .. .. .. .. .. .. ..$ : NULL
  .. .. .. .. .. .. ..@ x       : num [1:21013327] 1 1 1 1 1 1 1 1 1 6 ...
  .. .. .. .. .. .. ..@ factors : list()
  .. .. .. .. ..$ data.RNA      :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
  .. .. .. .. .. .. ..@ i       : int [1:21013327] 17 48 53 59 86 109 113 137 168 170 ...
  .. .. .. .. .. .. ..@ p       : int [1:10413] 0 3308 5204 8108 10390 11743 14804 16495 19523 21306 ...
  .. .. .. .. .. .. ..@ Dim     : int [1:2] 36601 10412
  .. .. .. .. .. .. ..@ Dimnames:List of 2
  .. .. .. .. .. .. .. ..$ : NULL
  .. .. .. .. .. .. .. ..$ : NULL
  .. .. .. .. .. .. ..@ x       : num [1:21013327] 0.785 0.785 0.785 0.785 0.785 ...
  .. .. .. .. .. .. ..@ factors : list()
  .. .. .. .. ..$ scale.data.RNA: num [1:2000, 1:10412] -0.2207 -0.6728 -0.1666 -0.1699 -0.0562 ...
  .. .. .. .. ..$ data.ATAC     :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
  .. .. .. .. .. .. ..@ i       : int [1:9629093] 1 5 7 10 18 19 24 25 29 30 ...
  .. .. .. .. .. .. ..@ p       : int [1:10413] 0 799 1659 2573 3383 4222 5070 5938 7021 7887 ...
  .. .. .. .. .. .. ..@ Dim     : int [1:2] 2000 10412
  .. .. .. .. .. .. ..@ Dimnames:List of 2
  .. .. .. .. .. .. .. ..$ : NULL
  .. .. .. .. .. .. .. ..$ : NULL
  .. .. .. .. .. .. ..@ x       : num [1:9629093] 0.2066 0.0154 0.1295 0.1415 0.0312 ...
  .. .. .. .. .. .. ..@ factors : list()
  .. .. .. .. ..$ scale.data    : num [1:2000, 1:20824] -0.05678 -0.48389 -0.03771 -0.04061 -0.00521 ...
  .. .. .. ..@ cells     :Formal class 'LogMap' [package "SeuratObject"] with 1 slot
  .. .. .. .. .. ..@ .Data: logi [1:20824, 1:5] TRUE TRUE TRUE TRUE TRUE TRUE ...
  .. .. .. .. .. .. ..- attr(*, "dimnames")=List of 2
  .. .. .. .. .. .. .. ..$ : chr [1:20824] "AAACAGCCAAGGAATC-1_1" "AAACAGCCAATCCCTT-1_1" "AAACAGCCAATGCGCT-1_1" "AAACAGCCACCAACCG-1_1" ...
  .. .. .. .. .. .. .. ..$ : chr [1:5] "counts.RNA" "data.RNA" "scale.data.RNA" "data.ATAC" ...
  .. .. .. .. .. ..$ dim     : int [1:2] 20824 5
  .. .. .. .. .. ..$ dimnames:List of 2
  .. .. .. .. .. .. ..$ : chr [1:20824] "AAACAGCCAAGGAATC-1_1" "AAACAGCCAATCCCTT-1_1" "AAACAGCCAATGCGCT-1_1" "AAACAGCCACCAACCG-1_1" ...
  .. .. .. .. .. .. ..$ : chr [1:5] "counts.RNA" "data.RNA" "scale.data.RNA" "data.ATAC" ...
  .. .. .. ..@ features  :Formal class 'LogMap' [package "SeuratObject"] with 1 slot
  .. .. .. .. .. ..@ .Data: logi [1:36601, 1:5] TRUE TRUE TRUE TRUE TRUE TRUE ...
  .. .. .. .. .. .. ..- attr(*, "dimnames")=List of 2
  .. .. .. .. .. .. .. ..$ : chr [1:36601] "MIR1302-2HG" "FAM138A" "OR4F5" "AL627309.1" ...
  .. .. .. .. .. .. .. ..$ : chr [1:5] "counts.RNA" "data.RNA" "scale.data.RNA" "data.ATAC" ...
  .. .. .. .. .. ..$ dim     : int [1:2] 36601 5
  .. .. .. .. .. ..$ dimnames:List of 2
  .. .. .. .. .. .. ..$ : chr [1:36601] "MIR1302-2HG" "FAM138A" "OR4F5" "AL627309.1" ...
  .. .. .. .. .. .. ..$ : chr [1:5] "counts.RNA" "data.RNA" "scale.data.RNA" "data.ATAC" ...
  .. .. .. ..@ default   : int 1
  .. .. .. ..@ assay.orig: chr(0) 
  .. .. .. ..@ meta.data :'data.frame': 36601 obs. of  8 variables:
  .. .. .. .. ..$ vf_vst_counts.RNA_mean                 : num [1:36601] 0 0 0 0.00759 0 ...
  .. .. .. .. ..$ vf_vst_counts.RNA_variance             : num [1:36601] 0 0 0 0.00868 0 ...
  .. .. .. .. ..$ vf_vst_counts.RNA_variance.expected    : num [1:36601] 0 0 0 0.00909 0 ...
  .. .. .. .. ..$ vf_vst_counts.RNA_variance.standardized: num [1:36601] 0 0 0 0.955 0 ...
  .. .. .. .. ..$ vf_vst_counts.RNA_variable             : logi [1:36601] FALSE FALSE FALSE FALSE FALSE FALSE ...
  .. .. .. .. ..$ vf_vst_counts.RNA_rank                 : int [1:36601] NA NA NA NA NA NA NA NA NA NA ...
  .. .. .. .. ..$ var.features                           : chr [1:36601] NA NA NA NA ...
  .. .. .. .. ..$ var.features.rank                      : int [1:36601] NA NA NA NA NA NA NA NA NA NA ...
  .. .. .. ..@ misc      : list()
  .. .. .. ..@ key       : chr "rna_"
  .. ..$ ATAC    :Formal class 'ChromatinAssay' [package "Signac"] with 16 slots
  .. .. .. ..@ ranges            :Formal class 'GRanges' [package "GenomicRanges"] with 7 slots
  .. .. .. .. .. ..@ seqnames       :Formal class 'Rle' [package "S4Vectors"] with 4 slots
  .. .. .. .. .. .. .. ..@ values         : Factor w/ 33 levels "chr1","chr2",..: 1 10 11 12 13 14 15 16 17 18 ...
  .. .. .. .. .. .. .. ..@ lengths        : int [1:33] 10538 4967 5249 5670 2410 3734 3513 4033 5766 2013 ...
  .. .. .. .. .. .. .. ..@ elementMetadata: NULL
  .. .. .. .. .. .. .. ..@ metadata       : list()
  .. .. .. .. .. ..@ ranges         :Formal class 'IRanges' [package "IRanges"] with 6 slots
  .. .. .. .. .. .. .. ..@ start          : int [1:108377] 10109 180730 191491 267816 586028 629721 633793 777634 816881 819912 ...
  .. .. .. .. .. .. .. ..@ width          : int [1:108377] 249 901 246 381 346 452 472 2293 767 3589 ...
  .. .. .. .. .. .. .. ..@ NAMES          : NULL
  .. .. .. .. .. .. .. ..@ elementType    : chr "ANY"
  .. .. .. .. .. .. .. ..@ elementMetadata: NULL
  .. .. .. .. .. .. .. ..@ metadata       : list()
  .. .. .. .. .. ..@ strand         :Formal class 'Rle' [package "S4Vectors"] with 4 slots
  .. .. .. .. .. .. .. ..@ values         : Factor w/ 3 levels "+","-","*": 3
  .. .. .. .. .. .. .. ..@ lengths        : int 108377
  .. .. .. .. .. .. .. ..@ elementMetadata: NULL
  .. .. .. .. .. .. .. ..@ metadata       : list()
  .. .. .. .. .. ..@ seqinfo        :Formal class 'Seqinfo' [package "GenomeInfoDb"] with 4 slots
  .. .. .. .. .. .. .. ..@ seqnames   : chr [1:33] "chr1" "chr2" "chr3" "chr4" ...
  .. .. .. .. .. .. .. ..@ seqlengths : int [1:33] NA NA NA NA NA NA NA NA NA NA ...
  .. .. .. .. .. .. .. ..@ is_circular: logi [1:33] NA NA NA NA NA NA ...
  .. .. .. .. .. .. .. ..@ genome     : chr [1:33] NA NA NA NA ...
  .. .. .. .. .. ..@ elementMetadata:Formal class 'DFrame' [package "S4Vectors"] with 6 slots
  .. .. .. .. .. .. .. ..@ rownames       : NULL
  .. .. .. .. .. .. .. ..@ nrows          : int 108377
  .. .. .. .. .. .. .. ..@ elementType    : chr "ANY"
  .. .. .. .. .. .. .. ..@ elementMetadata: NULL
  .. .. .. .. .. .. .. ..@ metadata       : list()
  .. .. .. .. .. .. .. ..@ listData       : Named list()
  .. .. .. .. .. ..@ elementType    : chr "ANY"
  .. .. .. .. .. ..@ metadata       : list()
  .. .. .. ..@ motifs            : NULL
  .. .. .. ..@ fragments         :List of 1
  .. .. .. .. ..$ :Formal class 'Fragment' [package "Signac"] with 3 slots
  .. .. .. .. .. .. ..@ path : chr "https://cf.10xgenomics.com/samples/cell-arc/1.0.0/pbmc_granulocyte_sorted_10k/pbmc_granulocyte_sorted_10k_atac_"| __truncated__
  .. .. .. .. .. .. ..@ hash : chr [1:2] NA NA
  .. .. .. .. .. .. ..@ cells: Named chr [1:10412] "AAACAGCCAAGGAATC-1" "AAACAGCCAATCCCTT-1" "AAACAGCCAATGCGCT-1" "AAACAGCCACCAACCG-1" ...
  .. .. .. .. .. .. .. ..- attr(*, "names")= chr [1:10412] "AAACAGCCAAGGAATC-1_2" "AAACAGCCAATCCCTT-1_2" "AAACAGCCAATGCGCT-1_2" "AAACAGCCACCAACCG-1_2" ...
  .. .. .. ..@ seqinfo           : NULL
  .. .. .. ..@ annotation        :Formal class 'GRanges' [package "GenomicRanges"] with 7 slots
  .. .. .. .. .. ..@ seqnames       :Formal class 'Rle' [package "S4Vectors"] with 4 slots
  .. .. .. .. .. .. .. ..@ values         : Factor w/ 25 levels "chrX","chr20",..: 1 2 3 4 5 6 7 8 9 10 ...
  .. .. .. .. .. .. .. ..@ lengths        : int [1:25] 92459 65287 276555 135207 190802 141516 182677 182178 116305 194669 ...
  .. .. .. .. .. .. .. ..@ elementMetadata: NULL
  .. .. .. .. .. .. .. ..@ metadata       : list()
  .. .. .. .. .. ..@ ranges         :Formal class 'IRanges' [package "IRanges"] with 6 slots
  .. .. .. .. .. .. .. ..@ start          : int [1:3021151] 276322 276324 276353 281055 281192 281194 281462 281482 281482 281482 ...
  .. .. .. .. .. .. .. ..@ width          : int [1:3021151] 73 71 42 67 493 63 223 203 203 203 ...
  .. .. .. .. .. .. .. ..@ NAMES          : chr [1:3021151] "ENSE00001489430" "ENSE00001536003" "ENSE00002160563" "ENSE00001750899" ...
  .. .. .. .. .. .. .. ..@ elementType    : chr "ANY"
  .. .. .. .. .. .. .. ..@ elementMetadata: NULL
  .. .. .. .. .. .. .. ..@ metadata       : list()
  .. .. .. .. .. ..@ strand         :Formal class 'Rle' [package "S4Vectors"] with 4 slots
  .. .. .. .. .. .. .. ..@ values         : Factor w/ 3 levels "+","-","*": 1 2 1 2 1 2 1 2 1 2 ...
  .. .. .. .. .. .. .. ..@ lengths        : int [1:108826] 57 70 26 32 119 1 72 12 5 50 ...
  .. .. .. .. .. .. .. ..@ elementMetadata: NULL
  .. .. .. .. .. .. .. ..@ metadata       : list()
  .. .. .. .. .. ..@ seqinfo        :Formal class 'Seqinfo' [package "GenomeInfoDb"] with 4 slots
  .. .. .. .. .. .. .. ..@ seqnames   : chr [1:25] "chrX" "chr20" "chr1" "chr6" ...
  .. .. .. .. .. .. .. ..@ seqlengths : int [1:25] 156040895 64444167 248956422 170805979 198295559 159345973 133275309 135086622 190214555 83257441 ...
  .. .. .. .. .. .. .. ..@ is_circular: logi [1:25] FALSE FALSE FALSE FALSE FALSE FALSE ...
  .. .. .. .. .. .. .. ..@ genome     : chr [1:25] "hg38" "hg38" "hg38" "hg38" ...
  .. .. .. .. .. ..@ elementMetadata:Formal class 'DFrame' [package "S4Vectors"] with 6 slots
  .. .. .. .. .. .. .. ..@ rownames       : NULL
  .. .. .. .. .. .. .. ..@ nrows          : int 3021151
  .. .. .. .. .. .. .. ..@ elementType    : chr "ANY"
  .. .. .. .. .. .. .. ..@ elementMetadata: NULL
  .. .. .. .. .. .. .. ..@ metadata       : list()
  .. .. .. .. .. .. .. ..@ listData       :List of 5
  .. .. .. .. .. .. .. .. ..$ tx_id       : chr [1:3021151] "ENST00000399012" "ENST00000484611" "ENST00000430923" "ENST00000445062" ...
  .. .. .. .. .. .. .. .. ..$ gene_name   : chr [1:3021151] "PLCXD1" "PLCXD1" "PLCXD1" "PLCXD1" ...
  .. .. .. .. .. .. .. .. ..$ gene_id     : chr [1:3021151] "ENSG00000182378" "ENSG00000182378" "ENSG00000182378" "ENSG00000182378" ...
  .. .. .. .. .. .. .. .. ..$ gene_biotype: chr [1:3021151] "protein_coding" "protein_coding" "protein_coding" "protein_coding" ...
  .. .. .. .. .. .. .. .. ..$ type        : Factor w/ 4 levels "cds","exon","gap",..: 2 2 2 2 2 2 2 2 2 2 ...
  .. .. .. .. .. ..@ elementType    : chr "ANY"
  .. .. .. .. .. ..@ metadata       : list()
  .. .. .. ..@ bias              : NULL
  .. .. .. ..@ positionEnrichment: list()
  .. .. .. ..@ links             :Formal class 'GRanges' [package "GenomicRanges"] with 7 slots
  .. .. .. .. .. ..@ seqnames       :Formal class 'Rle' [package "S4Vectors"] with 4 slots
  .. .. .. .. .. .. .. ..@ values         : Factor w/ 0 levels: 
  .. .. .. .. .. .. .. ..@ lengths        : int(0) 
  .. .. .. .. .. .. .. ..@ elementMetadata: NULL
  .. .. .. .. .. .. .. ..@ metadata       : list()
  .. .. .. .. .. ..@ ranges         :Formal class 'IRanges' [package "IRanges"] with 6 slots
  .. .. .. .. .. .. .. ..@ start          : int(0) 
  .. .. .. .. .. .. .. ..@ width          : int(0) 
  .. .. .. .. .. .. .. ..@ NAMES          : NULL
  .. .. .. .. .. .. .. ..@ elementType    : chr "ANY"
  .. .. .. .. .. .. .. ..@ elementMetadata: NULL
  .. .. .. .. .. .. .. ..@ metadata       : list()
  .. .. .. .. .. ..@ strand         :Formal class 'Rle' [package "S4Vectors"] with 4 slots
  .. .. .. .. .. .. .. ..@ values         : Factor w/ 3 levels "+","-","*": 
  .. .. .. .. .. .. .. ..@ lengths        : int(0) 
  .. .. .. .. .. .. .. ..@ elementMetadata: NULL
  .. .. .. .. .. .. .. ..@ metadata       : list()
  .. .. .. .. .. ..@ seqinfo        :Formal class 'Seqinfo' [package "GenomeInfoDb"] with 4 slots
  .. .. .. .. .. .. .. ..@ seqnames   : chr(0) 
  .. .. .. .. .. .. .. ..@ seqlengths : int(0) 
  .. .. .. .. .. .. .. ..@ is_circular: logi(0) 
  .. .. .. .. .. .. .. ..@ genome     : chr(0) 
  .. .. .. .. .. ..@ elementMetadata:Formal class 'DFrame' [package "S4Vectors"] with 6 slots
  .. .. .. .. .. .. .. ..@ rownames       : NULL
  .. .. .. .. .. .. .. ..@ nrows          : int 0
  .. .. .. .. .. .. .. ..@ elementType    : chr "ANY"
  .. .. .. .. .. .. .. ..@ elementMetadata: NULL
  .. .. .. .. .. .. .. ..@ metadata       : list()
  .. .. .. .. .. .. .. ..@ listData       : Named list()
  .. .. .. .. .. ..@ elementType    : chr "ANY"
  .. .. .. .. .. ..@ metadata       : list()
  .. .. .. ..@ counts            :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
  .. .. .. .. .. ..@ i       : int [1:81271603] 12 16 25 27 28 30 34 37 38 39 ...
  .. .. .. .. .. ..@ p       : int [1:10413] 0 13878 21131 27659 30982 35249 46882 54127 62729 70448 ...
  .. .. .. .. .. ..@ Dim     : int [1:2] 108377 10412
  .. .. .. .. .. ..@ Dimnames:List of 2
  .. .. .. .. .. .. ..$ : chr [1:108377] "chr1-10109-10357" "chr1-180730-181630" "chr1-191491-191736" "chr1-267816-268196" ...
  .. .. .. .. .. .. ..$ : chr [1:10412] "AAACAGCCAAGGAATC-1_2" "AAACAGCCAATCCCTT-1_2" "AAACAGCCAATGCGCT-1_2" "AAACAGCCACCAACCG-1_2" ...
  .. .. .. .. .. ..@ x       : num [1:81271603] 1 2 2 2 2 2 4 6 8 10 ...
  .. .. .. .. .. ..@ factors : list()
  .. .. .. ..@ data              :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
  .. .. .. .. .. ..@ i       : int [1:81271603] 12 16 25 27 28 30 34 37 38 39 ...
  .. .. .. .. .. ..@ p       : int [1:10413] 0 13878 21131 27659 30982 35249 46882 54127 62729 70448 ...
  .. .. .. .. .. ..@ Dim     : int [1:2] 108377 10412
  .. .. .. .. .. ..@ Dimnames:List of 2
  .. .. .. .. .. .. ..$ : chr [1:108377] "chr1-10109-10357" "chr1-180730-181630" "chr1-191491-191736" "chr1-267816-268196" ...
  .. .. .. .. .. .. ..$ : chr [1:10412] "AAACAGCCAAGGAATC-1_2" "AAACAGCCAATCCCTT-1_2" "AAACAGCCAATGCGCT-1_2" "AAACAGCCACCAACCG-1_2" ...
  .. .. .. .. .. ..@ x       : num [1:81271603] 1.331 0.799 0.589 0.258 0.71 ...
  .. .. .. .. .. ..@ factors : list()
  .. .. .. ..@ scale.data        : num[0 , 0 ] 
  .. .. .. ..@ assay.orig        : NULL
  .. .. .. ..@ var.features      : chr [1:108377] "chr19-1230013-1281741" "chr19-39381822-39438328" "chr17-7832492-7859042" "chr19-12774552-12797715" ...
  .. .. .. ..@ meta.features     :'data.frame': 108377 obs. of  2 variables:
  .. .. .. .. ..$ count     : num [1:108377] 119 304 78 222 156 ...
  .. .. .. .. ..$ percentile: num [1:108377] 0.1389 0.4029 0.0845 0.315 0.2134 ...
  .. .. .. ..@ misc              : list()
  .. .. .. ..@ key               : chr "atac_"
  .. ..$ ACTIVITY:Formal class 'Assay' [package "SeuratObject"] with 8 slots
  .. .. .. ..@ counts       :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
  .. .. .. .. .. ..@ i       : int [1:4600705] 12 24 32 33 36 37 39 41 44 45 ...
  .. .. .. .. .. ..@ p       : int [1:10413] 0 594 965 1341 1553 1784 2303 2670 3185 3570 ...
  .. .. .. .. .. ..@ Dim     : int [1:2] 1286 10412
  .. .. .. .. .. ..@ Dimnames:List of 2
  .. .. .. .. .. .. ..$ : chr [1:1286] "CSF2RA" "XG" "MID1" "NHS" ...
  .. .. .. .. .. .. ..$ : chr [1:10412] "AAACAGCCAAGGAATC-1_2" "AAACAGCCAATCCCTT-1_2" "AAACAGCCAATGCGCT-1_2" "AAACAGCCACCAACCG-1_2" ...
  .. .. .. .. .. ..@ x       : num [1:4600705] 1 3 1 3 1 1 4 1 3 6 ...
  .. .. .. .. .. ..@ factors : list()
  .. .. .. ..@ data         :Formal class 'dgCMatrix' [package "Matrix"] with 6 slots
  .. .. .. .. .. ..@ i       : int [1:4600705] 12 24 32 33 36 37 39 41 44 45 ...
  .. .. .. .. .. ..@ p       : int [1:10413] 0 594 965 1341 1553 1784 2303 2670 3185 3570 ...
  .. .. .. .. .. ..@ Dim     : int [1:2] 1286 10412
  .. .. .. .. .. ..@ Dimnames:List of 2
  .. .. .. .. .. .. ..$ : chr [1:1286] "CSF2RA" "XG" "MID1" "NHS" ...
  .. .. .. .. .. .. ..$ : chr [1:10412] "AAACAGCCAAGGAATC-1_2" "AAACAGCCAATCCCTT-1_2" "AAACAGCCAATGCGCT-1_2" "AAACAGCCACCAACCG-1_2" ...
  .. .. .. .. .. ..@ x       : num [1:4600705] 1.82 2.8 1.82 2.8 1.82 ...
  .. .. .. .. .. ..@ factors : list()
  .. .. .. ..@ scale.data   : num [1:1286, 1:10412] -0.596 -0.332 -0.366 -0.545 -0.382 ...
  .. .. .. .. ..- attr(*, "dimnames")=List of 2
  .. .. .. .. .. ..$ : chr [1:1286] "CSF2RA" "XG" "MID1" "NHS" ...
  .. .. .. .. .. ..$ : chr [1:10412] "AAACAGCCAAGGAATC-1_2" "AAACAGCCAATCCCTT-1_2" "AAACAGCCAATGCGCT-1_2" "AAACAGCCACCAACCG-1_2" ...
  .. .. .. ..@ assay.orig   : NULL
  .. .. .. ..@ var.features : logi(0) 
  .. .. .. ..@ meta.features:'data.frame':  1286 obs. of  0 variables
  .. .. .. ..@ misc         : Named list()
  .. .. .. ..@ key          : chr "activity_"
  ..@ meta.data   :'data.frame':    20824 obs. of  30 variables:
  .. ..$ orig.ident                     : chr [1:20824] "RNA" "RNA" "RNA" "RNA" ...
  .. ..$ nCount_RNA                     : num [1:20824] 8380 3771 6876 5415 2759 ...
  .. ..$ nFeature_RNA                   : int [1:20824] 3308 1896 2904 2282 1353 3061 1691 3028 1783 1442 ...
  .. ..$ seurat_annotations             : chr [1:20824] "CD4 Naive" "CD4 TCM" "CD4 Naive" "CD8 Naive" ...
  .. ..$ nCount_ATAC                    : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ nFeature_ATAC                  : int [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ nCount_ACTIVITY                : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ nFeature_ACTIVITY              : int [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ predicted.id                   : chr [1:20824] NA NA NA NA ...
  .. ..$ prediction.score.CD4.Naive     : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.CD4.TCM       : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.CD8.Naive     : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.CD16.Mono     : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.NK            : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.Treg          : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.CD14.Mono     : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.cDC           : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.CD8.TEM_1     : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.Intermediate.B: num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.Naive.B       : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.Plasma        : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.CD4.TEM       : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.MAIT          : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.Memory.B      : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.gdT           : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.pDC           : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.CD8.TEM_2     : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.HSPC          : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ prediction.score.max           : num [1:20824] NA NA NA NA NA NA NA NA NA NA ...
  .. ..$ annotation_correct             : logi [1:20824] NA NA NA NA NA NA ...
  ..@ active.assay: chr "RNA"
  ..@ active.ident: Factor w/ 2 levels "RNA","ATAC": 1 1 1 1 1 1 1 1 1 1 ...
  .. ..- attr(*, "names")= chr [1:20824] "AAACAGCCAAGGAATC-1_1" "AAACAGCCAATCCCTT-1_1" "AAACAGCCAATGCGCT-1_1" "AAACAGCCACCAACCG-1_1" ...
  ..@ graphs      : list()
  ..@ neighbors   : list()
  ..@ reductions  :List of 2
  .. ..$ pca :Formal class 'DimReduc' [package "SeuratObject"] with 9 slots
  .. .. .. ..@ cell.embeddings           : num [1:20824, 1:50] -8.22 -7.64 -7.69 -7.32 -8.21 ...
  .. .. .. .. ..- attr(*, "dimnames")=List of 2
  .. .. .. .. .. ..$ : chr [1:20824] "AAACAGCCAAGGAATC-1_1" "AAACAGCCAATCCCTT-1_1" "AAACAGCCAATGCGCT-1_1" "AAACAGCCACCAACCG-1_1" ...
  .. .. .. .. .. ..$ : chr [1:50] "PC_1" "PC_2" "PC_3" "PC_4" ...
  .. .. .. ..@ feature.loadings          : num [1:2000, 1:50] -0.000875 -0.001317 -0.0072 -0.001528 0.001147 ...
  .. .. .. .. ..- attr(*, "dimnames")=List of 2
  .. .. .. .. .. ..$ : chr [1:2000] "JCHAIN" "PTGDS" "IGKC" "IGHA1" ...
  .. .. .. .. .. ..$ : chr [1:50] "PC_1" "PC_2" "PC_3" "PC_4" ...
  .. .. .. ..@ feature.loadings.projected: num[0 , 0 ] 
  .. .. .. ..@ assay.used                : chr "RNA"
  .. .. .. ..@ global                    : logi FALSE
  .. .. .. ..@ stdev                     : num [1:50] 9.83 4.35 3.46 2.57 2.01 ...
  .. .. .. ..@ jackstraw                 :Formal class 'JackStrawData' [package "SeuratObject"] with 4 slots
  .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ] 
  .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ] 
  .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ] 
  .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ] 
  .. .. .. ..@ misc                      :List of 1
  .. .. .. .. ..$ total.variance: num 277
  .. .. .. ..@ key                       : chr "PC_"
  .. ..$ umap:Formal class 'DimReduc' [package "SeuratObject"] with 9 slots
  .. .. .. ..@ cell.embeddings           : num [1:20824, 1:2] -7.68 -1.53 -7.75 -4.82 -6.11 ...
  .. .. .. .. ..- attr(*, "scaled:center")= num [1:2] -0.237 -0.144
  .. .. .. .. ..- attr(*, "dimnames")=List of 2
  .. .. .. .. .. ..$ : chr [1:20824] "AAACAGCCAAGGAATC-1_1" "AAACAGCCAATCCCTT-1_1" "AAACAGCCAATGCGCT-1_1" "AAACAGCCACCAACCG-1_1" ...
  .. .. .. .. .. ..$ : chr [1:2] "umap_1" "umap_2"
  .. .. .. ..@ feature.loadings          : num[0 , 0 ] 
  .. .. .. ..@ feature.loadings.projected: num[0 , 0 ] 
  .. .. .. ..@ assay.used                : chr "RNA"
  .. .. .. ..@ global                    : logi TRUE
  .. .. .. ..@ stdev                     : num(0) 
  .. .. .. ..@ jackstraw                 :Formal class 'JackStrawData' [package "SeuratObject"] with 4 slots
  .. .. .. .. .. ..@ empirical.p.values     : num[0 , 0 ] 
  .. .. .. .. .. ..@ fake.reduction.scores  : num[0 , 0 ] 
  .. .. .. .. .. ..@ empirical.p.values.full: num[0 , 0 ] 
  .. .. .. .. .. ..@ overall.p.values       : num[0 , 0 ] 
  .. .. .. ..@ misc                      : list()
  .. .. .. ..@ key                       : chr "umap_"
  ..@ images      : list()
  ..@ project.name: chr "SeuratProject"
  ..@ misc        : list()
  ..@ version     :Classes 'package_version', 'numeric_version'  hidden list of 1
  .. ..$ : int [1:3] 5 0 1
  ..@ commands    :List of 3
  .. ..$ ScaleData.RNA  :Formal class 'SeuratCommand' [package "SeuratObject"] with 5 slots
  .. .. .. ..@ name       : chr "ScaleData.RNA"
  .. .. .. ..@ time.stamp : POSIXct[1:1], format: "2023-12-10 19:41:30"
  .. .. .. ..@ assay.used : chr "RNA"
  .. .. .. ..@ call.string: chr "ScaleData(coembed, features = genes.use, do.scale = FALSE)"
  .. .. .. ..@ params     :List of 10
  .. .. .. .. ..$ features          : chr [1:2000] "JCHAIN" "PTGDS" "IGKC" "IGHA1" ...
  .. .. .. .. ..$ assay             : chr "RNA"
  .. .. .. .. ..$ model.use         : chr "linear"
  .. .. .. .. ..$ use.umi           : logi FALSE
  .. .. .. .. ..$ do.scale          : logi FALSE
  .. .. .. .. ..$ do.center         : logi TRUE
  .. .. .. .. ..$ scale.max         : num 10
  .. .. .. .. ..$ block.size        : num 1000
  .. .. .. .. ..$ min.cells.to.block: num 3000
  .. .. .. .. ..$ verbose           : logi TRUE
  .. ..$ RunPCA.RNA     :Formal class 'SeuratCommand' [package "SeuratObject"] with 5 slots
  .. .. .. ..@ name       : chr "RunPCA.RNA"
  .. .. .. ..@ time.stamp : POSIXct[1:1], format: "2023-12-10 19:41:51"
  .. .. .. ..@ assay.used : chr "RNA"
  .. .. .. ..@ call.string: chr "RunPCA(coembed, features = genes.use, verbose = FALSE)"
  .. .. .. ..@ params     :List of 11
  .. .. .. .. ..$ assay          : chr "RNA"
  .. .. .. .. ..$ features       : chr [1:2000] "JCHAIN" "PTGDS" "IGKC" "IGHA1" ...
  .. .. .. .. ..$ npcs           : num 50
  .. .. .. .. ..$ rev.pca        : logi FALSE
  .. .. .. .. ..$ weight.by.var  : logi TRUE
  .. .. .. .. ..$ verbose        : logi FALSE
  .. .. .. .. ..$ ndims.print    : int [1:5] 1 2 3 4 5
  .. .. .. .. ..$ nfeatures.print: num 30
  .. .. .. .. ..$ reduction.name : chr "pca"
  .. .. .. .. ..$ reduction.key  : chr "PC_"
  .. .. .. .. ..$ seed.use       : num 42
  .. ..$ RunUMAP.RNA.pca:Formal class 'SeuratCommand' [package "SeuratObject"] with 5 slots
  .. .. .. ..@ name       : chr "RunUMAP.RNA.pca"
  .. .. .. ..@ time.stamp : POSIXct[1:1], format: "2023-12-10 19:42:37"
  .. .. .. ..@ assay.used : chr "RNA"
  .. .. .. ..@ call.string: chr "RunUMAP(coembed, dims = 1:30)"
  .. .. .. ..@ params     :List of 25
  .. .. .. .. ..$ dims                : int [1:30] 1 2 3 4 5 6 7 8 9 10 ...
  .. .. .. .. ..$ reduction           : chr "pca"
  .. .. .. .. ..$ assay               : chr "RNA"
  .. .. .. .. ..$ slot                : chr "data"
  .. .. .. .. ..$ umap.method         : chr "uwot"
  .. .. .. .. ..$ return.model        : logi FALSE
  .. .. .. .. ..$ n.neighbors         : int 30
  .. .. .. .. ..$ n.components        : int 2
  .. .. .. .. ..$ metric              : chr "cosine"
  .. .. .. .. ..$ learning.rate       : num 1
  .. .. .. .. ..$ min.dist            : num 0.3
  .. .. .. .. ..$ spread              : num 1
  .. .. .. .. ..$ set.op.mix.ratio    : num 1
  .. .. .. .. ..$ local.connectivity  : int 1
  .. .. .. .. ..$ repulsion.strength  : num 1
  .. .. .. .. ..$ negative.sample.rate: int 5
  .. .. .. .. ..$ uwot.sgd            : logi FALSE
  .. .. .. .. ..$ seed.use            : int 42
  .. .. .. .. ..$ angular.rp.forest   : logi FALSE
  .. .. .. .. ..$ densmap             : logi FALSE
  .. .. .. .. ..$ dens.lambda         : num 2
  .. .. .. .. ..$ dens.frac           : num 0.3
  .. .. .. .. ..$ dens.var.shift      : num 0.1
  .. .. .. .. ..$ verbose             : logi TRUE
  .. .. .. .. ..$ reduction.name      : chr "umap"
  ..@ tools       : list()</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb160"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb160-1"><a href="#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.3.2 (2023-10-31 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default


locale:
[1] LC_COLLATE=English_United States.utf8 
[2] LC_CTYPE=English_United States.utf8   
[3] LC_MONETARY=English_United States.utf8
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.utf8    

time zone: America/Tijuana
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] cowplot_1.1.1                 ggplot2_3.4.4                
 [3] EnsDb.Hsapiens.v86_2.99.0     ensembldb_2.26.0             
 [5] AnnotationFilter_1.26.0       GenomicFeatures_1.54.1       
 [7] AnnotationDbi_1.64.1          Biobase_2.62.0               
 [9] GenomicRanges_1.54.1          GenomeInfoDb_1.38.1          
[11] IRanges_2.36.0                S4Vectors_0.40.2             
[13] BiocGenerics_0.48.1           Signac_1.12.0                
[15] Seurat_5.0.1                  SeuratObject_5.0.1           
[17] sp_2.1-1                      pbmcMultiome.SeuratData_0.1.4
[19] SeuratData_0.2.2.9001        

loaded via a namespace (and not attached):
  [1] RcppAnnoy_0.0.21            splines_4.3.2              
  [3] later_1.3.1                 BiocIO_1.12.0              
  [5] bitops_1.0-7                filelock_1.0.2             
  [7] tibble_3.2.1                polyclip_1.10-6            
  [9] rpart_4.1.21                XML_3.99-0.15              
 [11] fastDummies_1.7.3           lifecycle_1.0.4            
 [13] globals_0.16.2              lattice_0.22-5             
 [15] MASS_7.3-60                 backports_1.4.1            
 [17] magrittr_2.0.3              Hmisc_5.1-1                
 [19] plotly_4.10.3               rmarkdown_2.25             
 [21] yaml_2.3.7                  httpuv_1.6.12              
 [23] sctransform_0.4.1           spam_2.10-0                
 [25] spatstat.sparse_3.0-3       reticulate_1.34.0          
 [27] pbapply_1.7-2               DBI_1.1.3                  
 [29] RColorBrewer_1.1-3          abind_1.4-5                
 [31] zlibbioc_1.48.0             Rtsne_0.16                 
 [33] purrr_1.0.2                 biovizBase_1.50.0          
 [35] RCurl_1.98-1.13             nnet_7.3-19                
 [37] VariantAnnotation_1.48.1    rappdirs_0.3.3             
 [39] GenomeInfoDbData_1.2.11     ggrepel_0.9.4              
 [41] irlba_2.3.5.1               listenv_0.9.0              
 [43] spatstat.utils_3.0-4        goftest_1.2-3              
 [45] RSpectra_0.16-1             spatstat.random_3.2-1      
 [47] fitdistrplus_1.1-11         parallelly_1.36.0          
 [49] DelayedArray_0.28.0         leiden_0.4.3.1             
 [51] codetools_0.2-19            RcppRoll_0.3.0             
 [53] xml2_1.3.5                  tidyselect_1.2.0           
 [55] farver_2.1.1                base64enc_0.1-3            
 [57] matrixStats_1.1.0           BiocFileCache_2.10.1       
 [59] spatstat.explore_3.2-5      GenomicAlignments_1.38.0   
 [61] jsonlite_1.8.7              Formula_1.2-5              
 [63] ellipsis_0.3.2              progressr_0.14.0           
 [65] ggridges_0.5.4              survival_3.5-7             
 [67] systemfonts_1.0.5           tools_4.3.2                
 [69] progress_1.2.2              ragg_1.2.6                 
 [71] ica_1.0-3                   Rcpp_1.0.11                
 [73] glue_1.6.2                  SparseArray_1.2.2          
 [75] gridExtra_2.3               xfun_0.41                  
 [77] MatrixGenerics_1.14.0       dplyr_1.1.4                
 [79] withr_2.5.2                 fastmap_1.1.1              
 [81] fansi_1.0.5                 digest_0.6.33              
 [83] R6_2.5.1                    mime_0.12                  
 [85] textshaping_0.3.7           colorspace_2.1-0           
 [87] scattermore_1.2             tensor_1.5                 
 [89] dichromat_2.0-0.1           spatstat.data_3.0-3        
 [91] biomaRt_2.58.0              RSQLite_2.3.3              
 [93] utf8_1.2.4                  tidyr_1.3.0                
 [95] generics_0.1.3              data.table_1.14.8          
 [97] rtracklayer_1.62.0          S4Arrays_1.2.0             
 [99] prettyunits_1.2.0           httr_1.4.7                 
[101] htmlwidgets_1.6.3           uwot_0.1.16                
[103] pkgconfig_2.0.3             gtable_0.3.4               
[105] blob_1.2.4                  lmtest_0.9-40              
[107] XVector_0.42.0              htmltools_0.5.7            
[109] dotCall64_1.1-0             ProtGenerics_1.34.0        
[111] scales_1.2.1                png_0.1-8                  
[113] knitr_1.45                  rstudioapi_0.15.0          
[115] rjson_0.2.21                reshape2_1.4.4             
[117] checkmate_2.3.0             nlme_3.1-163               
[119] curl_5.1.0                  zoo_1.8-12                 
[121] cachem_1.0.8                stringr_1.5.1              
[123] KernSmooth_2.23-22          parallel_4.3.2             
[125] miniUI_0.1.1.1              foreign_0.8-85             
[127] restfulr_0.0.15             pillar_1.9.0               
[129] grid_4.3.2                  vctrs_0.6.4                
[131] RANN_2.6.1                  promises_1.2.1             
[133] dbplyr_2.4.0                xtable_1.8-4               
[135] cluster_2.1.4               htmlTable_2.4.2            
[137] evaluate_0.23               cli_3.6.1                  
[139] compiler_4.3.2              Rsamtools_2.18.0           
[141] rlang_1.1.2                 crayon_1.5.2               
[143] future.apply_1.11.0         labeling_0.4.3             
[145] plyr_1.8.9                  stringi_1.8.1              
[147] viridisLite_0.4.2           deldir_2.0-2               
[149] BiocParallel_1.36.0         munsell_0.5.0              
[151] Biostrings_2.70.1           lazyeval_0.2.2             
[153] spatstat.geom_3.2-7         Matrix_1.6-3               
[155] BSgenome_1.70.1             RcppHNSW_0.5.0             
[157] hms_1.1.3                   patchwork_1.1.3            
[159] bit64_4.0.5                 future_1.33.0              
[161] KEGGREST_1.42.0             shiny_1.8.0                
[163] SummarizedExperiment_1.32.0 ROCR_1.0-11                
[165] igraph_1.5.1                memoise_2.0.1              
[167] fastmatch_1.1-4             bit_4.0.5                  </code></pre>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>