---
title: "RNA-seq differential analysis with CummeRbund"
author: "Lavinia Carabet"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Install the CummeRbund package

Analysis, exploration, manipulation, and visualization of Cufflinks high-throughput sequencing data

```{r, message=FALSE, warning=FALSE}
#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#BiocManager::install("cummeRbund")
```

## Load the CummeRbund package

```{r, message=FALSE, warning=FALSE}
library(cummeRbund)
```

## Create a CummeRbund database from the Cuffdiff output

```{r}
print(readCufflinks)
```

```{r}
cuff_data <- readCufflinks(dir = '../cuffdiff/diff_out/')
cuff_data@genes # or genes(cuff_data)
getLevels(cuff_data)
replicates(cuff_data)
cuff_data
```
## Plot the distribution of expression levels for all genes for each sample/condition

T - treated 
U - untreated

`csDensity` method creates a smoothed density plot, by sample, for log10 FPKM values from a cuffdiff run
FPKM - fragments per kilobase of transcript per million fragments mapped

```{r}
csDensity(cuff_data@genes, pseudocount=1.0)
```

## Compare the expression of each gene in the two conditions 

`csScatter` method creates a scatter plot comparing the FPKM values from the two samples in a cuffdiff run

```{r, message=FALSE, warnings=FALSE}
csScatter(cuff_data@genes, 'T', 'U', pseudocount=1.0, colorByStatus=TRUE, smooth=TRUE)
```

## Inspect differentially expressed genes 

`csVolcano` method creates a volcano plot of log fold change in expression vs significance (-log(pval)) for the two samples in a cuffdiff run

```{r, message=FALSE, warning=FALSE}
gene_diff_data <- diffData(cuff_data@genes, 'T', 'U')
gene_diff_data
#csVolcano(cuff_data@genes, 'T', 'U', alpha=0.05, showSignificant=T)
# The call to csVolcano fails with
#Error in `$<-.data.frame`(`*tmp*`, "significant", value = "no") : 
#replacement has 1 row, data has 0
```
### OOPS! No significantly differential expressed genes found?! Figuring out the issue ...

## Explore expression levels for some gene 

```{r, message=FALSE, warning=FALSE}
mygene <- getGene(cuff_data, 'SNRNP27')
expressionBarplot(mygene)
```

## Look at the `gene_exp.diff` and `iso_exp.diff` files generated by cuffdiff

### Issue found

```{r}
gene_expr_diff <- read.delim2('../cuffdiff/diff_out/gene_exp.diff')
head(gene_expr_diff)
# Issue: the 'T' character label is the files is converted behind the scenes to logical TRUE in data.frame(s) OOPS!
# Lesson learned: Never use 'T' or 'F' to label conditions/samples
gene_expr_diff$sample_1 <- 'T' # replace logical TRUE interpretation for the original character label T
iso_expr_diff <- read.delim2('../cuffdiff/diff_out/isoform_exp.diff')
iso_expr_diff$sample_1 <- 'T'

dim(gene_expr_diff[gene_expr_diff$status == 'OK' & gene_expr_diff$significant == 'yes', ])
head(gene_expr_diff[gene_expr_diff$status == 'OK' & gene_expr_diff$significant == 'yes', ])
dim(iso_expr_diff[iso_expr_diff$status == 'OK' & iso_expr_diff$significant == 'yes', ])
```

## Volcano Plot

```{r}

ged <- gene_expr_diff[gene_expr_diff$status == 'OK' & !is.infinite(as.numeric(gene_expr_diff$log2.fold_change.)), ]
ged$log2.fold.change <- as.numeric(ged$log2.fold_change.)
ged$adj.p.value <- as.numeric(ged$q_value)
head(ged)

h.line <- -log10(0.05) 
v.line <- log2(2) 

ged$diff.expr <- "No"
ged$diff.expr[ged$log2.fold.change > v.line & -log10(ged$adj.p.value) > h.line ] <- "Up"
ged$diff.expr[ged$log2.fold.change < -v.line & -log10(ged$adj.p.value) > h.line ] <- "Down"

head(ged)
```

```{r}
cols <- c("green", "red", "black")
names(cols) <- c("Down", "Up", "No")

vp <- ggplot(data=ged, aes(x=log2.fold.change, y=-log10(adj.p.value ), col=diff.expr)) + 
  geom_point() + theme_minimal() +
  scale_color_manual(values=cols) +
  geom_vline(xintercept=c(-v.line, v.line), col="red") + 
  geom_hline(yintercept=h.line, col="red") 
  

vp
```

## Top 10 down- and up-regulated genes

```{r}
ged.diff.expr <- ged[ged$diff.expr %in% c("Up", "Down"),]
dim(ged.diff.expr)
dim(ged.diff.expr[ged.diff.expr$significant == "no",])
ged.diff.expr.order <- ged.diff.expr[order(ged.diff.expr[, 15]),] 
head(ged.diff.expr.order, n=5)
tail(ged.diff.expr.order, n=5)
```