{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download data from SRA\n",
    "#### GEO95077 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/cv/RNA-seq/data\n"
     ]
    }
   ],
   "source": [
    "cd data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['SRR5272674', 'SRR5272675', 'SRR5272676', 'SRR5272677',\n",
       "       'SRR5272678', 'SRR5272679'], dtype=object)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "dat = pd.read_table(\"SRR_Acc_List.txt\", header=None).to_numpy().flatten()\n",
    "dat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-02-22 17:07:19,008 - SRR ids: ['SRR5272674']\n",
      "2022-02-22 17:07:19,009 - extra args: ['--split-files', '--gzip']\n",
      "2022-02-22 17:07:19,009 - tempdir: /tmp/pfd_ft4btce7\n",
      "2022-02-22 17:07:19,009 - CMD: sra-stat --meta --quick SRR5272674\n",
      "2022-02-22 17:07:21,081 - SRR5272674 spots: 50946237\n",
      "2022-02-22 17:07:21,081 - blocks: [[1, 6368279], [6368280, 12736558], [12736559, 19104837], [19104838, 25473116], [25473117, 31841395], [31841396, 38209674], [38209675, 44577953], [44577954, 50946237]]\n",
      "2022-02-22 17:07:21,082 - CMD: fastq-dump -N 1 -X 6368279 -O /tmp/pfd_ft4btce7/0 --split-files --gzip SRR5272674\n",
      "2022-02-22 17:07:21,103 - CMD: fastq-dump -N 6368280 -X 12736558 -O /tmp/pfd_ft4btce7/1 --split-files --gzip SRR5272674\n",
      "2022-02-22 17:07:21,106 - CMD: fastq-dump -N 12736559 -X 19104837 -O /tmp/pfd_ft4btce7/2 --split-files --gzip SRR5272674\n",
      "2022-02-22 17:07:21,110 - CMD: fastq-dump -N 19104838 -X 25473116 -O /tmp/pfd_ft4btce7/3 --split-files --gzip SRR5272674\n",
      "2022-02-22 17:07:21,114 - CMD: fastq-dump -N 25473117 -X 31841395 -O /tmp/pfd_ft4btce7/4 --split-files --gzip SRR5272674\n",
      "2022-02-22 17:07:21,118 - CMD: fastq-dump -N 31841396 -X 38209674 -O /tmp/pfd_ft4btce7/5 --split-files --gzip SRR5272674\n",
      "2022-02-22 17:07:21,121 - CMD: fastq-dump -N 38209675 -X 44577953 -O /tmp/pfd_ft4btce7/6 --split-files --gzip SRR5272674\n",
      "2022-02-22 17:07:21,170 - CMD: fastq-dump -N 44577954 -X 50946237 -O /tmp/pfd_ft4btce7/7 --split-files --gzip SRR5272674\n",
      "Read 6368279 spots for SRR5272674\n",
      "Written 6368279 spots for SRR5272674\n",
      "Read 6368279 spots for SRR5272674\n",
      "Written 6368279 spots for SRR5272674\n",
      "Read 6368279 spots for SRR5272674\n",
      "Written 6368279 spots for SRR5272674\n",
      "Read 6368279 spots for SRR5272674\n",
      "Written 6368279 spots for SRR5272674\n",
      "Read 6368279 spots for SRR5272674\n",
      "Written 6368279 spots for SRR5272674\n",
      "Read 6368279 spots for SRR5272674\n",
      "Written 6368279 spots for SRR5272674\n",
      "Read 6368284 spots for SRR5272674\n",
      "Written 6368284 spots for SRR5272674\n",
      "Read 6368279 spots for SRR5272674\n",
      "Written 6368279 spots for SRR5272674\n",
      "2022-02-22 17:36:37,978 - SRR ids: ['SRR5272675']\n",
      "2022-02-22 17:36:37,979 - extra args: ['--split-files', '--gzip']\n",
      "2022-02-22 17:36:37,998 - tempdir: /tmp/pfd_jjv9p9ln\n",
      "2022-02-22 17:36:37,998 - CMD: sra-stat --meta --quick SRR5272675\n",
      "2022-02-22 17:36:40,708 - SRR5272675 spots: 62571514\n",
      "2022-02-22 17:36:40,709 - blocks: [[1, 7821439], [7821440, 15642878], [15642879, 23464317], [23464318, 31285756], [31285757, 39107195], [39107196, 46928634], [46928635, 54750073], [54750074, 62571514]]\n",
      "2022-02-22 17:36:40,709 - CMD: fastq-dump -N 1 -X 7821439 -O /tmp/pfd_jjv9p9ln/0 --split-files --gzip SRR5272675\n",
      "2022-02-22 17:36:40,712 - CMD: fastq-dump -N 7821440 -X 15642878 -O /tmp/pfd_jjv9p9ln/1 --split-files --gzip SRR5272675\n",
      "2022-02-22 17:36:40,715 - CMD: fastq-dump -N 15642879 -X 23464317 -O /tmp/pfd_jjv9p9ln/2 --split-files --gzip SRR5272675\n",
      "2022-02-22 17:36:40,717 - CMD: fastq-dump -N 23464318 -X 31285756 -O /tmp/pfd_jjv9p9ln/3 --split-files --gzip SRR5272675\n",
      "2022-02-22 17:36:40,718 - CMD: fastq-dump -N 31285757 -X 39107195 -O /tmp/pfd_jjv9p9ln/4 --split-files --gzip SRR5272675\n",
      "2022-02-22 17:36:40,719 - CMD: fastq-dump -N 39107196 -X 46928634 -O /tmp/pfd_jjv9p9ln/5 --split-files --gzip SRR5272675\n",
      "2022-02-22 17:36:40,721 - CMD: fastq-dump -N 46928635 -X 54750073 -O /tmp/pfd_jjv9p9ln/6 --split-files --gzip SRR5272675\n",
      "2022-02-22 17:36:40,722 - CMD: fastq-dump -N 54750074 -X 62571514 -O /tmp/pfd_jjv9p9ln/7 --split-files --gzip SRR5272675\n",
      "Read 7821439 spots for SRR5272675\n",
      "Written 7821439 spots for SRR5272675\n",
      "Read 7821439 spots for SRR5272675\n",
      "Written 7821439 spots for SRR5272675\n",
      "Read 7821439 spots for SRR5272675\n",
      "Written 7821439 spots for SRR5272675\n",
      "Read 7821439 spots for SRR5272675\n",
      "Written 7821439 spots for SRR5272675\n",
      "Read 7821441 spots for SRR5272675\n",
      "Written 7821441 spots for SRR5272675\n",
      "Read 7821439 spots for SRR5272675\n",
      "Written 7821439 spots for SRR5272675\n",
      "Read 7821439 spots for SRR5272675\n",
      "Written 7821439 spots for SRR5272675\n",
      "Read 7821439 spots for SRR5272675\n",
      "Written 7821439 spots for SRR5272675\n",
      "2022-02-22 18:24:35,848 - SRR ids: ['SRR5272676']\n",
      "2022-02-22 18:24:35,848 - extra args: ['--split-files', '--gzip']\n",
      "2022-02-22 18:24:35,848 - tempdir: /tmp/pfd_nbm7mnl1\n",
      "2022-02-22 18:24:35,848 - CMD: sra-stat --meta --quick SRR5272676\n",
      "2022-02-22 18:24:38,611 - SRR5272676 spots: 77906369\n",
      "2022-02-22 18:24:38,612 - blocks: [[1, 9738296], [9738297, 19476592], [19476593, 29214888], [29214889, 38953184], [38953185, 48691480], [48691481, 58429776], [58429777, 68168072], [68168073, 77906369]]\n",
      "2022-02-22 18:24:38,612 - CMD: fastq-dump -N 1 -X 9738296 -O /tmp/pfd_nbm7mnl1/0 --split-files --gzip SRR5272676\n",
      "2022-02-22 18:24:38,614 - CMD: fastq-dump -N 9738297 -X 19476592 -O /tmp/pfd_nbm7mnl1/1 --split-files --gzip SRR5272676\n",
      "2022-02-22 18:24:38,617 - CMD: fastq-dump -N 19476593 -X 29214888 -O /tmp/pfd_nbm7mnl1/2 --split-files --gzip SRR5272676\n",
      "2022-02-22 18:24:38,619 - CMD: fastq-dump -N 29214889 -X 38953184 -O /tmp/pfd_nbm7mnl1/3 --split-files --gzip SRR5272676\n",
      "2022-02-22 18:24:38,620 - CMD: fastq-dump -N 38953185 -X 48691480 -O /tmp/pfd_nbm7mnl1/4 --split-files --gzip SRR5272676\n",
      "2022-02-22 18:24:38,621 - CMD: fastq-dump -N 48691481 -X 58429776 -O /tmp/pfd_nbm7mnl1/5 --split-files --gzip SRR5272676\n",
      "2022-02-22 18:24:38,622 - CMD: fastq-dump -N 58429777 -X 68168072 -O /tmp/pfd_nbm7mnl1/6 --split-files --gzip SRR5272676\n",
      "2022-02-22 18:24:38,623 - CMD: fastq-dump -N 68168073 -X 77906369 -O /tmp/pfd_nbm7mnl1/7 --split-files --gzip SRR5272676\n",
      "Read 9738296 spots for SRR5272676\n",
      "Written 9738296 spots for SRR5272676\n",
      "Read 9738296 spots for SRR5272676\n",
      "Written 9738296 spots for SRR5272676\n",
      "Read 9738296 spots for SRR5272676\n",
      "Written 9738296 spots for SRR5272676\n",
      "Read 9738296 spots for SRR5272676\n",
      "Written 9738296 spots for SRR5272676\n",
      "Read 9738296 spots for SRR5272676\n",
      "Written 9738296 spots for SRR5272676\n",
      "Read 9738296 spots for SRR5272676\n",
      "Written 9738296 spots for SRR5272676\n",
      "Read 9738297 spots for SRR5272676\n",
      "Written 9738297 spots for SRR5272676\n",
      "Read 9738296 spots for SRR5272676\n",
      "Written 9738296 spots for SRR5272676\n",
      "2022-02-22 19:30:50,622 - SRR ids: ['SRR5272677']\n",
      "2022-02-22 19:30:50,622 - extra args: ['--split-files', '--gzip']\n",
      "2022-02-22 19:30:50,625 - tempdir: /tmp/pfd_pmql9qra\n",
      "2022-02-22 19:30:50,625 - CMD: sra-stat --meta --quick SRR5272677\n",
      "2022-02-22 19:30:55,036 - SRR5272677 spots: 60733681\n",
      "2022-02-22 19:30:55,036 - blocks: [[1, 7591710], [7591711, 15183420], [15183421, 22775130], [22775131, 30366840], [30366841, 37958550], [37958551, 45550260], [45550261, 53141970], [53141971, 60733681]]\n",
      "2022-02-22 19:30:55,036 - CMD: fastq-dump -N 1 -X 7591710 -O /tmp/pfd_pmql9qra/0 --split-files --gzip SRR5272677\n",
      "2022-02-22 19:30:55,037 - CMD: fastq-dump -N 7591711 -X 15183420 -O /tmp/pfd_pmql9qra/1 --split-files --gzip SRR5272677\n",
      "2022-02-22 19:30:55,038 - CMD: fastq-dump -N 15183421 -X 22775130 -O /tmp/pfd_pmql9qra/2 --split-files --gzip SRR5272677\n",
      "2022-02-22 19:30:55,039 - CMD: fastq-dump -N 22775131 -X 30366840 -O /tmp/pfd_pmql9qra/3 --split-files --gzip SRR5272677\n",
      "2022-02-22 19:30:55,040 - CMD: fastq-dump -N 30366841 -X 37958550 -O /tmp/pfd_pmql9qra/4 --split-files --gzip SRR5272677\n",
      "2022-02-22 19:30:55,042 - CMD: fastq-dump -N 37958551 -X 45550260 -O /tmp/pfd_pmql9qra/5 --split-files --gzip SRR5272677\n",
      "2022-02-22 19:30:55,043 - CMD: fastq-dump -N 45550261 -X 53141970 -O /tmp/pfd_pmql9qra/6 --split-files --gzip SRR5272677\n",
      "2022-02-22 19:30:55,045 - CMD: fastq-dump -N 53141971 -X 60733681 -O /tmp/pfd_pmql9qra/7 --split-files --gzip SRR5272677\n",
      "Read 7591710 spots for SRR5272677\n",
      "Written 7591710 spots for SRR5272677\n",
      "Read 7591710 spots for SRR5272677\n",
      "Written 7591710 spots for SRR5272677\n",
      "Read 7591710 spots for SRR5272677\n",
      "Written 7591710 spots for SRR5272677\n",
      "Read 7591710 spots for SRR5272677\n",
      "Written 7591710 spots for SRR5272677\n",
      "Read 7591710 spots for SRR5272677\n",
      "Written 7591710 spots for SRR5272677\n",
      "Read 7591710 spots for SRR5272677\n",
      "Written 7591710 spots for SRR5272677\n",
      "Read 7591710 spots for SRR5272677\n",
      "Written 7591710 spots for SRR5272677\n",
      "Read 7591711 spots for SRR5272677\n",
      "Written 7591711 spots for SRR5272677\n",
      "2022-02-22 20:25:46,132 - SRR ids: ['SRR5272678']\n",
      "2022-02-22 20:25:46,132 - extra args: ['--split-files', '--gzip']\n",
      "2022-02-22 20:25:46,134 - tempdir: /tmp/pfd_7dhm5ol0\n",
      "2022-02-22 20:25:46,134 - CMD: sra-stat --meta --quick SRR5272678\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-02-22 20:25:48,917 - SRR5272678 spots: 66024375\n",
      "2022-02-22 20:25:48,917 - blocks: [[1, 8253046], [8253047, 16506092], [16506093, 24759138], [24759139, 33012184], [33012185, 41265230], [41265231, 49518276], [49518277, 57771322], [57771323, 66024375]]\n",
      "2022-02-22 20:25:48,918 - CMD: fastq-dump -N 1 -X 8253046 -O /tmp/pfd_7dhm5ol0/0 --split-files --gzip SRR5272678\n",
      "2022-02-22 20:25:48,920 - CMD: fastq-dump -N 8253047 -X 16506092 -O /tmp/pfd_7dhm5ol0/1 --split-files --gzip SRR5272678\n",
      "2022-02-22 20:25:48,923 - CMD: fastq-dump -N 16506093 -X 24759138 -O /tmp/pfd_7dhm5ol0/2 --split-files --gzip SRR5272678\n",
      "2022-02-22 20:25:48,925 - CMD: fastq-dump -N 24759139 -X 33012184 -O /tmp/pfd_7dhm5ol0/3 --split-files --gzip SRR5272678\n",
      "2022-02-22 20:25:48,926 - CMD: fastq-dump -N 33012185 -X 41265230 -O /tmp/pfd_7dhm5ol0/4 --split-files --gzip SRR5272678\n",
      "2022-02-22 20:25:48,928 - CMD: fastq-dump -N 41265231 -X 49518276 -O /tmp/pfd_7dhm5ol0/5 --split-files --gzip SRR5272678\n",
      "2022-02-22 20:25:48,929 - CMD: fastq-dump -N 49518277 -X 57771322 -O /tmp/pfd_7dhm5ol0/6 --split-files --gzip SRR5272678\n",
      "2022-02-22 20:25:48,930 - CMD: fastq-dump -N 57771323 -X 66024375 -O /tmp/pfd_7dhm5ol0/7 --split-files --gzip SRR5272678\n",
      "Read 8253046 spots for SRR5272678\n",
      "Written 8253046 spots for SRR5272678\n",
      "Read 8253046 spots for SRR5272678\n",
      "Written 8253046 spots for SRR5272678\n",
      "Read 8253053 spots for SRR5272678\n",
      "Written 8253053 spots for SRR5272678\n",
      "Read 8253046 spots for SRR5272678\n",
      "Written 8253046 spots for SRR5272678\n",
      "Read 8253046 spots for SRR5272678\n",
      "Written 8253046 spots for SRR5272678\n",
      "Read 8253046 spots for SRR5272678\n",
      "Written 8253046 spots for SRR5272678\n",
      "Read 8253046 spots for SRR5272678\n",
      "Written 8253046 spots for SRR5272678\n",
      "Read 8253046 spots for SRR5272678\n",
      "Written 8253046 spots for SRR5272678\n",
      "2022-02-22 21:20:51,230 - SRR ids: ['SRR5272679']\n",
      "2022-02-22 21:20:51,231 - extra args: ['--split-files', '--gzip']\n",
      "2022-02-22 21:20:51,232 - tempdir: /tmp/pfd_u89wzk2r\n",
      "2022-02-22 21:20:51,232 - CMD: sra-stat --meta --quick SRR5272679\n",
      "2022-02-22 21:20:53,858 - SRR5272679 spots: 45598113\n",
      "2022-02-22 21:20:53,858 - blocks: [[1, 5699764], [5699765, 11399528], [11399529, 17099292], [17099293, 22799056], [22799057, 28498820], [28498821, 34198584], [34198585, 39898348], [39898349, 45598113]]\n",
      "2022-02-22 21:20:53,859 - CMD: fastq-dump -N 1 -X 5699764 -O /tmp/pfd_u89wzk2r/0 --split-files --gzip SRR5272679\n",
      "2022-02-22 21:20:53,861 - CMD: fastq-dump -N 5699765 -X 11399528 -O /tmp/pfd_u89wzk2r/1 --split-files --gzip SRR5272679\n",
      "2022-02-22 21:20:53,863 - CMD: fastq-dump -N 11399529 -X 17099292 -O /tmp/pfd_u89wzk2r/2 --split-files --gzip SRR5272679\n",
      "2022-02-22 21:20:53,865 - CMD: fastq-dump -N 17099293 -X 22799056 -O /tmp/pfd_u89wzk2r/3 --split-files --gzip SRR5272679\n",
      "2022-02-22 21:20:53,867 - CMD: fastq-dump -N 22799057 -X 28498820 -O /tmp/pfd_u89wzk2r/4 --split-files --gzip SRR5272679\n",
      "2022-02-22 21:20:53,869 - CMD: fastq-dump -N 28498821 -X 34198584 -O /tmp/pfd_u89wzk2r/5 --split-files --gzip SRR5272679\n",
      "2022-02-22 21:20:53,871 - CMD: fastq-dump -N 34198585 -X 39898348 -O /tmp/pfd_u89wzk2r/6 --split-files --gzip SRR5272679\n",
      "2022-02-22 21:20:53,873 - CMD: fastq-dump -N 39898349 -X 45598113 -O /tmp/pfd_u89wzk2r/7 --split-files --gzip SRR5272679\n",
      "Read 5699764 spots for SRR5272679\n",
      "Written 5699764 spots for SRR5272679\n",
      "Read 5699764 spots for SRR5272679\n",
      "Written 5699764 spots for SRR5272679\n",
      "Read 5699764 spots for SRR5272679\n",
      "Written 5699764 spots for SRR5272679\n",
      "Read 5699764 spots for SRR5272679\n",
      "Written 5699764 spots for SRR5272679\n",
      "Read 5699764 spots for SRR5272679\n",
      "Written 5699764 spots for SRR5272679\n",
      "Read 5699764 spots for SRR5272679\n",
      "Written 5699764 spots for SRR5272679\n",
      "Read 5699765 spots for SRR5272679\n",
      "Written 5699765 spots for SRR5272679\n",
      "Read 5699764 spots for SRR5272679\n",
      "Written 5699764 spots for SRR5272679\n"
     ]
    }
   ],
   "source": [
    "# ! prefetch --option-file SRR_Acc_List.txt -slow\n",
    "\n",
    "for acc in dat:\n",
    "    ! parallel-fastq-dump --sra-id {acc} --threads 8 --split-files --gzip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 66000240\r\n",
      "      4 -rw-rw-r-- 1 cv cv       2254 Feb 22 17:46 SraRunTable.txt\r\n",
      "4376716 -rw-r--r-- 1 cv cv 4481751113 Feb 22 17:36 \u001b[0m\u001b[01;31mSRR5272674_1.fastq.gz\u001b[0m\r\n",
      "4281988 -rw-r--r-- 1 cv cv 4384747859 Feb 22 17:36 \u001b[01;31mSRR5272674_2.fastq.gz\u001b[0m\r\n",
      "5812284 -rw-r--r-- 1 cv cv 5951771178 Feb 22 18:24 \u001b[01;31mSRR5272675_1.fastq.gz\u001b[0m\r\n",
      "5739344 -rw-r--r-- 1 cv cv 5877081615 Feb 22 18:24 \u001b[01;31mSRR5272675_2.fastq.gz\u001b[0m\r\n",
      "7181148 -rw-r--r-- 1 cv cv 7353489858 Feb 22 19:30 \u001b[01;31mSRR5272676_1.fastq.gz\u001b[0m\r\n",
      "7075764 -rw-r--r-- 1 cv cv 7245574218 Feb 22 19:30 \u001b[01;31mSRR5272676_2.fastq.gz\u001b[0m\r\n",
      "5765128 -rw-r--r-- 1 cv cv 5903484559 Feb 22 20:25 \u001b[01;31mSRR5272677_1.fastq.gz\u001b[0m\r\n",
      "5704472 -rw-r--r-- 1 cv cv 5841371230 Feb 22 20:25 \u001b[01;31mSRR5272677_2.fastq.gz\u001b[0m\r\n",
      "6096444 -rw-r--r-- 1 cv cv 6242753659 Feb 22 21:20 \u001b[01;31mSRR5272678_1.fastq.gz\u001b[0m\r\n",
      "6041056 -rw-r--r-- 1 cv cv 6186033882 Feb 22 21:20 \u001b[01;31mSRR5272678_2.fastq.gz\u001b[0m\r\n",
      "4019336 -rw-r--r-- 1 cv cv 4115793689 Feb 22 22:11 \u001b[01;31mSRR5272679_1.fastq.gz\u001b[0m\r\n",
      "3906552 -rw-r--r-- 1 cv cv 4000304857 Feb 22 22:12 \u001b[01;31mSRR5272679_2.fastq.gz\u001b[0m\r\n",
      "      4 -rw-rw-r-- 1 cv cv         65 Feb 22 15:59 SRR_Acc_List.txt\r\n"
     ]
    }
   ],
   "source": [
    "ls -ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Cell_Line</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>Consent</th>\n",
       "      <th>...</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>LibrarySelection</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Platform</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>source_name</th>\n",
       "      <th>SRA Study</th>\n",
       "      <th>treated_with</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR5272674</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>202</td>\n",
       "      <td>10291139874</td>\n",
       "      <td>PRJNA375938</td>\n",
       "      <td>SAMN06348087</td>\n",
       "      <td>6378448380</td>\n",
       "      <td>KMS-12</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>...</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2018-09-11T00:00:00Z</td>\n",
       "      <td>GSM2495761</td>\n",
       "      <td>BM_DM</td>\n",
       "      <td>SRP100328</td>\n",
       "      <td>Untreated (DMSO)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR5272675</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>202</td>\n",
       "      <td>12639445828</td>\n",
       "      <td>PRJNA375938</td>\n",
       "      <td>SAMN06348086</td>\n",
       "      <td>8546403427</td>\n",
       "      <td>KMS-12</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>...</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2018-09-11T00:00:00Z</td>\n",
       "      <td>GSM2495762</td>\n",
       "      <td>BM_DM</td>\n",
       "      <td>SRP100328</td>\n",
       "      <td>Untreated (DMSO)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR5272676</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>202</td>\n",
       "      <td>15737086538</td>\n",
       "      <td>PRJNA375938</td>\n",
       "      <td>SAMN06348097</td>\n",
       "      <td>10552387868</td>\n",
       "      <td>KMS-12</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>...</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2018-09-11T00:00:00Z</td>\n",
       "      <td>GSM2495763</td>\n",
       "      <td>BM_DM</td>\n",
       "      <td>SRP100328</td>\n",
       "      <td>Untreated (DMSO)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR5272677</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>202</td>\n",
       "      <td>12268203562</td>\n",
       "      <td>PRJNA375938</td>\n",
       "      <td>SAMN06348096</td>\n",
       "      <td>8495701271</td>\n",
       "      <td>KMS-12</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>...</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2018-09-11T00:00:00Z</td>\n",
       "      <td>GSM2495764</td>\n",
       "      <td>BM_AMIL</td>\n",
       "      <td>SRP100328</td>\n",
       "      <td>Amiloride 0.1 mM 24h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR5272678</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>202</td>\n",
       "      <td>13336923750</td>\n",
       "      <td>PRJNA375938</td>\n",
       "      <td>SAMN06348095</td>\n",
       "      <td>8977732893</td>\n",
       "      <td>KMS-12</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>...</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2018-09-11T00:00:00Z</td>\n",
       "      <td>GSM2495765</td>\n",
       "      <td>BM_AMIL</td>\n",
       "      <td>SRP100328</td>\n",
       "      <td>Amiloride 0.1 mM 24h</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR5272679</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>202</td>\n",
       "      <td>9210818826</td>\n",
       "      <td>PRJNA375938</td>\n",
       "      <td>SAMN06348094</td>\n",
       "      <td>5849632492</td>\n",
       "      <td>KMS-12</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>...</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2018-09-11T00:00:00Z</td>\n",
       "      <td>GSM2495766</td>\n",
       "      <td>BM_AMIL</td>\n",
       "      <td>SRP100328</td>\n",
       "      <td>Amiloride 0.1 mM 24h</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows Ã— 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Run Assay Type  AvgSpotLen        Bases   BioProject     BioSample  \\\n",
       "0  SRR5272674    RNA-Seq         202  10291139874  PRJNA375938  SAMN06348087   \n",
       "1  SRR5272675    RNA-Seq         202  12639445828  PRJNA375938  SAMN06348086   \n",
       "2  SRR5272676    RNA-Seq         202  15737086538  PRJNA375938  SAMN06348097   \n",
       "3  SRR5272677    RNA-Seq         202  12268203562  PRJNA375938  SAMN06348096   \n",
       "4  SRR5272678    RNA-Seq         202  13336923750  PRJNA375938  SAMN06348095   \n",
       "5  SRR5272679    RNA-Seq         202   9210818826  PRJNA375938  SAMN06348094   \n",
       "\n",
       "         Bytes Cell_Line Center Name Consent  ... LibraryLayout  \\\n",
       "0   6378448380    KMS-12         GEO  public  ...        PAIRED   \n",
       "1   8546403427    KMS-12         GEO  public  ...        PAIRED   \n",
       "2  10552387868    KMS-12         GEO  public  ...        PAIRED   \n",
       "3   8495701271    KMS-12         GEO  public  ...        PAIRED   \n",
       "4   8977732893    KMS-12         GEO  public  ...        PAIRED   \n",
       "5   5849632492    KMS-12         GEO  public  ...        PAIRED   \n",
       "\n",
       "  LibrarySelection   LibrarySource      Organism  Platform  \\\n",
       "0             cDNA  TRANSCRIPTOMIC  Homo sapiens  ILLUMINA   \n",
       "1             cDNA  TRANSCRIPTOMIC  Homo sapiens  ILLUMINA   \n",
       "2             cDNA  TRANSCRIPTOMIC  Homo sapiens  ILLUMINA   \n",
       "3             cDNA  TRANSCRIPTOMIC  Homo sapiens  ILLUMINA   \n",
       "4             cDNA  TRANSCRIPTOMIC  Homo sapiens  ILLUMINA   \n",
       "5             cDNA  TRANSCRIPTOMIC  Homo sapiens  ILLUMINA   \n",
       "\n",
       "            ReleaseDate Sample Name source_name  SRA Study  \\\n",
       "0  2018-09-11T00:00:00Z  GSM2495761       BM_DM  SRP100328   \n",
       "1  2018-09-11T00:00:00Z  GSM2495762       BM_DM  SRP100328   \n",
       "2  2018-09-11T00:00:00Z  GSM2495763       BM_DM  SRP100328   \n",
       "3  2018-09-11T00:00:00Z  GSM2495764     BM_AMIL  SRP100328   \n",
       "4  2018-09-11T00:00:00Z  GSM2495765     BM_AMIL  SRP100328   \n",
       "5  2018-09-11T00:00:00Z  GSM2495766     BM_AMIL  SRP100328   \n",
       "\n",
       "           treated_with  \n",
       "0      Untreated (DMSO)  \n",
       "1      Untreated (DMSO)  \n",
       "2      Untreated (DMSO)  \n",
       "3  Amiloride 0.1 mM 24h  \n",
       "4  Amiloride 0.1 mM 24h  \n",
       "5  Amiloride 0.1 mM 24h  \n",
       "\n",
       "[6 rows x 27 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_dat = pd.read_csv(\"SraRunTable.txt\")\n",
    "meta_dat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download reference genome\n",
    "\n",
    "Homo sapiens NCBI assembly GRCh38 \n",
    "https://support.illumina.com/sequencing/sequencing_software/igenome.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Homo_sapiens/NCBI/GRCh38/\n",
      "Homo_sapiens/NCBI/GRCh38/Annotation/\n",
      "Homo_sapiens/NCBI/GRCh38/Annotation/Genes\n",
      "Homo_sapiens/NCBI/GRCh38/Annotation/SmallRNA\n",
      "Homo_sapiens/NCBI/GRCh38/Annotation/Genes.gencode\n",
      "Homo_sapiens/NCBI/GRCh38/Annotation/Archives/\n",
      "Homo_sapiens/NCBI/GRCh38/Annotation/Archives/archive-2015-08-11-09-31-31/\n",
      "Homo_sapiens/NCBI/GRCh38/Annotation/Archives/archive-2015-08-11-09-31-31/Genes/\n",
      "Homo_sapiens/NCBI/GRCh38/Annotation/Archives/archive-2015-08-11-09-31-31/Genes/genes.gtf\n",
      "Homo_sapiens/NCBI/GRCh38/Annotation/Archives/archive-2015-08-11-09-31-31/Genes/refFlat.txt.gz\n",
      "Homo_sapiens/NCBI/GRCh38/Annotation/Archives/archive-2015-08-11-09-31-31/README.txt\n",
      "Homo_sapiens/NCBI/GRCh38/Annotation/Archives/archive-2015-08-11-09-31-31/SmallRNA/\n",
      "Homo_sapiens/NCBI/GRCh38/Annotation/Archives/archive-2015-08-11-09-31-31/SmallRNA/mature.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Annotation/Archives/archive-2015-08-11-09-31-31/SmallRNA/hairpin.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Annotation/Archives/archive-2015-08-11-09-31-31/Genes.gencode/\n",
      "Homo_sapiens/NCBI/GRCh38/Annotation/Archives/archive-2015-08-11-09-31-31/Genes.gencode/genes.gtf\n",
      "Homo_sapiens/NCBI/GRCh38/Annotation/Archives/archive-2015-08-11-09-31-31/Genes.gencode/refFlat.txt.gz\n",
      "Homo_sapiens/NCBI/GRCh38/Annotation/Archives/archive-current\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Bowtie2Index/\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Bowtie2Index/genome.3.bt2\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Bowtie2Index/genome.1.bt2\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Bowtie2Index/genome.rev.2.bt2\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Bowtie2Index/genome.rev.1.bt2\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Bowtie2Index/genome.4.bt2\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Bowtie2Index/genome.2.bt2\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Bowtie2Index/genome.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/genome.fa.amb\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/genome.fa.sa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/version0.5.x/\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/version0.5.x/genome.fa.amb\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/version0.5.x/genome.fa.sa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/version0.5.x/genome.fa.bwt\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/version0.5.x/genome.fa.ann\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/version0.5.x/genome.fa.rbwt\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/version0.5.x/genome.fa.pac\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/version0.5.x/genome.fa.rpac\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/version0.5.x/genome.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/version0.5.x/genome.fa.rsa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/version0.6.0/\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/version0.6.0/genome.fa.amb\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/version0.6.0/genome.fa.sa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/version0.6.0/genome.fa.bwt\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/version0.6.0/genome.fa.ann\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/version0.6.0/genome.fa.pac\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/version0.6.0/genome.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/genome.fa.bwt\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/genome.fa.ann\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/genome.fa.pac\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BWAIndex/genome.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270388v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr22_KI270731v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270442v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrY.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr17_GL000205v2_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_GL000195v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270320v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr1_KI270714v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270753v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270393v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270516v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270743v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270391v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270528v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270757v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr9_KI270717v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270746v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270425v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270544v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_GL000216v2.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270303v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr21.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr5.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr1_KI270712v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270330v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270581v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr3.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr4_GL000008v2_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270588v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr14_KI270725v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr2.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270362v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270396v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270394v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr9_KI270718v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270752v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_GL000214v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270375v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_GL000213v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270467v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270310v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270747v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr11_KI270721v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr14_KI270724v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270395v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270338v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270336v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270417v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr22_KI270734v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrEBV.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr6.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270590v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270529v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270538v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270311v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr1_KI270710v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270412v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270389v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr16.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270591v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr22_KI270737v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_GL000220v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270316v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270748v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270304v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270548v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr1_KI270707v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270744v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr16_KI270728v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270580v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270422v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270329v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270312v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_GL000226v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270435v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270522v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr1_KI270711v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270305v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr20.fa\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270392v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270376v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270378v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270383v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr17_KI270729v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr14_KI270723v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270385v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270466v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270379v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr17_KI270730v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270366v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr15.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270374v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr12.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270517v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270515v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270381v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270364v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270414v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270539v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr14_KI270726v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270530v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270337v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrM.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270371v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr14_GL000225v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr15_KI270727v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr22_KI270735v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270448v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr4.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr2_KI270716v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270317v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr9_KI270719v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr14_GL000194v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr22_KI270739v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270751v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270584v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270340v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr3_GL000221v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr9.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270384v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270429v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr2_KI270715v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270372v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr18.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270519v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr1_KI270708v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270518v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270438v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270510v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr22_KI270736v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270509v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270507v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr5_GL000208v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270419v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270754v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270335v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270508v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270334v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr10.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270315v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270423v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270742v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr1_KI270713v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr22.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr14.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270511v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270593v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270579v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270468v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270583v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr9_KI270720v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrX.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270373v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270755v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr1_KI270709v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270521v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270756v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270589v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270411v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr11.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr13.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270386v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr22_KI270733v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270750v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr19.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270749v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270512v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr8.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr22_KI270732v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_GL000218v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270418v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrY_KI270740v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr17.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr14_GL000009v2_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270390v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270322v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270302v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270424v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr14_KI270722v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270741v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270745v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270363v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270465v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr7.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270587v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270582v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr1_KI270706v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270333v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270420v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_GL000219v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_GL000224v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chr22_KI270738v1_random.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270387v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/Chromosomes/chrUn_KI270382v1.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BowtieIndex/\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BowtieIndex/genome.2.ebwt\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BowtieIndex/genome.rev.2.ebwt\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BowtieIndex/genome.1.ebwt\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BowtieIndex/genome.3.ebwt\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BowtieIndex/genome.rev.1.ebwt\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BowtieIndex/genome.fa\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/BowtieIndex/genome.4.ebwt\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/WholeGenomeFasta/\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/WholeGenomeFasta/GenomeSize.xml\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/WholeGenomeFasta/genome.dict\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/WholeGenomeFasta/GenomeSize.xml.old\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/WholeGenomeFasta/genome.dict.old\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/WholeGenomeFasta/genome.fa.fai\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/WholeGenomeFasta/genome.fa\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Homo_sapiens/NCBI/GRCh38/Sequence/AbundantSequences/\r\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/AbundantSequences/polyA.fa\r\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/AbundantSequences/hum5SrDNA.fa\r\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/AbundantSequences/chrM.fa\r\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/AbundantSequences/polyC.fa\r\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/AbundantSequences/humRibosomal.fa\r\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/AbundantSequences/adapter_contam1.fa\r\n",
      "Homo_sapiens/NCBI/GRCh38/Sequence/AbundantSequences/phix.fa\r\n",
      "README.txt\r\n"
     ]
    }
   ],
   "source": [
    "! tar zxvf Homo_sapiens_NCBI_GRCh38.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data preprocessing with fastp\n",
    "\n",
    "fastp - fast all-in-one preprocessing for fastq files; performs quality control, adapter trimming, quality filtering, per-read quality pruning and many other operations with a single scan of the fastq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir preprocessing_fastp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: fastp --in1=string [options] ... \r\n",
      "options:\r\n",
      "  -i, --in1                            read1 input file name (string)\r\n",
      "  -o, --out1                           read1 output file name (string [=])\r\n",
      "  -I, --in2                            read2 input file name (string [=])\r\n",
      "  -O, --out2                           read2 output file name (string [=])\r\n",
      "  -6, --phred64                        indicates the input is using phred64 scoring (it'll be converted to phred33, so the output will still be phred33)\r\n",
      "  -z, --compression                    compression level for gzip output (1 ~ 9). 1 is fastest, 9 is smallest, default is 2. (int [=2])\r\n",
      "  -A, --disable_adapter_trimming       adapter trimming is enabled by default. If this option is specified, adapter trimming is disabled\r\n",
      "  -a, --adapter_sequence               the adapter for read1. For SE data, if not specified, the adapter will be auto-detected. For PE data, this is used if R1/R2 are found not overlapped. (string [=auto])\r\n",
      "      --adapter_sequence_r2            the adapter for read2 (PE data only). This is used if R1/R2 are found not overlapped. If not specified, it will be the same as <adapter_sequence> (string [=])\r\n",
      "  -f, --trim_front1                    trimming how many bases in front for read1, default is 0 (int [=0])\r\n",
      "  -t, --trim_tail1                     trimming how many bases in tail for read1, default is 0 (int [=0])\r\n",
      "  -F, --trim_front2                    trimming how many bases in front for read2. If it's not specified, it will follow read1's settings (int [=0])\r\n",
      "  -T, --trim_tail2                     trimming how many bases in tail for read2. If it's not specified, it will follow read1's settings (int [=0])\r\n",
      "  -g, --trim_poly_g                    force polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data\r\n",
      "      --poly_g_min_len                 the minimum length to detect polyG in the read tail. 10 by default. (int [=10])\r\n",
      "  -G, --disable_trim_poly_g            disable polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data\r\n",
      "  -5, --cut_by_quality5                enable per read cutting by quality in front (5'), default is disabled (WARNING: this will interfere deduplication for both PE/SE data)\r\n",
      "  -3, --cut_by_quality3                enable per read cutting by quality in tail (3'), default is disabled (WARNING: this will interfere deduplication for SE data)\r\n",
      "  -W, --cut_window_size                the size of the sliding window for sliding window trimming, default is 4 (int [=4])\r\n",
      "  -M, --cut_mean_quality               the bases in the sliding window with mean quality below cutting_quality will be cut, default is Q20 (int [=20])\r\n",
      "  -Q, --disable_quality_filtering      quality filtering is enabled by default. If this option is specified, quality filtering is disabled\r\n",
      "  -q, --qualified_quality_phred        the quality value that a base is qualified. Default 15 means phred quality >=Q15 is qualified. (int [=15])\r\n",
      "  -u, --unqualified_percent_limit      how many percents of bases are allowed to be unqualified (0~100). Default 40 means 40% (int [=40])\r\n",
      "  -n, --n_base_limit                   if one read's number of N base is >n_base_limit, then this read/pair is discarded. Default is 5 (int [=5])\r\n",
      "  -L, --disable_length_filtering       length filtering is enabled by default. If this option is specified, length filtering is disabled\r\n",
      "  -l, --length_required                reads shorter than length_required will be discarded, default is 15. (int [=15])\r\n",
      "  -c, --correction                     enable base correction in overlapped regions (only for PE data), default is disabled\r\n",
      "  -U, --umi                            enable unique molecular identifer (UMI) preprocessing\r\n",
      "      --umi_loc                        specify the location of UMI, can be (index1/index2/read1/read2/per_index/per_read, default is none (string [=])\r\n",
      "      --umi_len                        if the UMI is in read1/read2, its length should be provided (int [=0])\r\n",
      "      --umi_prefix                     if specified, an underline will be used to connect prefix and UMI (i.e. prefix=UMI, UMI=AATTCG, final=UMI_AATTCG). No prefix by default (string [=])\r\n",
      "  -p, --overrepresentation_analysis    enable overrepresented sequence analysis.\r\n",
      "  -P, --overrepresentation_sampling    one in (--overrepresentation_sampling) reads will be computed for overrepresentation analysis (1~10000), smaller is slower, default is 20. (int [=20])\r\n",
      "  -j, --json                           the json format report file name (string [=fastp.json])\r\n",
      "  -h, --html                           the html format report file name (string [=fastp.html])\r\n",
      "  -R, --report_title                   should be quoted with ' or \", default is \"fastp report\" (string [=fastp report])\r\n",
      "  -w, --thread                         worker thread number, default is 3 (int [=3])\r\n",
      "  -s, --split                          split output by limiting total split file number with this option (2~999), a sequential number prefix will be added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (int [=0])\r\n",
      "  -S, --split_by_lines                 split output by limiting lines of each file with this option(>=1000), a sequential number prefix will be added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (long [=0])\r\n",
      "  -d, --split_prefix_digits            the digits for the sequential number padding (1~10), default is 4, so the filename will be padded as 0001.xxx, 0 to disable padding (int [=4])\r\n",
      "  -?, --help                           print this message\r\n"
     ]
    }
   ],
   "source": [
    "! fastp --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read1 before filtering:\n",
      "total reads: 50946237\n",
      "total bases: 5145569937\n",
      "Q20 bases: 5000489942(97.1805%)\n",
      "Q30 bases: 4818750530(93.6485%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 48290890\n",
      "total bases: 4775433321\n",
      "Q20 bases: 4709426319(98.6178%)\n",
      "Q30 bases: 4555062674(95.3853%)\n",
      "\n",
      "Read2 before filtering:\n",
      "total reads: 50946237\n",
      "total bases: 5145569937\n",
      "Q20 bases: 4929215966(95.7953%)\n",
      "Q30 bases: 4756665433(92.442%)\n",
      "\n",
      "Read2 aftering filtering:\n",
      "total reads: 48290890\n",
      "total bases: 4775433321\n",
      "Q20 bases: 4720013684(98.8395%)\n",
      "Q30 bases: 4573445833(95.7703%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 96581780\n",
      "reads failed due to low quality: 5244928\n",
      "reads failed due to too many N: 65766\n",
      "reads failed due to too short: 0\n",
      "reads with adapter trimmed: 14197428\n",
      "bases trimmed due to adapters: 210919688\n",
      "\n",
      "JSON report: preprocessing_fastp/SRR5272674.json\n",
      "HTML report: preprocessing_fastp/SRR5272674.html\n",
      "\n",
      "fastp -i SRR5272674_1.fastq.gz -I SRR5272674_2.fastq.gz -o preprocessing_fastp/SRR5272674_1.fq.gz -O preprocessing_fastp/SRR5272674_2.fq.gz -w 8 -h preprocessing_fastp/SRR5272674.html -j preprocessing_fastp/SRR5272674.json \n",
      "fastp v0.12.4, time used: 588 seconds\n",
      "Read1 before filtering:\n",
      "total reads: 62571514\n",
      "total bases: 6319722914\n",
      "Q20 bases: 6032300530(95.452%)\n",
      "Q30 bases: 5663654573(89.6187%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 58312910\n",
      "total bases: 5845926569\n",
      "Q20 bases: 5707508400(97.6322%)\n",
      "Q30 bases: 5395029102(92.287%)\n",
      "\n",
      "Read2 before filtering:\n",
      "total reads: 62571514\n",
      "total bases: 6319722914\n",
      "Q20 bases: 5952011622(94.1815%)\n",
      "Q30 bases: 5578426246(88.2701%)\n",
      "\n",
      "Read2 aftering filtering:\n",
      "total reads: 58312910\n",
      "total bases: 5845926569\n",
      "Q20 bases: 5715806901(97.7742%)\n",
      "Q30 bases: 5390350224(92.2069%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 116625820\n",
      "reads failed due to low quality: 8450224\n",
      "reads failed due to too many N: 66984\n",
      "reads failed due to too short: 0\n",
      "reads with adapter trimmed: 7239672\n",
      "bases trimmed due to adapters: 92928380\n",
      "\n",
      "JSON report: preprocessing_fastp/SRR5272675.json\n",
      "HTML report: preprocessing_fastp/SRR5272675.html\n",
      "\n",
      "fastp -i SRR5272675_1.fastq.gz -I SRR5272675_2.fastq.gz -o preprocessing_fastp/SRR5272675_1.fq.gz -O preprocessing_fastp/SRR5272675_2.fq.gz -w 8 -h preprocessing_fastp/SRR5272675.html -j preprocessing_fastp/SRR5272675.json \n",
      "fastp v0.12.4, time used: 739 seconds\n",
      "Read1 before filtering:\n",
      "total reads: 77906369\n",
      "total bases: 7868543269\n",
      "Q20 bases: 7526909164(95.6582%)\n",
      "Q30 bases: 7092117821(90.1325%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 72963018\n",
      "total bases: 7295047383\n",
      "Q20 bases: 7130505712(97.7445%)\n",
      "Q30 bases: 6759985405(92.6654%)\n",
      "\n",
      "Read2 before filtering:\n",
      "total reads: 77906369\n",
      "total bases: 7868543269\n",
      "Q20 bases: 7459728819(94.8044%)\n",
      "Q30 bases: 7028471117(89.3237%)\n",
      "\n",
      "Read2 aftering filtering:\n",
      "total reads: 72963018\n",
      "total bases: 7295047383\n",
      "Q20 bases: 7150569078(98.0195%)\n",
      "Q30 bases: 6776421984(92.8907%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 145926036\n",
      "reads failed due to low quality: 9802654\n",
      "reads failed due to too many N: 84048\n",
      "reads failed due to too short: 0\n",
      "reads with adapter trimmed: 12029326\n",
      "bases trimmed due to adapters: 157090696\n",
      "\n",
      "JSON report: preprocessing_fastp/SRR5272676.json\n",
      "HTML report: preprocessing_fastp/SRR5272676.html\n",
      "\n",
      "fastp -i SRR5272676_1.fastq.gz -I SRR5272676_2.fastq.gz -o preprocessing_fastp/SRR5272676_1.fq.gz -O preprocessing_fastp/SRR5272676_2.fq.gz -w 8 -h preprocessing_fastp/SRR5272676.html -j preprocessing_fastp/SRR5272676.json \n",
      "fastp v0.12.4, time used: 933 seconds\n",
      "Read1 before filtering:\n",
      "total reads: 60733681\n",
      "total bases: 6134101781\n",
      "Q20 bases: 5792556154(94.432%)\n",
      "Q30 bases: 5379087540(87.6915%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 55261562\n",
      "total bases: 5483525039\n",
      "Q20 bases: 5332903067(97.2532%)\n",
      "Q30 bases: 5000480199(91.191%)\n",
      "\n",
      "Read2 before filtering:\n",
      "total reads: 60733681\n",
      "total bases: 6134101781\n",
      "Q20 bases: 5651939898(92.1396%)\n",
      "Q30 bases: 5203284869(84.8255%)\n",
      "\n",
      "Read2 aftering filtering:\n",
      "total reads: 55261562\n",
      "total bases: 5483525039\n",
      "Q20 bases: 5321468956(97.0447%)\n",
      "Q30 bases: 4944439635(90.169%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 110523124\n",
      "reads failed due to low quality: 10873116\n",
      "reads failed due to too many N: 71122\n",
      "reads failed due to too short: 0\n",
      "reads with adapter trimmed: 14885612\n",
      "bases trimmed due to adapters: 211602000\n",
      "\n",
      "JSON report: preprocessing_fastp/SRR5272677.json\n",
      "HTML report: preprocessing_fastp/SRR5272677.html\n",
      "\n",
      "fastp -i SRR5272677_1.fastq.gz -I SRR5272677_2.fastq.gz -o preprocessing_fastp/SRR5272677_1.fq.gz -O preprocessing_fastp/SRR5272677_2.fq.gz -w 8 -h preprocessing_fastp/SRR5272677.html -j preprocessing_fastp/SRR5272677.json \n",
      "fastp v0.12.4, time used: 713 seconds\n",
      "Read1 before filtering:\n",
      "total reads: 66024375\n",
      "total bases: 6668461875\n",
      "Q20 bases: 6379730594(95.6702%)\n",
      "Q30 bases: 6009627863(90.1202%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 61849224\n",
      "total bases: 6167171830\n",
      "Q20 bases: 6029940004(97.7748%)\n",
      "Q30 bases: 5716973653(92.7001%)\n",
      "\n",
      "Read2 before filtering:\n",
      "total reads: 66024375\n",
      "total bases: 6668461875\n",
      "Q20 bases: 6313729243(94.6804%)\n",
      "Q30 bases: 5934745030(88.9972%)\n",
      "\n",
      "Read2 aftering filtering:\n",
      "total reads: 61849224\n",
      "total bases: 6167171830\n",
      "Q20 bases: 6038188563(97.9086%)\n",
      "Q30 bases: 5710636830(92.5973%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 123698448\n",
      "reads failed due to low quality: 8278042\n",
      "reads failed due to too many N: 72260\n",
      "reads failed due to too short: 0\n",
      "reads with adapter trimmed: 12062368\n",
      "bases trimmed due to adapters: 168609468\n",
      "\n",
      "JSON report: preprocessing_fastp/SRR5272678.json\n",
      "HTML report: preprocessing_fastp/SRR5272678.html\n",
      "\n",
      "fastp -i SRR5272678_1.fastq.gz -I SRR5272678_2.fastq.gz -o preprocessing_fastp/SRR5272678_1.fq.gz -O preprocessing_fastp/SRR5272678_2.fq.gz -w 8 -h preprocessing_fastp/SRR5272678.html -j preprocessing_fastp/SRR5272678.json \n",
      "fastp v0.12.4, time used: 790 seconds\n",
      "Read1 before filtering:\n",
      "total reads: 45598113\n",
      "total bases: 4605409413\n",
      "Q20 bases: 4503556884(97.7884%)\n",
      "Q30 bases: 4334526500(94.1182%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 43980643\n",
      "total bases: 4413739413\n",
      "Q20 bases: 4353023611(98.6244%)\n",
      "Q30 bases: 4199927593(95.1558%)\n",
      "\n",
      "Read2 before filtering:\n",
      "total reads: 45598113\n",
      "total bases: 4605409413\n",
      "Q20 bases: 4453520793(96.702%)\n",
      "Q30 bases: 4305243961(93.4823%)\n",
      "\n",
      "Read2 aftering filtering:\n",
      "total reads: 43980643\n",
      "total bases: 4413739413\n",
      "Q20 bases: 4362779100(98.8454%)\n",
      "Q30 bases: 4227897947(95.7895%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 87961286\n",
      "reads failed due to low quality: 3131592\n",
      "reads failed due to too many N: 103348\n",
      "reads failed due to too short: 0\n",
      "reads with adapter trimmed: 5012196\n",
      "bases trimmed due to adapters: 57940838\n",
      "\n",
      "JSON report: preprocessing_fastp/SRR5272679.json\n",
      "HTML report: preprocessing_fastp/SRR5272679.html\n",
      "\n",
      "fastp -i SRR5272679_1.fastq.gz -I SRR5272679_2.fastq.gz -o preprocessing_fastp/SRR5272679_1.fq.gz -O preprocessing_fastp/SRR5272679_2.fq.gz -w 8 -h preprocessing_fastp/SRR5272679.html -j preprocessing_fastp/SRR5272679.json \n",
      "fastp v0.12.4, time used: 557 seconds\n"
     ]
    }
   ],
   "source": [
    "for acc in dat:\n",
    "    ! fastp -i {acc}_1.fastq.gz -I {acc}_2.fastq.gz \\\n",
    "            -o preprocessing_fastp/{acc}_1.fq.gz -O preprocessing_fastp/{acc}_2.fq.gz \\\n",
    "            -w 8 -h preprocessing_fastp/{acc}.html -j preprocessing_fastp/{acc}.json "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 69093780\r\n",
      "4640140 -rw-r--r-- 1 cv cv 4751498164 Feb 24 15:28 \u001b[0m\u001b[01;31mSRR5272674_1.fq.gz\u001b[0m\r\n",
      "4581852 -rw-r--r-- 1 cv cv 4691808539 Feb 24 15:28 \u001b[01;31mSRR5272674_2.fq.gz\u001b[0m\r\n",
      "    452 -rw-r--r-- 1 cv cv     461445 Feb 24 15:28 SRR5272674.html\r\n",
      "    120 -rw-r--r-- 1 cv cv     118986 Feb 24 15:28 SRR5272674.json\r\n",
      "6033156 -rw-r--r-- 1 cv cv 6177947454 Feb 24 15:40 \u001b[01;31mSRR5272675_1.fq.gz\u001b[0m\r\n",
      "6017440 -rw-r--r-- 1 cv cv 6161851804 Feb 24 15:40 \u001b[01;31mSRR5272675_2.fq.gz\u001b[0m\r\n",
      "    452 -rw-r--r-- 1 cv cv     461273 Feb 24 15:40 SRR5272675.html\r\n",
      "    120 -rw-r--r-- 1 cv cv     118873 Feb 24 15:40 SRR5272675.json\r\n",
      "7490216 -rw-r--r-- 1 cv cv 7669973380 Feb 24 15:56 \u001b[01;31mSRR5272676_1.fq.gz\u001b[0m\r\n",
      "7446176 -rw-r--r-- 1 cv cv 7624877199 Feb 24 15:56 \u001b[01;31mSRR5272676_2.fq.gz\u001b[0m\r\n",
      "    452 -rw-r--r-- 1 cv cv     461835 Feb 24 15:56 SRR5272676.html\r\n",
      "    120 -rw-r--r-- 1 cv cv     119429 Feb 24 15:56 SRR5272676.json\r\n",
      "5786188 -rw-r--r-- 1 cv cv 5925049400 Feb 24 16:07 \u001b[01;31mSRR5272677_1.fq.gz\u001b[0m\r\n",
      "5802464 -rw-r--r-- 1 cv cv 5941716733 Feb 24 16:07 \u001b[01;31mSRR5272677_2.fq.gz\u001b[0m\r\n",
      "    452 -rw-r--r-- 1 cv cv     461406 Feb 24 16:07 SRR5272677.html\r\n",
      "    120 -rw-r--r-- 1 cv cv     118987 Feb 24 16:07 SRR5272677.json\r\n",
      "6342560 -rw-r--r-- 1 cv cv 6494775609 Feb 24 16:21 \u001b[01;31mSRR5272678_1.fq.gz\u001b[0m\r\n",
      "6334988 -rw-r--r-- 1 cv cv 6487020692 Feb 24 16:21 \u001b[01;31mSRR5272678_2.fq.gz\u001b[0m\r\n",
      "    452 -rw-r--r-- 1 cv cv     461570 Feb 24 16:21 SRR5272678.html\r\n",
      "    120 -rw-r--r-- 1 cv cv     119135 Feb 24 16:21 SRR5272678.json\r\n",
      "4346924 -rw-r--r-- 1 cv cv 4451245964 Feb 24 16:30 \u001b[01;31mSRR5272679_1.fq.gz\u001b[0m\r\n",
      "4268248 -rw-r--r-- 1 cv cv 4370681213 Feb 24 16:30 \u001b[01;31mSRR5272679_2.fq.gz\u001b[0m\r\n",
      "    452 -rw-r--r-- 1 cv cv     460645 Feb 24 16:30 SRR5272679.html\r\n",
      "    116 -rw-r--r-- 1 cv cv     118396 Feb 24 16:30 SRR5272679.json\r\n"
     ]
    }
   ],
   "source": [
    "ls -ls preprocessing_fastp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/cv/RNA-seq/data/preprocessing_fastp\n"
     ]
    }
   ],
   "source": [
    "cd preprocessing_fastp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "! ln -s ../Homo_sapiens/NCBI/GRCh38/Annotation/Genes/genes.gtf ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "! ln -s ../Homo_sapiens/NCBI/GRCh38/Sequence/Bowtie2Index/genome.* ."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RNA-seq reads alignment to reference genome with TopHat\n",
    "\n",
    "TopHat uses Bowtie as the alignment engine for short reads and requires samtools as it outputs and handles alignments in binary alignment/mapping (BAM) format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TopHat v2.1.1\r\n"
     ]
    }
   ],
   "source": [
    "! tophat --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/cv/bin/bowtie2-align-s version 2.4.1\r\n",
      "64-bit\r\n",
      "Built on \r\n",
      "Fri Feb 28 22:21:25 UTC 2020\r\n",
      "Compiler: gcc version 7.3.1 20180303 (Red Hat 7.3.1-5) (GCC) \r\n",
      "Options: -O3 -msse2 -funroll-loops -g3 -g -O2 -fvisibility=hidden -I/hbb_exe_gc_hardened/include -ffunction-sections -fdata-sections -fstack-protector -D_FORTIFY_SOURCE=2 -fPIE -DPOPCNT_CAPABILITY -DWITH_TBB -std=c++11 -DNO_SPINLOCK -DWITH_QUEUELOCK=1\r\n",
      "Sizeof {int, long, long long, void*, size_t, off_t}: {4, 8, 8, 8, 8, 8}\r\n"
     ]
    }
   ],
   "source": [
    "! bowtie2 --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools 1.9\r\n",
      "Using htslib 1.9\r\n",
      "Copyright (C) 2018 Genome Research Ltd.\r\n"
     ]
    }
   ],
   "source": [
    "! samtools --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir -p ../../analysis/tophat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[2022-02-24 21:29:47] Beginning TopHat run (v2.1.1)\n",
      "-----------------------------------------------\n",
      "[2022-02-24 21:29:47] Checking for Bowtie\n",
      "\t\t  Bowtie version:\t 2.4.1.0\n",
      "[2022-02-24 21:29:47] Checking for Bowtie index files (genome)..\n",
      "[2022-02-24 21:29:47] Checking for reference FASTA file\n",
      "[2022-02-24 21:29:47] Generating SAM header for genome\n",
      "[2022-02-24 21:30:26] Reading known junctions from GTF file\n",
      "[2022-02-24 21:30:30] Preparing reads\n",
      "\t left reads: min. length=34, max. length=101, 48290497 kept reads (393 discarded)\n",
      "\tright reads: min. length=34, max. length=101, 48290404 kept reads (486 discarded)\n",
      "[2022-02-24 21:57:25] Building transcriptome data files ../../analysis/tophat/SRR5272674_thout/tmp/genes\n",
      "[2022-02-24 21:58:17] Building Bowtie index from genes.fa\n",
      "[2022-02-24 22:29:24] Mapping left_kept_reads to transcriptome genes with Bowtie2 \n",
      "[2022-02-24 23:38:56] Mapping right_kept_reads to transcriptome genes with Bowtie2 \n",
      "[2022-02-25 00:47:07] Resuming TopHat pipeline with unmapped reads\n",
      "[2022-02-25 00:47:07] Mapping left_kept_reads.m2g_um to genome genome with Bowtie2 \n",
      "[2022-02-25 01:07:55] Mapping left_kept_reads.m2g_um_seg1 to genome genome with Bowtie2 (1/4)\n",
      "[2022-02-25 01:10:19] Mapping left_kept_reads.m2g_um_seg2 to genome genome with Bowtie2 (2/4)\n",
      "[2022-02-25 01:12:28] Mapping left_kept_reads.m2g_um_seg3 to genome genome with Bowtie2 (3/4)\n",
      "[2022-02-25 01:14:39] Mapping left_kept_reads.m2g_um_seg4 to genome genome with Bowtie2 (4/4)\n",
      "[2022-02-25 01:17:04] Mapping right_kept_reads.m2g_um to genome genome with Bowtie2 \n",
      "[2022-02-25 01:36:31] Mapping right_kept_reads.m2g_um_seg1 to genome genome with Bowtie2 (1/4)\n",
      "[2022-02-25 01:38:09] Mapping right_kept_reads.m2g_um_seg2 to genome genome with Bowtie2 (2/4)\n",
      "[2022-02-25 01:39:47] Mapping right_kept_reads.m2g_um_seg3 to genome genome with Bowtie2 (3/4)\n",
      "[2022-02-25 01:41:37] Mapping right_kept_reads.m2g_um_seg4 to genome genome with Bowtie2 (4/4)\n",
      "[2022-02-25 01:43:24] Searching for junctions via segment mapping\n",
      "[2022-02-25 02:11:19] Retrieving sequences for splices\n",
      "[2022-02-25 02:12:33] Indexing splices\n",
      "Building a SMALL index\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2022-02-25 02:13:00] Mapping left_kept_reads.m2g_um_seg1 to genome segment_juncs with Bowtie2 (1/4)\n",
      "[2022-02-25 02:13:45] Mapping left_kept_reads.m2g_um_seg2 to genome segment_juncs with Bowtie2 (2/4)\n",
      "[2022-02-25 02:14:31] Mapping left_kept_reads.m2g_um_seg3 to genome segment_juncs with Bowtie2 (3/4)\n",
      "[2022-02-25 02:15:23] Mapping left_kept_reads.m2g_um_seg4 to genome segment_juncs with Bowtie2 (4/4)\n",
      "[2022-02-25 02:16:15] Joining segment hits\n",
      "[2022-02-25 02:18:18] Mapping right_kept_reads.m2g_um_seg1 to genome segment_juncs with Bowtie2 (1/4)\n",
      "[2022-02-25 02:19:09] Mapping right_kept_reads.m2g_um_seg2 to genome segment_juncs with Bowtie2 (2/4)\n",
      "[2022-02-25 02:19:57] Mapping right_kept_reads.m2g_um_seg3 to genome segment_juncs with Bowtie2 (3/4)\n",
      "[2022-02-25 02:20:51] Mapping right_kept_reads.m2g_um_seg4 to genome segment_juncs with Bowtie2 (4/4)\n",
      "[2022-02-25 02:21:44] Joining segment hits\n",
      "[2022-02-25 02:23:51] Reporting output tracks\n",
      "-----------------------------------------------\n",
      "[2022-02-25 03:14:57] A summary of the alignment counts can be found in ../../analysis/tophat/SRR5272674_thout/align_summary.txt\n",
      "[2022-02-25 03:14:57] Run complete: 05:45:09 elapsed\n",
      "\n",
      "[2022-02-25 03:15:00] Beginning TopHat run (v2.1.1)\n",
      "-----------------------------------------------\n",
      "[2022-02-25 03:15:00] Checking for Bowtie\n",
      "\t\t  Bowtie version:\t 2.4.1.0\n",
      "[2022-02-25 03:15:01] Checking for Bowtie index files (genome)..\n",
      "[2022-02-25 03:15:01] Checking for reference FASTA file\n",
      "[2022-02-25 03:15:01] Generating SAM header for genome\n",
      "[2022-02-25 03:15:42] Reading known junctions from GTF file\n",
      "[2022-02-25 03:15:47] Preparing reads\n",
      "\t left reads: min. length=34, max. length=101, 58312642 kept reads (268 discarded)\n",
      "\tright reads: min. length=34, max. length=101, 58312621 kept reads (289 discarded)\n",
      "[2022-02-25 03:46:30] Building transcriptome data files ../../analysis/tophat/SRR5272675_thout/tmp/genes\n",
      "[2022-02-25 03:47:16] Building Bowtie index from genes.fa\n",
      "[2022-02-25 04:17:35] Mapping left_kept_reads to transcriptome genes with Bowtie2 \n",
      "[2022-02-25 05:38:36] Mapping right_kept_reads to transcriptome genes with Bowtie2 \n",
      "[2022-02-25 06:59:43] Resuming TopHat pipeline with unmapped reads\n",
      "[2022-02-25 06:59:43] Mapping left_kept_reads.m2g_um to genome genome with Bowtie2 \n",
      "[2022-02-25 07:24:15] Mapping left_kept_reads.m2g_um_seg1 to genome genome with Bowtie2 (1/4)\n",
      "[2022-02-25 07:26:55] Mapping left_kept_reads.m2g_um_seg2 to genome genome with Bowtie2 (2/4)\n",
      "[2022-02-25 07:29:06] Mapping left_kept_reads.m2g_um_seg3 to genome genome with Bowtie2 (3/4)\n",
      "[2022-02-25 07:31:31] Mapping left_kept_reads.m2g_um_seg4 to genome genome with Bowtie2 (4/4)\n",
      "[2022-02-25 07:34:32] Mapping right_kept_reads.m2g_um to genome genome with Bowtie2 \n",
      "[2022-02-25 07:57:20] Mapping right_kept_reads.m2g_um_seg1 to genome genome with Bowtie2 (1/4)\n",
      "[2022-02-25 07:58:55] Mapping right_kept_reads.m2g_um_seg2 to genome genome with Bowtie2 (2/4)\n",
      "[2022-02-25 08:00:40] Mapping right_kept_reads.m2g_um_seg3 to genome genome with Bowtie2 (3/4)\n",
      "[2022-02-25 08:02:30] Mapping right_kept_reads.m2g_um_seg4 to genome genome with Bowtie2 (4/4)\n",
      "[2022-02-25 08:04:27] Searching for junctions via segment mapping\n",
      "[2022-02-25 08:32:26] Retrieving sequences for splices\n",
      "[2022-02-25 08:33:43] Indexing splices\n",
      "Building a SMALL index\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2022-02-25 08:34:12] Mapping left_kept_reads.m2g_um_seg1 to genome segment_juncs with Bowtie2 (1/4)\n",
      "[2022-02-25 08:35:01] Mapping left_kept_reads.m2g_um_seg2 to genome segment_juncs with Bowtie2 (2/4)\n",
      "[2022-02-25 08:35:54] Mapping left_kept_reads.m2g_um_seg3 to genome segment_juncs with Bowtie2 (3/4)\n",
      "[2022-02-25 08:36:52] Mapping left_kept_reads.m2g_um_seg4 to genome segment_juncs with Bowtie2 (4/4)\n",
      "[2022-02-25 08:37:54] Joining segment hits\n",
      "[2022-02-25 08:40:15] Mapping right_kept_reads.m2g_um_seg1 to genome segment_juncs with Bowtie2 (1/4)\n",
      "[2022-02-25 08:41:06] Mapping right_kept_reads.m2g_um_seg2 to genome segment_juncs with Bowtie2 (2/4)\n",
      "[2022-02-25 08:41:58] Mapping right_kept_reads.m2g_um_seg3 to genome segment_juncs with Bowtie2 (3/4)\n",
      "[2022-02-25 08:42:54] Mapping right_kept_reads.m2g_um_seg4 to genome segment_juncs with Bowtie2 (4/4)\n",
      "[2022-02-25 08:43:50] Joining segment hits\n",
      "[2022-02-25 08:46:25] Reporting output tracks\n",
      "-----------------------------------------------\n",
      "[2022-02-25 09:55:11] A summary of the alignment counts can be found in ../../analysis/tophat/SRR5272675_thout/align_summary.txt\n",
      "[2022-02-25 09:55:11] Run complete: 06:40:11 elapsed\n",
      "\n",
      "[2022-02-25 09:55:13] Beginning TopHat run (v2.1.1)\n",
      "-----------------------------------------------\n",
      "[2022-02-25 09:55:13] Checking for Bowtie\n",
      "\t\t  Bowtie version:\t 2.4.1.0\n",
      "[2022-02-25 09:55:14] Checking for Bowtie index files (genome)..\n",
      "[2022-02-25 09:55:14] Checking for reference FASTA file\n",
      "[2022-02-25 09:55:14] Generating SAM header for genome\n",
      "[2022-02-25 09:55:53] Reading known junctions from GTF file\n",
      "[2022-02-25 09:55:58] Preparing reads\n",
      "\t left reads: min. length=34, max. length=101, 72962581 kept reads (437 discarded)\n",
      "\tright reads: min. length=34, max. length=101, 72962046 kept reads (972 discarded)\n",
      "[2022-02-25 10:34:42] Building transcriptome data files ../../analysis/tophat/SRR5272676_thout/tmp/genes\n",
      "[2022-02-25 10:35:28] Building Bowtie index from genes.fa\n",
      "[2022-02-25 11:05:52] Mapping left_kept_reads to transcriptome genes with Bowtie2 \n",
      "[2022-02-25 12:47:01] Mapping right_kept_reads to transcriptome genes with Bowtie2 \n",
      "[2022-02-25 14:29:08] Resuming TopHat pipeline with unmapped reads\n",
      "[2022-02-25 14:29:09] Mapping left_kept_reads.m2g_um to genome genome with Bowtie2 \n",
      "[2022-02-25 15:00:43] Mapping left_kept_reads.m2g_um_seg1 to genome genome with Bowtie2 (1/4)\n",
      "[2022-02-25 15:04:09] Mapping left_kept_reads.m2g_um_seg2 to genome genome with Bowtie2 (2/4)\n",
      "[2022-02-25 15:06:57] Mapping left_kept_reads.m2g_um_seg3 to genome genome with Bowtie2 (3/4)\n",
      "[2022-02-25 15:10:08] Mapping left_kept_reads.m2g_um_seg4 to genome genome with Bowtie2 (4/4)\n",
      "[2022-02-25 15:14:03] Mapping right_kept_reads.m2g_um to genome genome with Bowtie2 \n",
      "[2022-02-25 15:43:53] Mapping right_kept_reads.m2g_um_seg1 to genome genome with Bowtie2 (1/4)\n",
      "[2022-02-25 15:45:57] Mapping right_kept_reads.m2g_um_seg2 to genome genome with Bowtie2 (2/4)\n",
      "[2022-02-25 15:48:05] Mapping right_kept_reads.m2g_um_seg3 to genome genome with Bowtie2 (3/4)\n",
      "[2022-02-25 15:50:28] Mapping right_kept_reads.m2g_um_seg4 to genome genome with Bowtie2 (4/4)\n",
      "[2022-02-25 15:52:59] Searching for junctions via segment mapping\n",
      "[2022-02-25 16:27:25] Retrieving sequences for splices\n",
      "[2022-02-25 16:28:40] Indexing splices\n",
      "Building a SMALL index\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2022-02-25 16:29:11] Mapping left_kept_reads.m2g_um_seg1 to genome segment_juncs with Bowtie2 (1/4)\n",
      "[2022-02-25 16:30:17] Mapping left_kept_reads.m2g_um_seg2 to genome segment_juncs with Bowtie2 (2/4)\n",
      "[2022-02-25 16:31:27] Mapping left_kept_reads.m2g_um_seg3 to genome segment_juncs with Bowtie2 (3/4)\n",
      "[2022-02-25 16:32:46] Mapping left_kept_reads.m2g_um_seg4 to genome segment_juncs with Bowtie2 (4/4)\n",
      "[2022-02-25 16:34:08] Joining segment hits\n",
      "[2022-02-25 16:36:47] Mapping right_kept_reads.m2g_um_seg1 to genome segment_juncs with Bowtie2 (1/4)\n",
      "[2022-02-25 16:37:58] Mapping right_kept_reads.m2g_um_seg2 to genome segment_juncs with Bowtie2 (2/4)\n",
      "[2022-02-25 16:39:06] Mapping right_kept_reads.m2g_um_seg3 to genome segment_juncs with Bowtie2 (3/4)\n",
      "[2022-02-25 16:40:18] Mapping right_kept_reads.m2g_um_seg4 to genome segment_juncs with Bowtie2 (4/4)\n",
      "[2022-02-25 16:41:29] Joining segment hits\n",
      "[2022-02-25 16:44:27] Reporting output tracks\n",
      "-----------------------------------------------\n",
      "[2022-02-25 18:12:01] A summary of the alignment counts can be found in ../../analysis/tophat/SRR5272676_thout/align_summary.txt\n",
      "[2022-02-25 18:12:01] Run complete: 08:16:47 elapsed\n",
      "\n",
      "[2022-02-25 18:12:02] Beginning TopHat run (v2.1.1)\n",
      "-----------------------------------------------\n",
      "[2022-02-25 18:12:02] Checking for Bowtie\n",
      "\t\t  Bowtie version:\t 2.4.1.0\n",
      "[2022-02-25 18:12:03] Checking for Bowtie index files (genome)..\n",
      "[2022-02-25 18:12:03] Checking for reference FASTA file\n",
      "[2022-02-25 18:12:03] Generating SAM header for genome\n",
      "[2022-02-25 18:12:43] Reading known junctions from GTF file\n",
      "[2022-02-25 18:12:47] Preparing reads\n",
      "\t left reads: min. length=34, max. length=101, 55261192 kept reads (370 discarded)\n",
      "\tright reads: min. length=34, max. length=101, 55261087 kept reads (475 discarded)\n",
      "[2022-02-25 18:41:16] Building transcriptome data files ../../analysis/tophat/SRR5272677_thout/tmp/genes\n",
      "[2022-02-25 18:42:04] Building Bowtie index from genes.fa\n",
      "[2022-02-25 19:12:12] Mapping left_kept_reads to transcriptome genes with Bowtie2 \n",
      "[2022-02-25 20:31:28] Mapping right_kept_reads to transcriptome genes with Bowtie2 \n",
      "[2022-02-25 21:50:44] Resuming TopHat pipeline with unmapped reads\n",
      "[2022-02-25 21:50:44] Mapping left_kept_reads.m2g_um to genome genome with Bowtie2 \n",
      "[2022-02-25 22:16:13] Mapping left_kept_reads.m2g_um_seg1 to genome genome with Bowtie2 (1/4)\n",
      "[2022-02-25 22:19:04] Mapping left_kept_reads.m2g_um_seg2 to genome genome with Bowtie2 (2/4)\n",
      "[2022-02-25 22:21:30] Mapping left_kept_reads.m2g_um_seg3 to genome genome with Bowtie2 (3/4)\n",
      "[2022-02-25 22:24:17] Mapping left_kept_reads.m2g_um_seg4 to genome genome with Bowtie2 (4/4)\n",
      "[2022-02-25 22:27:31] Mapping right_kept_reads.m2g_um to genome genome with Bowtie2 \n",
      "[2022-02-25 22:52:08] Mapping right_kept_reads.m2g_um_seg1 to genome genome with Bowtie2 (1/4)\n",
      "[2022-02-25 22:54:22] Mapping right_kept_reads.m2g_um_seg2 to genome genome with Bowtie2 (2/4)\n",
      "[2022-02-25 22:56:46] Mapping right_kept_reads.m2g_um_seg3 to genome genome with Bowtie2 (3/4)\n",
      "[2022-02-25 22:59:33] Mapping right_kept_reads.m2g_um_seg4 to genome genome with Bowtie2 (4/4)\n",
      "[2022-02-25 23:02:34] Searching for junctions via segment mapping\n",
      "[2022-02-25 23:29:07] Retrieving sequences for splices\n",
      "[2022-02-25 23:30:24] Indexing splices\n",
      "Building a SMALL index\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2022-02-25 23:30:55] Mapping left_kept_reads.m2g_um_seg1 to genome segment_juncs with Bowtie2 (1/4)\n",
      "[2022-02-25 23:31:45] Mapping left_kept_reads.m2g_um_seg2 to genome segment_juncs with Bowtie2 (2/4)\n",
      "[2022-02-25 23:32:42] Mapping left_kept_reads.m2g_um_seg3 to genome segment_juncs with Bowtie2 (3/4)\n",
      "[2022-02-25 23:33:45] Mapping left_kept_reads.m2g_um_seg4 to genome segment_juncs with Bowtie2 (4/4)\n",
      "[2022-02-25 23:34:50] Joining segment hits\n",
      "[2022-02-25 23:37:13] Mapping right_kept_reads.m2g_um_seg1 to genome segment_juncs with Bowtie2 (1/4)\n",
      "[2022-02-25 23:38:26] Mapping right_kept_reads.m2g_um_seg2 to genome segment_juncs with Bowtie2 (2/4)\n",
      "[2022-02-25 23:39:31] Mapping right_kept_reads.m2g_um_seg3 to genome segment_juncs with Bowtie2 (3/4)\n",
      "[2022-02-25 23:40:45] Mapping right_kept_reads.m2g_um_seg4 to genome segment_juncs with Bowtie2 (4/4)\n",
      "[2022-02-25 23:42:04] Joining segment hits\n",
      "[2022-02-25 23:44:30] Reporting output tracks\n",
      "-----------------------------------------------\n",
      "[2022-02-26 00:52:40] A summary of the alignment counts can be found in ../../analysis/tophat/SRR5272677_thout/align_summary.txt\n",
      "[2022-02-26 00:52:40] Run complete: 06:40:38 elapsed\n",
      "\n",
      "[2022-02-26 00:52:41] Beginning TopHat run (v2.1.1)\n",
      "-----------------------------------------------\n",
      "[2022-02-26 00:52:41] Checking for Bowtie\n",
      "\t\t  Bowtie version:\t 2.4.1.0\n",
      "[2022-02-26 00:52:44] Checking for Bowtie index files (genome)..\n",
      "[2022-02-26 00:52:44] Checking for reference FASTA file\n",
      "[2022-02-26 00:52:44] Generating SAM header for genome\n",
      "[2022-02-26 00:53:25] Reading known junctions from GTF file\n",
      "[2022-02-26 00:53:29] Preparing reads\n",
      "\t left reads: min. length=34, max. length=101, 61848675 kept reads (549 discarded)\n",
      "\tright reads: min. length=34, max. length=101, 61848520 kept reads (704 discarded)\n",
      "[2022-02-26 01:26:00] Building transcriptome data files ../../analysis/tophat/SRR5272678_thout/tmp/genes\n",
      "[2022-02-26 01:26:49] Building Bowtie index from genes.fa\n",
      "[2022-02-26 01:56:58] Mapping left_kept_reads to transcriptome genes with Bowtie2 \n",
      "[2022-02-26 03:24:42] Mapping right_kept_reads to transcriptome genes with Bowtie2 \n",
      "[2022-02-26 04:52:53] Resuming TopHat pipeline with unmapped reads\n",
      "[2022-02-26 04:52:53] Mapping left_kept_reads.m2g_um to genome genome with Bowtie2 \n",
      "[2022-02-26 05:20:01] Mapping left_kept_reads.m2g_um_seg1 to genome genome with Bowtie2 (1/4)\n",
      "[2022-02-26 05:23:02] Mapping left_kept_reads.m2g_um_seg2 to genome genome with Bowtie2 (2/4)\n",
      "[2022-02-26 05:25:37] Mapping left_kept_reads.m2g_um_seg3 to genome genome with Bowtie2 (3/4)\n",
      "[2022-02-26 05:28:33] Mapping left_kept_reads.m2g_um_seg4 to genome genome with Bowtie2 (4/4)\n",
      "[2022-02-26 05:32:06] Mapping right_kept_reads.m2g_um to genome genome with Bowtie2 \n",
      "[2022-02-26 05:57:21] Mapping right_kept_reads.m2g_um_seg1 to genome genome with Bowtie2 (1/4)\n",
      "[2022-02-26 05:59:12] Mapping right_kept_reads.m2g_um_seg2 to genome genome with Bowtie2 (2/4)\n",
      "[2022-02-26 06:01:17] Mapping right_kept_reads.m2g_um_seg3 to genome genome with Bowtie2 (3/4)\n",
      "[2022-02-26 06:03:40] Mapping right_kept_reads.m2g_um_seg4 to genome genome with Bowtie2 (4/4)\n",
      "[2022-02-26 06:06:07] Searching for junctions via segment mapping\n",
      "[2022-02-26 06:32:29] Retrieving sequences for splices\n",
      "[2022-02-26 06:33:45] Indexing splices\n",
      "Building a SMALL index\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2022-02-26 06:34:18] Mapping left_kept_reads.m2g_um_seg1 to genome segment_juncs with Bowtie2 (1/4)\n",
      "[2022-02-26 06:35:11] Mapping left_kept_reads.m2g_um_seg2 to genome segment_juncs with Bowtie2 (2/4)\n",
      "[2022-02-26 06:36:10] Mapping left_kept_reads.m2g_um_seg3 to genome segment_juncs with Bowtie2 (3/4)\n",
      "[2022-02-26 06:37:15] Mapping left_kept_reads.m2g_um_seg4 to genome segment_juncs with Bowtie2 (4/4)\n",
      "[2022-02-26 06:38:25] Joining segment hits\n",
      "[2022-02-26 06:40:54] Mapping right_kept_reads.m2g_um_seg1 to genome segment_juncs with Bowtie2 (1/4)\n",
      "[2022-02-26 06:42:00] Mapping right_kept_reads.m2g_um_seg2 to genome segment_juncs with Bowtie2 (2/4)\n",
      "[2022-02-26 06:43:01] Mapping right_kept_reads.m2g_um_seg3 to genome segment_juncs with Bowtie2 (3/4)\n",
      "[2022-02-26 06:44:09] Mapping right_kept_reads.m2g_um_seg4 to genome segment_juncs with Bowtie2 (4/4)\n",
      "[2022-02-26 06:45:16] Joining segment hits\n",
      "[2022-02-26 06:47:45] Reporting output tracks\n",
      "-----------------------------------------------\n",
      "[2022-02-26 08:01:07] A summary of the alignment counts can be found in ../../analysis/tophat/SRR5272678_thout/align_summary.txt\n",
      "[2022-02-26 08:01:07] Run complete: 07:08:25 elapsed\n",
      "\n",
      "[2022-02-26 08:01:08] Beginning TopHat run (v2.1.1)\n",
      "-----------------------------------------------\n",
      "[2022-02-26 08:01:08] Checking for Bowtie\n",
      "\t\t  Bowtie version:\t 2.4.1.0\n",
      "[2022-02-26 08:01:11] Checking for Bowtie index files (genome)..\n",
      "[2022-02-26 08:01:11] Checking for reference FASTA file\n",
      "[2022-02-26 08:01:11] Generating SAM header for genome\n",
      "[2022-02-26 08:01:50] Reading known junctions from GTF file\n",
      "[2022-02-26 08:01:55] Preparing reads\n",
      "\t left reads: min. length=34, max. length=101, 43979843 kept reads (800 discarded)\n",
      "\tright reads: min. length=34, max. length=101, 43979741 kept reads (902 discarded)\n",
      "[2022-02-26 08:25:57] Building transcriptome data files ../../analysis/tophat/SRR5272679_thout/tmp/genes\n",
      "[2022-02-26 08:26:46] Building Bowtie index from genes.fa\n",
      "[2022-02-26 08:56:54] Mapping left_kept_reads to transcriptome genes with Bowtie2 \n",
      "[2022-02-26 10:02:08] Mapping right_kept_reads to transcriptome genes with Bowtie2 \n",
      "[2022-02-26 11:07:51] Resuming TopHat pipeline with unmapped reads\n",
      "[2022-02-26 11:07:51] Mapping left_kept_reads.m2g_um to genome genome with Bowtie2 \n",
      "[2022-02-26 11:27:49] Mapping left_kept_reads.m2g_um_seg1 to genome genome with Bowtie2 (1/4)\n",
      "[2022-02-26 11:29:43] Mapping left_kept_reads.m2g_um_seg2 to genome genome with Bowtie2 (2/4)\n",
      "[2022-02-26 11:31:26] Mapping left_kept_reads.m2g_um_seg3 to genome genome with Bowtie2 (3/4)\n",
      "[2022-02-26 11:33:27] Mapping left_kept_reads.m2g_um_seg4 to genome genome with Bowtie2 (4/4)\n",
      "[2022-02-26 11:35:38] Mapping right_kept_reads.m2g_um to genome genome with Bowtie2 \n",
      "[2022-02-26 11:54:41] Mapping right_kept_reads.m2g_um_seg1 to genome genome with Bowtie2 (1/4)\n",
      "[2022-02-26 11:56:04] Mapping right_kept_reads.m2g_um_seg2 to genome genome with Bowtie2 (2/4)\n",
      "[2022-02-26 11:57:37] Mapping right_kept_reads.m2g_um_seg3 to genome genome with Bowtie2 (3/4)\n",
      "[2022-02-26 11:59:23] Mapping right_kept_reads.m2g_um_seg4 to genome genome with Bowtie2 (4/4)\n",
      "[2022-02-26 12:01:14] Searching for junctions via segment mapping\n",
      "[2022-02-26 12:24:51] Retrieving sequences for splices\n",
      "[2022-02-26 12:26:06] Indexing splices\n",
      "Building a SMALL index\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n",
      "Warning: Encountered reference sequence with only gaps\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n",
      "Warning: Encountered reference sequence with only gaps\r\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2022-02-26 12:26:33] Mapping left_kept_reads.m2g_um_seg1 to genome segment_juncs with Bowtie2 (1/4)\n",
      "[2022-02-26 12:27:10] Mapping left_kept_reads.m2g_um_seg2 to genome segment_juncs with Bowtie2 (2/4)\n",
      "[2022-02-26 12:27:49] Mapping left_kept_reads.m2g_um_seg3 to genome segment_juncs with Bowtie2 (3/4)\n",
      "[2022-02-26 12:28:35] Mapping left_kept_reads.m2g_um_seg4 to genome segment_juncs with Bowtie2 (4/4)\n",
      "[2022-02-26 12:29:21] Joining segment hits\n",
      "[2022-02-26 12:31:10] Mapping right_kept_reads.m2g_um_seg1 to genome segment_juncs with Bowtie2 (1/4)\n",
      "[2022-02-26 12:31:54] Mapping right_kept_reads.m2g_um_seg2 to genome segment_juncs with Bowtie2 (2/4)\n",
      "[2022-02-26 12:32:39] Mapping right_kept_reads.m2g_um_seg3 to genome segment_juncs with Bowtie2 (3/4)\n",
      "[2022-02-26 12:33:26] Mapping right_kept_reads.m2g_um_seg4 to genome segment_juncs with Bowtie2 (4/4)\n",
      "[2022-02-26 12:34:16] Joining segment hits\n",
      "[2022-02-26 12:36:14] Reporting output tracks\n",
      "-----------------------------------------------\n",
      "[2022-02-26 13:23:51] A summary of the alignment counts can be found in ../../analysis/tophat/SRR5272679_thout/align_summary.txt\n",
      "[2022-02-26 13:23:51] Run complete: 05:22:43 elapsed\n"
     ]
    }
   ],
   "source": [
    "for acc in dat:\n",
    "    ! tophat -p 8 -G genes.gtf -o ../../analysis/tophat/{acc}_thout genome {acc}_1.fq.gz {acc}_2.fq.gz "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Look at the summary of the alignment counts for the last sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Left reads:\r\n",
      "          Input     :  43980643\r\n",
      "           Mapped   :  42900081 (97.5% of input)\r\n",
      "            of these:    532987 ( 1.2%) have multiple alignments (4182 have >20)\r\n",
      "Right reads:\r\n",
      "          Input     :  43980643\r\n",
      "           Mapped   :  43103304 (98.0% of input)\r\n",
      "            of these:    537322 ( 1.2%) have multiple alignments (4188 have >20)\r\n",
      "97.8% overall read mapping rate.\r\n",
      "\r\n",
      "Aligned pairs:  42398148\r\n",
      "     of these:    522296 ( 1.2%) have multiple alignments\r\n",
      "                  623355 ( 1.5%) are discordant alignments\r\n",
      "95.0% concordant pair alignment rate.\r\n"
     ]
    }
   ],
   "source": [
    "! cat ../../analysis/tophat/SRR5272679_thout/align_summary.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Transcript assembly from RNA-seq reads aligned to reference genome with Cufflinks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir -p ../../analysis/cufflinks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cufflinks v2.2.1\r\n",
      "linked against Boost version 104700\r\n",
      "-----------------------------\r\n",
      "Usage:   cufflinks [options] <hits.sam>\r\n",
      "General Options:\r\n",
      "  -o/--output-dir              write all output files to this directory              [ default:     ./ ]\r\n",
      "  -p/--num-threads             number of threads used during analysis                [ default:      1 ]\r\n",
      "  --seed                       value of random number generator seed                 [ default:      0 ]\r\n",
      "  -G/--GTF                     quantitate against reference transcript annotations                      \r\n",
      "  -g/--GTF-guide               use reference transcript annotation to guide assembly                   \r\n",
      "  -M/--mask-file               ignore all alignment within transcripts in this file                     \r\n",
      "  -b/--frag-bias-correct       use bias correction - reference fasta required        [ default:   NULL ]\r\n",
      "  -u/--multi-read-correct      use 'rescue method' for multi-reads (more accurate)   [ default:  FALSE ]\r\n",
      "  --library-type               library prep used for input reads                     [ default:  below ]\r\n",
      "  --library-norm-method        Method used to normalize library sizes                [ default:  below ]\r\n",
      "\r\n",
      "Advanced Abundance Estimation Options:\r\n",
      "  -m/--frag-len-mean           average fragment length (unpaired reads only)         [ default:    200 ]\r\n",
      "  -s/--frag-len-std-dev        fragment length std deviation (unpaired reads only)   [ default:     80 ]\r\n",
      "  --max-mle-iterations         maximum iterations allowed for MLE calculation        [ default:   5000 ]\r\n",
      "  --compatible-hits-norm       count hits compatible with reference RNAs only        [ default:  FALSE ]\r\n",
      "  --total-hits-norm            count all hits for normalization                      [ default:  TRUE  ]\r\n",
      "  --num-frag-count-draws       Number of fragment generation samples                 [ default:    100 ]\r\n",
      "  --num-frag-assign-draws      Number of fragment assignment samples per generation  [ default:     50 ]\r\n",
      "  --max-frag-multihits         Maximum number of alignments allowed per fragment     [ default: unlim  ]\r\n",
      "  --no-effective-length-correction   No effective length correction                  [ default:  FALSE ]\r\n",
      "  --no-length-correction       No length correction                                  [ default:  FALSE ]\r\n",
      "  -N/--upper-quartile-norm     Deprecated, use --library-norm-method                 [    DEPRECATED   ]\r\n",
      "  --raw-mapped-norm            Deprecated, use --library-norm-method                 [    DEPRECATED   ]\r\n",
      "\r\n",
      "Advanced Assembly Options:\r\n",
      "  -L/--label                   assembled transcripts have this ID prefix             [ default:   CUFF ]\r\n",
      "  -F/--min-isoform-fraction    suppress transcripts below this abundance level       [ default:   0.10 ]\r\n",
      "  -j/--pre-mrna-fraction       suppress intra-intronic transcripts below this level  [ default:   0.15 ]\r\n",
      "  -I/--max-intron-length       ignore alignments with gaps longer than this          [ default: 300000 ]\r\n",
      "  -a/--junc-alpha              alpha for junction binomial test filter               [ default:  0.001 ]\r\n",
      "  -A/--small-anchor-fraction   percent read overhang taken as 'suspiciously small'   [ default:   0.09 ]\r\n",
      "  --min-frags-per-transfrag    minimum number of fragments needed for new transfrags [ default:     10 ]\r\n",
      "  --overhang-tolerance         number of terminal exon bp to tolerate in introns     [ default:      8 ]\r\n",
      "  --max-bundle-length          maximum genomic length allowed for a given bundle     [ default:3500000 ]\r\n",
      "  --max-bundle-frags           maximum fragments allowed in a bundle before skipping [ default: 500000 ]\r\n",
      "  --min-intron-length          minimum intron size allowed in genome                 [ default:     50 ]\r\n",
      "  --trim-3-avgcov-thresh       minimum avg coverage required to attempt 3' trimming  [ default:     10 ]\r\n",
      "  --trim-3-dropoff-frac        fraction of avg coverage below which to trim 3' end   [ default:    0.1 ]\r\n",
      "  --max-multiread-fraction     maximum fraction of allowed multireads per transcript [ default:   0.75 ]\r\n",
      "  --overlap-radius             maximum gap size to fill between transfrags (in bp)   [ default:     50 ]\r\n",
      "\r\n",
      "Advanced Reference Annotation Guided Assembly Options:\r\n",
      "  --no-faux-reads              disable tiling by faux reads                          [ default:  FALSE ]\r\n",
      "  --3-overhang-tolerance       overhang allowed on 3' end when merging with reference[ default:    600 ]\r\n",
      "  --intron-overhang-tolerance  overhang allowed inside reference intron when merging [ default:     30 ]\r\n",
      "\r\n",
      "Advanced Program Behavior Options:\r\n",
      "  -v/--verbose                 log-friendly verbose processing (no progress bar)     [ default:  FALSE ]\r\n",
      "  -q/--quiet                   log-friendly quiet processing (no progress bar)       [ default:  FALSE ]\r\n",
      "  --no-update-check            do not contact server to check for update availability[ default:  FALSE ]\r\n",
      "\r\n",
      "Supported library types:\r\n",
      "\tff-firststrand\r\n",
      "\tff-secondstrand\r\n",
      "\tff-unstranded\r\n",
      "\tfr-firststrand\r\n",
      "\tfr-secondstrand\r\n",
      "\tfr-unstranded (default)\r\n",
      "\ttransfrags\r\n",
      "\r\n",
      "Supported library normalization methods:\r\n",
      "\tclassic-fpkm\r\n"
     ]
    }
   ],
   "source": [
    "! cufflinks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Could not connect to update server to verify current version. Please check at the Cufflinks website (http://cufflinks.cbcb.umd.edu).\n",
      "[23:14:55] Inspecting reads and determining fragment length distribution.\n",
      "> Processed 118538 loci.                       [*************************] 100%\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 47187075.50\n",
      ">\tRaw Map Mass: 47187075.50\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 149.27\n",
      ">\t           Estimated Std Dev: 56.32\n",
      "[23:21:33] Assembling transcripts and estimating abundances.\n",
      "> Processed 130048 loci.                       [*************************] 100%\n",
      "Warning: Could not connect to update server to verify current version. Please check at the Cufflinks website (http://cufflinks.cbcb.umd.edu).\n",
      "[00:24:02] Inspecting reads and determining fragment length distribution.\n",
      "> Processed 115521 loci.                       [*************************] 100%\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 57471081.06\n",
      ">\tRaw Map Mass: 57471081.06\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 162.94\n",
      ">\t           Estimated Std Dev: 54.45\n",
      "[00:31:44] Assembling transcripts and estimating abundances.\n",
      "> Processed 127913 loci.                       [*************************] 100%\n",
      "Warning: Could not connect to update server to verify current version. Please check at the Cufflinks website (http://cufflinks.cbcb.umd.edu).\n",
      "[01:11:29] Inspecting reads and determining fragment length distribution.\n",
      "> Processed 129273 loci.                       [*************************] 100%\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 71809195.24\n",
      ">\tRaw Map Mass: 71809195.24\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 159.09\n",
      ">\t           Estimated Std Dev: 55.09\n",
      "[01:21:26] Assembling transcripts and estimating abundances.\n",
      "> Processed 143951 loci.                       [*************************] 100%\n",
      "Warning: Could not connect to update server to verify current version. Please check at the Cufflinks website (http://cufflinks.cbcb.umd.edu).\n",
      "[02:18:45] Inspecting reads and determining fragment length distribution.\n",
      "> Processed 126614 loci.                       [*************************] 100%\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 54227694.63\n",
      ">\tRaw Map Mass: 54227694.63\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 150.73\n",
      ">\t           Estimated Std Dev: 54.64\n",
      "[02:25:39] Assembling transcripts and estimating abundances.\n",
      "> Processed 142302 loci.                       [*************************] 100%\n",
      "Warning: Could not connect to update server to verify current version. Please check at the Cufflinks website (http://cufflinks.cbcb.umd.edu).\n",
      "[03:04:47] Inspecting reads and determining fragment length distribution.\n",
      "> Processed 119045 loci.                       [*************************] 100%\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 60961910.44\n",
      ">\tRaw Map Mass: 60961910.44\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 159.51\n",
      ">\t           Estimated Std Dev: 57.96\n",
      "[03:12:39] Assembling transcripts and estimating abundances.\n",
      "> Processed 135871 loci.                       [*************************] 100%\n",
      "Warning: Could not connect to update server to verify current version. Please check at the Cufflinks website (http://cufflinks.cbcb.umd.edu).\n",
      "[03:58:15] Inspecting reads and determining fragment length distribution.\n",
      "> Processed 104029 loci.                       [*************************] 100%\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 43149026.46\n",
      ">\tRaw Map Mass: 43149026.46\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 172.13\n",
      ">\t           Estimated Std Dev: 65.07\n",
      "[04:03:44] Assembling transcripts and estimating abundances.\n",
      "> Processed 115953 loci.                       [*************************] 100%\n"
     ]
    }
   ],
   "source": [
    "for acc in dat:\n",
    "    ! cufflinks -p 8 -o ../../analysis/cufflinks/{acc}_clout ../../analysis/tophat/{acc}_thout/accepted_hits.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR5272674_clout  SRR5272676_clout  SRR5272678_clout\r\n",
      "SRR5272675_clout  SRR5272677_clout  SRR5272679_clout\r\n"
     ]
    }
   ],
   "source": [
    "! ls ../../analysis/cufflinks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 59560\r\n",
      " 1688 -rw-r--r-- 1 cv cv  1722693 Feb 27 04:34 genes.fpkm_tracking\r\n",
      " 2880 -rw-r--r-- 1 cv cv  2941187 Feb 27 04:34 isoforms.fpkm_tracking\r\n",
      "    0 -rw-r--r-- 1 cv cv        0 Feb 27 04:03 skipped.gtf\r\n",
      "54992 -rw-r--r-- 1 cv cv 56311039 Feb 27 04:34 transcripts.gtf\r\n"
     ]
    }
   ],
   "source": [
    "! ls -ls ../../analysis/cufflinks/SRR5272679_clout"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Group samples based on condition: untreated (U) or treated (T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/cv/RNA-seq/analysis/tophat\n"
     ]
    }
   ],
   "source": [
    "cd ../../analysis/tophat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "for acc in dat[:3]:\n",
    "    ! mv {acc}_thout U_{acc}_thout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "for acc in dat[3:]:\n",
    "    ! mv {acc}_thout T_{acc}_thout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "T_SRR5272677_thout  T_SRR5272679_thout\tU_SRR5272675_thout\r\n",
      "T_SRR5272678_thout  U_SRR5272674_thout\tU_SRR5272676_thout\r\n"
     ]
    }
   ],
   "source": [
    "! ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/cv/RNA-seq/analysis/cufflinks\n"
     ]
    }
   ],
   "source": [
    "cd ../../analysis/cufflinks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "for acc in dat[:3]:\n",
    "    ! mv {acc}_clout U_{acc}_clout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "for acc in dat[3:]:\n",
    "    ! mv {acc}_clout T_{acc}_clout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "T_SRR5272677_clout  T_SRR5272679_clout\tU_SRR5272675_clout\r\n",
      "T_SRR5272678_clout  U_SRR5272674_clout\tU_SRR5272676_clout\r\n"
     ]
    }
   ],
   "source": [
    "! ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a file listing the assembly file for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "! ls *_clout/transcripts.gtf > assemblies.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "T_SRR5272677_clout/transcripts.gtf\r\n",
      "T_SRR5272678_clout/transcripts.gtf\r\n",
      "T_SRR5272679_clout/transcripts.gtf\r\n",
      "U_SRR5272674_clout/transcripts.gtf\r\n",
      "U_SRR5272675_clout/transcripts.gtf\r\n",
      "U_SRR5272676_clout/transcripts.gtf\r\n"
     ]
    }
   ],
   "source": [
    "! cat assemblies.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge the transcript assemblies using cuffmerge to create a single merged transcriptome annotation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cuffmerge: \r\n",
      "cuffmerge takes two or more Cufflinks GTF files and merges them into a\r\n",
      "single unified transcript catalog.  Optionally, you can provide the script\r\n",
      "with a reference GTF, and the script will use it to attach gene names and other\r\n",
      "metadata to the merged catalog.\r\n",
      "\r\n",
      "Usage:\r\n",
      "    cuffmerge [Options] <assembly_GTF_list.txt>\r\n",
      "\r\n",
      "Options:\r\n",
      "    -h/--help                               Prints the help message and exits\r\n",
      "    -o                     <output_dir>     Directory where merged assembly will be written  [ default: ./merged_asm  ]\r\n",
      "    -g/--ref-gtf                            An optional \"reference\" annotation GTF.\r\n",
      "    -s/--ref-sequence      <seq_dir>/<seq_fasta> Genomic DNA sequences for the reference.\r\n",
      "    --min-isoform-fraction <0-1.0>          Discard isoforms with abundance below this       [ default:           0.05 ]\r\n",
      "    -p/--num-threads       <int>            Use this many threads to merge assemblies.       [ default:             1  ]\r\n",
      "    --keep-tmp                              Keep all intermediate files during merge\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "! cuffmerge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[Sun Feb 27 18:01:23 2022] Beginning transcriptome assembly merge\n",
      "-------------------------------------------\n",
      "\n",
      "[Sun Feb 27 18:01:23 2022] Preparing output location ./merged_asm/\n",
      "[Sun Feb 27 18:01:30 2022] Converting GTF files to SAM\n",
      "[18:01:30] Loading reference annotation.\n",
      "[18:01:31] Loading reference annotation.\n",
      "[18:01:33] Loading reference annotation.\n",
      "[18:01:34] Loading reference annotation.\n",
      "[18:01:36] Loading reference annotation.\n",
      "[18:01:37] Loading reference annotation.\n",
      "[Sun Feb 27 18:01:39 2022] Quantitating transcripts\n",
      "Warning: Could not connect to update server to verify current version. Please check at the Cufflinks website (http://cufflinks.cbcb.umd.edu).\n",
      "Command line:\n",
      "cufflinks -o ./merged_asm/ -F 0.05 -g ../../data/preprocessing_fastp/genes.gtf -q --overhang-tolerance 200 --library-type=transfrags -A 0.0 --min-frags-per-transfrag 0 --no-5-extend -p 8 ./merged_asm/tmp/mergeSam_filetTovwM \n",
      "[bam_header_read] EOF marker is absent. The input is probably truncated.\n",
      "[bam_header_read] invalid BAM binary header (this is not a BAM file).\n",
      "File ./merged_asm/tmp/mergeSam_filetTovwM doesn't appear to be a valid BAM file, trying SAM...\n",
      "[18:01:40] Loading reference annotation.\n",
      "[18:01:45] Inspecting reads and determining fragment length distribution.\n",
      "Processed 28737 loci.                       \n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 195628.00\n",
      ">\tRaw Map Mass: 195628.00\n",
      ">\tFragment Length Distribution: Truncated Gaussian (default)\n",
      ">\t              Default Mean: 200\n",
      ">\t           Default Std Dev: 80\n",
      "[18:01:46] Assembling transcripts and estimating abundances.\n",
      "Processed 28737 loci.                       \n",
      "[Sun Feb 27 18:04:28 2022] Comparing against reference file ../../data/preprocessing_fastp/genes.gtf\n",
      "Warning: Could not connect to update server to verify current version. Please check at the Cufflinks website (http://cufflinks.cbcb.umd.edu).\n",
      "No fasta index found for ../../data/preprocessing_fastp/genome.fa. Rebuilding, please wait..\n",
      "Fasta index rebuilt.\n",
      "Warning: couldn't find fasta record for 'chr10_GL383545v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr10_GL383546v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr10_KI270824v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr10_KI270825v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_JH159136v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_JH159137v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270827v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270829v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270830v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270831v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270832v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270902v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270903v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270927v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL383550v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL383551v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL383553v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL877875v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL877876v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270833v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270834v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270835v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270837v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270904v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr13_KI270838v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr13_KI270839v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr13_KI270840v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr13_KI270842v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr14_KI270844v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr14_KI270845v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr14_KI270846v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr14_KI270847v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_GL383554v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_GL383555v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270848v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270849v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270850v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270851v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270852v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270905v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270906v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_GL383556v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_GL383557v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_KI270853v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_KI270854v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_KI270855v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_KI270856v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_GL000258v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_GL383563v3_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_GL383564v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_GL383566v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_JH159146v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_JH159147v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_JH159148v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270857v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270860v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270861v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270862v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270907v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270908v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270909v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270910v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_GL383567v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_GL383571v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_GL383572v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_KI270863v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_KI270911v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_KI270912v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL000209v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL383573v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL383574v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL383575v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL383576v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949746v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949747v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949748v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949749v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949750v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949751v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949752v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949753v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270865v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270866v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270867v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270868v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270882v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270883v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270884v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270885v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270886v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270887v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270888v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270889v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270890v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270891v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270914v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270915v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270916v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270917v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270918v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270919v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270920v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270921v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270922v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270923v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270929v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270930v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270931v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270932v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270933v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270938v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_GL383518v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_GL383519v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_GL383520v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270759v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270761v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270762v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270763v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270765v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270766v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270892v1_alt'!\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: couldn't find fasta record for 'chr20_KI270869v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr20_KI270870v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr20_KI270871v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_GL383579v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_GL383580v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_GL383581v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_KI270872v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_KI270873v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_GL383582v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_GL383583v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KB663609v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270875v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270876v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270877v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270878v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270879v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270928v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_GL383521v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_GL383522v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_GL582966v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270767v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270768v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270769v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270770v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270772v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270773v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270774v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270776v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270893v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270894v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_GL383526v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_JH636055v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270777v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270779v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270780v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270782v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270783v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270784v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270895v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270924v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270934v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270935v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270936v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270937v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_GL000257v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_GL383527v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270786v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270789v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270790v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270896v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270925v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_GL339449v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_GL949742v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270791v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270792v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270793v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270794v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270795v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270796v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270897v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270898v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000250v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000251v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000252v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000253v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000254v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000255v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000256v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL383533v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270758v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270797v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270798v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270800v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270801v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_GL383534v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270803v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270804v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270806v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270808v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270809v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270899v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270811v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270812v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270813v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270814v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270815v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270816v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270817v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270818v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270819v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270820v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270821v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270822v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270900v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270926v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr9_GL383540v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr9_GL383541v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr9_KI270823v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chrX_KI270880v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chrX_KI270881v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chrX_KI270913v1_alt'!\n",
      "[Sun Feb 27 18:05:59 2022] Comparing against reference file ../../data/preprocessing_fastp/genes.gtf\n",
      "Warning: Could not connect to update server to verify current version. Please check at the Cufflinks website (http://cufflinks.cbcb.umd.edu).\n",
      "Warning: couldn't find fasta record for 'chr10_GL383545v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr10_GL383546v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr10_KI270824v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr10_KI270825v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_JH159136v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_JH159137v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270827v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270829v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270830v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270831v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270832v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270902v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270903v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270927v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL383550v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL383551v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL383553v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL877875v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL877876v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270833v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270834v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270835v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270837v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270904v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr13_KI270838v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr13_KI270839v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr13_KI270840v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr13_KI270842v1_alt'!\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: couldn't find fasta record for 'chr14_KI270844v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr14_KI270845v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr14_KI270846v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr14_KI270847v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_GL383554v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_GL383555v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270848v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270849v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270850v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270851v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270852v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270905v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270906v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_GL383556v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_GL383557v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_KI270853v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_KI270854v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_KI270855v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_KI270856v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_GL000258v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_GL383563v3_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_GL383564v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_GL383566v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_JH159146v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_JH159147v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_JH159148v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270857v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270860v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270861v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270862v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270907v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270908v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270909v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270910v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_GL383567v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_GL383571v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_GL383572v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_KI270863v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_KI270911v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_KI270912v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL000209v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL383573v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL383574v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL383575v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL383576v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949746v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949747v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949748v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949749v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949750v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949751v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949752v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949753v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270865v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270866v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270867v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270868v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270882v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270883v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270884v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270885v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270886v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270887v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270888v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270889v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270890v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270891v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270914v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270915v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270916v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270917v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270918v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270919v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270920v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270921v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270922v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270923v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270929v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270930v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270931v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270932v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270933v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270938v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_GL383518v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_GL383519v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_GL383520v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270759v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270761v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270762v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270763v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270765v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270766v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270892v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr20_KI270869v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr20_KI270870v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr20_KI270871v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_GL383579v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_GL383580v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_GL383581v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_KI270872v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_KI270873v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_GL383582v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_GL383583v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KB663609v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270875v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270876v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270877v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270878v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270879v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270928v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_GL383521v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_GL383522v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_GL582966v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270767v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270768v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270769v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270770v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270772v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270773v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270774v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270776v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270893v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270894v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_GL383526v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_JH636055v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270777v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270779v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270780v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270782v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270783v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270784v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270895v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270924v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270934v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270935v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270936v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270937v1_alt'!\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: couldn't find fasta record for 'chr4_GL000257v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_GL383527v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270786v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270789v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270790v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270896v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270925v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_GL339449v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_GL949742v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270791v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270792v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270793v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270794v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270795v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270796v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270897v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270898v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000250v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000251v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000252v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000253v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000254v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000255v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000256v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL383533v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270758v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270797v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270798v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270800v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270801v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_GL383534v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270803v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270804v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270806v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270808v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270809v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270899v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270811v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270812v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270813v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270814v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270815v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270816v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270817v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270818v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270819v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270820v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270821v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270822v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270900v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270926v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr9_GL383540v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr9_GL383541v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr9_KI270823v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chrX_KI270880v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chrX_KI270881v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chrX_KI270913v1_alt'!\n"
     ]
    }
   ],
   "source": [
    "! cuffmerge -g ../../data/preprocessing_fastp/genes.gtf \\\n",
    "            -s ../../data/preprocessing_fastp/genome.fa \\\n",
    "            -p 8 assemblies.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 343492\r\n",
      "     4 drwxr-xr-x 2 cv cv      4096 Feb 27 18:01 logs\r\n",
      "343488 -rw-r--r-- 1 cv cv 351727186 Feb 27 18:06 merged.gtf\r\n"
     ]
    }
   ],
   "source": [
    "! ls -ls ./merged_asm/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrY\tCufflinks\texon\t25527647\t25527799\t.\t-\t.\tgene_id \"XLOC_037176\"; transcript_id \"TCONS_00135957\"; exon_number \"4\"; gene_name \"LOC102724004\"; oId \"rna87623\"; nearest_ref \"rna87623\"; class_code \"=\"; tss_id \"TSS62410\";\r\n",
      "chrY\tCufflinks\texon\t25517743\t25517798\t.\t-\t.\tgene_id \"XLOC_037176\"; transcript_id \"TCONS_00135956\"; exon_number \"1\"; gene_name \"LOC102724004\"; oId \"rna87622\"; nearest_ref \"rna87622\"; class_code \"=\"; tss_id \"TSS62410\";\r\n",
      "chrY\tCufflinks\texon\t25524026\t25524110\t.\t-\t.\tgene_id \"XLOC_037176\"; transcript_id \"TCONS_00135956\"; exon_number \"2\"; gene_name \"LOC102724004\"; oId \"rna87622\"; nearest_ref \"rna87622\"; class_code \"=\"; tss_id \"TSS62410\";\r\n",
      "chrY\tCufflinks\texon\t25527647\t25527799\t.\t-\t.\tgene_id \"XLOC_037176\"; transcript_id \"TCONS_00135956\"; exon_number \"3\"; gene_name \"LOC102724004\"; oId \"rna87622\"; nearest_ref \"rna87622\"; class_code \"=\"; tss_id \"TSS62410\";\r\n",
      "chrY\tCufflinks\texon\t25517743\t25517798\t.\t-\t.\tgene_id \"XLOC_037176\"; transcript_id \"TCONS_00135955\"; exon_number \"1\"; gene_name \"LOC102724004\"; oId \"rna87621\"; nearest_ref \"rna87621\"; class_code \"=\"; tss_id \"TSS62410\";\r\n",
      "chrY\tCufflinks\texon\t25524026\t25524110\t.\t-\t.\tgene_id \"XLOC_037176\"; transcript_id \"TCONS_00135955\"; exon_number \"2\"; gene_name \"LOC102724004\"; oId \"rna87621\"; nearest_ref \"rna87621\"; class_code \"=\"; tss_id \"TSS62410\";\r\n",
      "chrY\tCufflinks\texon\t25525204\t25525288\t.\t-\t.\tgene_id \"XLOC_037176\"; transcript_id \"TCONS_00135955\"; exon_number \"3\"; gene_name \"LOC102724004\"; oId \"rna87621\"; nearest_ref \"rna87621\"; class_code \"=\"; tss_id \"TSS62410\";\r\n",
      "chrY\tCufflinks\texon\t25527649\t25527799\t.\t-\t.\tgene_id \"XLOC_037176\"; transcript_id \"TCONS_00135955\"; exon_number \"4\"; gene_name \"LOC102724004\"; oId \"rna87621\"; nearest_ref \"rna87621\"; class_code \"=\"; tss_id \"TSS62410\";\r\n",
      "chrY\tCufflinks\texon\t56854931\t56855161\t.\t.\t.\tgene_id \"XLOC_037177\"; transcript_id \"TCONS_00135958\"; exon_number \"1\"; oId \"CUFF.22258.1\"; class_code \"u\"; tss_id \"TSS62411\";\r\n",
      "chrY\tCufflinks\texon\t56856859\t56857612\t.\t.\t.\tgene_id \"XLOC_037178\"; transcript_id \"TCONS_00135959\"; exon_number \"1\"; oId \"CUFF.22257.1\"; class_code \"u\"; tss_id \"TSS62412\";\r\n"
     ]
    }
   ],
   "source": [
    "! tail -n 10 ./merged_asm/merged.gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Differentially expressed genes and transcripts with cuffdiff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cuffdiff v2.2.1 (4237)\r\n",
      "-----------------------------\r\n",
      "Usage:   cuffdiff [options] <transcripts.gtf> <sample1_hits.sam> <sample2_hits.sam> [... sampleN_hits.sam]\r\n",
      "   Supply replicate SAMs as comma separated lists for each condition: sample1_rep1.sam,sample1_rep2.sam,...sample1_repM.sam\r\n",
      "General Options:\r\n",
      "  -o/--output-dir              write all output files to this directory              [ default:     ./ ]\r\n",
      "  -L/--labels                  comma-separated list of condition labels\r\n",
      "  --FDR                        False discovery rate used in testing                  [ default:   0.05 ]\r\n",
      "  -M/--mask-file               ignore all alignment within transcripts in this file  [ default:   NULL ]\r\n",
      "  -C/--contrast-file           Perform the constrasts specified in this file         [ default:   NULL ]\r\n",
      "  -b/--frag-bias-correct       use bias correction - reference fasta required        [ default:   NULL ]\r\n",
      "  -u/--multi-read-correct      use 'rescue method' for multi-reads                   [ default:  FALSE ]\r\n",
      "  -p/--num-threads             number of threads used during quantification          [ default:      1 ]\r\n",
      "  --no-diff                    Don't generate differential analysis files            [ default:  FALSE ]\r\n",
      "  --no-js-tests                Don't perform isoform switching tests                 [ default:  FALSE ]\r\n",
      "  -T/--time-series             treat samples as a time-series                        [ default:  FALSE ]\r\n",
      "  --library-type               Library prep used for input reads                     [ default:  below ]\r\n",
      "  --dispersion-method          Method used to estimate dispersion models             [ default:  below ]\r\n",
      "  --library-norm-method        Method used to normalize library sizes                [ default:  below ]\r\n",
      "\r\n",
      "Advanced Options:\r\n",
      "  -m/--frag-len-mean           average fragment length (unpaired reads only)         [ default:    200 ]\r\n",
      "  -s/--frag-len-std-dev        fragment length std deviation (unpaired reads only)   [ default:     80 ]\r\n",
      "  -c/--min-alignment-count     minimum number of alignments in a locus for testing   [ default:   10 ]\r\n",
      "  --max-mle-iterations         maximum iterations allowed for MLE calculation        [ default:   5000 ]\r\n",
      "  --compatible-hits-norm       count hits compatible with reference RNAs only        [ default:   TRUE ]\r\n",
      "  --total-hits-norm            count all hits for normalization                      [ default:  FALSE ]\r\n",
      "  -v/--verbose                 log-friendly verbose processing (no progress bar)     [ default:  FALSE ]\r\n",
      "  -q/--quiet                   log-friendly quiet processing (no progress bar)       [ default:  FALSE ]\r\n",
      "  --seed                       value of random number generator seed                 [ default:      0 ]\r\n",
      "  --no-update-check            do not contact server to check for update availability[ default:  FALSE ]\r\n",
      "  --emit-count-tables          print count tables used to fit overdispersion         [    DEPRECATED   ]\r\n",
      "  --max-bundle-frags           maximum fragments allowed in a bundle before skipping [ default: 500000 ]\r\n",
      "  --num-frag-count-draws       Number of fragment generation samples                 [ default:    100 ]\r\n",
      "  --num-frag-assign-draws      Number of fragment assignment samples per generation  [ default:     50 ]\r\n",
      "  --max-frag-multihits         Maximum number of alignments allowed per fragment     [ default: unlim  ]\r\n",
      "  --min-outlier-p              Min replicate p value to admit for testing            [    DEPRECATED   ]\r\n",
      "  --min-reps-for-js-test       Replicates needed for relative isoform shift testing  [ default:      3 ]\r\n",
      "  --no-effective-length-correction   No effective length correction                  [ default:  FALSE ]\r\n",
      "  --no-length-correction       No length correction                                  [ default:  FALSE ]\r\n",
      "  -N/--upper-quartile-norm     Deprecated, use --library-norm-method                 [    DEPRECATED   ]\r\n",
      "  --geometric-norm             Deprecated, use --library-norm-method                 [    DEPRECATED   ]\r\n",
      "  --raw-mapped-norm            Deprecated, use --library-norm-method                 [    DEPRECATED   ]\r\n",
      "  --poisson-dispersion         Deprecated, use --dispersion-method                   [    DEPRECATED   ]\r\n",
      "\r\n",
      "Debugging use only:\r\n",
      "  --read-skip-fraction         Skip a random subset of reads this size               [ default:    0.0 ]\r\n",
      "  --no-read-pairs              Break all read pairs                                  [ default:  FALSE ]\r\n",
      "  --trim-read-length           Trim reads to be this long (keep 5' end)              [ default:   none ]\r\n",
      "  --no-scv-correction          Disable SCV correction                                [ default:  FALSE ]\r\n",
      "\r\n",
      "Supported library types:\r\n",
      "\tff-firststrand\r\n",
      "\tff-secondstrand\r\n",
      "\tff-unstranded\r\n",
      "\tfr-firststrand\r\n",
      "\tfr-secondstrand\r\n",
      "\tfr-unstranded (default)\r\n",
      "\ttransfrags\r\n",
      "\r\n",
      "Supported dispersion methods:\r\n",
      "\tblind\r\n",
      "\tper-condition\r\n",
      "\tpoisson\r\n",
      "\tpooled (default)\r\n",
      "\r\n",
      "Supported library normalization methods:\r\n",
      "\tclassic-fpkm\r\n",
      "\tgeometric (default)\r\n",
      "\tgeometric\r\n",
      "\tquartile\r\n"
     ]
    }
   ],
   "source": [
    "! cuffdiff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir -p ../cuffdiff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Could not connect to update server to verify current version. Please check at the Cufflinks website (http://cufflinks.cbcb.umd.edu).\n",
      "[18:58:21] Loading reference annotation and sequence.\n",
      "Warning: couldn't find fasta record for 'chr10_GL383545v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr10_GL383546v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr10_KI270824v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr10_KI270825v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_JH159136v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_JH159137v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_KI270827v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_KI270829v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_KI270830v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_KI270831v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_KI270832v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_KI270902v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_KI270903v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_KI270927v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_GL383550v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_GL383551v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_GL383553v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_GL877875v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_GL877876v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_KI270833v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_KI270834v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_KI270835v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_KI270837v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_KI270904v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr13_KI270838v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr13_KI270839v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr13_KI270840v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr13_KI270842v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr14_KI270844v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr14_KI270845v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr14_KI270846v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr14_KI270847v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_GL383554v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_GL383555v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_KI270848v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_KI270849v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_KI270850v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_KI270851v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_KI270852v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_KI270905v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_KI270906v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr16_GL383556v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr16_GL383557v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr16_KI270853v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr16_KI270854v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr16_KI270855v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr16_KI270856v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_GL000258v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_GL383563v3_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_GL383564v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_GL383566v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_JH159146v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_JH159147v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_JH159148v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_KI270857v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_KI270860v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_KI270861v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_KI270862v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_KI270907v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_KI270908v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_KI270909v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_KI270910v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr18_GL383567v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr18_GL383571v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr18_GL383572v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr18_KI270863v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr18_KI270911v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr18_KI270912v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL000209v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL383573v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL383574v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL383575v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL383576v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL949746v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL949747v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL949748v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL949749v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL949750v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL949751v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL949752v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL949753v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270865v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270866v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270867v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270868v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270882v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270883v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270884v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270885v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270886v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270887v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270888v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270889v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270890v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270891v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270914v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270915v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270916v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270917v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270918v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270919v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270920v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270921v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270922v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270923v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270929v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270930v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270931v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270932v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270933v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270938v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_GL383518v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_GL383519v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_GL383520v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_KI270759v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_KI270761v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_KI270762v1_alt'!\n",
      "This contig will not be bias corrected.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: couldn't find fasta record for 'chr1_KI270763v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_KI270765v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_KI270766v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_KI270892v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr20_KI270869v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr20_KI270870v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr20_KI270871v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr21_GL383579v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr21_GL383580v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr21_GL383581v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr21_KI270872v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr21_KI270873v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_GL383582v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_GL383583v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_KB663609v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_KI270875v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_KI270876v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_KI270877v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_KI270878v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_KI270879v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_KI270928v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_GL383521v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_GL383522v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_GL582966v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270767v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270768v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270769v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270770v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270772v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270773v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270774v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270776v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270893v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270894v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_GL383526v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_JH636055v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270777v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270779v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270780v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270782v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270783v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270784v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270895v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270924v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270934v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270935v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270936v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270937v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr4_GL000257v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr4_GL383527v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr4_KI270786v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr4_KI270789v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr4_KI270790v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr4_KI270896v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr4_KI270925v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_GL339449v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_GL949742v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_KI270791v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_KI270792v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_KI270793v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_KI270794v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_KI270795v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_KI270796v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_KI270897v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_KI270898v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_GL000250v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_GL000251v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_GL000252v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_GL000253v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_GL000254v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_GL000255v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_GL000256v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_GL383533v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_KI270758v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_KI270797v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_KI270798v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_KI270800v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_KI270801v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr7_GL383534v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr7_KI270803v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr7_KI270804v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr7_KI270806v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr7_KI270808v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr7_KI270809v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr7_KI270899v1_alt'!\n",
      "This contig will not be bias corrected.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: couldn't find fasta record for 'chr8_KI270811v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270812v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270813v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270814v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270815v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270816v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270817v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270818v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270819v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270820v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270821v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270822v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270900v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270926v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr9_GL383540v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr9_GL383541v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr9_KI270823v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chrX_KI270880v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chrX_KI270881v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chrX_KI270913v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "[18:58:58] Inspecting maps and determining fragment length distributions.\n",
      "[19:36:24] Modeling fragment count overdispersion.\n",
      "[19:37:32] Modeling fragment count overdispersion.\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 52101370.14\n",
      ">\tRaw Map Mass: 51716986.47\n",
      ">\tNumber of Multi-Reads: 686511 (with 1698915 total hits)\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 151.88\n",
      ">\t           Estimated Std Dev: 55.05\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 52101370.14\n",
      ">\tRaw Map Mass: 58200743.15\n",
      ">\tNumber of Multi-Reads: 736552 (with 1825077 total hits)\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 160.84\n",
      ">\t           Estimated Std Dev: 58.18\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 52101370.14\n",
      ">\tRaw Map Mass: 41277762.88\n",
      ">\tNumber of Multi-Reads: 510013 (with 1287989 total hits)\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 173.73\n",
      ">\t           Estimated Std Dev: 65.14\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 52101370.14\n",
      ">\tRaw Map Mass: 44710193.07\n",
      ">\tNumber of Multi-Reads: 753717 (with 1730869 total hits)\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 151.03\n",
      ">\t           Estimated Std Dev: 57.04\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 52101370.14\n",
      ">\tRaw Map Mass: 54362219.21\n",
      ">\tNumber of Multi-Reads: 813050 (with 1841792 total hits)\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 165.08\n",
      ">\t           Estimated Std Dev: 55.06\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 52101370.14\n",
      ">\tRaw Map Mass: 67321974.12\n",
      ">\tNumber of Multi-Reads: 1068413 (with 2423096 total hits)\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 161.12\n",
      ">\t           Estimated Std Dev: 55.51\n",
      "[19:38:43] Calculating preliminary abundance estimates\n",
      "> Processed 27980 loci.                        [*************************] 100%\n",
      "[21:37:11] Learning bias parameters.\n",
      "[22:05:34] Testing for differential expression and regulation in locus.\n",
      "> Processed 27980 loci.                        [*************************] 100%\n",
      "Performed 55462 isoform-level transcription difference tests\n",
      "Performed 28564 tss-level transcription difference tests\n",
      "Performed 17442 gene-level transcription difference tests\n",
      "Performed 23289 CDS-level transcription difference tests\n",
      "Performed 10914 splicing tests\n",
      "Performed 6535 promoter preference tests\n",
      "Performing 5988 relative CDS output tests\n",
      "Writing isoform-level FPKM tracking\n",
      "Writing TSS group-level FPKM tracking\n",
      "Writing gene-level FPKM tracking\n",
      "Writing CDS-level FPKM tracking\n",
      "Writing isoform-level count tracking\n",
      "Writing TSS group-level count tracking\n",
      "Writing gene-level count tracking\n",
      "Writing CDS-level count tracking\n",
      "Writing isoform-level read group tracking\n",
      "Writing TSS group-level read group tracking\n",
      "Writing gene-level read group tracking\n",
      "Writing CDS-level read group tracking\n",
      "Writing read group info\n",
      "Writing run info\n"
     ]
    }
   ],
   "source": [
    "! cuffdiff -o ../cuffdiff/diff_out -b ../../data/preprocessing_fastp/genome.fa -p 8 \\\n",
    "           -L T,U -u ./merged_asm/merged.gtf \\\n",
    "           ../tophat/T_SRR5272677_thout/accepted_hits.bam,../tophat/T_SRR5272678_thout/accepted_hits.bam,../tophat/T_SRR5272679_thout/accepted_hits.bam \\\n",
    "           ../tophat/U_SRR5272674_thout/accepted_hits.bam,../tophat/U_SRR5272675_thout/accepted_hits.bam,../tophat/U_SRR5272676_thout/accepted_hits.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 162692\r\n",
      "  392 -rw-r--r-- 1 cv cv   401149 Feb 28 14:39 bias_params.info\r\n",
      " 2704 -rw-r--r-- 1 cv cv  2767992 Feb 28 14:39 cds.count_tracking\r\n",
      " 1600 -rw-r--r-- 1 cv cv  1634886 Feb 28 14:39 cds.diff\r\n",
      " 5076 -rw-r--r-- 1 cv cv  5195782 Feb 28 14:39 cds_exp.diff\r\n",
      " 5652 -rw-r--r-- 1 cv cv  5783723 Feb 28 14:39 cds.fpkm_tracking\r\n",
      "11916 -rw-r--r-- 1 cv cv 12198064 Feb 28 14:39 cds.read_group_tracking\r\n",
      " 3732 -rw-r--r-- 1 cv cv  3820390 Feb 28 14:39 gene_exp.diff\r\n",
      " 1996 -rw-r--r-- 1 cv cv  2043006 Feb 28 14:39 genes.count_tracking\r\n",
      " 4328 -rw-r--r-- 1 cv cv  4430483 Feb 28 14:39 genes.fpkm_tracking\r\n",
      " 8944 -rw-r--r-- 1 cv cv  9156977 Feb 28 14:39 genes.read_group_tracking\r\n",
      "13820 -rw-r--r-- 1 cv cv 14151027 Feb 28 14:39 isoform_exp.diff\r\n",
      " 7688 -rw-r--r-- 1 cv cv  7872392 Feb 28 14:39 isoforms.count_tracking\r\n",
      "17276 -rw-r--r-- 1 cv cv 17686778 Feb 28 14:39 isoforms.fpkm_tracking\r\n",
      "35884 -rw-r--r-- 1 cv cv 36743825 Feb 28 14:39 isoforms.read_group_tracking\r\n",
      " 3088 -rw-r--r-- 1 cv cv  3161303 Feb 28 14:39 promoters.diff\r\n",
      "    4 -rw-r--r-- 1 cv cv      590 Feb 28 14:39 read_groups.info\r\n",
      "    4 -rw-r--r-- 1 cv cv      473 Feb 28 14:39 run.info\r\n",
      " 4964 -rw-r--r-- 1 cv cv  5082441 Feb 28 14:39 splicing.diff\r\n",
      " 6068 -rw-r--r-- 1 cv cv  6212411 Feb 28 14:39 tss_group_exp.diff\r\n",
      " 3236 -rw-r--r-- 1 cv cv  3310028 Feb 28 14:39 tss_groups.count_tracking\r\n",
      " 6668 -rw-r--r-- 1 cv cv  6826315 Feb 28 14:39 tss_groups.fpkm_tracking\r\n",
      "14424 -rw-r--r-- 1 cv cv 14766860 Feb 28 14:39 tss_groups.read_group_tracking\r\n",
      " 3228 -rw-r--r-- 1 cv cv  3301611 Feb 27 19:38 var_model.info\r\n"
     ]
    }
   ],
   "source": [
    "! ls -ls ../cuffdiff/diff_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/cv/RNA-seq/analysis/cufflinks2\n"
     ]
    }
   ],
   "source": [
    "cd ../../analysis/cufflinks2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C1_R1_SRR5272674_clout\tC1_R3_SRR5272676_clout\tC2_R2_SRR5272678_clout\r\n",
      "C1_R2_SRR5272675_clout\tC2_R1_SRR5272677_clout\tC2_R3_SRR5272679_clout\r\n"
     ]
    }
   ],
   "source": [
    "! ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "! ls *_clout/transcripts.gtf > assemblies2.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C1_R1_SRR5272674_clout/transcripts.gtf\r\n",
      "C1_R2_SRR5272675_clout/transcripts.gtf\r\n",
      "C1_R3_SRR5272676_clout/transcripts.gtf\r\n",
      "C2_R1_SRR5272677_clout/transcripts.gtf\r\n",
      "C2_R2_SRR5272678_clout/transcripts.gtf\r\n",
      "C2_R3_SRR5272679_clout/transcripts.gtf\r\n"
     ]
    }
   ],
   "source": [
    "! cat assemblies2.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C1_R1_SRR5272674_thout\tC1_R3_SRR5272676_thout\tC2_R2_SRR5272678_thout\r\n",
      "C1_R2_SRR5272675_thout\tC2_R1_SRR5272677_thout\tC2_R3_SRR5272679_thout\r\n"
     ]
    }
   ],
   "source": [
    "! ls ../../analysis/tophat2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[Thu Mar  3 00:09:52 2022] Beginning transcriptome assembly merge\n",
      "-------------------------------------------\n",
      "\n",
      "[Thu Mar  3 00:09:52 2022] Preparing output location ./merged_asm/\n",
      "[Thu Mar  3 00:09:57 2022] Converting GTF files to SAM\n",
      "[00:09:57] Loading reference annotation.\n",
      "[00:09:59] Loading reference annotation.\n",
      "[00:10:00] Loading reference annotation.\n",
      "[00:10:02] Loading reference annotation.\n",
      "[00:10:03] Loading reference annotation.\n",
      "[00:10:05] Loading reference annotation.\n",
      "[Thu Mar  3 00:10:07 2022] Quantitating transcripts\n",
      "Warning: Could not connect to update server to verify current version. Please check at the Cufflinks website (http://cufflinks.cbcb.umd.edu).\n",
      "Command line:\n",
      "cufflinks -o ./merged_asm/ -F 0.05 -g ../../data/preprocessing_fastp/genes.gtf -q --overhang-tolerance 200 --library-type=transfrags -A 0.0 --min-frags-per-transfrag 0 --no-5-extend -p 8 ./merged_asm/tmp/mergeSam_fileexHs8S \n",
      "[bam_header_read] EOF marker is absent. The input is probably truncated.\n",
      "[bam_header_read] invalid BAM binary header (this is not a BAM file).\n",
      "File ./merged_asm/tmp/mergeSam_fileexHs8S doesn't appear to be a valid BAM file, trying SAM...\n",
      "[00:10:08] Loading reference annotation.\n",
      "[00:10:13] Inspecting reads and determining fragment length distribution.\n",
      "Processed 28737 loci.                       \n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 195628.00\n",
      ">\tRaw Map Mass: 195628.00\n",
      ">\tFragment Length Distribution: Truncated Gaussian (default)\n",
      ">\t              Default Mean: 200\n",
      ">\t           Default Std Dev: 80\n",
      "[00:10:15] Assembling transcripts and estimating abundances.\n",
      "Processed 28737 loci.                       \n",
      "[Thu Mar  3 00:12:52 2022] Comparing against reference file ../../data/preprocessing_fastp/genes.gtf\n",
      "Warning: Could not connect to update server to verify current version. Please check at the Cufflinks website (http://cufflinks.cbcb.umd.edu).\n",
      "Warning: couldn't find fasta record for 'chr10_GL383545v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr10_GL383546v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr10_KI270824v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr10_KI270825v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_JH159136v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_JH159137v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270827v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270829v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270830v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270831v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270832v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270902v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270903v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270927v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL383550v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL383551v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL383553v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL877875v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL877876v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270833v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270834v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270835v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270837v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270904v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr13_KI270838v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr13_KI270839v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr13_KI270840v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr13_KI270842v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr14_KI270844v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr14_KI270845v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr14_KI270846v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr14_KI270847v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_GL383554v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_GL383555v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270848v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270849v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270850v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270851v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270852v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270905v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270906v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_GL383556v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_GL383557v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_KI270853v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_KI270854v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_KI270855v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_KI270856v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_GL000258v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_GL383563v3_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_GL383564v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_GL383566v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_JH159146v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_JH159147v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_JH159148v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270857v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270860v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270861v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270862v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270907v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270908v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270909v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270910v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_GL383567v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_GL383571v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_GL383572v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_KI270863v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_KI270911v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_KI270912v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL000209v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL383573v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL383574v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL383575v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL383576v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949746v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949747v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949748v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949749v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949750v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949751v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949752v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949753v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270865v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270866v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270867v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270868v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270882v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270883v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270884v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270885v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270886v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270887v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270888v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270889v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270890v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270891v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270914v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270915v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270916v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270917v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270918v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270919v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270920v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270921v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270922v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270923v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270929v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270930v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270931v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270932v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270933v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270938v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_GL383518v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_GL383519v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_GL383520v2_alt'!\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: couldn't find fasta record for 'chr1_KI270759v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270761v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270762v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270763v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270765v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270766v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270892v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr20_KI270869v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr20_KI270870v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr20_KI270871v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_GL383579v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_GL383580v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_GL383581v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_KI270872v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_KI270873v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_GL383582v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_GL383583v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KB663609v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270875v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270876v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270877v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270878v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270879v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270928v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_GL383521v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_GL383522v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_GL582966v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270767v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270768v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270769v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270770v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270772v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270773v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270774v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270776v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270893v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270894v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_GL383526v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_JH636055v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270777v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270779v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270780v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270782v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270783v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270784v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270895v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270924v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270934v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270935v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270936v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270937v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_GL000257v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_GL383527v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270786v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270789v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270790v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270896v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270925v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_GL339449v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_GL949742v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270791v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270792v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270793v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270794v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270795v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270796v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270897v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270898v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000250v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000251v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000252v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000253v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000254v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000255v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000256v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL383533v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270758v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270797v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270798v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270800v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270801v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_GL383534v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270803v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270804v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270806v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270808v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270809v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270899v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270811v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270812v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270813v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270814v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270815v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270816v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270817v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270818v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270819v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270820v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270821v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270822v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270900v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270926v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr9_GL383540v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr9_GL383541v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr9_KI270823v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chrX_KI270880v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chrX_KI270881v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chrX_KI270913v1_alt'!\n",
      "[Thu Mar  3 00:13:54 2022] Comparing against reference file ../../data/preprocessing_fastp/genes.gtf\n",
      "Warning: Could not connect to update server to verify current version. Please check at the Cufflinks website (http://cufflinks.cbcb.umd.edu).\n",
      "Warning: couldn't find fasta record for 'chr10_GL383545v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr10_GL383546v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr10_KI270824v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr10_KI270825v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_JH159136v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_JH159137v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270827v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270829v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270830v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270831v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270832v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270902v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270903v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr11_KI270927v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL383550v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL383551v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL383553v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL877875v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_GL877876v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270833v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270834v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270835v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270837v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr12_KI270904v1_alt'!\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: couldn't find fasta record for 'chr13_KI270838v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr13_KI270839v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr13_KI270840v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr13_KI270842v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr14_KI270844v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr14_KI270845v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr14_KI270846v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr14_KI270847v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_GL383554v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_GL383555v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270848v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270849v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270850v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270851v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270852v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270905v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr15_KI270906v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_GL383556v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_GL383557v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_KI270853v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_KI270854v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_KI270855v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr16_KI270856v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_GL000258v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_GL383563v3_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_GL383564v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_GL383566v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_JH159146v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_JH159147v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_JH159148v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270857v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270860v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270861v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270862v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270907v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270908v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270909v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr17_KI270910v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_GL383567v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_GL383571v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_GL383572v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_KI270863v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_KI270911v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr18_KI270912v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL000209v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL383573v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL383574v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL383575v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL383576v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949746v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949747v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949748v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949749v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949750v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949751v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949752v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_GL949753v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270865v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270866v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270867v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270868v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270882v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270883v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270884v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270885v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270886v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270887v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270888v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270889v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270890v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270891v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270914v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270915v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270916v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270917v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270918v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270919v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270920v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270921v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270922v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270923v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270929v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270930v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270931v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270932v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270933v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr19_KI270938v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_GL383518v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_GL383519v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_GL383520v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270759v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270761v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270762v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270763v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270765v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270766v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr1_KI270892v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr20_KI270869v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr20_KI270870v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr20_KI270871v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_GL383579v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_GL383580v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_GL383581v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_KI270872v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr21_KI270873v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_GL383582v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_GL383583v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KB663609v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270875v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270876v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270877v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270878v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270879v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr22_KI270928v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_GL383521v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_GL383522v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_GL582966v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270767v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270768v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270769v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270770v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270772v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270773v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270774v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270776v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270893v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr2_KI270894v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_GL383526v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_JH636055v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270777v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270779v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270780v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270782v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270783v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270784v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270895v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270924v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270934v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270935v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270936v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr3_KI270937v1_alt'!\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: couldn't find fasta record for 'chr4_GL000257v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_GL383527v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270786v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270789v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270790v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270896v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr4_KI270925v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_GL339449v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_GL949742v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270791v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270792v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270793v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270794v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270795v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270796v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270897v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr5_KI270898v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000250v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000251v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000252v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000253v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000254v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000255v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL000256v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_GL383533v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270758v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270797v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270798v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270800v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr6_KI270801v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_GL383534v2_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270803v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270804v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270806v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270808v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270809v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr7_KI270899v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270811v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270812v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270813v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270814v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270815v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270816v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270817v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270818v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270819v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270820v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270821v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270822v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270900v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr8_KI270926v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr9_GL383540v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr9_GL383541v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chr9_KI270823v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chrX_KI270880v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chrX_KI270881v1_alt'!\n",
      "Warning: couldn't find fasta record for 'chrX_KI270913v1_alt'!\n"
     ]
    }
   ],
   "source": [
    "! cuffmerge -g ../../data/preprocessing_fastp/genes.gtf \\\n",
    "            -s ../../data/preprocessing_fastp/genome.fa \\\n",
    "            -p 8 assemblies2.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir -p ../cuffdiff2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrY\tCufflinks\texon\t25527647\t25527799\t.\t-\t.\tgene_id \"XLOC_037165\"; transcript_id \"TCONS_00135945\"; exon_number \"4\"; gene_name \"LOC102724004\"; oId \"rna87623\"; nearest_ref \"rna87623\"; class_code \"=\"; tss_id \"TSS62399\";\r\n",
      "chrY\tCufflinks\texon\t25517743\t25517798\t.\t-\t.\tgene_id \"XLOC_037165\"; transcript_id \"TCONS_00135944\"; exon_number \"1\"; gene_name \"LOC102724004\"; oId \"rna87622\"; nearest_ref \"rna87622\"; class_code \"=\"; tss_id \"TSS62399\";\r\n",
      "chrY\tCufflinks\texon\t25524026\t25524110\t.\t-\t.\tgene_id \"XLOC_037165\"; transcript_id \"TCONS_00135944\"; exon_number \"2\"; gene_name \"LOC102724004\"; oId \"rna87622\"; nearest_ref \"rna87622\"; class_code \"=\"; tss_id \"TSS62399\";\r\n",
      "chrY\tCufflinks\texon\t25527647\t25527799\t.\t-\t.\tgene_id \"XLOC_037165\"; transcript_id \"TCONS_00135944\"; exon_number \"3\"; gene_name \"LOC102724004\"; oId \"rna87622\"; nearest_ref \"rna87622\"; class_code \"=\"; tss_id \"TSS62399\";\r\n",
      "chrY\tCufflinks\texon\t25517743\t25517798\t.\t-\t.\tgene_id \"XLOC_037165\"; transcript_id \"TCONS_00135943\"; exon_number \"1\"; gene_name \"LOC102724004\"; oId \"rna87621\"; nearest_ref \"rna87621\"; class_code \"=\"; tss_id \"TSS62399\";\r\n",
      "chrY\tCufflinks\texon\t25524026\t25524110\t.\t-\t.\tgene_id \"XLOC_037165\"; transcript_id \"TCONS_00135943\"; exon_number \"2\"; gene_name \"LOC102724004\"; oId \"rna87621\"; nearest_ref \"rna87621\"; class_code \"=\"; tss_id \"TSS62399\";\r\n",
      "chrY\tCufflinks\texon\t25525204\t25525288\t.\t-\t.\tgene_id \"XLOC_037165\"; transcript_id \"TCONS_00135943\"; exon_number \"3\"; gene_name \"LOC102724004\"; oId \"rna87621\"; nearest_ref \"rna87621\"; class_code \"=\"; tss_id \"TSS62399\";\r\n",
      "chrY\tCufflinks\texon\t25527649\t25527799\t.\t-\t.\tgene_id \"XLOC_037165\"; transcript_id \"TCONS_00135943\"; exon_number \"4\"; gene_name \"LOC102724004\"; oId \"rna87621\"; nearest_ref \"rna87621\"; class_code \"=\"; tss_id \"TSS62399\";\r\n",
      "chrY\tCufflinks\texon\t56854931\t56855161\t.\t.\t.\tgene_id \"XLOC_037166\"; transcript_id \"TCONS_00135946\"; exon_number \"1\"; oId \"CUFF.22256.1\"; class_code \"u\"; tss_id \"TSS62400\";\r\n",
      "chrY\tCufflinks\texon\t56856859\t56857612\t.\t.\t.\tgene_id \"XLOC_037167\"; transcript_id \"TCONS_00135947\"; exon_number \"1\"; oId \"CUFF.22257.1\"; class_code \"u\"; tss_id \"TSS62401\";\r\n"
     ]
    }
   ],
   "source": [
    "! tail -n 10 ./merged_asm/merged.gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/cv/RNA-seq/analysis/cufflinks2\r\n"
     ]
    }
   ],
   "source": [
    "! pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Could not connect to update server to verify current version. Please check at the Cufflinks website (http://cufflinks.cbcb.umd.edu).\n",
      "[15:36:58] Loading reference annotation and sequence.\n",
      "Warning: couldn't find fasta record for 'chr10_GL383545v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr10_GL383546v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr10_KI270824v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr10_KI270825v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_JH159136v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_JH159137v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_KI270827v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_KI270829v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_KI270830v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_KI270831v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_KI270832v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_KI270902v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_KI270903v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr11_KI270927v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_GL383550v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_GL383551v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_GL383553v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_GL877875v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_GL877876v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_KI270833v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_KI270834v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_KI270835v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_KI270837v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr12_KI270904v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr13_KI270838v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr13_KI270839v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr13_KI270840v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr13_KI270842v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr14_KI270844v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr14_KI270845v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr14_KI270846v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr14_KI270847v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_GL383554v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_GL383555v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_KI270848v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_KI270849v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_KI270850v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_KI270851v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_KI270852v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_KI270905v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr15_KI270906v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr16_GL383556v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr16_GL383557v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr16_KI270853v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr16_KI270854v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr16_KI270855v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr16_KI270856v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_GL000258v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_GL383563v3_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_GL383564v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_GL383566v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_JH159146v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_JH159147v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_JH159148v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_KI270857v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_KI270860v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_KI270861v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_KI270862v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_KI270907v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_KI270908v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_KI270909v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr17_KI270910v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr18_GL383567v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr18_GL383571v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr18_GL383572v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr18_KI270863v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr18_KI270911v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr18_KI270912v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL000209v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL383573v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL383574v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL383575v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL383576v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL949746v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL949747v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL949748v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL949749v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL949750v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL949751v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL949752v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_GL949753v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270865v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270866v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270867v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270868v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270882v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270883v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270884v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270885v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270886v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270887v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270888v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270889v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270890v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270891v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270914v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270915v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270916v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270917v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270918v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270919v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270920v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270921v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270922v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270923v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270929v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270930v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270931v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270932v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270933v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr19_KI270938v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_GL383518v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_GL383519v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_GL383520v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_KI270759v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_KI270761v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_KI270762v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_KI270763v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_KI270765v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_KI270766v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr1_KI270892v1_alt'!\n",
      "This contig will not be bias corrected.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: couldn't find fasta record for 'chr20_KI270869v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr20_KI270870v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr20_KI270871v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr21_GL383579v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr21_GL383580v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr21_GL383581v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr21_KI270872v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr21_KI270873v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_GL383582v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_GL383583v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_KB663609v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_KI270875v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_KI270876v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_KI270877v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_KI270878v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_KI270879v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr22_KI270928v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_GL383521v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_GL383522v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_GL582966v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270767v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270768v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270769v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270770v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270772v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270773v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270774v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270776v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270893v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr2_KI270894v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_GL383526v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_JH636055v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270777v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270779v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270780v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270782v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270783v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270784v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270895v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270924v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270934v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270935v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270936v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr3_KI270937v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr4_GL000257v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr4_GL383527v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr4_KI270786v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr4_KI270789v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr4_KI270790v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr4_KI270896v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr4_KI270925v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_GL339449v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_GL949742v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_KI270791v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_KI270792v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_KI270793v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_KI270794v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_KI270795v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_KI270796v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_KI270897v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr5_KI270898v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_GL000250v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_GL000251v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_GL000252v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_GL000253v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_GL000254v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_GL000255v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_GL000256v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_GL383533v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_KI270758v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_KI270797v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_KI270798v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_KI270800v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr6_KI270801v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr7_GL383534v2_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr7_KI270803v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr7_KI270804v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr7_KI270806v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr7_KI270808v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr7_KI270809v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr7_KI270899v1_alt'!\n",
      "This contig will not be bias corrected.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: couldn't find fasta record for 'chr8_KI270811v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270812v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270813v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270814v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270815v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270816v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270817v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270818v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270819v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270820v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270821v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270822v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270900v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr8_KI270926v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr9_GL383540v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr9_GL383541v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chr9_KI270823v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chrX_KI270880v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chrX_KI270881v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "Warning: couldn't find fasta record for 'chrX_KI270913v1_alt'!\n",
      "This contig will not be bias corrected.\n",
      "[15:37:35] Inspecting maps and determining fragment length distributions.\n",
      "[16:14:46] Modeling fragment count overdispersion.\n",
      "[16:15:55] Modeling fragment count overdispersion.\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 52100656.33\n",
      ">\tRaw Map Mass: 44710143.07\n",
      ">\tNumber of Multi-Reads: 753717 (with 1730869 total hits)\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 151.03\n",
      ">\t           Estimated Std Dev: 57.04\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 52100656.33\n",
      ">\tRaw Map Mass: 54362143.21\n",
      ">\tNumber of Multi-Reads: 813050 (with 1841792 total hits)\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 165.08\n",
      ">\t           Estimated Std Dev: 55.06\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 52100656.33\n",
      ">\tRaw Map Mass: 67321884.62\n",
      ">\tNumber of Multi-Reads: 1068413 (with 2423096 total hits)\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 161.12\n",
      ">\t           Estimated Std Dev: 55.51\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 52100656.33\n",
      ">\tRaw Map Mass: 51716885.47\n",
      ">\tNumber of Multi-Reads: 686511 (with 1698915 total hits)\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 151.88\n",
      ">\t           Estimated Std Dev: 55.05\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 52100656.33\n",
      ">\tRaw Map Mass: 58200605.15\n",
      ">\tNumber of Multi-Reads: 736552 (with 1825077 total hits)\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 160.84\n",
      ">\t           Estimated Std Dev: 58.18\n",
      "> Map Properties:\n",
      ">\tNormalized Map Mass: 52100656.33\n",
      ">\tRaw Map Mass: 41277688.38\n",
      ">\tNumber of Multi-Reads: 510013 (with 1287989 total hits)\n",
      ">\tFragment Length Distribution: Empirical (learned)\n",
      ">\t              Estimated Mean: 173.73\n",
      ">\t           Estimated Std Dev: 65.14\n",
      "[16:17:05] Calculating preliminary abundance estimates\n",
      "> Processed 27980 loci.                        [*************************] 100%\n",
      "[18:13:46] Learning bias parameters.\n",
      "[18:42:14] Testing for differential expression and regulation in locus.\n",
      "> Processed 27980 loci.                        [*************************] 100%\n",
      "Performed 55456 isoform-level transcription difference tests\n",
      "Performed 28556 tss-level transcription difference tests\n",
      "Performed 17435 gene-level transcription difference tests\n",
      "Performed 23289 CDS-level transcription difference tests\n",
      "Performed 10915 splicing tests\n",
      "Performed 6535 promoter preference tests\n",
      "Performing 5988 relative CDS output tests\n",
      "Writing isoform-level FPKM tracking\n",
      "Writing TSS group-level FPKM tracking\n",
      "Writing gene-level FPKM tracking\n",
      "Writing CDS-level FPKM tracking\n",
      "Writing isoform-level count tracking\n",
      "Writing TSS group-level count tracking\n",
      "Writing gene-level count tracking\n",
      "Writing CDS-level count tracking\n",
      "Writing isoform-level read group tracking\n",
      "Writing TSS group-level read group tracking\n",
      "Writing gene-level read group tracking\n",
      "Writing CDS-level read group tracking\n",
      "Writing read group info\n",
      "Writing run info\n"
     ]
    }
   ],
   "source": [
    "! cuffdiff -o ../cuffdiff2/diff_out -b ../../data/preprocessing_fastp/genome.fa -p 8 \\\n",
    "           -L C1,C2 -u ./merged_asm/merged.gtf \\\n",
    "           ../tophat2/C1_R1_SRR5272674_thout/accepted_hits.bam,../tophat2/C1_R2_SRR5272675_thout/accepted_hits.bam,../tophat2/C1_R3_SRR5272676_thout/accepted_hits.bam \\\n",
    "           ../tophat2/C2_R1_SRR5272677_thout/accepted_hits.bam,../tophat2/C2_R2_SRR5272678_thout/accepted_hits.bam,../tophat2/C2_R3_SRR5272679_thout/accepted_hits.bam \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
