<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Gene Expression Mixed Effects Modeling with limma</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="limma_mixed_effects_model_files/libs/clipboard/clipboard.min.js"></script>
<script src="limma_mixed_effects_model_files/libs/quarto-html/quarto.js"></script>
<script src="limma_mixed_effects_model_files/libs/quarto-html/popper.min.js"></script>
<script src="limma_mixed_effects_model_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="limma_mixed_effects_model_files/libs/quarto-html/anchor.min.js"></script>
<link href="limma_mixed_effects_model_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="limma_mixed_effects_model_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="limma_mixed_effects_model_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="limma_mixed_effects_model_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="limma_mixed_effects_model_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Gene Expression Mixed Effects Modeling with limma</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="load-libraries" class="level2">
<h2 class="anchored" data-anchor-id="load-libraries">Load libraries</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readxl)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(limma)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-data" class="level2">
<h2 class="anchored" data-anchor-id="load-data">Load Data</h2>
<p><em>Toy study of treatment effects with gene expression measurements taken from patients at multiple time points. Three patients were administered treatment D1, and another three were administered treatment D2. Gene expression measurements were taken for the patients at two timepoints, T1 (baseline) and T2.</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> readxl<span class="sc">::</span><span class="fu">read_excel</span>(<span class="st">'./data/expression.xlsx'</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>dat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12 × 4
   expression id       treatment timepoint
        &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;     &lt;chr&gt;    
 1       1.01 PATIENT1 D1        T1       
 2       1.04 PATIENT2 D1        T1       
 3       1.04 PATIENT3 D1        T1       
 4       1.99 PATIENT1 D1        T2       
 5       2.36 PATIENT2 D1        T2       
 6       2    PATIENT3 D1        T2       
 7       2.89 PATIENT4 D2        T1       
 8       3.12 PATIENT5 D2        T1       
 9       2.98 PATIENT6 D2        T1       
10       5    PATIENT4 D2        T2       
11       4.92 PATIENT5 D2        T2       
12       4.78 PATIENT6 D2        T2       </code></pre>
</div>
</div>
</section>
<section id="mixed-effects-modeling-with-limma" class="level2">
<h2 class="anchored" data-anchor-id="mixed-effects-modeling-with-limma">Mixed Effects Modeling with limma</h2>
<p><em>What is the difference between timepoint T2 and T1 in treatment D1? What is the difference between timepoint T2 and T1 in treatment D2? What is the overall difference between treatment D1 and D2?</em></p>
<p><em>Considering that specific differences between patients are not of direct interest, relative to differences in treatments, and at different time points within treatment, patient ids can be included as random effects in a mixed effects model. Individual patient effects should still be accounted for when calculating the relative difference between T1 and T2 timepoints.</em></p>
<p>Create design matrix such that <code>treatment</code> and <code>timepoint</code> are fitted as fixed effects in the model, and incorporate patient <code>ids</code> as a random effect.</p>
<p>Merge <code>treatment</code> and <code>timepoint</code> factors into a single <code>group</code> factor.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> dat <span class="sc">%&gt;%</span> </span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unite</span>(<span class="st">"group"</span>, treatment<span class="sc">:</span>timepoint, <span class="at">sep=</span><span class="st">"_"</span>, <span class="at">remove=</span><span class="cn">FALSE</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>dat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12 × 5
   expression id       group treatment timepoint
        &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt;     &lt;chr&gt;    
 1       1.01 PATIENT1 D1_T1 D1        T1       
 2       1.04 PATIENT2 D1_T1 D1        T1       
 3       1.04 PATIENT3 D1_T1 D1        T1       
 4       1.99 PATIENT1 D1_T2 D1        T2       
 5       2.36 PATIENT2 D1_T2 D1        T2       
 6       2    PATIENT3 D1_T2 D1        T2       
 7       2.89 PATIENT4 D2_T1 D2        T1       
 8       3.12 PATIENT5 D2_T1 D2        T1       
 9       2.98 PATIENT6 D2_T1 D2        T1       
10       5    PATIENT4 D2_T2 D2        T2       
11       4.92 PATIENT5 D2_T2 D2        T2       
12       4.78 PATIENT6 D2_T2 D2        T2       </code></pre>
</div>
</div>
<section id="create-design-matrix" class="level3">
<h3 class="anchored" data-anchor-id="create-design-matrix">Create design matrix</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>design <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span><span class="dv">0</span><span class="sc">+</span>group, <span class="at">data=</span>dat)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>design</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   groupD1_T1 groupD1_T2 groupD2_T1 groupD2_T2
1           1          0          0          0
2           1          0          0          0
3           1          0          0          0
4           0          1          0          0
5           0          1          0          0
6           0          1          0          0
7           0          0          1          0
8           0          0          1          0
9           0          0          1          0
10          0          0          0          1
11          0          0          0          1
12          0          0          0          1
attr(,"assign")
[1] 1 1 1 1
attr(,"contrasts")
attr(,"contrasts")$group
[1] "contr.treatment"</code></pre>
</div>
</div>
</section>
<section id="approximation-of-mixed-effects-in-limma" class="level3">
<h3 class="anchored" data-anchor-id="approximation-of-mixed-effects-in-limma">Approximation of mixed effects in limma</h3>
<p>Estimate the mean correlation of gene expression between repeated patient measurements using the <code>duplicateCorrelation</code> function in <strong>limma</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>corr <span class="ot">&lt;-</span>  <span class="fu">duplicateCorrelation</span>(<span class="at">object=</span>dat<span class="sc">$</span>expression, </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>                              <span class="at">design=</span>design, </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">block=</span>dat<span class="sc">$</span>id)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>corr<span class="sc">$</span>consensus.correlation </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] -0.03401039</code></pre>
</div>
</div>
<p>The correlation between measurements taken from the same patient is estimated as -0.03, which is a very small correlation value that could be removed from the model as its contribution to the overall model fit would be minor. Range of correlation estimates {‘very small’:&lt;0.01, ‘high’:0.7-0.9}</p>
</section>
<section id="linear-modeling" class="level3">
<h3 class="anchored" data-anchor-id="linear-modeling">Linear Modeling</h3>
<p>Using <strong>limma</strong>’s linear modeling <code>lmFit</code> function, fit the random effects into the linear model by setting the <em>correlation</em> argument to the estimated mean gene expression correlation between paired samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span>  <span class="fu">lmFit</span>(<span class="at">object=</span>dat<span class="sc">$</span>expression, </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">design=</span>design, </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">block=</span>dat<span class="sc">$</span>id, </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">correlation =</span> corr<span class="sc">$</span>consensus.correlation)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>fit<span class="sc">$</span>coefficients</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     groupD1_T1 groupD1_T2 groupD2_T1 groupD2_T2
[1,]       1.03   2.116667   2.996667        4.9</code></pre>
</div>
</div>
<p>The mixed effects model estimates the mean gene expression of patient receiving treatment D1 at timepoint T1 to be 1.03, treatment D1 at timepoint T2 to be 2.12, treatment D2 at timepoint T1 to be 3, and treatment D2 at timepoint T2 to be 4.9.</p>
</section>
<section id="create-contrasts-matrix" class="level3">
<h3 class="anchored" data-anchor-id="create-contrasts-matrix">Create contrasts matrix</h3>
<p>Create contrasts matrix using <code>makeContrasts</code> function to obtain estimates for the comparisons of interest</p>
<ul>
<li><p><em>difference between timepoint T2 and T1 in treatment D1</em></p></li>
<li><p><em>difference between timepoint T2 and T1 in treatment D2</em></p></li>
<li><p><em>overall difference between treatment D1 and D2</em></p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>contrasts <span class="ot">&lt;-</span> <span class="fu">makeContrasts</span>(</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">D1_T2vsT1 =</span> groupD1_T2 <span class="sc">-</span> groupD1_T1,</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">D2_T2vsT1 =</span> groupD2_T2 <span class="sc">-</span> groupD2_T1,</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">D1vsD2 =</span> (groupD1_T2 <span class="sc">-</span> groupD1_T1) <span class="sc">-</span> (groupD2_T2 <span class="sc">-</span> groupD2_T1),</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">levels =</span> <span class="fu">colnames</span>(design))</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>contrasts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            Contrasts
Levels       D1_T2vsT1 D2_T2vsT1 D1vsD2
  groupD1_T1        -1         0     -1
  groupD1_T2         1         0      1
  groupD2_T1         0        -1      1
  groupD2_T2         0         1     -1</code></pre>
</div>
</div>
</section>
<section id="fitting-multiple-contrasts" class="level3">
<h3 class="anchored" data-anchor-id="fitting-multiple-contrasts">Fitting multiple contrasts</h3>
<p>Incorporate the contrast matrix into the <strong>limma</strong> pipeline using the <code>contrasts.fit</code> function</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">contrasts.fit</span>(fit, contrasts)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>fit<span class="sc">$</span>coefficients</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Contrasts
       D1_T2vsT1 D2_T2vsT1     D1vsD2
  [1,]  1.086667  1.903333 -0.8166667</code></pre>
</div>
</div>
<ul>
<li><p><em>gene expression difference between timepoint T2 and T1 for treatment D1 &gt;= 1.09</em></p></li>
<li><p><em>gene expression difference between timepoint T2 and T1 for treatment D2 &gt;= 1.9</em></p></li>
<li><p><em>overall gene expression difference between treatment D1 and D2 = -0.82</em></p></li>
</ul>
<p>Greater gene expression changes under D2 and at T2.</p>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>