<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>MultiNicheNet: differential cell-cell communication analysis from single-cell transcriptomics data with complex multi-sample, multi-condition designs</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="multinichenet1_files/libs/clipboard/clipboard.min.js"></script>
<script src="multinichenet1_files/libs/quarto-html/quarto.js"></script>
<script src="multinichenet1_files/libs/quarto-html/popper.min.js"></script>
<script src="multinichenet1_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="multinichenet1_files/libs/quarto-html/anchor.min.js"></script>
<link href="multinichenet1_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="multinichenet1_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="multinichenet1_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="multinichenet1_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="multinichenet1_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">MultiNicheNet: differential cell-cell communication analysis from single-cell transcriptomics data with complex multi-sample, multi-condition designs</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="install-multinichenet" class="level2">
<h2 class="anchored" data-anchor-id="install-multinichenet">Install MultiNicheNet</h2>
<p><a href="https://github.com/saeyslab/multinichenetr/tree/main">MultiNicheNet</a> aims to find which ligand-receptor interactions are differentially expressed and differentially active between conditions of interest (such as patient groups) from data with complex experimental designs such as those with multiple samples and conditions, and multiple receiver cell types of interest.</p>
<p>MultiNicheNet can exploit the flexibility of generalized linear models in the Muscat pseudobulk + edgeR framework to handle complex multifactor experimental designs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages("devtools")</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># devtools::install_github("saeyslab/nichenetr")</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># devtools::install_github("saeyslab/multinichenetr")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-libraries" class="level2">
<h2 class="anchored" data-anchor-id="load-libraries">Load Libraries</h2>
<p>Load required libraries:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SingleCellExperiment)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: SummarizedExperiment</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: MatrixGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: matrixStats</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'matrixStats' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'MatrixGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:matrixStats':

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomicRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: stats4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BiocGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,
    table, tapply, union, unique, unsplit, which.max, which.min</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:utils':

    findMatches</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: IRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'IRanges'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:grDevices':

    windows</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomeInfoDb</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'GenomeInfoDb' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Biobase</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'Biobase'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:MatrixGenerics':

    rowMedians</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:matrixStats':

    anyMissing, rowMedians</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'dplyr' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:Biobase':

    combine</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:GenomicRanges':

    intersect, setdiff, union</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:GenomeInfoDb':

    intersect</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:IRanges':

    collapse, desc, intersect, setdiff, slice, union</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:S4Vectors':

    first, intersect, rename, setdiff, setequal, union</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:BiocGenerics':

    combine, intersect, setdiff, union</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:matrixStats':

    count</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'ggplot2' was built under R version 4.3.2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(multinichenetr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Registered S3 method overwritten by 'lava':
  method         from    
  print.estimate EnvStats</code></pre>
</div>
</div>
</section>
<section id="single-cell-expression-data-of-interacting-cells" class="level2">
<h2 class="anchored" data-anchor-id="single-cell-expression-data-of-interacting-cells">Single-cell Expression Data of Interacting Cells</h2>
<p>scRNA-seq data from breast cancer biopsies of patients receiving <a href="https://www.cancer.gov/about-cancer/treatment/types/immunotherapy/checkpoint-inhibitors">anti-PD1</a> immune-checkpoint blockade therapy. From each patient, one tumor biopsy was collected before anti-PD1 therapy (“pre-treatment”) and one during subsequent surgery (“on-treatment”). Through scTCR-seq, one group of patients was identified with <a href="https://www.sc-best-practices.org/air_repertoire/clonotype.html">clonotype expansion</a> (“E”) as response to the therapy and another group with minimal or no clonotype expansion (“NE”).</p>
<p>Source: Bassez et al - <a href="https://www.nature.com/articles/s41591-021-01323-8">A single-cell map of intratumoral changes during anti-PD1 treatment of patients with breast cancer</a>.</p>
<p>Want to compare cell-cell interaction changes during anti-PD1 therapy (“on” versus “pre”) between the “E” patients and the “NE” patients.</p>
</section>
<section id="multinichenet-analysis-workflow" class="level2">
<h2 class="anchored" data-anchor-id="multinichenet-analysis-workflow">MultiNicheNet Analysis Workflow</h2>
<ul>
<li><p><a href="#step-1-analysis-preparation---nichenet-ligand-receptor-network-ligand-target-matrix-single-cell-expression-data-and-definition-of-main-settings-of-the-analysis">Step 1: Analysis Preparation - NicheNet ligand-receptor network &amp; ligand-target matrix, single-cell expression data, and definition of main settings of the analysis</a></p></li>
<li><p><a href="#step2-cell-type-abundance-and-expression-information-extraction-from-sender-and-receiver-cell-types-and-link-expression-information-for-ligands-of-the-sender-cell-types-to-the-corresponding-receptors-of-the-receiver-cell-types">Step2: Cell type abundance and expression information extraction from sender and receiver cell types, and link expression information for ligands of the sender cell types to the corresponding receptors of the receiver cell types</a></p></li>
<li><p><a href="#step3-genome-wide-differential-expression-analysis-of-sender-and-receiver-cell-types-to-define-degs-differentially-expressed-genes-between-the-conditions-of-interest-define-logfcp-value-of-ligands-in-senders-and-receptors-in-receiver-and-define-the-set-of-affected-target-genes-in-the-receiver">Step3: Genome-wide differential expression analysis of sender and receiver cell types to define DEGs (differentially expressed genes) between the conditions of interest; define logFC/p-value of ligands in senders and receptors in receiver, and define the set of affected target genes in the receiver</a></p></li>
<li><p><a href="#step-4-nichenet-prediction-of-ligand-activities-and-ligand-target-links-inference-based-on-de-results">Step 4: NicheNet prediction of ligand activities and ligand-target links inference based on DE results</a></p></li>
<li><p><a href="#step-5-prioritization-of-all-sender-ligand---receiver-receptor-pairs">Step 5: Prioritization of all sender-ligand - receiver-receptor pairs</a></p></li>
<li><p><a href="#step-6-correlation-between-ligand-receptor-pairs-and-their-predicted-target-genes-expression">Step 6: Correlation between ligand-receptor pairs and their predicted target genes expression</a></p></li>
<li><p><a href="#step-7-save-multinichenet-cell-cell-communication-analysis">Step 7: Save MultiNicheNet cell-cell communication analysis</a></p></li>
<li><p><a href="#step-8-visualization-of-multinichenet-cell-cell-communication-analysis">Step 8: Visualization of MultiNicheNet cell-cell communication analysis</a></p></li>
</ul>
</section>
<section id="step-1-analysis-preparation---nichenet-ligand-receptor-network-ligand-target-matrix-single-cell-expression-data-and-definition-of-main-settings-of-the-analysis" class="level2">
<h2 class="anchored" data-anchor-id="step-1-analysis-preparation---nichenet-ligand-receptor-network-ligand-target-matrix-single-cell-expression-data-and-definition-of-main-settings-of-the-analysis">Step 1: Analysis Preparation - NicheNet ligand-receptor network &amp; ligand-target matrix, single-cell expression data, and definition of main settings of the analysis</h2>
<section id="nichenet-v2-ligand-receptor-network-and-ligand-target-matrix-for-human" class="level3">
<h3 class="anchored" data-anchor-id="nichenet-v2-ligand-receptor-network-and-ligand-target-matrix-for-human">1.1 NicheNet v2 ligand-receptor network and ligand-target matrix for human</h3>
<p>Load <strong>ligand-receptor network - prior knowledge ligand-receptor network</strong>. A matrix with ligand genes in columns and receptor genes in rows.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>lr_network <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">url</span>(<span class="st">"https://zenodo.org/record/7074291/files/lr_network_human_21122021.rds"</span>))</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(lr_network)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  from  to     database             source              
  &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;                &lt;chr&gt;               
1 A2M   MMP2   omnipath             omnipath            
2 A2M   MMP9   omnipath             omnipath            
3 A2M   LRP1   omnipath             omnipath            
4 A2M   KLK3   omnipath             omnipath            
5 AANAT MTNR1A nichenet_verschueren nichenet_verschueren
6 AANAT MTNR1B nichenet_verschueren nichenet_verschueren</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(lr_network)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4986    4</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>lr_network <span class="ot">=</span> lr_network <span class="sc">%&gt;%</span> </span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">ligand=</span>from, <span class="at">receptor=</span>to) <span class="sc">%&gt;%</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(ligand, receptor) <span class="sc">%&gt;%</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">ligand=</span><span class="fu">make.names</span>(ligand), </span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">receptor=</span><span class="fu">make.names</span>(receptor))</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(lr_network)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 2
  ligand receptor
  &lt;chr&gt;  &lt;chr&gt;   
1 A2M    MMP2    
2 A2M    MMP9    
3 A2M    LRP1    
4 A2M    KLK3    
5 AANAT  MTNR1A  
6 AANAT  MTNR1B  </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(lr_network)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4986    2</code></pre>
</div>
</div>
<p>Load <strong>ligand-target_matrix - prior knowledge model of ligand-target regulatory potential</strong>. A matrix with ligand genes in columns and target genes in rows.</p>
<p>The ligand-target matrix denotes the prior potential that particular ligands might regulate the expression of particular target genes. This matrix is necessary to prioritize possible ligand-receptor interactions based on observed gene expression effects (i.e.&nbsp;NicheNet’s ligand activity analysis) and infer affected target genes of these prioritized ligands.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>ligand_target_matrix <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">url</span>(<span class="st">"https://zenodo.org/record/7074291/files/ligand_target_matrix_nsga2r_final.rds"</span>))</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ligand_target_matrix) <span class="ot">=</span> <span class="fu">colnames</span>(ligand_target_matrix) <span class="sc">%&gt;%</span> <span class="fu">make.names</span>()</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ligand_target_matrix) <span class="ot">=</span> <span class="fu">rownames</span>(ligand_target_matrix) <span class="sc">%&gt;%</span> <span class="fu">make.names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(ligand_target_matrix)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 33354  1226</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>ligand_target_matrix[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                    A2M        AANAT        ABCA1          ACE        ACE2
A.GAMMA3.E 0.0000000000 0.0000000000 0.0000000000 0.0000000000 0.000000000
A1BG       0.0018503922 0.0011108718 0.0014225077 0.0028594037 0.001139013
A1BG.AS1   0.0007400797 0.0004677614 0.0005193137 0.0007836698 0.000375007
A1CF       0.0024799266 0.0013026348 0.0020420890 0.0047921048 0.003273375
A2M        0.0084693452 0.0040689323 0.0064256379 0.0105191365 0.005719199</code></pre>
</div>
</div>
</section>
<section id="load-sender-and-receiver-cell-types-in-singlecellexperiment-object" class="level3">
<h3 class="anchored" data-anchor-id="load-sender-and-receiver-cell-types-in-singlecellexperiment-object">1.2 Load sender and receiver cell types in SingleCellExperiment object</h3>
<p>Load subset (3 cell types) of breast cancer scRNA-seq data in <a href="https://bioconductor.org/books/3.13/OSCA.intro/the-singlecellexperiment-class.html">SingleCellExperiment</a> object</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">url</span>(<span class="st">"https://zenodo.org/record/8010790/files/sce_subset_breastcancer.rds"</span>))</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>sce</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SingleCellExperiment 
dim: 24414 48708 
metadata(0):
assays(2): counts logcounts
rownames(24414): A1BG A1BG-AS1 ... BPIFA1 PROKR1
rowData names(0):
colnames(48708): BIOKEY_10_Pre_AAACCTGGTTCTCATT-1
  BIOKEY_10_Pre_AAAGATGAGCTCCTTC-1 ... BIOKEY_24_On_TTTGTCAGTCGCGAAA-1
  BIOKEY_24_On_TTTGTCATCGGTTAAC-1
colData names(18): Cell orig.ident ... ang_response cellSubType
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">@</span>colData <span class="co"># or colData(sce)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 48708 rows and 18 columns
                                                   Cell orig.ident nCount_RNA
                                            &lt;character&gt;   &lt;factor&gt;  &lt;numeric&gt;
BIOKEY_10_Pre_AAACCTGGTTCTCATT-1 BIOKEY_10_Pre_AAACCT..     BIOKEY       4798
BIOKEY_10_Pre_AAAGATGAGCTCCTTC-1 BIOKEY_10_Pre_AAAGAT..     BIOKEY       1524
BIOKEY_10_Pre_AAAGATGTCTGCTGCT-1 BIOKEY_10_Pre_AAAGAT..     BIOKEY       6009
BIOKEY_10_Pre_AAAGCAAAGTCTCCTC-1 BIOKEY_10_Pre_AAAGCA..     BIOKEY      24516
BIOKEY_10_Pre_AAAGTAGCAAGTAGTA-1 BIOKEY_10_Pre_AAAGTA..     BIOKEY       1274
...                                                 ...        ...        ...
BIOKEY_24_On_TTTGCGCGTAGCTAAA-1  BIOKEY_24_On_TTTGCGC..     BIOKEY       9982
BIOKEY_24_On_TTTGTCAAGCCACCTG-1  BIOKEY_24_On_TTTGTCA..     BIOKEY        480
BIOKEY_24_On_TTTGTCAAGGAATCGC-1  BIOKEY_24_On_TTTGTCA..     BIOKEY        950
BIOKEY_24_On_TTTGTCAGTCGCGAAA-1  BIOKEY_24_On_TTTGTCA..     BIOKEY       6228
BIOKEY_24_On_TTTGTCATCGGTTAAC-1  BIOKEY_24_On_TTTGTCA..     BIOKEY       6143
                                 nFeature_RNA nCount_SCT nFeature_SCT
                                    &lt;integer&gt;  &lt;numeric&gt;    &lt;integer&gt;
BIOKEY_10_Pre_AAACCTGGTTCTCATT-1         2115       4084         2115
BIOKEY_10_Pre_AAAGATGAGCTCCTTC-1          888       3064          902
BIOKEY_10_Pre_AAAGATGTCTGCTGCT-1         2257       4298         2227
BIOKEY_10_Pre_AAAGCAAAGTCTCCTC-1         4494       4654         1676
BIOKEY_10_Pre_AAAGTAGCAAGTAGTA-1          751       2660          821
...                                       ...        ...          ...
BIOKEY_24_On_TTTGCGCGTAGCTAAA-1          2695       4284         1730
BIOKEY_24_On_TTTGTCAAGCCACCTG-1           324       2364          542
BIOKEY_24_On_TTTGTCAAGGAATCGC-1           394       2719          464
BIOKEY_24_On_TTTGTCAGTCGCGAAA-1          1810       4260         1766
BIOKEY_24_On_TTTGTCATCGGTTAAC-1          2066       4256         2037
                                  patient_id   timepoint   expansion
                                 &lt;character&gt; &lt;character&gt; &lt;character&gt;
BIOKEY_10_Pre_AAACCTGGTTCTCATT-1   BIOKEY_10         Pre           E
BIOKEY_10_Pre_AAAGATGAGCTCCTTC-1   BIOKEY_10         Pre           E
BIOKEY_10_Pre_AAAGATGTCTGCTGCT-1   BIOKEY_10         Pre           E
BIOKEY_10_Pre_AAAGCAAAGTCTCCTC-1   BIOKEY_10         Pre           E
BIOKEY_10_Pre_AAAGTAGCAAGTAGTA-1   BIOKEY_10         Pre           E
...                                      ...         ...         ...
BIOKEY_24_On_TTTGCGCGTAGCTAAA-1    BIOKEY_24          On          NE
BIOKEY_24_On_TTTGTCAAGCCACCTG-1    BIOKEY_24          On          NE
BIOKEY_24_On_TTTGTCAAGGAATCGC-1    BIOKEY_24          On          NE
BIOKEY_24_On_TTTGTCAGTCGCGAAA-1    BIOKEY_24          On          NE
BIOKEY_24_On_TTTGTCATCGGTTAAC-1    BIOKEY_24          On          NE
                                     BC_type     cellType          cohort
                                 &lt;character&gt;  &lt;character&gt;     &lt;character&gt;
BIOKEY_10_Pre_AAACCTGGTTCTCATT-1        TNBC       T_cell treatment_naive
BIOKEY_10_Pre_AAAGATGAGCTCCTTC-1        TNBC       T_cell treatment_naive
BIOKEY_10_Pre_AAAGATGTCTGCTGCT-1        TNBC   Fibroblast treatment_naive
BIOKEY_10_Pre_AAAGCAAAGTCTCCTC-1        TNBC Myeloid_cell treatment_naive
BIOKEY_10_Pre_AAAGTAGCAAGTAGTA-1        TNBC       T_cell treatment_naive
...                                      ...          ...             ...
BIOKEY_24_On_TTTGCGCGTAGCTAAA-1          ER+   Fibroblast treatment_naive
BIOKEY_24_On_TTTGTCAAGCCACCTG-1          ER+   Fibroblast treatment_naive
BIOKEY_24_On_TTTGTCAAGGAATCGC-1          ER+   Fibroblast treatment_naive
BIOKEY_24_On_TTTGTCAGTCGCGAAA-1          ER+   Fibroblast treatment_naive
BIOKEY_24_On_TTTGTCATCGGTTAAC-1          ER+ Myeloid_cell treatment_naive
                                 expansion_timepoint    sample_id     subType
                                         &lt;character&gt;  &lt;character&gt; &lt;character&gt;
BIOKEY_10_Pre_AAACCTGGTTCTCATT-1                PreE BIOKEY_10Pre        CD4T
BIOKEY_10_Pre_AAAGATGAGCTCCTTC-1                PreE BIOKEY_10Pre        CD4T
BIOKEY_10_Pre_AAAGATGTCTGCTGCT-1                PreE BIOKEY_10Pre  Fibroblast
BIOKEY_10_Pre_AAAGCAAAGTCTCCTC-1                PreE BIOKEY_10Pre macrophages
BIOKEY_10_Pre_AAAGTAGCAAGTAGTA-1                PreE BIOKEY_10Pre        CD4T
...                                              ...          ...         ...
BIOKEY_24_On_TTTGCGCGTAGCTAAA-1                 OnNE  BIOKEY_24On  Fibroblast
BIOKEY_24_On_TTTGTCAAGCCACCTG-1                 OnNE  BIOKEY_24On  Fibroblast
BIOKEY_24_On_TTTGTCAAGGAATCGC-1                 OnNE  BIOKEY_24On  Fibroblast
BIOKEY_24_On_TTTGTCAGTCGCGAAA-1                 OnNE  BIOKEY_24On  Fibroblast
BIOKEY_24_On_TTTGTCATCGGTTAAC-1                 OnNE  BIOKEY_24On macrophages
                                       ident ang_response cellSubType
                                    &lt;factor&gt;  &lt;character&gt; &lt;character&gt;
BIOKEY_10_Pre_AAACCTGGTTCTCATT-1 CD4T        undetermined      CD4_EX
BIOKEY_10_Pre_AAAGATGAGCTCCTTC-1 CD4T        undetermined      CD4_EX
BIOKEY_10_Pre_AAAGATGTCTGCTGCT-1 Fibroblast  undetermined  Fibroblast
BIOKEY_10_Pre_AAAGCAAAGTCTCCTC-1 macrophages undetermined    C5_CCL18
BIOKEY_10_Pre_AAAGTAGCAAGTAGTA-1 CD4T        undetermined      CD4_EX
...                                      ...          ...         ...
BIOKEY_24_On_TTTGCGCGTAGCTAAA-1  Fibroblast         angio  Fibroblast
BIOKEY_24_On_TTTGTCAAGCCACCTG-1  Fibroblast         angio  Fibroblast
BIOKEY_24_On_TTTGTCAAGGAATCGC-1  Fibroblast         angio  Fibroblast
BIOKEY_24_On_TTTGTCAGTCGCGAAA-1  Fibroblast         angio  Fibroblast
BIOKEY_24_On_TTTGTCATCGGTTAAC-1  macrophages        angio   C10_LYVE1</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># cells</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">length</span>(<span class="fu">colnames</span>(sce)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 48708</code></pre>
</div>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sce)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "BIOKEY_10_Pre_AAACCTGGTTCTCATT-1" "BIOKEY_10_Pre_AAAGATGAGCTCCTTC-1"
[3] "BIOKEY_10_Pre_AAAGATGTCTGCTGCT-1" "BIOKEY_10_Pre_AAAGCAAAGTCTCCTC-1"
[5] "BIOKEY_10_Pre_AAAGTAGCAAGTAGTA-1"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co"># genes</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 24414 rows and 0 columns</code></pre>
</div>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(<span class="fu">rowData</span>(sce))[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "A1BG"     "A1BG-AS1" "A2M"      "A2M-AS1"  "A4GALT"  </code></pre>
</div>
</div>
<p>Convert gene aliases used in expression data to official gene symbols</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">=</span> <span class="fu">alias_to_symbol_SCE</span>(sce, <span class="st">"human"</span>) <span class="sc">%&gt;%</span> <span class="fu">makenames_SCE</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Want to study differences in therapy-induced cell-cell communication changes (On-vs-Pre therapy) between two patient groups (E vs NE: patients).</strong></p>
<p>Both therapy-timepoint and patient group are indicated in the <code>expansion_timepoint</code> metadata column and has 4 different values: PreE, OnE, PreNE, OnNE.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(<span class="fu">colData</span>(sce)<span class="sc">$</span>expansion_timepoint)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "PreE"  "OnE"   "PreNE" "OnNE" </code></pre>
</div>
</div>
<p>Cell types annotations are indicated in the <code>subType</code> column</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(<span class="fu">colData</span>(sce)<span class="sc">$</span>subType)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "CD4T"        "Fibroblast"  "macrophages"</code></pre>
</div>
</div>
<p>Samples are indicated in the <code>sample_id</code> column</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(<span class="fu">colData</span>(sce)<span class="sc">$</span>sample_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "BIOKEY_10Pre" "BIOKEY_10On"  "BIOKEY_16Pre" "BIOKEY_16On"  "BIOKEY_14Pre"
 [6] "BIOKEY_14On"  "BIOKEY_19Pre" "BIOKEY_19On"  "BIOKEY_23Pre" "BIOKEY_23On" 
[11] "BIOKEY_26Pre" "BIOKEY_26On"  "BIOKEY_28Pre" "BIOKEY_28On"  "BIOKEY_3Pre" 
[16] "BIOKEY_3On"   "BIOKEY_15Pre" "BIOKEY_15On"  "BIOKEY_8Pre"  "BIOKEY_8On"  
[21] "BIOKEY_5Pre"  "BIOKEY_5On"   "BIOKEY_30Pre" "BIOKEY_30On"  "BIOKEY_12Pre"
[26] "BIOKEY_12On"  "BIOKEY_31Pre" "BIOKEY_31On"  "BIOKEY_20Pre" "BIOKEY_20On" 
[31] "BIOKEY_22Pre" "BIOKEY_22On"  "BIOKEY_25Pre" "BIOKEY_25On"  "BIOKEY_21Pre"
[36] "BIOKEY_21On"  "BIOKEY_29Pre" "BIOKEY_29On"  "BIOKEY_4Pre"  "BIOKEY_4On"  
[41] "BIOKEY_9Pre"  "BIOKEY_9On"   "BIOKEY_18Pre" "BIOKEY_18On"  "BIOKEY_11Pre"
[46] "BIOKEY_11On"  "BIOKEY_7Pre"  "BIOKEY_7On"   "BIOKEY_2Pre"  "BIOKEY_2On"  
[51] "BIOKEY_6Pre"  "BIOKEY_6On"   "BIOKEY_17Pre" "BIOKEY_17On"  "BIOKEY_27Pre"
[56] "BIOKEY_27On"  "BIOKEY_24Pre" "BIOKEY_24On" </code></pre>
</div>
</div>
<p>Check number of cells per cell type-condition/sample combination: there should be sufficient cells per celltype-sample combination for subsequent DE analysis</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># build contingency table of the counts at each combination of factor levels</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="co"># using the cross-classifying factors</span></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a><span class="co"># cell types vs samples</span></span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(SummarizedExperiment<span class="sc">::</span><span class="fu">colData</span>(sce)<span class="sc">$</span>subType, </span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>      SummarizedExperiment<span class="sc">::</span><span class="fu">colData</span>(sce)<span class="sc">$</span>sample_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             
              BIOKEY_10On BIOKEY_10Pre BIOKEY_11On BIOKEY_11Pre BIOKEY_12On
  CD4T               1600          497         659          182        1022
  Fibroblast          220          365          46           38         253
  macrophages         326           77         162           94         125
             
              BIOKEY_12Pre BIOKEY_14On BIOKEY_14Pre BIOKEY_15On BIOKEY_15Pre
  CD4T                 972          85          407        1324          509
  Fibroblast           460         115          475          85          329
  macrophages           77          21           32         117          224
             
              BIOKEY_16On BIOKEY_16Pre BIOKEY_17On BIOKEY_17Pre BIOKEY_18On
  CD4T                543          614          34           31         757
  Fibroblast           41           63        1663          664         480
  macrophages         778          511         138           67          56
             
              BIOKEY_18Pre BIOKEY_19On BIOKEY_19Pre BIOKEY_20On BIOKEY_20Pre
  CD4T                  78         377          818         396           51
  Fibroblast           473         567          741         401          145
  macrophages           29          73           40          84            4
             
              BIOKEY_21On BIOKEY_21Pre BIOKEY_22On BIOKEY_22Pre BIOKEY_23On
  CD4T                 15          143          16           41         281
  Fibroblast          136          274        1762          243         623
  macrophages          98           29          32            6         105
             
              BIOKEY_23Pre BIOKEY_24On BIOKEY_24Pre BIOKEY_25On BIOKEY_25Pre
  CD4T                  10         110          134         190           43
  Fibroblast            55         557          267          27           53
  macrophages            4         152           82           8           14
             
              BIOKEY_26On BIOKEY_26Pre BIOKEY_27On BIOKEY_27Pre BIOKEY_28On
  CD4T                  6           38         257          295         394
  Fibroblast           63          136         751         1069          37
  macrophages          89           54          48           23         793
             
              BIOKEY_28Pre BIOKEY_29On BIOKEY_29Pre BIOKEY_2On BIOKEY_2Pre
  CD4T                 163          77           60       1006         616
  Fibroblast           196          36          148        261         268
  macrophages           86           8           24        105         166
             
              BIOKEY_30On BIOKEY_30Pre BIOKEY_31On BIOKEY_31Pre BIOKEY_3On
  CD4T                 37           28         507          162         50
  Fibroblast          177          146         145           45       1543
  macrophages          77           67         347          258        154
             
              BIOKEY_3Pre BIOKEY_4On BIOKEY_4Pre BIOKEY_5On BIOKEY_5Pre
  CD4T                116        521        1039        192         828
  Fibroblast         2312         30         201         63          40
  macrophages         144         31          65         13          10
             
              BIOKEY_6On BIOKEY_6Pre BIOKEY_7On BIOKEY_7Pre BIOKEY_8On
  CD4T               348         247          5          31        108
  Fibroblast         716         801       1176         108          4
  macrophages        141         116        158          17        211
             
              BIOKEY_8Pre BIOKEY_9On BIOKEY_9Pre
  CD4T                 39         80          50
  Fibroblast            8        146         324
  macrophages          50         31          47</code></pre>
</div>
</div>
<p>Check number of patients/expansion_timepoint per condition/sample</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co"># cell types vs samples</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(SummarizedExperiment<span class="sc">::</span><span class="fu">colData</span>(sce)<span class="sc">$</span>subType, </span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>      SummarizedExperiment<span class="sc">::</span><span class="fu">colData</span>(sce)<span class="sc">$</span>expansion_timepoint)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             
                OnE  OnNE  PreE PreNE
  CD4T         6998  3999  4005  4237
  Fibroblast   1370 10754  2009  8438
  macrophages  2717  1764  1366  1051</code></pre>
</div>
</div>
</section>
<section id="define-settings-for-cell-cell-communication-analysis" class="level3">
<h3 class="anchored" data-anchor-id="define-settings-for-cell-cell-communication-analysis">1.3 Define settings for cell-cell communication analysis</h3>
<p>The minimum metadata for a MultiNicheNet cell-cell communication analysis (aka lookup for cell-cell communications between cell types for each sample, and compare the cell-cell communication patterns between groups) are <code>group</code>, <code>sample</code>, and <code>cell type</code> metadata columns.</p>
<p>Define metadata columns in which group, sample and cell type IDs are found.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>sample_id <span class="ot">=</span> <span class="st">"sample_id"</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>group_id <span class="ot">=</span> <span class="st">"expansion_timepoint"</span></span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>celltype_id <span class="ot">=</span> <span class="st">"subType"</span></span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>covariates <span class="ot">=</span> <span class="cn">NA</span></span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a>batches <span class="ot">=</span> <span class="cn">NA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Define sender and receiver cell types of interest (_oi). In this All-vs-All analysis both are all cell types in the dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>senders_oi <span class="ot">=</span> SummarizedExperiment<span class="sc">::</span><span class="fu">colData</span>(sce)[, celltype_id] <span class="sc">%&gt;%</span> <span class="fu">unique</span>()</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>receivers_oi <span class="ot">=</span> SummarizedExperiment<span class="sc">::</span><span class="fu">colData</span>(sce)[, celltype_id] <span class="sc">%&gt;%</span> <span class="fu">unique</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Subset senders and receiver, if needed.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">=</span> sce[, SummarizedExperiment<span class="sc">::</span><span class="fu">colData</span>(sce)[,celltype_id] <span class="sc">%in%</span> <span class="fu">c</span>(senders_oi, receivers_oi)]</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>sce</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SingleCellExperiment 
dim: 24414 48708 
metadata(0):
assays(2): counts logcounts
rownames(24414): A1BG A1BG.AS1 ... BPIFA1 PROKR1
rowData names(0):
colnames(48708): BIOKEY_10_Pre_AAACCTGGTTCTCATT-1
  BIOKEY_10_Pre_AAAGATGAGCTCCTTC-1 ... BIOKEY_24_On_TTTGTCAGTCGCGAAA-1
  BIOKEY_24_On_TTTGTCATCGGTTAAC-1
colData names(18): Cell orig.ident ... ang_response cellSubType
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
</section>
</section>
<section id="step2-cell-type-abundance-and-expression-information-extraction-from-sender-and-receiver-cell-types-and-link-expression-information-for-ligands-of-the-sender-cell-types-to-the-corresponding-receptors-of-the-receiver-cell-types" class="level2">
<h2 class="anchored" data-anchor-id="step2-cell-type-abundance-and-expression-information-extraction-from-sender-and-receiver-cell-types-and-link-expression-information-for-ligands-of-the-sender-cell-types-to-the-corresponding-receptors-of-the-receiver-cell-types">Step2: Cell type abundance and expression information extraction from sender and receiver cell types, and link expression information for ligands of the sender cell types to the corresponding receptors of the receiver cell types</h2>
<p>MultiNicheNet infers group differences at the sample level for each cell type via Muscat - pseudobulking + EdgeR. For this, a minimum number of cells per sample of a cell type needs to be set.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>min_cells <span class="ot">=</span> <span class="dv">10</span> <span class="co"># recommended</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="calculate-abundance-and-expression-information-for-each-cell-typesamplegroup-combination" class="level3">
<h3 class="anchored" data-anchor-id="calculate-abundance-and-expression-information-for-each-cell-typesamplegroup-combination">2.1 Calculate abundance and expression information for each cell type/sample/group combination</h3>
<p>get_abundance_expression_info function calculates the average and fraction of expression of each gene per sample and per group as well as calculates relative abundances of cell types.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info <span class="ot">=</span> <span class="fu">get_abundance_expression_info</span>(</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">sce =</span> sce,</span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample_id =</span> sample_id,</span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">group_id =</span> group_id,</span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">celltype_id =</span> celltype_id,</span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">min_cells =</span> min_cells,</span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">senders_oi =</span> senders_oi,</span>
<span id="cb86-8"><a href="#cb86-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">receivers_oi =</span> receivers_oi,</span>
<span id="cb86-9"><a href="#cb86-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">lr_network =</span> lr_network,</span>
<span id="cb86-10"><a href="#cb86-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">batches =</span> batches)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co">#abundance_expression_info </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="visualization-and-diagnostic-of-cell-type-abundance-information" class="level3">
<h3 class="anchored" data-anchor-id="visualization-and-diagnostic-of-cell-type-abundance-information">2.2 Visualization and diagnostic of cell type abundance information</h3>
<p>The abund_plot_sample shows the number of cells per celltype-sample combination, and indicates which combinations are removed during DE analysis because there are less then min_cells in the celltype-sample combination.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info<span class="sc">$</span>abund_plot_sample</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The abund_plot_group shows the differential abundance between conditions/groups. There should not be huge differences in abundances of a cell type between different groups.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info<span class="sc">$</span>abund_plot_group</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The abund_barplot shows the relative abundances of cell types per sample</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info<span class="sc">$</span>abund_barplot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Sender and receiver cell numbers behind the plots. In the case of an all-vs-all analysis both cell numbers are the same.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info<span class="sc">$</span>abundance_data_receiver <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
# Groups:   sample, receiver [6]
  sample       receiver    n_cells_receiver group keep_receiver
  &lt;chr&gt;        &lt;chr&gt;                  &lt;int&gt; &lt;chr&gt;         &lt;dbl&gt;
1 BIOKEY_10On  CD4T                    1600 OnE               1
2 BIOKEY_10On  Fibroblast               220 OnE               1
3 BIOKEY_10On  macrophages              326 OnE               1
4 BIOKEY_10Pre CD4T                     497 PreE              1
5 BIOKEY_10Pre Fibroblast               365 PreE              1
6 BIOKEY_10Pre macrophages               77 PreE              1</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info<span class="sc">$</span>abundance_data_sender <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
# Groups:   sample, sender [6]
  sample       sender      n_cells_sender group keep_sender
  &lt;chr&gt;        &lt;chr&gt;                &lt;int&gt; &lt;chr&gt;       &lt;dbl&gt;
1 BIOKEY_10On  CD4T                  1600 OnE             1
2 BIOKEY_10On  Fibroblast             220 OnE             1
3 BIOKEY_10On  macrophages            326 OnE             1
4 BIOKEY_10Pre CD4T                   497 PreE            1
5 BIOKEY_10Pre Fibroblast             365 PreE            1
6 BIOKEY_10Pre macrophages             77 PreE            1</code></pre>
</div>
</div>
</section>
<section id="expression-information" class="level3">
<h3 class="anchored" data-anchor-id="expression-information">2.3 Expression information</h3>
<p>Check the average expression for each gene per sample</p>
<p>Normalized expression value</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info<span class="sc">$</span>celltype_info<span class="sc">$</span>avg_df <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  gene     sample       average_sample celltype
  &lt;chr&gt;    &lt;chr&gt;                 &lt;dbl&gt; &lt;fct&gt;   
1 A1BG     BIOKEY_10Pre        0.0972  CD4T    
2 A1BG.AS1 BIOKEY_10Pre        0.0351  CD4T    
3 A2M      BIOKEY_10Pre        0.0195  CD4T    
4 A2M.AS1  BIOKEY_10Pre        0.00139 CD4T    
5 A4GALT   BIOKEY_10Pre        0.00139 CD4T    
6 AAAS     BIOKEY_10Pre        0.0797  CD4T    </code></pre>
</div>
</div>
<p>Fraction of expressing cells with non-zero counts</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info<span class="sc">$</span>celltype_info<span class="sc">$</span>frq_df <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  gene     sample    fraction_sample celltype group expressed_sample n_expressed
  &lt;chr&gt;    &lt;chr&gt;               &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt; &lt;lgl&gt;                  &lt;int&gt;
1 A1BG     BIOKEY_1…         0.129   CD4T     PreE  TRUE                      55
2 A1BG.AS1 BIOKEY_1…         0.0483  CD4T     PreE  TRUE                      22
3 A2M      BIOKEY_1…         0.0262  CD4T     PreE  TRUE                      14
4 A2M.AS1  BIOKEY_1…         0.00201 CD4T     PreE  FALSE                      3
5 A4GALT   BIOKEY_1…         0.00201 CD4T     PreE  FALSE                      0
6 AAAS     BIOKEY_1…         0.113   CD4T     PreE  TRUE                      50
# ℹ 1 more variable: expressed_celltype &lt;lgl&gt;</code></pre>
</div>
</div>
<p>logCPM-pseudocounts</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info<span class="sc">$</span>celltype_info<span class="sc">$</span>pb_df <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  gene     sample      pb_sample celltype
  &lt;chr&gt;    &lt;chr&gt;           &lt;dbl&gt; &lt;fct&gt;   
1 A1BG     BIOKEY_10On     4.85  CD4T    
2 A1BG.AS1 BIOKEY_10On     3.47  CD4T    
3 A2M      BIOKEY_10On     2.70  CD4T    
4 A2M.AS1  BIOKEY_10On     2.32  CD4T    
5 A4GALT   BIOKEY_10On     0.374 CD4T    
6 AAAS     BIOKEY_10On     4.38  CD4T    </code></pre>
</div>
</div>
<p>Check the average expression for each gene per group</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info<span class="sc">$</span>celltype_info<span class="sc">$</span>avg_df_group <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
# Groups:   group, celltype [1]
  group celltype gene      average_group
  &lt;chr&gt; &lt;fct&gt;    &lt;chr&gt;             &lt;dbl&gt;
1 OnE   CD4T     A1BG           0.0870  
2 OnE   CD4T     A1BG.AS1       0.0179  
3 OnE   CD4T     A2M            0.0130  
4 OnE   CD4T     A2M.AS1        0.00494 
5 OnE   CD4T     A2ML1          0.000390
6 OnE   CD4T     A2ML1.AS1      0       </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info<span class="sc">$</span>celltype_info<span class="sc">$</span>frq_df_group <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
# Groups:   group, celltype [1]
  group celltype gene      fraction_group
  &lt;chr&gt; &lt;chr&gt;    &lt;chr&gt;              &lt;dbl&gt;
1 OnE   CD4T     A1BG            0.117   
2 OnE   CD4T     A1BG.AS1        0.0253  
3 OnE   CD4T     A2M             0.0166  
4 OnE   CD4T     A2M.AS1         0.00704 
5 OnE   CD4T     A2ML1           0.000563
6 OnE   CD4T     A2ML1.AS1       0       </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info<span class="sc">$</span>celltype_info<span class="sc">$</span>pb_df_group <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
# Groups:   group, celltype [1]
  group celltype gene      pb_group
  &lt;chr&gt; &lt;fct&gt;    &lt;chr&gt;        &lt;dbl&gt;
1 OnE   CD4T     A1BG         5.23 
2 OnE   CD4T     A1BG.AS1     3.03 
3 OnE   CD4T     A2M          2.29 
4 OnE   CD4T     A2M.AS1      1.51 
5 OnE   CD4T     A2ML1        0.153
6 OnE   CD4T     A2ML1.AS1    0    </code></pre>
</div>
</div>
</section>
<section id="link-expression-information-for-ligands-of-the-sender-cell-types-to-the-corresponding-receptors-of-the-receiver-cell-types" class="level3">
<h3 class="anchored" data-anchor-id="link-expression-information-for-ligands-of-the-sender-cell-types-to-the-corresponding-receptors-of-the-receiver-cell-types">2.4 Link expression information for ligands of the sender cell types to the corresponding receptors of the receiver cell types</h3>
<p>Combine expression information for each ligand-receptor pair gene combination for each sender-receiver cell type combination either sample-based or group-based.</p>
<p>Sample-based</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info<span class="sc">$</span>sender_receiver_info<span class="sc">$</span>avg_df <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  sample       sender      receiver    ligand  receptor avg_ligand avg_receptor
  &lt;chr&gt;        &lt;fct&gt;       &lt;fct&gt;       &lt;chr&gt;   &lt;chr&gt;         &lt;dbl&gt;        &lt;dbl&gt;
1 BIOKEY_24Pre macrophages macrophages HLA.DMA CD74           1.97         4.96
2 BIOKEY_7On   Fibroblast  macrophages MIF     CD74           2.26         4.29
3 BIOKEY_27Pre macrophages macrophages HLA.DMA CD74           1.96         4.78
4 BIOKEY_6Pre  macrophages macrophages HLA.DMA CD74           1.87         5.00
5 BIOKEY_28On  Fibroblast  macrophages MIF     CD74           2.58         3.56
6 BIOKEY_14Pre macrophages macrophages HLA.DMA CD74           1.82         4.87
# ℹ 1 more variable: ligand_receptor_prod &lt;dbl&gt;</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info<span class="sc">$</span>sender_receiver_info<span class="sc">$</span>frq_df <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  sample       sender receiver ligand receptor fraction_ligand fraction_receptor
  &lt;chr&gt;        &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt;              &lt;dbl&gt;             &lt;dbl&gt;
1 BIOKEY_23Pre CD4T   CD4T     CD99   CD99                   1                 1
2 BIOKEY_23Pre CD4T   macroph… CD99   CD99                   1                 1
3 BIOKEY_26On  CD4T   macroph… MIF    CD74                   1                 1
4 BIOKEY_21On  CD4T   CD4T     MIF    CXCR4                  1                 1
5 BIOKEY_21On  CD4T   macroph… MIF    CD74                   1                 1
6 BIOKEY_29On  CD4T   macroph… HLA.A  APLP2                  1                 1
# ℹ 1 more variable: ligand_receptor_fraction_prod &lt;dbl&gt;</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info<span class="sc">$</span>sender_receiver_info<span class="sc">$</span>pb_df <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  sample       sender      receiver    ligand  receptor pb_ligand pb_receptor
  &lt;chr&gt;        &lt;fct&gt;       &lt;fct&gt;       &lt;chr&gt;   &lt;chr&gt;        &lt;dbl&gt;       &lt;dbl&gt;
1 BIOKEY_28On  macrophages macrophages MIF     CD74          12.2        14.3
2 BIOKEY_10On  macrophages macrophages HLA.DMA CD74          11.0        15.6
3 BIOKEY_28On  Fibroblast  macrophages MIF     CD74          11.9        14.3
4 BIOKEY_11On  macrophages macrophages HLA.DMA CD74          10.8        15.6
5 BIOKEY_7On   Fibroblast  macrophages MIF     CD74          11.5        14.5
6 BIOKEY_31Pre macrophages macrophages MIF     CD74          12.0        13.8
# ℹ 1 more variable: ligand_receptor_pb_prod &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Group-based</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info<span class="sc">$</span>sender_receiver_info<span class="sc">$</span>avg_df_group <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
# Groups:   group, sender [6]
  group sender      receiver ligand receptor avg_ligand_group avg_receptor_group
  &lt;chr&gt; &lt;fct&gt;       &lt;fct&gt;    &lt;chr&gt;  &lt;chr&gt;               &lt;dbl&gt;              &lt;dbl&gt;
1 OnE   Fibroblast  macroph… MIF    CD74                 1.49               4.46
2 OnE   macrophages macroph… HLA.D… CD74                 1.43               4.46
3 PreNE macrophages macroph… HLA.D… CD74                 1.45               4.39
4 PreNE Fibroblast  macroph… MIF    CD74                 1.35               4.39
5 OnNE  macrophages macroph… HLA.D… CD74                 1.34               4.15
6 PreE  macrophages macroph… MIF    CD74                 1.34               4.14
# ℹ 1 more variable: ligand_receptor_prod_group &lt;dbl&gt;</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info<span class="sc">$</span>sender_receiver_info<span class="sc">$</span>frq_df_group <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
# Groups:   group, sender [5]
  group sender      receiver    ligand  receptor fraction_ligand_group
  &lt;chr&gt; &lt;chr&gt;       &lt;chr&gt;       &lt;chr&gt;   &lt;chr&gt;                    &lt;dbl&gt;
1 OnE   Fibroblast  macrophages MIF     CD74                     0.952
2 OnE   macrophages macrophages HLA.DRA CD63                     0.986
3 OnNE  Fibroblast  macrophages MIF     CD74                     0.924
4 PreNE Fibroblast  macrophages MIF     CD74                     0.921
5 OnE   macrophages Fibroblast  HLA.DRA CD63                     0.986
6 PreE  Fibroblast  macrophages MIF     CD74                     0.919
# ℹ 2 more variables: fraction_receptor_group &lt;dbl&gt;,
#   ligand_receptor_fraction_prod_group &lt;dbl&gt;</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a>abundance_expression_info<span class="sc">$</span>sender_receiver_info<span class="sc">$</span>pb_df_group <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
# Groups:   group, sender [5]
  group sender      receiver   ligand receptor pb_ligand_group pb_receptor_group
  &lt;chr&gt; &lt;fct&gt;       &lt;fct&gt;      &lt;chr&gt;  &lt;chr&gt;              &lt;dbl&gt;             &lt;dbl&gt;
1 OnE   macrophages macrophag… HLA.D… CD74               10.3               15.0
2 OnE   Fibroblast  macrophag… MIF    CD74               10.1               15.0
3 OnE   macrophages macrophag… MIF    CD74                9.90              15.0
4 OnE   CD4T        macrophag… MIF    CD74                9.77              15.0
5 PreE  macrophages macrophag… MIF    CD74               10.0               14.5
6 PreNE Fibroblast  macrophag… MIF    CD74                9.98              14.4
# ℹ 1 more variable: ligand_receptor_pb_prod_group &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
</section>
<section id="step3-genome-wide-differential-expression-analysis-of-sender-and-receiver-cell-types-to-define-degs-differentially-expressed-genes-between-the-conditions-of-interest-define-logfcp-value-of-ligands-in-senders-and-receptors-in-receiver-and-define-the-set-of-affected-target-genes-in-the-receiver" class="level2">
<h2 class="anchored" data-anchor-id="step3-genome-wide-differential-expression-analysis-of-sender-and-receiver-cell-types-to-define-degs-differentially-expressed-genes-between-the-conditions-of-interest-define-logfcp-value-of-ligands-in-senders-and-receptors-in-receiver-and-define-the-set-of-affected-target-genes-in-the-receiver">Step3: Genome-wide differential expression analysis of sender and receiver cell types to define DEGs (differentially expressed genes) between the conditions of interest; define logFC/p-value of ligands in senders and receptors in receiver, and define the set of affected target genes in the receiver</h2>
<p>Multi-group, multi-sample differential expression (DE) analysis also called differential state analysis in Muscat. Use <code>get_DE_info</code> function to perform DE analysis via Muscat - pseudobulking approach and visualize p-values distribution. First, define comparisons/contrasts of interest.</p>
<section id="define-contrasts-and-covariates-of-interest-for-the-de-analysis" class="level3">
<h3 class="anchored" data-anchor-id="define-contrasts-and-covariates-of-interest-for-the-de-analysis">3.1 Define contrasts and covariates of interest for the DE analysis</h3>
<p>Want to compare how cell-cell communication changes On-vs-Pre therapy are different between responder/expander patients vs non-responder/non-expander patients. Want to study how both patient groups (E and NE) react differently to therapy (on-vs-pre).</p>
<p>If wanting to compare group A vs B: ‘contrasts_oi = c(“‘A-B’”)’<br>
If wanting to compare group A vs B &amp; B vs A: ‘contrasts_oi = c(“‘A-B’,‘B-A’”)’</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a>contrasts_oi <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"'(OnE-PreE)-(OnNE-PreNE)','(OnNE-PreNE)-(OnE-PreE)'"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a>contrast_tbl <span class="ot">=</span> <span class="fu">tibble</span>(<span class="at">contrast =</span></span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a>                        <span class="fu">c</span>(<span class="st">"(OnE-PreE)-(OnNE-PreNE)"</span>, <span class="st">"(OnNE-PreNE)-(OnE-PreE)"</span>),</span>
<span id="cb120-3"><a href="#cb120-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">group =</span> <span class="fu">c</span>(<span class="st">"OnE"</span>,<span class="st">"OnNE"</span>)) </span>
<span id="cb120-4"><a href="#cb120-4" aria-hidden="true" tabindex="-1"></a>contrast_tbl</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  contrast                group
  &lt;chr&gt;                   &lt;chr&gt;
1 (OnE-PreE)-(OnNE-PreNE) OnE  
2 (OnNE-PreNE)-(OnE-PreE) OnNE </code></pre>
</div>
</div>
</section>
<section id="de-analysis-for-each-cell-type" class="level3">
<h3 class="anchored" data-anchor-id="de-analysis-for-each-cell-type">3.2 DE analysis for each cell type</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a>DE_info <span class="ot">=</span> <span class="fu">get_DE_info</span>(<span class="at">sce=</span>sce,</span>
<span id="cb122-2"><a href="#cb122-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">sample_id =</span> sample_id,</span>
<span id="cb122-3"><a href="#cb122-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">group_id =</span> group_id, </span>
<span id="cb122-4"><a href="#cb122-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">celltype_id =</span> celltype_id,</span>
<span id="cb122-5"><a href="#cb122-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">batches =</span> batches,</span>
<span id="cb122-6"><a href="#cb122-6" aria-hidden="true" tabindex="-1"></a>                      <span class="at">covariates =</span> covariates,</span>
<span id="cb122-7"><a href="#cb122-7" aria-hidden="true" tabindex="-1"></a>                      <span class="at">contrasts_oi =</span> contrasts_oi,</span>
<span id="cb122-8"><a href="#cb122-8" aria-hidden="true" tabindex="-1"></a>                      <span class="at">min_cells =</span> min_cells)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "DE analysis is done:"
[1] "included cell types are:"
[1] "CD4T"        "Fibroblast"  "macrophages"</code></pre>
</div>
</div>
</section>
<section id="de-statistics-check-logfc-and-p-values" class="level3">
<h3 class="anchored" data-anchor-id="de-statistics-check-logfc-and-p-values">3.3 DE statistics check: logFC and p-values</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DE_info </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>LogFC and p-values for each gene-celltype-contrast</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a>DE_info<span class="sc">$</span>celltype_de<span class="sc">$</span>de_output_tidy <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(p_val) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 9
  gene     cluster_id  logFC logCPM     F     p_val p_adj.loc  p_adj contrast   
  &lt;chr&gt;    &lt;chr&gt;       &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;      
1 IGKV1.5  macrophages  7.64   5.5   25.7 0.0000046    0.0334 0.0334 (OnE-PreE)…
2 IGKV1.5  macrophages -7.64   5.5   25.7 0.0000046    0.0334 0.0334 (OnNE-PreN…
3 IGHV4.39 CD4T         5.17   4.17  14.8 0.000287     1      1      (OnE-PreE)…
4 IGHV4.39 CD4T        -5.17   4.17  14.8 0.000287     1      1      (OnNE-PreN…
5 IGLV3.1  Fibroblast  -6.93   8.81  14   0.000428     1      1      (OnE-PreE)…
6 IGLV3.1  Fibroblast   6.93   8.81  14   0.000428     1      1      (OnNE-PreN…</code></pre>
</div>
</div>
<p>Distributions of DE analysis p-values - to be trustworthy p-value distributions should be uniform distributions, with a peak allowed between 0 and 0.05 if there would be a clear biological effect in the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a>DE_info<span class="sc">$</span>hist_pvals</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-48-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Correct for suboptimal p-value distribution using empirical Null procedure which define empirical p-values based on the observed distribution of the test statistic (here: logFC) and not based on the theoretical distribution.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>empirical_pval <span class="ot">=</span> <span class="cn">TRUE</span></span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(empirical_pval <span class="sc">==</span> <span class="cn">TRUE</span>){</span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a>  DE_info_emp <span class="ot">=</span> <span class="fu">get_empirical_pvals</span>(DE_info<span class="sc">$</span>celltype_de<span class="sc">$</span>de_output_tidy)</span>
<span id="cb128-4"><a href="#cb128-4" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Corrected, empirical logFC and p-values for each gene-celltype-constrast</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a>DE_info_emp<span class="sc">$</span>de_output_tidy_emp <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(p_emp) <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 11
  gene  cluster_id logFC logCPM     F   p_val p_adj.loc  p_adj contrast    p_emp
  &lt;chr&gt; &lt;chr&gt;      &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;       &lt;dbl&gt;
1 IGKV… macrophag… -7.64   5.5   25.7 4.6 e-6    0.0334 0.0334 (OnNE-P… 8.17e-11
2 IGKV… macrophag…  7.64   5.5   25.7 4.6 e-6    0.0334 0.0334 (OnE-Pr… 8.17e-11
3 IGHV… CD4T       -5.17   4.17  14.8 2.87e-4    1      1      (OnNE-P… 1.52e- 8
4 IGHV… CD4T        5.17   4.17  14.8 2.87e-4    1      1      (OnE-Pr… 1.52e- 8
5 CCL4… CD4T        3.65   6.22  12   1   e-3    1      1      (OnNE-P… 2.96e- 7
6 CCL4… CD4T       -3.65   6.22  12   1   e-3    1      1      (OnE-Pr… 2.96e- 7
# ℹ 1 more variable: p_adj_emp &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Distributions of corrected, empirical p-values:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a>DE_info_emp<span class="sc">$</span>hist_pvals_emp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Check the empirical distributions of z-scores - the green density curves should fit well with the histogram, otherwise there might be some issues with the DE model definition.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a>DE_info_emp<span class="sc">$</span>z_distr_plots_emp_pval</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$`CD4T.(OnE-PreE)-(OnNE-PreNE)`</code></pre>
</div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-52-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
$`CD4T.(OnNE-PreNE)-(OnE-PreE)`</code></pre>
</div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-52-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
$`Fibroblast.(OnE-PreE)-(OnNE-PreNE)`</code></pre>
</div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-52-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
$`Fibroblast.(OnNE-PreNE)-(OnE-PreE)`</code></pre>
</div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-52-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
$`macrophages.(OnE-PreE)-(OnNE-PreNE)`</code></pre>
</div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-52-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
$`macrophages.(OnNE-PreNE)-(OnE-PreE)`</code></pre>
</div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-52-6.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Continue DE analysis with the estimated, empirical p-values</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a>empirical_pval <span class="ot">=</span> <span class="cn">TRUE</span></span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (empirical_pval <span class="sc">==</span><span class="cn">FALSE</span>) {</span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a>  celltype_de <span class="ot">=</span> DE_info<span class="sc">$</span>celltype_de<span class="sc">$</span>de_output_tidy</span>
<span id="cb139-4"><a href="#cb139-4" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb139-5"><a href="#cb139-5" aria-hidden="true" tabindex="-1"></a>  celltype_de <span class="ot">=</span> DE_info_emp<span class="sc">$</span>de_output_tidy_emp <span class="sc">%&gt;%</span></span>
<span id="cb139-6"><a href="#cb139-6" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>p_val, <span class="sc">-</span>p_adj) <span class="sc">%&gt;%</span></span>
<span id="cb139-7"><a href="#cb139-7" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">p_val =</span> p_emp, <span class="at">p_adj =</span> p_adj_emp)</span>
<span id="cb139-8"><a href="#cb139-8" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a>celltype_de <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 9
  gene     cluster_id   logFC logCPM      F p_adj.loc contrast       p_val p_adj
  &lt;chr&gt;    &lt;chr&gt;        &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;          &lt;dbl&gt; &lt;dbl&gt;
1 A1BG     CD4T       -0.192    5.59 0.549          1 (OnE-PreE)-(… 0.253  0.827
2 A1BG.AS1 CD4T       -0.695    3.18 3.09           1 (OnE-PreE)-(… 0.0292 0.451
3 A2M      CD4T        0.621    3.17 0.602          1 (OnE-PreE)-(… 0.228  0.803
4 AAAS     CD4T       -0.107    4.35 0.131          1 (OnE-PreE)-(… 0.578  0.970
5 PRXL2C   CD4T       -0.967    3.46 2.63           1 (OnE-PreE)-(… 0.0129 0.343
6 AAGAB    CD4T       -0.0602   5.01 0.0685         1 (OnE-PreE)-(… 0.687  0.992</code></pre>
</div>
</div>
</section>
<section id="combine-de-information-for-ligand-senders-and-receptor-receivers" class="level3">
<h3 class="anchored" data-anchor-id="combine-de-information-for-ligand-senders-and-receptor-receivers">3.4 Combine DE information for ligand-senders and receptor-receivers</h3>
<p>Combine Muscat differential expression output for senders and receivers celltypes by linking gene ligands in the sender celltypes to their cognate gene receptors in the receiver celltypes based on the prior knowledge ligand-receptor network.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a>sender_receiver_de <span class="ot">=</span> <span class="fu">combine_sender_receiver_de</span>(</span>
<span id="cb142-2"><a href="#cb142-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">sender_de =</span> celltype_de,</span>
<span id="cb142-3"><a href="#cb142-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">receiver_de =</span> celltype_de,</span>
<span id="cb142-4"><a href="#cb142-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">senders_oi =</span> senders_oi,</span>
<span id="cb142-5"><a href="#cb142-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">receivers_oi =</span> receivers_oi,</span>
<span id="cb142-6"><a href="#cb142-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">lr_network =</span> lr_network</span>
<span id="cb142-7"><a href="#cb142-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a>sender_receiver_de <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 20 × 12
   contrast              sender receiver ligand receptor lfc_ligand lfc_receptor
   &lt;chr&gt;                 &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt;         &lt;dbl&gt;        &lt;dbl&gt;
 1 (OnNE-PreNE)-(OnE-Pr… macro… CD4T     CCL17  CCR4           5.62       0.339 
 2 (OnNE-PreNE)-(OnE-Pr… Fibro… CD4T     PIP    NPTN           4.4        0.329 
 3 (OnNE-PreNE)-(OnE-Pr… macro… Fibrobl… CXCL13 ACKR4          4.04       0.529 
 4 (OnNE-PreNE)-(OnE-Pr… Fibro… macroph… PIP    NPTN           4.4        0.162 
 5 (OnNE-PreNE)-(OnE-Pr… Fibro… Fibrobl… PIP    NPTN           4.4        0.125 
 6 (OnNE-PreNE)-(OnE-Pr… macro… CD4T     CXCL13 CXCR5          4.04       0.461 
 7 (OnNE-PreNE)-(OnE-Pr… Fibro… Fibrobl… PIP    CD4            4.4        0.0816
 8 (OnNE-PreNE)-(OnE-Pr… Fibro… CD4T     PIP    CD4            4.4       -0.0236
 9 (OnNE-PreNE)-(OnE-Pr… macro… Fibrobl… CXCL13 CCR10          4.04       0.245 
10 (OnNE-PreNE)-(OnE-Pr… CD4T   CD4T     CCL4L2 CCR1           3.65       0.494 
11 (OnNE-PreNE)-(OnE-Pr… Fibro… macroph… PIP    CD4            4.4       -0.328 
12 (OnE-PreE)-(OnNE-Pre… CD4T   Fibrobl… CXCL14 CXCR4          2.65       1.28  
13 (OnE-PreE)-(OnNE-Pre… Fibro… Fibrobl… CDH2   CDH2           1.96       1.96  
14 (OnNE-PreNE)-(OnE-Pr… macro… CD4T     CXCL13 CXCR3          4.04      -0.142 
15 (OnNE-PreNE)-(OnE-Pr… Fibro… macroph… SAA1   CD36           2.77       1.06  
16 (OnNE-PreNE)-(OnE-Pr… CD4T   macroph… CCL4L2 CCR5           3.65       0.119 
17 (OnE-PreE)-(OnNE-Pre… Fibro… Fibrobl… HLA.D… CD37           1.7        2.02  
18 (OnE-PreE)-(OnNE-Pre… Fibro… Fibrobl… AREG   MMP9           2.67       1.02  
19 (OnE-PreE)-(OnNE-Pre… Fibro… Fibrobl… PENK   OGFR           3.05       0.469 
20 (OnE-PreE)-(OnNE-Pre… Fibro… Fibrobl… MMP7   SDC1           3.2        0.312 
# ℹ 5 more variables: ligand_receptor_lfc_avg &lt;dbl&gt;, p_val_ligand &lt;dbl&gt;,
#   p_adj_ligand &lt;dbl&gt;, p_val_receptor &lt;dbl&gt;, p_adj_receptor &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
</section>
<section id="step-4-nichenet-prediction-of-ligand-activities-and-ligand-target-links-inference-based-on-de-results" class="level2">
<h2 class="anchored" data-anchor-id="step-4-nichenet-prediction-of-ligand-activities-and-ligand-target-links-inference-based-on-de-results">Step 4: NicheNet prediction of ligand activities and ligand-target links inference based on DE results</h2>
<section id="define-parameters-for-nichenet-ligand-activity-analysis" class="level3">
<h3 class="anchored" data-anchor-id="define-parameters-for-nichenet-ligand-activity-analysis">4.1 Define parameters for NicheNet ligand activity analysis</h3>
<p>NicheNet ligand activity is calculated via <code>get_ligand_activities_targets_DEgenes</code> function as the enrichment of predicted target genes of ligands in the set of DE genes compared to the genomic background.&nbsp;The function requires a number of parameters to be define the gene set of interest for NicheNet ligand activity.</p>
<section id="define-the-thresholds-that-will-be-used-to-consider-genes-as-differentially-expressed-or-not-logfc-p-value-minimum-fraction-of-cells-that-should-express-a-gene-in-at-least-one-sample-in-a-group." class="level4">
<h4 class="anchored" data-anchor-id="define-the-thresholds-that-will-be-used-to-consider-genes-as-differentially-expressed-or-not-logfc-p-value-minimum-fraction-of-cells-that-should-express-a-gene-in-at-least-one-sample-in-a-group.">4.1.1 Define the thresholds that will be used to consider genes as differentially expressed or not (logFC, p-value, minimum fraction of cells that should express a gene in at least one sample in a group).</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a>logFC_threshold <span class="ot">=</span> <span class="fl">0.50</span></span>
<span id="cb145-2"><a href="#cb145-2" aria-hidden="true" tabindex="-1"></a>p_val_threshold <span class="ot">=</span> <span class="fl">0.05</span></span>
<span id="cb145-3"><a href="#cb145-3" aria-hidden="true" tabindex="-1"></a>fraction_cutoff <span class="ot">=</span> <span class="fl">0.05</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="decide-whether-to-use-adjusted-or-normal-p-values" class="level4">
<h4 class="anchored" data-anchor-id="decide-whether-to-use-adjusted-or-normal-p-values">4.1.2 Decide whether to use adjusted or normal p-values</h4>
<p>Apply the p-value cutoff on the normal p-values, and not on the p-values corrected for multiple testing because the dataset has only a few samples per group and might lack of statistical power due to pseudobulking. In case of more samples per group, and a sufficient high number of DE genes per group-celltype (&gt; 20-50), use adjusted p-values instead.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a>p_val_adj <span class="ot">=</span> <span class="cn">FALSE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="select-top-n-of-the-predicted-target-genes-per-ligand-to-consider-for-nichenet-ligand-target-inference" class="level4">
<h4 class="anchored" data-anchor-id="select-top-n-of-the-predicted-target-genes-per-ligand-to-consider-for-nichenet-ligand-target-inference">4.1.3 Select top n of the predicted target genes per ligand to consider for NicheNet ligand-target inference</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a>top_n_target <span class="ot">=</span> <span class="dv">250</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="set-the-number-of-cores-to-run-nichenet-ligand-activity-analysis-in-parallel-for-each-receiver-celltype-if-there-are-many-receiver-celltypes" class="level4">
<h4 class="anchored" data-anchor-id="set-the-number-of-cores-to-run-nichenet-ligand-activity-analysis-in-parallel-for-each-receiver-celltype-if-there-are-many-receiver-celltypes">4.1.4 Set the number of cores to run NicheNet ligand activity analysis in parallel for each receiver celltype, if there are many receiver celltypes</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a>verbose <span class="ot">=</span> <span class="cn">TRUE</span></span>
<span id="cb148-2"><a href="#cb148-2" aria-hidden="true" tabindex="-1"></a>cores_system <span class="ot">=</span> <span class="dv">8</span></span>
<span id="cb148-3"><a href="#cb148-3" aria-hidden="true" tabindex="-1"></a>n.cores <span class="ot">=</span> <span class="fu">min</span>(cores_system, sender_receiver_de<span class="sc">$</span>receiver <span class="sc">%&gt;%</span> <span class="fu">unique</span>() <span class="sc">%&gt;%</span> <span class="fu">length</span>())</span>
<span id="cb148-4"><a href="#cb148-4" aria-hidden="true" tabindex="-1"></a>n.cores</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3</code></pre>
</div>
</div>
</section>
</section>
<section id="run-nichenet-ligand-activity-analysis" class="level3">
<h3 class="anchored" data-anchor-id="run-nichenet-ligand-activity-analysis">4.2 Run NicheNet ligand activity analysis</h3>
<p>Use <code>get_ligand_activities_targets_DEgenes</code> function to predict NicheNet ligand activities and ligand-target links for the receiver cell types.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a>ligand_activities_targets_DEgenes <span class="ot">=</span> <span class="fu">suppressMessages</span>(<span class="fu">suppressWarnings</span>(</span>
<span id="cb150-2"><a href="#cb150-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">get_ligand_activities_targets_DEgenes</span>(</span>
<span id="cb150-3"><a href="#cb150-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">receiver_de =</span> celltype_de, <span class="co"># DE analysis output for receiver cell types</span></span>
<span id="cb150-4"><a href="#cb150-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">receivers_oi =</span> receivers_oi, <span class="co"># names of the receiver cell types of interest</span></span>
<span id="cb150-5"><a href="#cb150-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">ligand_target_matrix =</span> ligand_target_matrix, <span class="co"># prior knowlegde model of ligand-target regulatory potential</span></span>
<span id="cb150-6"><a href="#cb150-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">logFC_threshold =</span> logFC_threshold, <span class="co"># minimum logFC a gene should have to belong to the gene set of interest for NicheNet ligand activity</span></span>
<span id="cb150-7"><a href="#cb150-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">p_val_threshold =</span> p_val_threshold, <span class="co"># maximim p_value a gene should have</span></span>
<span id="cb150-8"><a href="#cb150-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">p_val_adj =</span> p_val_adj, <span class="co"># should the p-value corrected for multiple testing should be used for defining the gene set of interest</span></span>
<span id="cb150-9"><a href="#cb150-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">top_n_target =</span> top_n_target, <span class="co"># which top n predicted target genes</span></span>
<span id="cb150-10"><a href="#cb150-10" aria-hidden="true" tabindex="-1"></a>      <span class="at">verbose =</span> verbose,</span>
<span id="cb150-11"><a href="#cb150-11" aria-hidden="true" tabindex="-1"></a>      <span class="at">n.cores =</span> n.cores</span>
<span id="cb150-12"><a href="#cb150-12" aria-hidden="true" tabindex="-1"></a>    )))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="de-genes-used-for-the-activity-analysis" class="level4">
<h4 class="anchored" data-anchor-id="de-genes-used-for-the-activity-analysis">4.2.1 DE genes used for the activity analysis</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(ligand_activities_targets_DEgenes<span class="sc">$</span>de_genes_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3428    6</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a>ligand_activities_targets_DEgenes<span class="sc">$</span>de_genes_df <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 20 × 6
   gene       receiver logFC    p_val p_adj contrast               
   &lt;chr&gt;      &lt;chr&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;                  
 1 AC007388.1 CD4T     0.507 0.0471   0.532 (OnE-PreE)-(OnNE-PreNE)
 2 AC008124.1 CD4T     0.541 0.00911  0.313 (OnE-PreE)-(OnNE-PreNE)
 3 AC084033.3 CD4T     0.544 0.0116   0.334 (OnE-PreE)-(OnNE-PreNE)
 4 AC245014.3 CD4T     0.682 0.0228   0.412 (OnE-PreE)-(OnNE-PreNE)
 5 ACOT4      CD4T     0.95  0.00502  0.253 (OnE-PreE)-(OnNE-PreNE)
 6 ACOX3      CD4T     0.58  0.0266   0.436 (OnE-PreE)-(OnNE-PreNE)
 7 AGPAT5     CD4T     0.628 0.00230  0.182 (OnE-PreE)-(OnNE-PreNE)
 8 AL133415.1 CD4T     0.691 0.0320   0.467 (OnE-PreE)-(OnNE-PreNE)
 9 ALG11      CD4T     0.711 0.00696  0.282 (OnE-PreE)-(OnNE-PreNE)
10 ANTXR2     CD4T     0.587 0.00123  0.136 (OnE-PreE)-(OnNE-PreNE)
11 AP4E1      CD4T     0.7   0.00600  0.268 (OnE-PreE)-(OnNE-PreNE)
12 AP4M1      CD4T     0.519 0.0320   0.467 (OnE-PreE)-(OnNE-PreNE)
13 APOBR      CD4T     0.889 0.000768 0.113 (OnE-PreE)-(OnNE-PreNE)
14 APOC1      CD4T     0.908 0.0198   0.399 (OnE-PreE)-(OnNE-PreNE)
15 APOE       CD4T     1.07  0.00255  0.189 (OnE-PreE)-(OnNE-PreNE)
16 ARL5B      CD4T     0.754 0.00151  0.156 (OnE-PreE)-(OnNE-PreNE)
17 ARSG       CD4T     0.549 0.0180   0.387 (OnE-PreE)-(OnNE-PreNE)
18 ATN1       CD4T     0.544 0.0192   0.394 (OnE-PreE)-(OnNE-PreNE)
19 AUH        CD4T     0.528 0.00297  0.210 (OnE-PreE)-(OnNE-PreNE)
20 BCS1L      CD4T     0.629 0.0245   0.425 (OnE-PreE)-(OnNE-PreNE)</code></pre>
</div>
</div>
</section>
<section id="ligand-activities-and-ligand-target-links-analysis-output" class="level4">
<h4 class="anchored" data-anchor-id="ligand-activities-and-ligand-target-links-analysis-output">4.2.2 Ligand activities and ligand-target links analysis output</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb155"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(ligand_activities_targets_DEgenes<span class="sc">$</span>ligand_activities)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 202804      8</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb157"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a>ligand_activities_targets_DEgenes<span class="sc">$</span>ligand_activities <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 20 × 8
# Groups:   receiver, contrast [1]
   ligand activity contrast                target  ligand_target_weight receiver
   &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt;                   &lt;chr&gt;                  &lt;dbl&gt; &lt;chr&gt;   
 1 A2M    -0.00263 (OnE-PreE)-(OnNE-PreNE) CD70                 0.00689 CD4T    
 2 A2M    -0.00263 (OnE-PreE)-(OnNE-PreNE) CEBPD                0.00752 CD4T    
 3 A2M    -0.00263 (OnE-PreE)-(OnNE-PreNE) IL2RA                0.00743 CD4T    
 4 A2M    -0.00263 (OnE-PreE)-(OnNE-PreNE) SHC1                 0.00671 CD4T    
 5 A2M    -0.00263 (OnE-PreE)-(OnNE-PreNE) SLC1A5               0.00710 CD4T    
 6 A2M    -0.00263 (OnE-PreE)-(OnNE-PreNE) SLC2A1               0.00739 CD4T    
 7 A2M    -0.00263 (OnE-PreE)-(OnNE-PreNE) THBS1                0.0104  CD4T    
 8 AANAT  -0.00248 (OnE-PreE)-(OnNE-PreNE) CEBPD                0.00408 CD4T    
 9 AANAT  -0.00248 (OnE-PreE)-(OnNE-PreNE) IL2RA                0.00425 CD4T    
10 AANAT  -0.00248 (OnE-PreE)-(OnNE-PreNE) NBPF1                0.00445 CD4T    
11 AANAT  -0.00248 (OnE-PreE)-(OnNE-PreNE) SHC1                 0.00405 CD4T    
12 AANAT  -0.00248 (OnE-PreE)-(OnNE-PreNE) SLC1A5               0.00416 CD4T    
13 AANAT  -0.00248 (OnE-PreE)-(OnNE-PreNE) SLC2A1               0.00481 CD4T    
14 AANAT  -0.00248 (OnE-PreE)-(OnNE-PreNE) THBS1                0.00599 CD4T    
15 ABCA1  -0.00202 (OnE-PreE)-(OnNE-PreNE) ACOX3                0.0447  CD4T    
16 ABCA1  -0.00202 (OnE-PreE)-(OnNE-PreNE) EIF4EN…              0.0436  CD4T    
17 ABCA1  -0.00202 (OnE-PreE)-(OnNE-PreNE) HLA.DRA              0.0464  CD4T    
18 ABCA1  -0.00202 (OnE-PreE)-(OnNE-PreNE) TSTD2                0.0432  CD4T    
19 ABCA1  -0.00202 (OnE-PreE)-(OnNE-PreNE) VKORC1               0.0438  CD4T    
20 ABCA1  -0.00202 (OnE-PreE)-(OnNE-PreNE) ZNF235               0.0425  CD4T    
# ℹ 2 more variables: direction_regulation &lt;fct&gt;, activity_scaled &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="step-5-prioritization-of-all-sender-ligand---receiver-receptor-pairs" class="level2">
<h2 class="anchored" data-anchor-id="step-5-prioritization-of-all-sender-ligand---receiver-receptor-pairs">Step 5: Prioritization of all sender-ligand - receiver-receptor pairs</h2>
<p>MultiNicheNet prioritizes ligands according to their activity (i.e., how well they predict observed changes in gene expression in the receiver cell) and looks for affected targets with high potential to be regulated by these prioritized ligands.</p>
<p><code>generate_prioritization_tables</code> function performs the MultiNicheNet prioritization of cell-cell interactions using the previously calculated pseudobulk expression, differential expression and NicheNet activity information along with a set of user-defined prioritization weights and grouping tables.</p>
<section id="define-prioritization-weights" class="level3">
<h3 class="anchored" data-anchor-id="define-prioritization-weights">5.1 Define prioritization weights</h3>
<p>The user can define the prioritizing weights (importance) of each of the following criteria to prioritize ligand-receptor interactions:</p>
<ul>
<li><p>differential expression of ligand and receptor - indicates upregulation of the ligand in a sender cell type and/or upregulation of the receptor in a receiver cell type in the condition of interest compared to other conditions: <code>de_ligand</code> and <code>de_receptor</code>.</p></li>
<li><p>cell-type and condition specific expression of ligand in the sender cell type and receptor in the receiver cell type (to mitigate the influence of upregulated but still relatively weakly expressed ligands/receptors): <code>exprs_ligand</code> and <code>exprs_receptor</code>.</p></li>
<li><p>high NicheNet ligand activity, to further prioritize ligand-receptor pairs based in their predicted effect of the ligand-receptor interaction on the gene expression in the receiver cell type; indicates active signaling of a ligand in a receiver cell type in the condition of interest compared with other conditions: <code>activity_scaled</code>.</p></li>
<li><p>fraction of samples in a group that show high enough expression of a specific senderLigand-receiverReceptor pair (to mitigate the influence of outlier samples): <code>frac_exprs_ligand_receptor</code>.</p></li>
<li><p>high relative cell type abundance of sender/receiver in the condition of interest: <code>abund_sender</code> and <code>abund_receiver</code>.</p></li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb159"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb159-1"><a href="#cb159-1" aria-hidden="true" tabindex="-1"></a>prioritizing_weights_DE <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"de_ligand"</span> <span class="ot">=</span> <span class="dv">1</span>,</span>
<span id="cb159-2"><a href="#cb159-2" aria-hidden="true" tabindex="-1"></a>                            <span class="st">"de_receptor"</span> <span class="ot">=</span> <span class="dv">1</span>)</span>
<span id="cb159-3"><a href="#cb159-3" aria-hidden="true" tabindex="-1"></a>prioritizing_weights_activity <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"activity_scaled"</span> <span class="ot">=</span> <span class="dv">2</span>)</span>
<span id="cb159-4"><a href="#cb159-4" aria-hidden="true" tabindex="-1"></a>prioritizing_weights_expression_specificity <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"exprs_ligand"</span> <span class="ot">=</span> <span class="dv">2</span>,</span>
<span id="cb159-5"><a href="#cb159-5" aria-hidden="true" tabindex="-1"></a>                                                <span class="st">"exprs_receptor"</span> <span class="ot">=</span> <span class="dv">2</span>)</span>
<span id="cb159-6"><a href="#cb159-6" aria-hidden="true" tabindex="-1"></a>prioritizing_weights_expression_sufficiency <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"frac_exprs_ligand_receptor"</span> <span class="ot">=</span> <span class="dv">1</span>)</span>
<span id="cb159-7"><a href="#cb159-7" aria-hidden="true" tabindex="-1"></a>prioritizing_weights_relative_abundance <span class="ot">=</span> <span class="fu">c</span>( <span class="st">"abund_sender"</span> <span class="ot">=</span> <span class="dv">0</span>,</span>
<span id="cb159-8"><a href="#cb159-8" aria-hidden="true" tabindex="-1"></a>                                             <span class="st">"abund_receiver"</span> <span class="ot">=</span> <span class="dv">0</span>)</span>
<span id="cb159-9"><a href="#cb159-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb159-10"><a href="#cb159-10" aria-hidden="true" tabindex="-1"></a>prioritizing_weights <span class="ot">=</span> <span class="fu">c</span>(prioritizing_weights_DE, </span>
<span id="cb159-11"><a href="#cb159-11" aria-hidden="true" tabindex="-1"></a>                         prioritizing_weights_activity, </span>
<span id="cb159-12"><a href="#cb159-12" aria-hidden="true" tabindex="-1"></a>                         prioritizing_weights_expression_specificity,</span>
<span id="cb159-13"><a href="#cb159-13" aria-hidden="true" tabindex="-1"></a>                         prioritizing_weights_expression_sufficiency, </span>
<span id="cb159-14"><a href="#cb159-14" aria-hidden="true" tabindex="-1"></a>                         prioritizing_weights_relative_abundance)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="define-grouping-data-frames" class="level3">
<h3 class="anchored" data-anchor-id="define-grouping-data-frames">5.2 Define grouping data frames</h3>
<p>Define dataframe with all sender-receiver cell type combinations, with columns: sender and receiver</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb160"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb160-1"><a href="#cb160-1" aria-hidden="true" tabindex="-1"></a>sender_receiver_tbl <span class="ot">=</span> sender_receiver_de <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">distinct</span>(sender, receiver)</span>
<span id="cb160-2"><a href="#cb160-2" aria-hidden="true" tabindex="-1"></a>sender_receiver_tbl</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 9 × 2
  sender      receiver   
  &lt;chr&gt;       &lt;chr&gt;      
1 macrophages CD4T       
2 Fibroblast  CD4T       
3 macrophages Fibroblast 
4 Fibroblast  macrophages
5 Fibroblast  Fibroblast 
6 CD4T        CD4T       
7 CD4T        Fibroblast 
8 CD4T        macrophages
9 macrophages macrophages</code></pre>
</div>
</div>
<p>Define dataframe showing the groups of each sample (and batches per sample if applicable, with columns: sample and group, and if applicable all batches of interest.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb162"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a>metadata_combined <span class="ot">=</span> SummarizedExperiment<span class="sc">::</span><span class="fu">colData</span>(sce) <span class="sc">%&gt;%</span> tibble<span class="sc">::</span><span class="fu">as.tibble</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: `as.tibble()` was deprecated in tibble 2.0.0.
ℹ Please use `as_tibble()` instead.
ℹ The signature and semantics have changed, see `?as_tibble`.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb164"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb164-1"><a href="#cb164-1" aria-hidden="true" tabindex="-1"></a>metadata_combined <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 18
  Cell     orig.ident nCount_RNA nFeature_RNA nCount_SCT nFeature_SCT patient_id
  &lt;chr&gt;    &lt;fct&gt;           &lt;dbl&gt;        &lt;int&gt;      &lt;dbl&gt;        &lt;int&gt; &lt;chr&gt;     
1 BIOKEY_… BIOKEY           4798         2115       4084         2115 BIOKEY_10 
2 BIOKEY_… BIOKEY           1524          888       3064          902 BIOKEY_10 
3 BIOKEY_… BIOKEY           6009         2257       4298         2227 BIOKEY_10 
4 BIOKEY_… BIOKEY          24516         4494       4654         1676 BIOKEY_10 
5 BIOKEY_… BIOKEY           1274          751       2660          821 BIOKEY_10 
6 BIOKEY_… BIOKEY            684          517       2210          714 BIOKEY_10 
# ℹ 11 more variables: timepoint &lt;chr&gt;, expansion &lt;chr&gt;, BC_type &lt;chr&gt;,
#   cellType &lt;chr&gt;, cohort &lt;chr&gt;, expansion_timepoint &lt;chr&gt;, sample_id &lt;chr&gt;,
#   subType &lt;chr&gt;, ident &lt;fct&gt;, ang_response &lt;chr&gt;, cellSubType &lt;chr&gt;</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb166"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb166-1"><a href="#cb166-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.na</span>(batches)) {</span>
<span id="cb166-2"><a href="#cb166-2" aria-hidden="true" tabindex="-1"></a>  grouping_tbl <span class="ot">=</span> metadata_combined[, <span class="fu">c</span>(sample_id, group_id, batches)] <span class="sc">%&gt;%</span></span>
<span id="cb166-3"><a href="#cb166-3" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">as.tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb166-4"><a href="#cb166-4" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">distinct</span>()</span>
<span id="cb166-5"><a href="#cb166-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(grouping_tbl) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"sample"</span>, <span class="st">"group"</span>, batches)</span>
<span id="cb166-6"><a href="#cb166-6" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb166-7"><a href="#cb166-7" aria-hidden="true" tabindex="-1"></a>  grouping_tbl <span class="ot">=</span> metadata_combined[, <span class="fu">c</span>(sample_id, group_id)] <span class="sc">%&gt;%</span></span>
<span id="cb166-8"><a href="#cb166-8" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">as.tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb166-9"><a href="#cb166-9" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">distinct</span>()</span>
<span id="cb166-10"><a href="#cb166-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(grouping_tbl) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"sample"</span>, <span class="st">"group"</span>)</span>
<span id="cb166-11"><a href="#cb166-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb166-12"><a href="#cb166-12" aria-hidden="true" tabindex="-1"></a>grouping_tbl <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 2
  sample       group
  &lt;chr&gt;        &lt;chr&gt;
1 BIOKEY_10Pre PreE 
2 BIOKEY_10On  OnE  
3 BIOKEY_16Pre PreE 
4 BIOKEY_16On  OnE  
5 BIOKEY_14Pre PreNE
6 BIOKEY_14On  OnNE </code></pre>
</div>
</div>
<p>Recall contrasts table for consistency in naming of the group column</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb168"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb168-1"><a href="#cb168-1" aria-hidden="true" tabindex="-1"></a>contrast_tbl <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  contrast                group
  &lt;chr&gt;                   &lt;chr&gt;
1 (OnE-PreE)-(OnNE-PreNE) OnE  
2 (OnNE-PreNE)-(OnE-PreE) OnNE </code></pre>
</div>
</div>
</section>
<section id="run-the-prioritization" class="level3">
<h3 class="anchored" data-anchor-id="run-the-prioritization">5.3 Run the prioritization</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb170"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb170-1"><a href="#cb170-1" aria-hidden="true" tabindex="-1"></a>prioritization_tables <span class="ot">=</span> <span class="fu">suppressMessages</span>(</span>
<span id="cb170-2"><a href="#cb170-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">generate_prioritization_tables</span>(</span>
<span id="cb170-3"><a href="#cb170-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">sender_receiver_info =</span> abundance_expression_info<span class="sc">$</span>sender_receiver_info,</span>
<span id="cb170-4"><a href="#cb170-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">sender_receiver_de =</span> sender_receiver_de,</span>
<span id="cb170-5"><a href="#cb170-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">ligand_activities_targets_DEgenes =</span> ligand_activities_targets_DEgenes,</span>
<span id="cb170-6"><a href="#cb170-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">contrast_tbl =</span> contrast_tbl,</span>
<span id="cb170-7"><a href="#cb170-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">sender_receiver_tbl =</span> sender_receiver_tbl,</span>
<span id="cb170-8"><a href="#cb170-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">grouping_tbl =</span> grouping_tbl,</span>
<span id="cb170-9"><a href="#cb170-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">prioritizing_weights =</span> prioritizing_weights,</span>
<span id="cb170-10"><a href="#cb170-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">fraction_cutoff =</span> fraction_cutoff,</span>
<span id="cb170-11"><a href="#cb170-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">abundance_data_receiver =</span> abundance_expression_info<span class="sc">$</span>abundance_data_receiver,</span>
<span id="cb170-12"><a href="#cb170-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">abundance_data_sender =</span> abundance_expression_info<span class="sc">$</span>abundance_data_sender</span>
<span id="cb170-13"><a href="#cb170-13" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb170-14"><a href="#cb170-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Prioritized senderLigand-receiverReceptor interactions</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb171"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb171-1"><a href="#cb171-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(prioritization_tables)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 3
 $ group_prioritization_tbl       : tibble [23,520 × 59] (S3: tbl_df/tbl/data.frame)
  ..$ contrast                           : chr [1:23520] "(OnE-PreE)-(OnNE-PreNE)" "(OnE-PreE)-(OnNE-PreNE)" "(OnE-PreE)-(OnNE-PreNE)" "(OnE-PreE)-(OnNE-PreNE)" ...
  ..$ group                              : chr [1:23520] "OnE" "OnE" "OnE" "OnE" ...
  ..$ sender                             : chr [1:23520] "macrophages" "macrophages" "Fibroblast" "Fibroblast" ...
  ..$ receiver                           : chr [1:23520] "Fibroblast" "Fibroblast" "CD4T" "CD4T" ...
  ..$ ligand                             : chr [1:23520] "TNF" "TNF" "TGM2" "TGM2" ...
  ..$ receptor                           : chr [1:23520] "LTBR" "LTBR" "ITGA4" "ITGA4" ...
  ..$ lfc_ligand                         : num [1:23520] 0.845 0.845 1.41 1.41 0.845 0.845 0.845 0.845 1.61 1.61 ...
  ..$ lfc_receptor                       : num [1:23520] 0.181 0.181 0.311 0.311 0.184 0.184 0.285 0.285 0.142 0.142 ...
  ..$ ligand_receptor_lfc_avg            : num [1:23520] 0.513 0.513 0.86 0.86 0.514 ...
  ..$ p_val_ligand                       : num [1:23520] 0.0658 0.0658 0.0182 0.0182 0.0658 ...
  ..$ p_adj_ligand                       : num [1:23520] 0.547 0.547 0.422 0.422 0.547 ...
  ..$ p_val_receptor                     : num [1:23520] 0.2143 0.2143 0.0401 0.0401 0.2598 ...
  ..$ p_adj_receptor                     : num [1:23520] 0.794 0.794 0.495 0.495 0.825 ...
  ..$ activity                           : num [1:23520] 0.09813 -0.00362 -0.0019 0.01035 0.02971 ...
  ..$ direction_regulation               : Factor w/ 2 levels "up","down": 1 2 1 2 1 2 1 2 1 2 ...
  ..$ activity_scaled                    : num [1:23520] 3.302 0.194 0.221 2.477 4.855 ...
  ..$ lr_interaction                     : chr [1:23520] "TNF_LTBR" "TNF_LTBR" "TGM2_ITGA4" "TGM2_ITGA4" ...
  ..$ id                                 : chr [1:23520] "TNF_LTBR_macrophages_Fibroblast" "TNF_LTBR_macrophages_Fibroblast" "TGM2_ITGA4_Fibroblast_CD4T" "TGM2_ITGA4_Fibroblast_CD4T" ...
  ..$ avg_ligand_group                   : num [1:23520] 0.245 0.245 0.199 0.199 0.245 ...
  ..$ avg_receptor_group                 : num [1:23520] 0.294 0.294 0.303 0.303 0.242 ...
  ..$ ligand_receptor_prod_group         : num [1:23520] 0.072 0.072 0.0604 0.0604 0.0591 ...
  ..$ fraction_ligand_group              : num [1:23520] 0.219 0.219 0.203 0.203 0.219 ...
  ..$ fraction_receptor_group            : num [1:23520] 0.379 0.379 0.345 0.345 0.327 ...
  ..$ ligand_receptor_fraction_prod_group: num [1:23520] 0.0828 0.0828 0.0701 0.0701 0.0715 ...
  ..$ rel_abundance_scaled_sender        : num [1:23520] 1.001 1.001 0.001 0.001 1.001 ...
  .. ..- attr(*, "addend")= Named num -0.0744
  .. .. ..- attr(*, "names")= chr "0%"
  .. ..- attr(*, "multiplier")= num 3.08
  ..$ rel_abundance_scaled_receiver      : num [1:23520] 0.001 0.001 0.898 0.898 1.001 ...
  .. ..- attr(*, "addend")= Named num -0.0744
  .. .. ..- attr(*, "names")= chr "0%"
  .. ..- attr(*, "multiplier")= num 3.08
  ..$ sender_receiver_rel_abundance_avg  : num [1:23520] 0.501 0.501 0.45 0.45 1.001 ...
  .. ..- attr(*, "addend")= Named num -0.0744
  .. .. ..- attr(*, "names")= chr "0%"
  .. ..- attr(*, "multiplier")= num 3.08
  ..$ lfc_pval_ligand                    : num [1:23520] 0.999 0.999 2.454 2.454 0.999 ...
  ..$ p_val_ligand_adapted               : num [1:23520] 1.18 1.18 1.74 1.74 1.18 ...
  ..$ scaled_lfc_ligand                  : num [1:23520] 0.926 0.926 0.974 0.974 0.926 ...
  ..$ scaled_p_val_ligand                : num [1:23520] 0.837 0.837 0.926 0.926 0.837 ...
  ..$ scaled_lfc_pval_ligand             : num [1:23520] 0.932 0.932 0.974 0.974 0.932 ...
  ..$ scaled_p_val_ligand_adapted        : num [1:23520] 0.919 0.919 0.963 0.963 0.919 ...
  ..$ lfc_pval_receptor                  : num [1:23520] 0.121 0.121 0.434 0.434 0.108 ...
  ..$ p_val_receptor_adapted             : num [1:23520] 0.669 0.669 1.397 1.397 0.585 ...
  ..$ scaled_lfc_receptor                : num [1:23520] 0.691 0.691 0.789 0.789 0.694 ...
  ..$ scaled_p_val_receptor              : num [1:23520] 0.694 0.694 0.885 0.885 0.648 ...
  ..$ scaled_lfc_pval_receptor           : num [1:23520] 0.769 0.769 0.887 0.887 0.76 ...
  ..$ scaled_p_val_receptor_adapted      : num [1:23520] 0.847 0.847 0.943 0.943 0.824 ...
  ..$ activity_up                        : num [1:23520] 0.0981 0.0981 -0.0019 -0.0019 0.0297 ...
  ..$ activity_scaled_up                 : num [1:23520] 3.302 3.302 0.221 0.221 4.855 ...
  ..$ scaled_activity_scaled_up          : num [1:23520] 1.001 1.001 0.469 0.469 1.001 ...
  .. ..- attr(*, "addend")= Named num 2.29
  .. .. ..- attr(*, "names")= chr "1%"
  .. ..- attr(*, "multiplier")= num 0.186
  ..$ scaled_activity_up                 : num [1:23520] 1.001 1.001 0.0491 0.0491 0.5106 ...
  .. ..- attr(*, "addend")= Named num 0.0052
  .. .. ..- attr(*, "names")= chr "1%"
  .. ..- attr(*, "multiplier")= num 14.6
  ..$ activity_down                      : num [1:23520] -0.00362 -0.00362 0.01035 0.01035 0.00181 ...
  ..$ activity_scaled_down               : num [1:23520] 0.194 0.194 2.477 2.477 -2.144 ...
  ..$ scaled_activity_scaled_down        : num [1:23520] 0.4639 0.4639 0.8896 0.8896 0.0279 ...
  .. ..- attr(*, "addend")= Named num 2.29
  .. .. ..- attr(*, "names")= chr "1%"
  .. ..- attr(*, "multiplier")= num 0.186
  ..$ scaled_activity_down               : num [1:23520] 0.0241 0.0241 0.2279 0.2279 0.1034 ...
  .. ..- attr(*, "addend")= Named num 0.0052
  .. .. ..- attr(*, "names")= chr "1%"
  .. ..- attr(*, "multiplier")= num 14.6
  ..$ scaled_avg_exprs_ligand            : num [1:23520] 1 1 1 1 1 ...
  ..$ scaled_avg_frq_ligand              : num [1:23520] 1.001 1.001 0.857 0.857 1.001 ...
  ..$ pb_ligand_group                    : num [1:23520] 7.32 7.32 6.07 6.07 7.32 ...
  ..$ scaled_pb_ligand                   : num [1:23520] 1 1 1 1 1 ...
  ..$ scaled_avg_exprs_receptor          : num [1:23520] 1 1 1 1 0.82 ...
  ..$ scaled_avg_frq_receptor            : num [1:23520] 1.001 1.001 1.001 1.001 0.863 ...
  ..$ pb_receptor_group                  : num [1:23520] 6.89 6.89 7.25 7.25 6.76 ...
  ..$ scaled_pb_receptor                 : num [1:23520] 1 1 1 1 0.98 ...
  ..$ fraction_expressing_ligand_receptor: num [1:23520] 1 1 0.889 0.889 1 ...
  ..$ max_scaled_activity                : num [1:23520] 1 1 0.89 0.89 1 ...
  .. ..- attr(*, "addend")= Named num 2.29
  .. .. ..- attr(*, "names")= chr "1%"
  .. ..- attr(*, "multiplier")= num 0.186
  ..$ prioritization_score               : num [1:23520] 0.944 0.944 0.94 0.94 0.939 ...
  .. ..- attr(*, "addend")= Named num 2.29
  .. .. ..- attr(*, "names")= chr "1%"
  .. ..- attr(*, "multiplier")= num 0.186
  ..$ top_group                          : chr [1:23520] "OnE" "OnE" "OnE" "OnE" ...
 $ sample_prioritization_tbl      : tibble [2,309,328 × 26] (S3: tbl_df/tbl/data.frame)
  ..$ sample                       : chr [1:2309328] "BIOKEY_24Pre" "BIOKEY_7On" "BIOKEY_27Pre" "BIOKEY_6Pre" ...
  ..$ sender                       : chr [1:2309328] "macrophages" "Fibroblast" "macrophages" "macrophages" ...
  ..$ receiver                     : chr [1:2309328] "macrophages" "macrophages" "macrophages" "macrophages" ...
  ..$ ligand                       : chr [1:2309328] "HLA.DMA" "MIF" "HLA.DMA" "HLA.DMA" ...
  ..$ receptor                     : chr [1:2309328] "CD74" "CD74" "CD74" "CD74" ...
  ..$ avg_ligand                   : num [1:2309328] 1.97 2.26 1.96 1.87 2.58 ...
  ..$ avg_receptor                 : num [1:2309328] 4.96 4.29 4.78 5 3.56 ...
  ..$ ligand_receptor_prod         : num [1:2309328] 9.79 9.68 9.39 9.33 9.18 ...
  ..$ fraction_ligand              : num [1:2309328] 0.976 0.992 0.957 0.966 1 ...
  ..$ fraction_receptor            : num [1:2309328] 1 1 1 1 0.997 ...
  ..$ ligand_receptor_fraction_prod: num [1:2309328] 0.976 0.992 0.957 0.966 0.997 ...
  ..$ pb_ligand                    : num [1:2309328] 10.6 11.5 10.8 10.7 11.9 ...
  ..$ pb_receptor                  : num [1:2309328] 15.1 14.5 15 15.4 14.3 ...
  ..$ ligand_receptor_pb_prod      : num [1:2309328] 161 167 161 165 170 ...
  ..$ group                        : chr [1:2309328] "PreNE" "OnNE" "PreNE" "PreNE" ...
  ..$ prioritization_score         : num [1:2309328] NA 0.479 NA NA 0.869 ...
  .. ..- attr(*, "addend")= Named num 2.29
  .. .. ..- attr(*, "names")= chr "1%"
  .. ..- attr(*, "multiplier")= num 0.186
  ..$ lr_interaction               : chr [1:2309328] "HLA.DMA_CD74" "MIF_CD74" "HLA.DMA_CD74" "HLA.DMA_CD74" ...
  ..$ id                           : chr [1:2309328] "HLA.DMA_CD74_macrophages_macrophages" "MIF_CD74_Fibroblast_macrophages" "HLA.DMA_CD74_macrophages_macrophages" "HLA.DMA_CD74_macrophages_macrophages" ...
  ..$ scaled_LR_prod               : num [1:2309328] 1.91 3.32 1.7 1.68 2.9 ...
  ..$ scaled_LR_frac               : num [1:2309328] 1.064 1.381 0.889 0.972 1.484 ...
  ..$ scaled_LR_pb_prod            : num [1:2309328] 1.1 2.48 1.13 1.35 2.75 ...
  ..$ n_cells_receiver             : num [1:2309328] 82 158 23 116 793 32 17 77 67 29 ...
  ..$ keep_receiver                : num [1:2309328] 1 1 1 1 1 1 1 1 1 1 ...
  ..$ n_cells_sender               : num [1:2309328] 82 1176 23 116 37 ...
  ..$ keep_sender                  : num [1:2309328] 1 1 1 1 1 1 1 1 1 1 ...
  ..$ keep_sender_receiver         : Factor w/ 4 levels "Sender &amp; Receiver absent",..: 4 4 4 4 4 4 4 4 4 4 ...
 $ ligand_activities_target_de_tbl: gropd_df [202,802 × 11] (S3: grouped_df/tbl_df/tbl/data.frame)
  ..$ contrast            : chr [1:202802] "(OnE-PreE)-(OnNE-PreNE)" "(OnE-PreE)-(OnNE-PreNE)" "(OnE-PreE)-(OnNE-PreNE)" "(OnE-PreE)-(OnNE-PreNE)" ...
  ..$ receiver            : chr [1:202802] "CD4T" "CD4T" "CD4T" "CD4T" ...
  ..$ ligand              : chr [1:202802] "A2M" "A2M" "A2M" "A2M" ...
  ..$ activity            : num [1:202802] -0.00263 -0.00263 -0.00263 -0.00263 -0.00263 ...
  ..$ activity_scaled     : num [1:202802] -0.412 -0.412 -0.412 -0.412 -0.412 ...
  ..$ target              : chr [1:202802] "CD70" "CEBPD" "IL2RA" "SHC1" ...
  ..$ ligand_target_weight: Named num [1:202802] 0.00689 0.00752 0.00743 0.00671 0.0071 ...
  .. ..- attr(*, "names")= chr [1:202802] "CD70" "CEBPD" "IL2RA" "SHC1" ...
  ..$ logFC               : num [1:202802] 0.646 1.08 1.37 0.563 0.563 0.723 1.19 1.08 1.37 0.846 ...
  ..$ p_val               : num [1:202802] 2.11e-02 2.18e-02 8.46e-05 1.20e-03 1.02e-02 ...
  ..$ p_val_adj           : num [1:202802] 0.4101 0.4113 0.0432 0.1358 0.321 ...
  ..$ direction_regulation: Factor w/ 2 levels "up","down": 1 1 1 1 1 1 1 1 1 1 ...
  ..- attr(*, "groups")= tibble [6 × 3] (S3: tbl_df/tbl/data.frame)
  .. ..$ receiver: chr [1:6] "CD4T" "CD4T" "Fibroblast" "Fibroblast" ...
  .. ..$ contrast: chr [1:6] "(OnE-PreE)-(OnNE-PreNE)" "(OnNE-PreNE)-(OnE-PreE)" "(OnE-PreE)-(OnNE-PreNE)" "(OnNE-PreNE)-(OnE-PreE)" ...
  .. ..$ .rows   : list&lt;int&gt; [1:6] 
  .. .. ..$ : int [1:23061] 1 2 3 4 5 6 7 8 9 10 ...
  .. .. ..$ : int [1:23061] 23062 23063 23064 23065 23066 23067 23068 23069 23070 23071 ...
  .. .. ..$ : int [1:45380] 46123 46124 46125 46126 46127 46128 46129 46130 46131 46132 ...
  .. .. ..$ : int [1:45380] 91503 91504 91505 91506 91507 91508 91509 91510 91511 91512 ...
  .. .. ..$ : int [1:32960] 136883 136884 136885 136886 136887 136888 136889 136890 136891 136892 ...
  .. .. ..$ : int [1:32960] 169843 169844 169845 169846 169847 169848 169849 169850 169851 169852 ...
  .. .. ..@ ptype: int(0) 
  .. ..- attr(*, ".drop")= logi TRUE</code></pre>
</div>
</div>
<section id="group-based-summary-table" class="level4">
<h4 class="anchored" data-anchor-id="group-based-summary-table">5.3.1 Group-based summary table</h4>
<p>Contains expression information of each LR pair per group</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb173"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb173-1"><a href="#cb173-1" aria-hidden="true" tabindex="-1"></a>prioritization_tables<span class="sc">$</span>group_prioritization_tbl <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 20 × 59
   contrast        group sender receiver ligand receptor lfc_ligand lfc_receptor
   &lt;chr&gt;           &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt;         &lt;dbl&gt;        &lt;dbl&gt;
 1 (OnE-PreE)-(On… OnE   macro… Fibrobl… TNF    LTBR          0.845       0.181 
 2 (OnE-PreE)-(On… OnE   macro… Fibrobl… TNF    LTBR          0.845       0.181 
 3 (OnE-PreE)-(On… OnE   Fibro… CD4T     TGM2   ITGA4         1.41        0.311 
 4 (OnE-PreE)-(On… OnE   Fibro… CD4T     TGM2   ITGA4         1.41        0.311 
 5 (OnE-PreE)-(On… OnE   macro… macroph… TNF    LTBR          0.845       0.184 
 6 (OnE-PreE)-(On… OnE   macro… macroph… TNF    LTBR          0.845       0.184 
 7 (OnE-PreE)-(On… OnE   macro… CD4T     TNF    TRAF2         0.845       0.285 
 8 (OnE-PreE)-(On… OnE   macro… CD4T     TNF    TRAF2         0.845       0.285 
 9 (OnE-PreE)-(On… OnE   macro… CD4T     CXCL9  CXCR3         1.61        0.142 
10 (OnE-PreE)-(On… OnE   macro… CD4T     CXCL9  CXCR3         1.61        0.142 
11 (OnE-PreE)-(On… OnE   macro… Fibrobl… TNF    TNFRSF1A      0.845       0.0877
12 (OnE-PreE)-(On… OnE   macro… Fibrobl… TNF    TNFRSF1A      0.845       0.0877
13 (OnE-PreE)-(On… OnE   Fibro… CD4T     CXCL9  CXCR3         2.01        0.142 
14 (OnE-PreE)-(On… OnE   Fibro… CD4T     CXCL9  CXCR3         2.01        0.142 
15 (OnNE-PreNE)-(… OnNE  CD4T   CD4T     PTPRC  DPP4          0.146       0.379 
16 (OnNE-PreNE)-(… OnNE  CD4T   CD4T     PTPRC  DPP4          0.146       0.379 
17 (OnE-PreE)-(On… OnE   Fibro… macroph… CSF1   CSF3R         0.431       0.695 
18 (OnE-PreE)-(On… OnE   Fibro… macroph… CSF1   CSF3R         0.431       0.695 
19 (OnE-PreE)-(On… OnE   macro… CD4T     TNF    TNFRSF1B      0.845       0.148 
20 (OnE-PreE)-(On… OnE   macro… CD4T     TNF    TNFRSF1B      0.845       0.148 
# ℹ 51 more variables: ligand_receptor_lfc_avg &lt;dbl&gt;, p_val_ligand &lt;dbl&gt;,
#   p_adj_ligand &lt;dbl&gt;, p_val_receptor &lt;dbl&gt;, p_adj_receptor &lt;dbl&gt;,
#   activity &lt;dbl&gt;, direction_regulation &lt;fct&gt;, activity_scaled &lt;dbl&gt;,
#   lr_interaction &lt;chr&gt;, id &lt;chr&gt;, avg_ligand_group &lt;dbl&gt;,
#   avg_receptor_group &lt;dbl&gt;, ligand_receptor_prod_group &lt;dbl&gt;,
#   fraction_ligand_group &lt;dbl&gt;, fraction_receptor_group &lt;dbl&gt;,
#   ligand_receptor_fraction_prod_group &lt;dbl&gt;, …</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb175"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb175-1"><a href="#cb175-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(prioritization_tables<span class="sc">$</span>group_prioritization_tbl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "contrast"                            "group"                              
 [3] "sender"                              "receiver"                           
 [5] "ligand"                              "receptor"                           
 [7] "lfc_ligand"                          "lfc_receptor"                       
 [9] "ligand_receptor_lfc_avg"             "p_val_ligand"                       
[11] "p_adj_ligand"                        "p_val_receptor"                     
[13] "p_adj_receptor"                      "activity"                           
[15] "direction_regulation"                "activity_scaled"                    
[17] "lr_interaction"                      "id"                                 
[19] "avg_ligand_group"                    "avg_receptor_group"                 
[21] "ligand_receptor_prod_group"          "fraction_ligand_group"              
[23] "fraction_receptor_group"             "ligand_receptor_fraction_prod_group"
[25] "rel_abundance_scaled_sender"         "rel_abundance_scaled_receiver"      
[27] "sender_receiver_rel_abundance_avg"   "lfc_pval_ligand"                    
[29] "p_val_ligand_adapted"                "scaled_lfc_ligand"                  
[31] "scaled_p_val_ligand"                 "scaled_lfc_pval_ligand"             
[33] "scaled_p_val_ligand_adapted"         "lfc_pval_receptor"                  
[35] "p_val_receptor_adapted"              "scaled_lfc_receptor"                
[37] "scaled_p_val_receptor"               "scaled_lfc_pval_receptor"           
[39] "scaled_p_val_receptor_adapted"       "activity_up"                        
[41] "activity_scaled_up"                  "scaled_activity_scaled_up"          
[43] "scaled_activity_up"                  "activity_down"                      
[45] "activity_scaled_down"                "scaled_activity_scaled_down"        
[47] "scaled_activity_down"                "scaled_avg_exprs_ligand"            
[49] "scaled_avg_frq_ligand"               "pb_ligand_group"                    
[51] "scaled_pb_ligand"                    "scaled_avg_exprs_receptor"          
[53] "scaled_avg_frq_receptor"             "pb_receptor_group"                  
[55] "scaled_pb_receptor"                  "fraction_expressing_ligand_receptor"
[57] "max_scaled_activity"                 "prioritization_score"               
[59] "top_group"                          </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb177"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb177-1"><a href="#cb177-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(prioritization_tables<span class="sc">$</span>group_prioritization_tbl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 23520    59</code></pre>
</div>
</div>
</section>
<section id="sample-based-summary-table" class="level4">
<h4 class="anchored" data-anchor-id="sample-based-summary-table">5.3.2 Sample-based summary table</h4>
<p>Contains expression information of each LR pair per sample</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb179"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb179-1"><a href="#cb179-1" aria-hidden="true" tabindex="-1"></a>prioritization_tables<span class="sc">$</span>sample_prioritization_tbl <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 20 × 26
   sample       sender      receiver    ligand  receptor avg_ligand avg_receptor
   &lt;chr&gt;        &lt;chr&gt;       &lt;chr&gt;       &lt;chr&gt;   &lt;chr&gt;         &lt;dbl&gt;        &lt;dbl&gt;
 1 BIOKEY_24Pre macrophages macrophages HLA.DMA CD74           1.97         4.96
 2 BIOKEY_7On   Fibroblast  macrophages MIF     CD74           2.26         4.29
 3 BIOKEY_27Pre macrophages macrophages HLA.DMA CD74           1.96         4.78
 4 BIOKEY_6Pre  macrophages macrophages HLA.DMA CD74           1.87         5.00
 5 BIOKEY_28On  Fibroblast  macrophages MIF     CD74           2.58         3.56
 6 BIOKEY_14Pre macrophages macrophages HLA.DMA CD74           1.82         4.87
 7 BIOKEY_7Pre  Fibroblast  macrophages MIF     CD74           2.10         4.22
 8 BIOKEY_30On  macrophages macrophages HLA.DMA CD74           1.80         4.90
 9 BIOKEY_30Pre macrophages macrophages HLA.DMA CD74           1.72         4.90
10 BIOKEY_18Pre macrophages macrophages HLA.DMA CD74           1.71         4.90
11 BIOKEY_20Pre macrophages macrophages HLA.DMA CD74           1.68         4.93
12 BIOKEY_18On  macrophages macrophages HLA.DMA CD74           1.67         4.90
13 BIOKEY_31Pre macrophages macrophages MIF     CD74           2.41         3.38
14 BIOKEY_5On   Fibroblast  macrophages MIF     CD74           1.86         4.37
15 BIOKEY_10On  macrophages macrophages HLA.DMA CD74           1.68         4.74
16 BIOKEY_12On  macrophages macrophages HLA.DMA CD74           1.65         4.81
17 BIOKEY_4Pre  macrophages macrophages HLA.DMA CD74           1.68         4.66
18 BIOKEY_19Pre macrophages macrophages HLA.DMA CD74           1.63         4.81
19 BIOKEY_28On  macrophages macrophages MIF     CD74           2.17         3.56
20 BIOKEY_27On  macrophages macrophages HLA.DMA CD74           1.77         4.34
# ℹ 19 more variables: ligand_receptor_prod &lt;dbl&gt;, fraction_ligand &lt;dbl&gt;,
#   fraction_receptor &lt;dbl&gt;, ligand_receptor_fraction_prod &lt;dbl&gt;,
#   pb_ligand &lt;dbl&gt;, pb_receptor &lt;dbl&gt;, ligand_receptor_pb_prod &lt;dbl&gt;,
#   group &lt;chr&gt;, prioritization_score &lt;dbl&gt;, lr_interaction &lt;chr&gt;, id &lt;chr&gt;,
#   scaled_LR_prod &lt;dbl&gt;, scaled_LR_frac &lt;dbl&gt;, scaled_LR_pb_prod &lt;dbl&gt;,
#   n_cells_receiver &lt;dbl&gt;, keep_receiver &lt;dbl&gt;, n_cells_sender &lt;dbl&gt;,
#   keep_sender &lt;dbl&gt;, keep_sender_receiver &lt;fct&gt;</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb181"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb181-1"><a href="#cb181-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(prioritization_tables<span class="sc">$</span>sample_prioritization_tbl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "sample"                        "sender"                       
 [3] "receiver"                      "ligand"                       
 [5] "receptor"                      "avg_ligand"                   
 [7] "avg_receptor"                  "ligand_receptor_prod"         
 [9] "fraction_ligand"               "fraction_receptor"            
[11] "ligand_receptor_fraction_prod" "pb_ligand"                    
[13] "pb_receptor"                   "ligand_receptor_pb_prod"      
[15] "group"                         "prioritization_score"         
[17] "lr_interaction"                "id"                           
[19] "scaled_LR_prod"                "scaled_LR_frac"               
[21] "scaled_LR_pb_prod"             "n_cells_receiver"             
[23] "keep_receiver"                 "n_cells_sender"               
[25] "keep_sender"                   "keep_sender_receiver"         </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb183"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb183-1"><a href="#cb183-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(prioritization_tables<span class="sc">$</span>sample_prioritization_tbl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2309328      26</code></pre>
</div>
</div>
</section>
<section id="ligand-activities-and-ligand-target-links" class="level4">
<h4 class="anchored" data-anchor-id="ligand-activities-and-ligand-target-links">5.3.3 Ligand activities and ligand-target links</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb185"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb185-1"><a href="#cb185-1" aria-hidden="true" tabindex="-1"></a>prioritization_tables<span class="sc">$</span>ligand_activities_target_de_tbl <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 20 × 11
# Groups:   receiver, contrast [1]
   contrast receiver ligand activity activity_scaled target ligand_target_weight
   &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;     &lt;dbl&gt;           &lt;dbl&gt; &lt;chr&gt;                 &lt;dbl&gt;
 1 (OnE-Pr… CD4T     A2M    -0.00263          -0.412 CD70                0.00689
 2 (OnE-Pr… CD4T     A2M    -0.00263          -0.412 CEBPD               0.00752
 3 (OnE-Pr… CD4T     A2M    -0.00263          -0.412 IL2RA               0.00743
 4 (OnE-Pr… CD4T     A2M    -0.00263          -0.412 SHC1                0.00671
 5 (OnE-Pr… CD4T     A2M    -0.00263          -0.412 SLC1A5              0.00710
 6 (OnE-Pr… CD4T     A2M    -0.00263          -0.412 SLC2A1              0.00739
 7 (OnE-Pr… CD4T     A2M    -0.00263          -0.412 THBS1               0.0104 
 8 (OnE-Pr… CD4T     AANAT  -0.00248          -0.280 CEBPD               0.00408
 9 (OnE-Pr… CD4T     AANAT  -0.00248          -0.280 IL2RA               0.00425
10 (OnE-Pr… CD4T     AANAT  -0.00248          -0.280 NBPF1               0.00445
11 (OnE-Pr… CD4T     AANAT  -0.00248          -0.280 SHC1                0.00405
12 (OnE-Pr… CD4T     AANAT  -0.00248          -0.280 SLC1A5              0.00416
13 (OnE-Pr… CD4T     AANAT  -0.00248          -0.280 SLC2A1              0.00481
14 (OnE-Pr… CD4T     AANAT  -0.00248          -0.280 THBS1               0.00599
15 (OnE-Pr… CD4T     ABCA1  -0.00202           0.122 ACOX3               0.0447 
16 (OnE-Pr… CD4T     ABCA1  -0.00202           0.122 EIF4E…              0.0436 
17 (OnE-Pr… CD4T     ABCA1  -0.00202           0.122 HLA.D…              0.0464 
18 (OnE-Pr… CD4T     ABCA1  -0.00202           0.122 TSTD2               0.0432 
19 (OnE-Pr… CD4T     ABCA1  -0.00202           0.122 VKORC1              0.0438 
20 (OnE-Pr… CD4T     ABCA1  -0.00202           0.122 ZNF235              0.0425 
# ℹ 4 more variables: logFC &lt;dbl&gt;, p_val &lt;dbl&gt;, p_val_adj &lt;dbl&gt;,
#   direction_regulation &lt;fct&gt;</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb187"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb187-1"><a href="#cb187-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(prioritization_tables<span class="sc">$</span>ligand_activities_target_de_tbl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "contrast"             "receiver"             "ligand"              
 [4] "activity"             "activity_scaled"      "target"              
 [7] "ligand_target_weight" "logFC"                "p_val"               
[10] "p_val_adj"            "direction_regulation"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb189"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb189-1"><a href="#cb189-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(prioritization_tables<span class="sc">$</span>ligand_activities_target_de_tbl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 202802     11</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="step-6-correlation-between-ligand-receptor-pairs-and-their-predicted-target-genes-expression" class="level2">
<h2 class="anchored" data-anchor-id="step-6-correlation-between-ligand-receptor-pairs-and-their-predicted-target-genes-expression">Step 6: Correlation between ligand-receptor pairs and their predicted target genes expression</h2>
<p><code>lr_target_prior_cor_inference</code> function calculates the Pearson and Spearman expression correlation between LR pair pseudobulk expression products and DE gene expression product adding the NicheNet ligand-target regulatory potential score as well as prior knowledge support measures for the LR-to-target links.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb191"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb191-1"><a href="#cb191-1" aria-hidden="true" tabindex="-1"></a>lr_target_prior_cor <span class="ot">=</span> <span class="fu">lr_target_prior_cor_inference</span>(</span>
<span id="cb191-2"><a href="#cb191-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">receivers_oi =</span> prioritization_tables<span class="sc">$</span>group_prioritization_tbl<span class="sc">$</span>receiver <span class="sc">%&gt;%</span> </span>
<span id="cb191-3"><a href="#cb191-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unique</span>(),</span>
<span id="cb191-4"><a href="#cb191-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">abundance_expression_info =</span> abundance_expression_info,</span>
<span id="cb191-5"><a href="#cb191-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">celltype_de =</span> celltype_de,</span>
<span id="cb191-6"><a href="#cb191-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">grouping_tbl =</span> grouping_tbl,</span>
<span id="cb191-7"><a href="#cb191-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">prioritization_tables =</span> prioritization_tables,</span>
<span id="cb191-8"><a href="#cb191-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">ligand_target_matrix =</span> ligand_target_matrix,</span>
<span id="cb191-9"><a href="#cb191-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">logFC_threshold =</span> logFC_threshold,</span>
<span id="cb191-10"><a href="#cb191-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">p_val_threshold =</span> p_val_threshold,</span>
<span id="cb191-11"><a href="#cb191-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">p_val_adj =</span> p_val_adj</span>
<span id="cb191-12"><a href="#cb191-12" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb192"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb192-1"><a href="#cb192-1" aria-hidden="true" tabindex="-1"></a>lr_target_prior_cor <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 20 × 14
   sender   receiver ligand receptor id    target  pearson pearson_pval spearman
   &lt;chr&gt;    &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt;     &lt;dbl&gt;        &lt;dbl&gt;    &lt;dbl&gt;
 1 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… AADAT   0.357     0.00688      0.405 
 2 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… AARS2  -0.0147    0.914       -0.0276
 3 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… ABCD3   0.0290    0.832        0.0446
 4 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… ABHD1… -0.476     0.000208    -0.455 
 5 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… ABHD8  -0.173     0.201       -0.219 
 6 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… ACKR3  -0.0344    0.801       -0.0729
 7 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… ACTN1   0.557     0.00000816   0.528 
 8 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… ACTR8   0.185     0.173        0.125 
 9 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… ADAL    0.0675    0.621        0.0569
10 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… ADAM33 -0.148     0.275       -0.248 
11 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… ADAMT…  0.141     0.299       -0.0704
12 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… ADAMT… -0.0260    0.849       -0.0339
13 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… ADAT1  -0.0861    0.528       -0.0488
14 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… ADCY6  -0.163     0.231       -0.181 
15 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… ADCYA… -0.273     0.0421      -0.200 
16 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… AGAP1   0.257     0.0563       0.229 
17 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… AGAP6   0.0957    0.483        0.101 
18 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… AGL     0.00374   0.978       -0.103 
19 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… AGTPB…  0.247     0.0660       0.234 
20 Fibrobl… Fibrobl… COL1A1 ITGB1    COL1… AGTR1  -0.133     0.330       -0.196 
# ℹ 5 more variables: spearman_pval &lt;dbl&gt;, prior_score &lt;dbl&gt;,
#   rank_of_target &lt;int&gt;, rank_of_ligand &lt;int&gt;, id_target &lt;chr&gt;</code></pre>
</div>
</div>
</section>
<section id="step-7-save-multinichenet-cell-cell-communication-analysis" class="level2">
<h2 class="anchored" data-anchor-id="step-7-save-multinichenet-cell-cell-communication-analysis">Step 7: Save MultiNicheNet cell-cell communication analysis</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb194"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb194-1"><a href="#cb194-1" aria-hidden="true" tabindex="-1"></a>path <span class="ot">=</span> <span class="st">"./"</span></span>
<span id="cb194-2"><a href="#cb194-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb194-3"><a href="#cb194-3" aria-hidden="true" tabindex="-1"></a>multinichenet_output <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb194-4"><a href="#cb194-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">celltype_info =</span> abundance_expression_info<span class="sc">$</span>celltype_info,</span>
<span id="cb194-5"><a href="#cb194-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">celltype_de =</span> celltype_de,</span>
<span id="cb194-6"><a href="#cb194-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">sender_receiver_info =</span> abundance_expression_info<span class="sc">$</span>sender_receiver_info,</span>
<span id="cb194-7"><a href="#cb194-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">sender_receiver_de =</span>  sender_receiver_de,</span>
<span id="cb194-8"><a href="#cb194-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">ligand_activities_targets_DEgenes =</span> ligand_activities_targets_DEgenes,</span>
<span id="cb194-9"><a href="#cb194-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">prioritization_tables =</span> prioritization_tables,</span>
<span id="cb194-10"><a href="#cb194-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">grouping_tbl =</span> grouping_tbl,</span>
<span id="cb194-11"><a href="#cb194-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">lr_target_prior_cor =</span> lr_target_prior_cor</span>
<span id="cb194-12"><a href="#cb194-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb194-13"><a href="#cb194-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb194-14"><a href="#cb194-14" aria-hidden="true" tabindex="-1"></a>multinichenet_output <span class="ot">=</span> <span class="fu">make_lite_output</span>(multinichenet_output)</span>
<span id="cb194-15"><a href="#cb194-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb194-16"><a href="#cb194-16" aria-hidden="true" tabindex="-1"></a>save <span class="ot">=</span> <span class="cn">TRUE</span></span>
<span id="cb194-17"><a href="#cb194-17" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(save <span class="sc">==</span> <span class="cn">TRUE</span>){</span>
<span id="cb194-18"><a href="#cb194-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(multinichenet_output, <span class="fu">paste0</span>(path, <span class="st">"multinichenet_output.rds"</span>))</span>
<span id="cb194-19"><a href="#cb194-19" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb194-20"><a href="#cb194-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb194-21"><a href="#cb194-21" aria-hidden="true" tabindex="-1"></a><span class="co"># multinichenet_output &lt;- readRDS("D:/MultiNicheNet/multinichenet_output.rds")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="step-8-visualization-of-multinichenet-cell-cell-communication-analysis" class="level2">
<h2 class="anchored" data-anchor-id="step-8-visualization-of-multinichenet-cell-cell-communication-analysis">Step 8: Visualization of MultiNicheNet cell-cell communication analysis</h2>
<section id="circos-plots-of-condition-specific-top-prioritized-lr-interactions" class="level3">
<h3 class="anchored" data-anchor-id="circos-plots-of-condition-specific-top-prioritized-lr-interactions">8.1 Circos plots of condition-specific top prioritized LR interactions</h3>
<p>Top 50 predictions across all contrasts, senders and receivers of interest.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb195"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb195-1"><a href="#cb195-1" aria-hidden="true" tabindex="-1"></a>prioritized_tbl_oi_all <span class="ot">=</span> <span class="fu">get_top_n_lr_pairs</span>(</span>
<span id="cb195-2"><a href="#cb195-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">prioritization_tables =</span> multinichenet_output<span class="sc">$</span>prioritization_tables,</span>
<span id="cb195-3"><a href="#cb195-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">top_n =</span> <span class="dv">50</span>,</span>
<span id="cb195-4"><a href="#cb195-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">rank_per_group =</span> <span class="cn">FALSE</span> <span class="co"># top n given over all groups</span></span>
<span id="cb195-5"><a href="#cb195-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb196"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb196-1"><a href="#cb196-1" aria-hidden="true" tabindex="-1"></a>prioritized_tbl_oi_all <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  group sender      receiver    ligand receptor id          prioritization_score
  &lt;chr&gt; &lt;chr&gt;       &lt;chr&gt;       &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;                      &lt;dbl&gt;
1 OnE   macrophages Fibroblast  TNF    LTBR     TNF_LTBR_m…                0.944
2 OnE   Fibroblast  CD4T        TGM2   ITGA4    TGM2_ITGA4…                0.940
3 OnE   macrophages macrophages TNF    LTBR     TNF_LTBR_m…                0.939
4 OnE   macrophages CD4T        TNF    TRAF2    TNF_TRAF2_…                0.937
5 OnE   macrophages CD4T        CXCL9  CXCR3    CXCL9_CXCR…                0.933
6 OnE   macrophages Fibroblast  TNF    TNFRSF1A TNF_TNFRSF…                0.923
# ℹ 1 more variable: prioritization_rank &lt;dbl&gt;</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb198"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb198-1"><a href="#cb198-1" aria-hidden="true" tabindex="-1"></a>prioritized_tbl_oi <span class="ot">=</span> multinichenet_output<span class="sc">$</span>prioritization_tables<span class="sc">$</span>group_prioritization_tbl <span class="sc">%&gt;%</span></span>
<span id="cb198-2"><a href="#cb198-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(id <span class="sc">%in%</span> prioritized_tbl_oi_all<span class="sc">$</span>id) <span class="sc">%&gt;%</span></span>
<span id="cb198-3"><a href="#cb198-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(id, sender, receiver, ligand, receptor, group) <span class="sc">%&gt;%</span></span>
<span id="cb198-4"><a href="#cb198-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(prioritized_tbl_oi_all)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(id, sender, receiver, ligand, receptor, group)`</code></pre>
</div>
<div class="sourceCode cell-code" id="cb200"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb200-1"><a href="#cb200-1" aria-hidden="true" tabindex="-1"></a>prioritized_tbl_oi<span class="sc">$</span>prioritization_score[</span>
<span id="cb200-2"><a href="#cb200-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">is.na</span>(prioritized_tbl_oi<span class="sc">$</span>prioritization_score)] <span class="ot">=</span> <span class="dv">0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb201"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb201-1"><a href="#cb201-1" aria-hidden="true" tabindex="-1"></a>prioritized_tbl_oi <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  id                  sender receiver ligand receptor group prioritization_score
  &lt;chr&gt;               &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;                &lt;dbl&gt;
1 TNF_LTBR_macrophag… macro… Fibrobl… TNF    LTBR     OnE                  0.944
2 TGM2_ITGA4_Fibrobl… Fibro… CD4T     TGM2   ITGA4    OnE                  0.940
3 TNF_LTBR_macrophag… macro… macroph… TNF    LTBR     OnE                  0.939
4 TNF_TRAF2_macropha… macro… CD4T     TNF    TRAF2    OnE                  0.937
5 CXCL9_CXCR3_macrop… macro… CD4T     CXCL9  CXCR3    OnE                  0.933
6 TNF_TNFRSF1A_macro… macro… Fibrobl… TNF    TNFRSF1A OnE                  0.923
# ℹ 1 more variable: prioritization_rank &lt;dbl&gt;</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb203"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb203-1"><a href="#cb203-1" aria-hidden="true" tabindex="-1"></a>senders_receivers <span class="ot">=</span> <span class="fu">union</span>(prioritized_tbl_oi<span class="sc">$</span>sender <span class="sc">%&gt;%</span> <span class="fu">unique</span>(),</span>
<span id="cb203-2"><a href="#cb203-2" aria-hidden="true" tabindex="-1"></a>                          prioritized_tbl_oi<span class="sc">$</span>receiver <span class="sc">%&gt;%</span> <span class="fu">unique</span>()) <span class="sc">%&gt;%</span> <span class="fu">sort</span>()</span>
<span id="cb203-3"><a href="#cb203-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb203-4"><a href="#cb203-4" aria-hidden="true" tabindex="-1"></a>colors_sender <span class="ot">=</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(</span>
<span id="cb203-5"><a href="#cb203-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">n =</span> <span class="fu">length</span>(senders_receivers), <span class="at">name =</span> <span class="st">'Spectral'</span></span>
<span id="cb203-6"><a href="#cb203-6" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span> magrittr<span class="sc">::</span><span class="fu">set_names</span>(senders_receivers)</span>
<span id="cb203-7"><a href="#cb203-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb203-8"><a href="#cb203-8" aria-hidden="true" tabindex="-1"></a>colors_receiver <span class="ot">=</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(</span>
<span id="cb203-9"><a href="#cb203-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">n =</span> <span class="fu">length</span>(senders_receivers), <span class="at">name =</span> <span class="st">'Spectral'</span></span>
<span id="cb203-10"><a href="#cb203-10" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span> magrittr<span class="sc">::</span><span class="fu">set_names</span>(senders_receivers)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb204"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb204-1"><a href="#cb204-1" aria-hidden="true" tabindex="-1"></a>senders_receivers </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "CD4T"        "Fibroblast"  "macrophages"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb206"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb206-1"><a href="#cb206-1" aria-hidden="true" tabindex="-1"></a>colors_sender</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       CD4T  Fibroblast macrophages 
  "#FC8D59"   "#FFFFBF"   "#99D594" </code></pre>
</div>
</div>
<p><code>make_circos_group_comparison</code> makes a circos plot with top prioritized ligand-receptor interaction for each group of interest. In each circos, all possible LR pairs are shown, but arrows are only drawn between the ones belonging to the group of interest.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb208"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb208-1"><a href="#cb208-1" aria-hidden="true" tabindex="-1"></a>circos_list <span class="ot">=</span> <span class="fu">make_circos_group_comparison</span>(</span>
<span id="cb208-2"><a href="#cb208-2" aria-hidden="true" tabindex="-1"></a>  prioritized_tbl_oi, colors_sender, colors_receiver</span>
<span id="cb208-3"><a href="#cb208-3" aria-hidden="true" tabindex="-1"></a>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(sender)`
Joining with `by = join_by(receiver)`
Joining with `by = join_by(sender)`
Joining with `by = join_by(receiver)`</code></pre>
</div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-92-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-92-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-92-3.png" class="img-fluid" width="672"></p>
</div>
</div>
<p><code>make_circos_one_group</code> makes full circos plots for one group of interest for top n LR pairs per group.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb210"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb210-1"><a href="#cb210-1" aria-hidden="true" tabindex="-1"></a>prioritized_tbl_oi_E_30 <span class="ot">=</span> <span class="fu">get_top_n_lr_pairs</span>( multinichenet_output<span class="sc">$</span>prioritization_tables, <span class="dv">30</span>, <span class="at">groups_oi =</span> <span class="st">"OnE"</span>)</span>
<span id="cb210-2"><a href="#cb210-2" aria-hidden="true" tabindex="-1"></a>prioritized_tbl_oi_NE_30 <span class="ot">=</span> <span class="fu">get_top_n_lr_pairs</span>( multinichenet_output<span class="sc">$</span>prioritization_tables, <span class="dv">30</span>, <span class="at">groups_oi =</span> <span class="st">"OnNE"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb211"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb211-1"><a href="#cb211-1" aria-hidden="true" tabindex="-1"></a>circos_E <span class="ot">=</span> <span class="fu">make_circos_one_group</span>(</span>
<span id="cb211-2"><a href="#cb211-2" aria-hidden="true" tabindex="-1"></a>  prioritized_tbl_oi_E_30, colors_sender, colors_receiver)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(sender)`
Joining with `by = join_by(receiver)`</code></pre>
</div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-94-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-94-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb213"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb213-1"><a href="#cb213-1" aria-hidden="true" tabindex="-1"></a>circos_NE <span class="ot">=</span> <span class="fu">make_circos_one_group</span>(</span>
<span id="cb213-2"><a href="#cb213-2" aria-hidden="true" tabindex="-1"></a>  prioritized_tbl_oi_NE_30, colors_sender, colors_receiver)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(sender)`
Joining with `by = join_by(receiver)`</code></pre>
</div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-95-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-95-2.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="dotplotheatmap-visualization-of-scaled-ligand-receptor-pseudobulk-expression-product-per-sample-and-ligand-activity" class="level3">
<h3 class="anchored" data-anchor-id="dotplotheatmap-visualization-of-scaled-ligand-receptor-pseudobulk-expression-product-per-sample-and-ligand-activity">8.2 Dotplot/Heatmap visualization of scaled ligand-receptor pseudobulk expression product per sample and ligand activity</h3>
<p><code>make_sample_lr_prod_activity_plots</code> used to visualize the scaled product of Ligand-Receptor (pseudobulk) expression per sample, and compare the different groups. In addition, shows the NicheNet ligand activities in each receiver-celltype combination.</p>
<p>Dotplot/heatmap of top 30 interactions of which the On-vs-Pre change is stronger in group E than the NE group</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb215"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb215-1"><a href="#cb215-1" aria-hidden="true" tabindex="-1"></a>plot_oi <span class="ot">=</span> <span class="fu">make_sample_lr_prod_activity_plots</span>(</span>
<span id="cb215-2"><a href="#cb215-2" aria-hidden="true" tabindex="-1"></a>  multinichenet_output<span class="sc">$</span>prioritization_tables,</span>
<span id="cb215-3"><a href="#cb215-3" aria-hidden="true" tabindex="-1"></a>  prioritized_tbl_oi_E_30)</span>
<span id="cb215-4"><a href="#cb215-4" aria-hidden="true" tabindex="-1"></a>plot_oi </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-96-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Dotplot/heatmap of top 30 interactions of which the On-vs-Pre change is stronger in group NE than the E group</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb216"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb216-1"><a href="#cb216-1" aria-hidden="true" tabindex="-1"></a>plot_oi <span class="ot">=</span> <span class="fu">make_sample_lr_prod_activity_plots</span>(</span>
<span id="cb216-2"><a href="#cb216-2" aria-hidden="true" tabindex="-1"></a>  multinichenet_output<span class="sc">$</span>prioritization_tables,</span>
<span id="cb216-3"><a href="#cb216-3" aria-hidden="true" tabindex="-1"></a>  prioritized_tbl_oi_NE_30)</span>
<span id="cb216-4"><a href="#cb216-4" aria-hidden="true" tabindex="-1"></a>plot_oi</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-97-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Typically, there are way more than 30 differentially expressed and active ligand-receptor pairs per group across all sender-receiver combinations. Therefore it might be useful to zoom in on specific cell types as senders/receivers. Eg macrophages as receiver</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb217"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb217-1"><a href="#cb217-1" aria-hidden="true" tabindex="-1"></a>group_oi <span class="ot">=</span> <span class="st">"OnE"</span></span>
<span id="cb217-2"><a href="#cb217-2" aria-hidden="true" tabindex="-1"></a>prioritized_tbl_oi_top_50 <span class="ot">=</span> <span class="fu">get_top_n_lr_pairs</span>(</span>
<span id="cb217-3"><a href="#cb217-3" aria-hidden="true" tabindex="-1"></a>  multinichenet_output<span class="sc">$</span>prioritization_tables, <span class="dv">50</span>, </span>
<span id="cb217-4"><a href="#cb217-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">groups_oi =</span> group_oi, <span class="at">receivers_oi =</span> <span class="st">"macrophages"</span>)</span>
<span id="cb217-5"><a href="#cb217-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb217-6"><a href="#cb217-6" aria-hidden="true" tabindex="-1"></a>plot_oi <span class="ot">=</span> <span class="fu">make_sample_lr_prod_activity_plots</span>(</span>
<span id="cb217-7"><a href="#cb217-7" aria-hidden="true" tabindex="-1"></a>  multinichenet_output<span class="sc">$</span>prioritization_tables,</span>
<span id="cb217-8"><a href="#cb217-8" aria-hidden="true" tabindex="-1"></a>  prioritized_tbl_oi_top_50)</span>
<span id="cb217-9"><a href="#cb217-9" aria-hidden="true" tabindex="-1"></a>plot_oi</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-98-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="intercellular-regulatory-network-systems-view" class="level3">
<h3 class="anchored" data-anchor-id="intercellular-regulatory-network-systems-view">8.3 Intercellular regulatory network systems view</h3>
<p>Generate a ‘systems’ view of the intercellular feedback and cascade processes than can be occuring between the different cell populations involved. In this plot, links are drawn between ligands of sender cell types and their ligand/receptor-annotated target genes in receiver cell types. So links are ligand-target links (= gene regulatory links) and not ligand-receptor protein-protein interactions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb218"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb218-1"><a href="#cb218-1" aria-hidden="true" tabindex="-1"></a>prioritized_tbl_oi <span class="ot">=</span> <span class="fu">get_top_n_lr_pairs</span>(</span>
<span id="cb218-2"><a href="#cb218-2" aria-hidden="true" tabindex="-1"></a>  multinichenet_output<span class="sc">$</span>prioritization_tables, <span class="dv">100</span>, <span class="at">rank_per_group =</span> <span class="cn">FALSE</span>)</span>
<span id="cb218-3"><a href="#cb218-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb218-4"><a href="#cb218-4" aria-hidden="true" tabindex="-1"></a>lr_target_prior_cor_filtered <span class="ot">=</span> multinichenet_output<span class="sc">$</span>prioritization_tables<span class="sc">$</span>group_prioritization_tbl<span class="sc">$</span>group <span class="sc">%&gt;%</span></span>
<span id="cb218-5"><a href="#cb218-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unique</span>() <span class="sc">%&gt;%</span></span>
<span id="cb218-6"><a href="#cb218-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(<span class="cf">function</span>(group_oi){</span>
<span id="cb218-7"><a href="#cb218-7" aria-hidden="true" tabindex="-1"></a>    lr_target_prior_cor_filtered <span class="ot">=</span> multinichenet_output<span class="sc">$</span>lr_target_prior_cor <span class="sc">%&gt;%</span></span>
<span id="cb218-8"><a href="#cb218-8" aria-hidden="true" tabindex="-1"></a>      <span class="fu">inner_join</span>(multinichenet_output<span class="sc">$</span>ligand_activities_targets_DEgenes<span class="sc">$</span>ligand_activities <span class="sc">%&gt;%</span> </span>
<span id="cb218-9"><a href="#cb218-9" aria-hidden="true" tabindex="-1"></a>      <span class="fu">distinct</span>(ligand, target, direction_regulation, contrast)) <span class="sc">%&gt;%</span></span>
<span id="cb218-10"><a href="#cb218-10" aria-hidden="true" tabindex="-1"></a>      <span class="fu">inner_join</span>(contrast_tbl) <span class="sc">%&gt;%</span> </span>
<span id="cb218-11"><a href="#cb218-11" aria-hidden="true" tabindex="-1"></a>      <span class="fu">filter</span>(group <span class="sc">==</span> group_oi)</span>
<span id="cb218-12"><a href="#cb218-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb218-13"><a href="#cb218-13" aria-hidden="true" tabindex="-1"></a>    lr_target_prior_cor_filtered_up <span class="ot">=</span> lr_target_prior_cor_filtered <span class="sc">%&gt;%</span> </span>
<span id="cb218-14"><a href="#cb218-14" aria-hidden="true" tabindex="-1"></a>      <span class="fu">filter</span>(direction_regulation <span class="sc">==</span> <span class="st">"up"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb218-15"><a href="#cb218-15" aria-hidden="true" tabindex="-1"></a>      <span class="fu">filter</span>( (rank_of_target <span class="sc">&lt;</span> top_n_target) <span class="sc">&amp;</span> (pearson <span class="sc">&gt;</span> <span class="fl">0.50</span> <span class="sc">|</span> spearman <span class="sc">&gt;</span> <span class="fl">0.50</span>))</span>
<span id="cb218-16"><a href="#cb218-16" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb218-17"><a href="#cb218-17" aria-hidden="true" tabindex="-1"></a>    lr_target_prior_cor_filtered_down <span class="ot">=</span> lr_target_prior_cor_filtered <span class="sc">%&gt;%</span> </span>
<span id="cb218-18"><a href="#cb218-18" aria-hidden="true" tabindex="-1"></a>      <span class="fu">filter</span>(direction_regulation <span class="sc">==</span> <span class="st">"down"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb218-19"><a href="#cb218-19" aria-hidden="true" tabindex="-1"></a>      <span class="fu">filter</span>( (rank_of_target <span class="sc">&lt;</span> top_n_target) <span class="sc">&amp;</span> (pearson <span class="sc">&lt;</span> <span class="sc">-</span><span class="fl">0.50</span> <span class="sc">|</span> spearman <span class="sc">&lt;</span> <span class="sc">-</span><span class="fl">0.50</span>))</span>
<span id="cb218-20"><a href="#cb218-20" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb218-21"><a href="#cb218-21" aria-hidden="true" tabindex="-1"></a>     lr_target_prior_cor_filtered <span class="ot">=</span> <span class="fu">bind_rows</span>(</span>
<span id="cb218-22"><a href="#cb218-22" aria-hidden="true" tabindex="-1"></a>       lr_target_prior_cor_filtered_up, </span>
<span id="cb218-23"><a href="#cb218-23" aria-hidden="true" tabindex="-1"></a>       lr_target_prior_cor_filtered_down)</span>
<span id="cb218-24"><a href="#cb218-24" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">%&gt;%</span> <span class="fu">bind_rows</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(receiver, ligand, target)`</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in inner_join(., multinichenet_output$ligand_activities_targets_DEgenes$ligand_activities %&gt;% : Detected an unexpected many-to-many relationship between `x` and `y`.
ℹ Row 47 of `x` matches multiple rows in `y`.
ℹ Row 54644 of `y` matches multiple rows in `x`.
ℹ If a many-to-many relationship is expected, set `relationship =
  "many-to-many"` to silence this warning.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(contrast)`
Joining with `by = join_by(receiver, ligand, target)`</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in inner_join(., multinichenet_output$ligand_activities_targets_DEgenes$ligand_activities %&gt;% : Detected an unexpected many-to-many relationship between `x` and `y`.
ℹ Row 47 of `x` matches multiple rows in `y`.
ℹ Row 54644 of `y` matches multiple rows in `x`.
ℹ If a many-to-many relationship is expected, set `relationship =
  "many-to-many"` to silence this warning.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(contrast)`</code></pre>
</div>
</div>
<p><code>make_ggraph_ligand_target_links</code> makes a network showing the gene regulatory links between ligands from sender cell types to their induced ligands/receptors in receiver cell types. Links are only drawn if the ligand/receptor in the receiver is a potential downstream target of the ligand based on prior knowledge and sufficient correlation in expression acrosss the different samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb224"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb224-1"><a href="#cb224-1" aria-hidden="true" tabindex="-1"></a>colors_sender[<span class="st">"Fibroblast"</span>] <span class="ot">=</span> <span class="st">"pink"</span> <span class="co"># the  original yellow with white font is not very readable</span></span>
<span id="cb224-2"><a href="#cb224-2" aria-hidden="true" tabindex="-1"></a>graph_plot <span class="ot">=</span> <span class="fu">make_ggraph_ligand_target_links</span>(</span>
<span id="cb224-3"><a href="#cb224-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">lr_target_prior_cor_filtered =</span> lr_target_prior_cor_filtered,</span>
<span id="cb224-4"><a href="#cb224-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">prioritized_tbl_oi =</span> prioritized_tbl_oi, </span>
<span id="cb224-5"><a href="#cb224-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">colors =</span> colors_sender)</span>
<span id="cb224-6"><a href="#cb224-6" aria-hidden="true" tabindex="-1"></a>graph_plot<span class="sc">$</span>plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)): font family
not found in Windows font database</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, : font
family not found in Windows font database</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)): font family
not found in Windows font database

Warning in grid.Call(C_stringMetric, as.graphicsAnnot(x$label)): font family
not found in Windows font database</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, : font
family not found in Windows font database

Warning in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, : font
family not found in Windows font database</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in grid.Call.graphics(C_text, as.graphicsAnnot(x$label), x$x, x$y, :
font family not found in Windows font database</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, : font
family not found in Windows font database

Warning in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, : font
family not found in Windows font database

Warning in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, : font
family not found in Windows font database

Warning in grid.Call(C_textBounds, as.graphicsAnnot(x$label), x$x, x$y, : font
family not found in Windows font database</code></pre>
</div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-100-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>LR-target links in the systems plot</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb231"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb231-1"><a href="#cb231-1" aria-hidden="true" tabindex="-1"></a>graph_plot<span class="sc">$</span>source_df_lt </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 31 × 6
   sender              receiver          direction_regulation group type  weight
   &lt;chr&gt;               &lt;chr&gt;             &lt;fct&gt;                &lt;chr&gt; &lt;chr&gt;  &lt;dbl&gt;
 1 Fibroblast_TIMP1    Fibroblast_TIMP1  up                   OnE   Liga…      1
 2 Fibroblast_FN1      Fibroblast_SERPI… up                   OnE   Liga…      1
 3 Fibroblast_FN1      Fibroblast_VEGFA  up                   OnE   Liga…      1
 4 Fibroblast_TIMP3    Fibroblast_VEGFA  up                   OnE   Liga…      1
 5 Fibroblast_SERPINE1 Fibroblast_SERPI… up                   OnE   Liga…      1
 6 Fibroblast_SERPINE1 Fibroblast_VEGFA  up                   OnE   Liga…      1
 7 Fibroblast_CCN1     Fibroblast_SERPI… up                   OnE   Liga…      1
 8 Fibroblast_CCN1     Fibroblast_VEGFA  up                   OnE   Liga…      1
 9 macrophages_ITGB2   Fibroblast_TIMP1  up                   OnE   Liga…      1
10 Fibroblast_ANGPTL4  Fibroblast_SERPI… up                   OnE   Liga…      1
# ℹ 21 more rows</code></pre>
</div>
</div>
<p>The arrows in the plot</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb233"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb233-1"><a href="#cb233-1" aria-hidden="true" tabindex="-1"></a>graph_plot<span class="sc">$</span>source_df_lt <span class="sc">%&gt;%</span> <span class="fu">filter</span>(<span class="fu">row_number</span>() <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">20</span>,<span class="dv">25</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 6
  sender           receiver   direction_regulation group type          weight
  &lt;chr&gt;            &lt;chr&gt;      &lt;fct&gt;                &lt;chr&gt; &lt;chr&gt;          &lt;dbl&gt;
1 Fibroblast_ICAM1 CD4T_IL2RA up                   OnE   Ligand-Target      1
2 CD4T_PTPRC       CD4T_IL2RA down                 OnNE  Ligand-Target      1</code></pre>
</div>
</div>
<p>Nodes in the systems plot</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb235"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb235-1"><a href="#cb235-1" aria-hidden="true" tabindex="-1"></a>graph_plot<span class="sc">$</span>nodes_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                   node    celltype     gene       type_gene
macrophages_ITGB2     macrophages_ITGB2 macrophages    ITGB2 ligand/receptor
Fibroblast_ICAM1       Fibroblast_ICAM1  Fibroblast    ICAM1 ligand/receptor
Fibroblast_VCAM1       Fibroblast_VCAM1  Fibroblast    VCAM1 ligand/receptor
Fibroblast_TIMP1       Fibroblast_TIMP1  Fibroblast    TIMP1          ligand
Fibroblast_FN1           Fibroblast_FN1  Fibroblast      FN1          ligand
Fibroblast_TIMP3       Fibroblast_TIMP3  Fibroblast    TIMP3          ligand
Fibroblast_SERPINE1 Fibroblast_SERPINE1  Fibroblast SERPINE1          ligand
Fibroblast_CCN1         Fibroblast_CCN1  Fibroblast     CCN1          ligand
Fibroblast_ANGPTL4   Fibroblast_ANGPTL4  Fibroblast  ANGPTL4          ligand
Fibroblast_TGFBI       Fibroblast_TGFBI  Fibroblast    TGFBI          ligand
macrophages_S100A8   macrophages_S100A8 macrophages   S100A8          ligand
macrophages_HLA.A     macrophages_HLA.A macrophages    HLA.A          ligand
macrophages_CD72       macrophages_CD72 macrophages     CD72          ligand
CD4T_PTPRC                   CD4T_PTPRC        CD4T    PTPRC          ligand
macrophages_ICAM1     macrophages_ICAM1 macrophages    ICAM1          ligand
macrophages_IL15       macrophages_IL15 macrophages     IL15          ligand
macrophages_LGALS9   macrophages_LGALS9 macrophages   LGALS9          ligand
Fibroblast_CCL2         Fibroblast_CCL2  Fibroblast     CCL2          ligand
CD4T_ITGA4                   CD4T_ITGA4        CD4T    ITGA4          ligand
macrophages_TNF         macrophages_TNF macrophages      TNF          ligand
Fibroblast_ITGA5       Fibroblast_ITGA5  Fibroblast    ITGA5        receptor
CD4T_IL2RA                   CD4T_IL2RA        CD4T    IL2RA        receptor
macrophages_CCR2       macrophages_CCR2 macrophages     CCR2        receptor
Fibroblast_VEGFA       Fibroblast_VEGFA  Fibroblast    VEGFA          ligand
CD4T_IL2                       CD4T_IL2        CD4T      IL2          ligand
macrophages_CSF3R     macrophages_CSF3R macrophages    CSF3R        receptor
macrophages_SLAMF7   macrophages_SLAMF7 macrophages   SLAMF7 ligand/receptor</code></pre>
</div>
</div>
<p>The graph object</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb237"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb237-1"><a href="#cb237-1" aria-hidden="true" tabindex="-1"></a>graph_plot<span class="sc">$</span>graph</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tbl_graph: 27 nodes and 31 edges
#
# A directed multigraph with 4 components
#
# A tibble: 27 × 4
  name              celltype    gene  type_gene      
  &lt;chr&gt;             &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;          
1 macrophages_ITGB2 macrophages ITGB2 ligand/receptor
2 Fibroblast_ICAM1  Fibroblast  ICAM1 ligand/receptor
3 Fibroblast_VCAM1  Fibroblast  VCAM1 ligand/receptor
4 Fibroblast_TIMP1  Fibroblast  TIMP1 ligand         
5 Fibroblast_FN1    Fibroblast  FN1   ligand         
6 Fibroblast_TIMP3  Fibroblast  TIMP3 ligand         
# ℹ 21 more rows
#
# A tibble: 31 × 6
   from    to direction_regulation group type          weight
  &lt;int&gt; &lt;int&gt; &lt;chr&gt;                &lt;chr&gt; &lt;chr&gt;          &lt;dbl&gt;
1     4     4 up                   OnE   Ligand-Target      1
2     5     7 up                   OnE   Ligand-Target      1
3     5    24 up                   OnE   Ligand-Target      1
# ℹ 28 more rows</code></pre>
</div>
</div>
<p>The source 100 prioritized LR Links from which the systems view was generated</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb239"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb239-1"><a href="#cb239-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(graph_plot<span class="sc">$</span>source_df_lr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 100   7</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb241"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb241-1"><a href="#cb241-1" aria-hidden="true" tabindex="-1"></a>graph_plot<span class="sc">$</span>source_df_lr <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 20 × 7
   group sender      receiver  celltype_ligand celltype_receptor ligand receptor
   &lt;chr&gt; &lt;chr&gt;       &lt;chr&gt;     &lt;chr&gt;           &lt;chr&gt;             &lt;chr&gt;  &lt;chr&gt;   
 1 OnE   macrophages Fibrobla… macrophages_TNF Fibroblast_LTBR   TNF    LTBR    
 2 OnE   Fibroblast  CD4T      Fibroblast_TGM2 CD4T_ITGA4        TGM2   ITGA4   
 3 OnE   macrophages macropha… macrophages_TNF macrophages_LTBR  TNF    LTBR    
 4 OnE   macrophages CD4T      macrophages_TNF CD4T_TRAF2        TNF    TRAF2   
 5 OnE   macrophages CD4T      macrophages_CX… CD4T_CXCR3        CXCL9  CXCR3   
 6 OnE   macrophages Fibrobla… macrophages_TNF Fibroblast_TNFRS… TNF    TNFRSF1A
 7 OnE   Fibroblast  CD4T      Fibroblast_CXC… CD4T_CXCR3        CXCL9  CXCR3   
 8 OnNE  CD4T        CD4T      CD4T_PTPRC      CD4T_DPP4         PTPRC  DPP4    
 9 OnE   Fibroblast  macropha… Fibroblast_CSF1 macrophages_CSF3R CSF1   CSF3R   
10 OnE   macrophages CD4T      macrophages_TNF CD4T_TNFRSF1B     TNF    TNFRSF1B
11 OnE   macrophages macropha… macrophages_IL… macrophages_IL15… IL15   IL15RA  
12 OnE   macrophages CD4T      macrophages_CD… CD4T_CD28         CD80   CD28    
13 OnNE  CD4T        Fibrobla… CD4T_CMTM8      Fibroblast_EGFR   CMTM8  EGFR    
14 OnNE  CD4T        macropha… CD4T_TNF        macrophages_TNFR… TNF    TNFRSF21
15 OnE   macrophages Fibrobla… macrophages_TNF Fibroblast_TNFRS… TNF    TNFRSF21
16 OnE   Fibroblast  CD4T      Fibroblast_CCL… CD4T_CXCR3        CCL19  CXCR3   
17 OnE   macrophages macropha… macrophages_HL… macrophages_CD74  HLA.D… CD74    
18 OnE   Fibroblast  Fibrobla… Fibroblast_SER… Fibroblast_PLAU   SERPI… PLAU    
19 OnE   macrophages CD4T      macrophages_PT… CD4T_CD247        PTPRC  CD247   
20 OnE   Fibroblast  Fibrobla… Fibroblast_ANG… Fibroblast_ITGA5  ANGPT… ITGA5   </code></pre>
</div>
</div>
</section>
<section id="heatmap-of-ligand-activities-for-each-receiver-group-combination" class="level3">
<h3 class="anchored" data-anchor-id="heatmap-of-ligand-activities-for-each-receiver-group-combination">8.4 Heatmap of ligand activities for each receiver-group combination</h3>
<p>Plot all the ligand activities (both scaled and absolute activities) of each receiver-group combination. This gives some insights in active signaling pathways across groups. Thus, show top ligands based on ligand activity - agnostic of expression in sender, however.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb243"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb243-1"><a href="#cb243-1" aria-hidden="true" tabindex="-1"></a>ligands_oi <span class="ot">=</span> multinichenet_output<span class="sc">$</span>prioritization_tables<span class="sc">$</span>ligand_activities_target_de_tbl <span class="sc">%&gt;%</span> </span>
<span id="cb243-2"><a href="#cb243-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">inner_join</span>(contrast_tbl) <span class="sc">%&gt;%</span> </span>
<span id="cb243-3"><a href="#cb243-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(group, receiver) <span class="sc">%&gt;%</span> </span>
<span id="cb243-4"><a href="#cb243-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">distinct</span>(ligand, receiver, group, activity) <span class="sc">%&gt;%</span> </span>
<span id="cb243-5"><a href="#cb243-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">top_n</span>(<span class="dv">5</span>, activity) <span class="sc">%&gt;%</span></span>
<span id="cb243-6"><a href="#cb243-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(ligand) <span class="sc">%&gt;%</span></span>
<span id="cb243-7"><a href="#cb243-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unique</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(contrast)`</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb245"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb245-1"><a href="#cb245-1" aria-hidden="true" tabindex="-1"></a>ligands_oi </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "C5"    "FAM3B" "NMU"   "RLN1"  "ST14"  "IFNA1" "IFNB1" "IFNG"  "IFNL1"
[10] "IL27"  "IL15"  "IL4"   "TNF"  </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb247"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb247-1"><a href="#cb247-1" aria-hidden="true" tabindex="-1"></a>plot_oi <span class="ot">=</span> <span class="fu">make_ligand_activity_plots</span>(</span>
<span id="cb247-2"><a href="#cb247-2" aria-hidden="true" tabindex="-1"></a>  multinichenet_output<span class="sc">$</span>prioritization_tables, </span>
<span id="cb247-3"><a href="#cb247-3" aria-hidden="true" tabindex="-1"></a>  ligands_oi, </span>
<span id="cb247-4"><a href="#cb247-4" aria-hidden="true" tabindex="-1"></a>  contrast_tbl, </span>
<span id="cb247-5"><a href="#cb247-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">widths =</span> <span class="cn">NULL</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(contrast)`</code></pre>
</div>
<div class="sourceCode cell-code" id="cb249"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb249-1"><a href="#cb249-1" aria-hidden="true" tabindex="-1"></a>plot_oi</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-109-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The plot shows strong interferon signature in all 3 cell types with stronger On-vs-Pre in the E vs NE group.</p>
</section>
<section id="violin-plots-for-specific-ligand-receptor-interactions-in-single-cell-expression-data" class="level3">
<h3 class="anchored" data-anchor-id="violin-plots-for-specific-ligand-receptor-interactions-in-single-cell-expression-data">8.5 Violin plots for specific ligand-receptor interactions in single-cell expression data</h3>
<p>Single-cell-based Violin plots of ligand-receptor interaction of interest: <code>make_ligand_receptor_violin_plot -</code>useful to zoom in on specific ligand-receptor interactions of interest by looking in more detail to their expression at the single cell level.</p>
<p>Recall prioritized_tbl_oi with the highest scoring links based in the general prioritization score.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb250"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb250-1"><a href="#cb250-1" aria-hidden="true" tabindex="-1"></a>prioritized_tbl_oi <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 8
   group sender      receiver    ligand receptor id         prioritization_score
   &lt;chr&gt; &lt;chr&gt;       &lt;chr&gt;       &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;                     &lt;dbl&gt;
 1 OnE   macrophages Fibroblast  TNF    LTBR     TNF_LTBR_…                0.944
 2 OnE   Fibroblast  CD4T        TGM2   ITGA4    TGM2_ITGA…                0.940
 3 OnE   macrophages macrophages TNF    LTBR     TNF_LTBR_…                0.939
 4 OnE   macrophages CD4T        TNF    TRAF2    TNF_TRAF2…                0.937
 5 OnE   macrophages CD4T        CXCL9  CXCR3    CXCL9_CXC…                0.933
 6 OnE   macrophages Fibroblast  TNF    TNFRSF1A TNF_TNFRS…                0.923
 7 OnE   Fibroblast  CD4T        CXCL9  CXCR3    CXCL9_CXC…                0.921
 8 OnNE  CD4T        CD4T        PTPRC  DPP4     PTPRC_DPP…                0.921
 9 OnE   Fibroblast  macrophages CSF1   CSF3R    CSF1_CSF3…                0.920
10 OnE   macrophages CD4T        TNF    TNFRSF1B TNF_TNFRS…                0.920
# ℹ 1 more variable: prioritization_rank &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Choose one of those to visualize.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb252"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb252-1"><a href="#cb252-1" aria-hidden="true" tabindex="-1"></a>ligand_oi <span class="ot">=</span> <span class="st">"CXCL9"</span></span>
<span id="cb252-2"><a href="#cb252-2" aria-hidden="true" tabindex="-1"></a>receptor_oi <span class="ot">=</span> <span class="st">"CXCR3"</span></span>
<span id="cb252-3"><a href="#cb252-3" aria-hidden="true" tabindex="-1"></a>group_oi <span class="ot">=</span> <span class="st">"OnE"</span></span>
<span id="cb252-4"><a href="#cb252-4" aria-hidden="true" tabindex="-1"></a>sender_oi <span class="ot">=</span> <span class="st">"macrophages"</span></span>
<span id="cb252-5"><a href="#cb252-5" aria-hidden="true" tabindex="-1"></a>receiver_oi <span class="ot">=</span> <span class="st">"CD4T"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb253"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb253-1"><a href="#cb253-1" aria-hidden="true" tabindex="-1"></a>p_violin <span class="ot">=</span> <span class="fu">make_ligand_receptor_violin_plot</span>(<span class="at">sce_sender =</span> sce, <span class="at">sce_receiver =</span> sce,</span>
<span id="cb253-2"><a href="#cb253-2" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">ligand_oi=</span>ligand_oi, <span class="at">receptor_oi =</span> receptor_oi,</span>
<span id="cb253-3"><a href="#cb253-3" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">group_oi =</span> group_oi, <span class="at">group_id =</span> group_id,</span>
<span id="cb253-4"><a href="#cb253-4" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">sender_oi =</span> sender_oi, <span class="at">receiver_oi =</span> receiver_oi,</span>
<span id="cb253-5"><a href="#cb253-5" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">sample_id =</span> sample_id,</span>
<span id="cb253-6"><a href="#cb253-6" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">celltype_id_sender =</span> celltype_id,</span>
<span id="cb253-7"><a href="#cb253-7" aria-hidden="true" tabindex="-1"></a>                                            <span class="at">celltype_id_receiver =</span> celltype_id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(cell)`
Joining with `by = join_by(cell)`</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb255"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb255-1"><a href="#cb255-1" aria-hidden="true" tabindex="-1"></a>p_violin</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-113-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="plots-of-cell-cell-communication-changes" class="level3">
<h3 class="anchored" data-anchor-id="plots-of-cell-cell-communication-changes">8.6 Plots of cell-cell communication changes</h3>
<p>Plots that showcase differences in therapy-induced cell-cell communication changes, thus specific to the multifactorial design and complex contrast in this single-cell data and analysis. The plots allow visualization of ligand-receptor expression prioritized by MultiNicheNet based on expression and also ligand activity.</p>
<section id="prepare-difference-plots" class="level4">
<h4 class="anchored" data-anchor-id="prepare-difference-plots">8.6.1 Prepare difference plots</h4>
<p>Create a <code>sample-level</code> data frame which stores for each sample the following information:</p>
<ul>
<li><p>which patient it came from</p></li>
<li><p>was it On or Pre therapy</p></li>
<li><p>was the patient from the E or NE group</p></li>
<li><p>what is the pseudobulk expression product for a ligand-receptor pair</p></li>
</ul>
<p>Use top 10 interactions with stronger On-Pre differences in the E group vs the NE group.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb256"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb256-1"><a href="#cb256-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get prioritized interactions</span></span>
<span id="cb256-2"><a href="#cb256-2" aria-hidden="true" tabindex="-1"></a>prioritized_tbl_oi <span class="ot">=</span> <span class="fu">get_top_n_lr_pairs</span>(multinichenet_output<span class="sc">$</span>prioritization_tables, <span class="dv">10</span>, <span class="at">rank_per_group =</span> <span class="cn">TRUE</span>, <span class="at">groups_oi =</span> <span class="st">"OnE"</span>)</span>
<span id="cb256-3"><a href="#cb256-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb256-4"><a href="#cb256-4" aria-hidden="true" tabindex="-1"></a><span class="co"># create sample-level data frame for prioritized interactions</span></span>
<span id="cb256-5"><a href="#cb256-5" aria-hidden="true" tabindex="-1"></a>sample_data <span class="ot">=</span> multinichenet_output<span class="sc">$</span>prioritization_tables<span class="sc">$</span>sample_prioritization_tbl <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(id <span class="sc">%in%</span> prioritized_tbl_oi<span class="sc">$</span>id) <span class="sc">%&gt;%</span></span>
<span id="cb256-6"><a href="#cb256-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">sender_receiver =</span> <span class="fu">paste</span>(sender, receiver, <span class="at">sep =</span> <span class="st">" --&gt;"</span>),</span>
<span id="cb256-7"><a href="#cb256-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">lr_interaction =</span> <span class="fu">paste</span>(ligand, receptor, <span class="at">sep =</span> <span class="st">" - "</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb256-8"><a href="#cb256-8" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(receiver) <span class="sc">%&gt;%</span></span>
<span id="cb256-9"><a href="#cb256-9" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(receiver) <span class="sc">%&gt;%</span></span>
<span id="cb256-10"><a href="#cb256-10" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(sender, <span class="at">.by_group =</span> <span class="cn">TRUE</span>)</span>
<span id="cb256-11"><a href="#cb256-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb256-12"><a href="#cb256-12" aria-hidden="true" tabindex="-1"></a>sample_data <span class="ot">=</span> sample_data <span class="sc">%&gt;%</span></span>
<span id="cb256-13"><a href="#cb256-13" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">sender_receiver =</span> </span>
<span id="cb256-14"><a href="#cb256-14" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">factor</span>(sender_receiver, </span>
<span id="cb256-15"><a href="#cb256-15" aria-hidden="true" tabindex="-1"></a>                         <span class="at">levels =</span> sample_data<span class="sc">$</span>sender_receiver <span class="sc">%&gt;%</span> <span class="fu">unique</span>()))</span>
<span id="cb256-16"><a href="#cb256-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb256-17"><a href="#cb256-17" aria-hidden="true" tabindex="-1"></a><span class="co"># define time point and group and link all together</span></span>
<span id="cb256-18"><a href="#cb256-18" aria-hidden="true" tabindex="-1"></a>grouping_tbl2 <span class="ot">=</span> multinichenet_output<span class="sc">$</span>grouping_tbl <span class="sc">%&gt;%</span></span>
<span id="cb256-19"><a href="#cb256-19" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">inner_join</span>(multinichenet_output<span class="sc">$</span>prioritization_tables<span class="sc">$</span>sample_prioritization_tbl <span class="sc">%&gt;%</span></span>
<span id="cb256-20"><a href="#cb256-20" aria-hidden="true" tabindex="-1"></a>                      dplyr<span class="sc">::</span><span class="fu">distinct</span>(sample, keep_receiver, keep_sender))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(sample)`</code></pre>
</div>
<div class="sourceCode cell-code" id="cb258"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb258-1"><a href="#cb258-1" aria-hidden="true" tabindex="-1"></a>grouping_tbl2 <span class="ot">=</span> grouping_tbl2 <span class="sc">%&gt;%</span></span>
<span id="cb258-2"><a href="#cb258-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">inner_join</span>(<span class="fu">tibble</span>(<span class="at">group =</span> <span class="fu">c</span>(<span class="st">"PreE"</span>, <span class="st">"PreNE"</span>, <span class="st">"OnE"</span>, <span class="st">"OnNE"</span>),</span>
<span id="cb258-3"><a href="#cb258-3" aria-hidden="true" tabindex="-1"></a>                           <span class="at">contrast =</span> <span class="fu">c</span>(<span class="st">"E"</span>, <span class="st">"NE"</span>, <span class="st">"E"</span>, <span class="st">"NE"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(group)`</code></pre>
</div>
<div class="sourceCode cell-code" id="cb260"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb260-1"><a href="#cb260-1" aria-hidden="true" tabindex="-1"></a>grouping_tbl2<span class="sc">$</span>on_pre <span class="ot">=</span> <span class="st">"On"</span></span>
<span id="cb260-2"><a href="#cb260-2" aria-hidden="true" tabindex="-1"></a>grouping_tbl2<span class="sc">$</span>on_pre[grouping_tbl2<span class="sc">$</span>group <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"PreE"</span>, <span class="st">"PreNE"</span>)] <span class="ot">=</span> <span class="st">"Pre"</span></span>
<span id="cb260-3"><a href="#cb260-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb260-4"><a href="#cb260-4" aria-hidden="true" tabindex="-1"></a>sample_data <span class="ot">=</span> sample_data <span class="sc">%&gt;%</span> <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span></span>
<span id="cb260-5"><a href="#cb260-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">patient =</span> sample_data<span class="sc">$</span>sample <span class="sc">%&gt;%</span> stringr<span class="sc">::</span><span class="fu">str_split</span>(<span class="st">"Pre"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb260-6"><a href="#cb260-6" aria-hidden="true" tabindex="-1"></a>           <span class="fu">sapply</span>(<span class="cf">function</span>(x){x[<span class="dv">1</span>]}) <span class="sc">%&gt;%</span> stringr<span class="sc">::</span><span class="fu">str_split</span>(<span class="st">"On"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb260-7"><a href="#cb260-7" aria-hidden="true" tabindex="-1"></a>           <span class="fu">sapply</span>(<span class="cf">function</span>(x){x[<span class="dv">1</span>]})) <span class="sc">%&gt;%</span></span>
<span id="cb260-8"><a href="#cb260-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">inner_join</span>(grouping_tbl2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(sample, group, keep_receiver, keep_sender)`</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb262"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb262-1"><a href="#cb262-1" aria-hidden="true" tabindex="-1"></a>sample_data <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 30
  sample       sender     receiver ligand receptor avg_ligand avg_receptor
  &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt;         &lt;dbl&gt;        &lt;dbl&gt;
1 BIOKEY_12On  Fibroblast CD4T     CXCL9  CXCR3         1.46         0.366
2 BIOKEY_4Pre  Fibroblast CD4T     CXCL9  CXCR3         1.07         0.465
3 BIOKEY_10On  Fibroblast CD4T     CXCL9  CXCR3         1.13         0.407
4 BIOKEY_15On  Fibroblast CD4T     CXCL9  CXCR3         1.86         0.223
5 BIOKEY_8On   Fibroblast CD4T     CXCL9  CXCR3         0.815        0.490
6 BIOKEY_11Pre Fibroblast CD4T     CXCL9  CXCR3         0.840        0.467
# ℹ 23 more variables: ligand_receptor_prod &lt;dbl&gt;, fraction_ligand &lt;dbl&gt;,
#   fraction_receptor &lt;dbl&gt;, ligand_receptor_fraction_prod &lt;dbl&gt;,
#   pb_ligand &lt;dbl&gt;, pb_receptor &lt;dbl&gt;, ligand_receptor_pb_prod &lt;dbl&gt;,
#   group &lt;chr&gt;, prioritization_score &lt;dbl&gt;, lr_interaction &lt;chr&gt;, id &lt;chr&gt;,
#   scaled_LR_prod &lt;dbl&gt;, scaled_LR_frac &lt;dbl&gt;, scaled_LR_pb_prod &lt;dbl&gt;,
#   n_cells_receiver &lt;dbl&gt;, keep_receiver &lt;dbl&gt;, n_cells_sender &lt;dbl&gt;,
#   keep_sender &lt;dbl&gt;, keep_sender_receiver &lt;fct&gt;, sender_receiver &lt;fct&gt;, …</code></pre>
</div>
</div>
<p>Remove samples where sender and/or receiver was missing and calculate the On-vs-Pre difference in pseudobulk expression (absolute and relative, named respectively <code>diff</code> and <code>lfc</code>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb264"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb264-1"><a href="#cb264-1" aria-hidden="true" tabindex="-1"></a>sample_data <span class="ot">=</span> sample_data <span class="sc">%&gt;%</span> <span class="fu">filter</span>(keep_sender <span class="sc">&amp;</span> keep_receiver) <span class="sc">%&gt;%</span></span>
<span id="cb264-2"><a href="#cb264-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">group =</span> <span class="fu">factor</span>(group, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"PreNE"</span>, <span class="st">"PreE"</span>, <span class="st">"OnNE"</span>, <span class="st">"OnE"</span>)),</span>
<span id="cb264-3"><a href="#cb264-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">on_pre =</span> <span class="fu">factor</span>(on_pre, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Pre"</span>, <span class="st">"On"</span>)))</span>
<span id="cb264-4"><a href="#cb264-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb264-5"><a href="#cb264-5" aria-hidden="true" tabindex="-1"></a>sample_data <span class="ot">=</span> sample_data <span class="sc">%&gt;%</span> <span class="fu">inner_join</span>(</span>
<span id="cb264-6"><a href="#cb264-6" aria-hidden="true" tabindex="-1"></a>  sample_data <span class="sc">%&gt;%</span> <span class="fu">filter</span>(keep_receiver <span class="sc">==</span> <span class="dv">1</span> <span class="sc">&amp;</span> keep_sender <span class="sc">==</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span> <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span></span>
<span id="cb264-7"><a href="#cb264-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(id, patient, on_pre, ligand_receptor_pb_prod) <span class="sc">%&gt;%</span> <span class="fu">distinct</span>() <span class="sc">%&gt;%</span></span>
<span id="cb264-8"><a href="#cb264-8" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">spread</span>(on_pre, ligand_receptor_pb_prod) <span class="sc">%&gt;%</span></span>
<span id="cb264-9"><a href="#cb264-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">diff =</span> On<span class="sc">-</span>Pre, <span class="at">fc =</span> On<span class="sc">/</span>Pre) <span class="sc">%&gt;%</span></span>
<span id="cb264-10"><a href="#cb264-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">lfc =</span> <span class="fu">log</span>(fc)) <span class="sc">%&gt;%</span></span>
<span id="cb264-11"><a href="#cb264-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">arrange</span>(<span class="sc">-</span>lfc)</span>
<span id="cb264-12"><a href="#cb264-12" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(id, patient)`</code></pre>
</div>
<div class="sourceCode cell-code" id="cb266"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb266-1"><a href="#cb266-1" aria-hidden="true" tabindex="-1"></a>sample_data <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 35
  sample       sender     receiver ligand receptor avg_ligand avg_receptor
  &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt;         &lt;dbl&gt;        &lt;dbl&gt;
1 BIOKEY_12On  Fibroblast CD4T     CXCL9  CXCR3         1.46         0.366
2 BIOKEY_4Pre  Fibroblast CD4T     CXCL9  CXCR3         1.07         0.465
3 BIOKEY_10On  Fibroblast CD4T     CXCL9  CXCR3         1.13         0.407
4 BIOKEY_15On  Fibroblast CD4T     CXCL9  CXCR3         1.86         0.223
5 BIOKEY_11Pre Fibroblast CD4T     CXCL9  CXCR3         0.840        0.467
6 BIOKEY_10Pre Fibroblast CD4T     CXCL9  CXCR3         0.521        0.750
# ℹ 28 more variables: ligand_receptor_prod &lt;dbl&gt;, fraction_ligand &lt;dbl&gt;,
#   fraction_receptor &lt;dbl&gt;, ligand_receptor_fraction_prod &lt;dbl&gt;,
#   pb_ligand &lt;dbl&gt;, pb_receptor &lt;dbl&gt;, ligand_receptor_pb_prod &lt;dbl&gt;,
#   group &lt;fct&gt;, prioritization_score &lt;dbl&gt;, lr_interaction &lt;chr&gt;, id &lt;chr&gt;,
#   scaled_LR_prod &lt;dbl&gt;, scaled_LR_frac &lt;dbl&gt;, scaled_LR_pb_prod &lt;dbl&gt;,
#   n_cells_receiver &lt;dbl&gt;, keep_receiver &lt;dbl&gt;, n_cells_sender &lt;dbl&gt;,
#   keep_sender &lt;dbl&gt;, keep_sender_receiver &lt;fct&gt;, sender_receiver &lt;fct&gt;, …</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb268"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb268-1"><a href="#cb268-1" aria-hidden="true" tabindex="-1"></a>order_patients <span class="ot">=</span> sample_data <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(patient) <span class="sc">%&gt;%</span></span>
<span id="cb268-2"><a href="#cb268-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">sum_diff =</span> <span class="fu">sum</span>(diff, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb268-3"><a href="#cb268-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="sc">-</span>sum_diff) <span class="sc">%&gt;%</span></span>
<span id="cb268-4"><a href="#cb268-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(patient)</span>
<span id="cb268-5"><a href="#cb268-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb268-6"><a href="#cb268-6" aria-hidden="true" tabindex="-1"></a>order_patients <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "BIOKEY_5"  "BIOKEY_11" "BIOKEY_23" "BIOKEY_6"  "BIOKEY_30" "BIOKEY_12"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb270"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb270-1"><a href="#cb270-1" aria-hidden="true" tabindex="-1"></a>order_patients </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "BIOKEY_5"  "BIOKEY_11" "BIOKEY_23" "BIOKEY_6"  "BIOKEY_30" "BIOKEY_12"
 [7] "BIOKEY_31" "BIOKEY_10" "BIOKEY_28" "BIOKEY_16" "BIOKEY_26" "BIOKEY_18"
[13] "BIOKEY_17" "BIOKEY_20" "BIOKEY_22" "BIOKEY_7"  "BIOKEY_19" "BIOKEY_29"
[19] "BIOKEY_2"  "BIOKEY_25" "BIOKEY_15" "BIOKEY_21" "BIOKEY_8"  "BIOKEY_3" 
[25] "BIOKEY_24" "BIOKEY_14" "BIOKEY_27" "BIOKEY_9"  "BIOKEY_4" </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb272"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb272-1"><a href="#cb272-1" aria-hidden="true" tabindex="-1"></a>order_samples <span class="ot">=</span> sample_data <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(patient) <span class="sc">%&gt;%</span></span>
<span id="cb272-2"><a href="#cb272-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">sum_diff =</span> <span class="fu">sum</span>(diff, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb272-3"><a href="#cb272-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">inner_join</span>(sample_data) <span class="sc">%&gt;%</span></span>
<span id="cb272-4"><a href="#cb272-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="sc">-</span>sum_diff) <span class="sc">%&gt;%</span></span>
<span id="cb272-5"><a href="#cb272-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(sample) <span class="sc">%&gt;%</span> <span class="fu">unique</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Joining with `by = join_by(patient)`</code></pre>
</div>
<div class="sourceCode cell-code" id="cb274"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb274-1"><a href="#cb274-1" aria-hidden="true" tabindex="-1"></a>order_samples <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "BIOKEY_5On"   "BIOKEY_5Pre"  "BIOKEY_11Pre" "BIOKEY_11On"  "BIOKEY_23On" 
[6] "BIOKEY_23Pre"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb276"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb276-1"><a href="#cb276-1" aria-hidden="true" tabindex="-1"></a>order_samples </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "BIOKEY_5On"   "BIOKEY_5Pre"  "BIOKEY_11Pre" "BIOKEY_11On"  "BIOKEY_23On" 
 [6] "BIOKEY_23Pre" "BIOKEY_6Pre"  "BIOKEY_6On"   "BIOKEY_30Pre" "BIOKEY_30On" 
[11] "BIOKEY_12On"  "BIOKEY_12Pre" "BIOKEY_31Pre" "BIOKEY_31On"  "BIOKEY_10On" 
[16] "BIOKEY_10Pre" "BIOKEY_28On"  "BIOKEY_28Pre" "BIOKEY_16On"  "BIOKEY_16Pre"
[21] "BIOKEY_26Pre" "BIOKEY_26On"  "BIOKEY_18On"  "BIOKEY_18Pre" "BIOKEY_17Pre"
[26] "BIOKEY_17On"  "BIOKEY_20On"  "BIOKEY_20Pre" "BIOKEY_22Pre" "BIOKEY_22On" 
[31] "BIOKEY_7Pre"  "BIOKEY_7On"   "BIOKEY_19On"  "BIOKEY_19Pre" "BIOKEY_29Pre"
[36] "BIOKEY_29On"  "BIOKEY_2Pre"  "BIOKEY_2On"   "BIOKEY_25Pre" "BIOKEY_25On" 
[41] "BIOKEY_15On"  "BIOKEY_15Pre" "BIOKEY_21Pre" "BIOKEY_21On"  "BIOKEY_8On"  
[46] "BIOKEY_8Pre"  "BIOKEY_3Pre"  "BIOKEY_3On"   "BIOKEY_24Pre" "BIOKEY_24On" 
[51] "BIOKEY_14On"  "BIOKEY_14Pre" "BIOKEY_27Pre" "BIOKEY_27On"  "BIOKEY_9Pre" 
[56] "BIOKEY_9On"   "BIOKEY_4Pre"  "BIOKEY_4On"  </code></pre>
</div>
</div>
</section>
<section id="boxplots" class="level4">
<h4 class="anchored" data-anchor-id="boxplots">8.6.2 Boxplots</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb278"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb278-1"><a href="#cb278-1" aria-hidden="true" tabindex="-1"></a>p_lr_prod_change_boxplot <span class="ot">=</span> sample_data <span class="sc">%&gt;%</span></span>
<span id="cb278-2"><a href="#cb278-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">patient =</span> <span class="fu">factor</span>(patient, <span class="at">levels =</span> order_patients)) <span class="sc">%&gt;%</span></span>
<span id="cb278-3"><a href="#cb278-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> contrast, <span class="at">y =</span> ligand_receptor_pb_prod, </span>
<span id="cb278-4"><a href="#cb278-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">fill =</span> on_pre, <span class="at">group =</span> group)) <span class="sc">+</span></span>
<span id="cb278-5"><a href="#cb278-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb278-6"><a href="#cb278-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(id<span class="sc">~</span>.) <span class="sc">+</span></span>
<span id="cb278-7"><a href="#cb278-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb278-8"><a href="#cb278-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">""</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">""</span>)</span>
<span id="cb278-9"><a href="#cb278-9" aria-hidden="true" tabindex="-1"></a>p_lr_prod_change_boxplot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-121-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The boxplots reflect the <em>average group differences</em> inferred by the DE model underlying MultiNicheNet.</p>
</section>
<section id="bubble-plots" class="level4">
<h4 class="anchored" data-anchor-id="bubble-plots">8.6.3 Bubble plots</h4>
<p>Bubble plots are used to visualize and gain insights into the potential inter-sample heterogeneity (that boxplots don’t illustrate).</p>
<p>On-vs-Pre <em>absolute difference</em> in pseudobulk ligand-receptor expression product for group E</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb279"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb279-1"><a href="#cb279-1" aria-hidden="true" tabindex="-1"></a>max_diff <span class="ot">=</span> <span class="fu">abs</span>(sample_data<span class="sc">$</span>diff) <span class="sc">%&gt;%</span> <span class="fu">max</span>(<span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb279-2"><a href="#cb279-2" aria-hidden="true" tabindex="-1"></a>custom_scale_color <span class="ot">=</span> <span class="fu">scale_color_gradientn</span>(</span>
<span id="cb279-3"><a href="#cb279-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">colours =</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="at">n =</span> <span class="dv">7</span>, <span class="at">name =</span> <span class="st">"RdBu"</span>) <span class="sc">%&gt;%</span> <span class="fu">rev</span>(), </span>
<span id="cb279-4"><a href="#cb279-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">values =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.30</span>, <span class="fl">0.425</span>, <span class="fl">0.5</span>, <span class="fl">0.575</span>, <span class="fl">0.70</span>, <span class="dv">1</span>), </span>
<span id="cb279-5"><a href="#cb279-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">limits =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span> <span class="sc">*</span> max_diff, max_diff))</span>
<span id="cb279-6"><a href="#cb279-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb279-7"><a href="#cb279-7" aria-hidden="true" tabindex="-1"></a>p_lr_prod_change <span class="ot">=</span> sample_data <span class="sc">%&gt;%</span> </span>
<span id="cb279-8"><a href="#cb279-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">patient =</span> <span class="fu">factor</span>(patient, <span class="at">levels =</span> order_patients)) <span class="sc">%&gt;%</span></span>
<span id="cb279-9"><a href="#cb279-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(patient, lr_interaction, <span class="at">color =</span> diff)) <span class="sc">+</span></span>
<span id="cb279-10"><a href="#cb279-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb279-11"><a href="#cb279-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(sender_receiver<span class="sc">~</span>contrast, </span>
<span id="cb279-12"><a href="#cb279-12" aria-hidden="true" tabindex="-1"></a>             <span class="at">scales =</span> <span class="st">"free"</span>, <span class="at">space =</span> <span class="st">"free"</span>, <span class="at">switch =</span> <span class="st">"y"</span>) <span class="sc">+</span></span>
<span id="cb279-13"><a href="#cb279-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_light</span>() <span class="sc">+</span>  </span>
<span id="cb279-14"><a href="#cb279-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.ticks =</span> <span class="fu">element_blank</span>(), <span class="at">axis.title =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb279-15"><a href="#cb279-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">face =</span> <span class="st">"bold.italic"</span>, <span class="at">size =</span> <span class="dv">9</span>), </span>
<span id="cb279-16"><a href="#cb279-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">9</span>, <span class="at">angle =</span> <span class="dv">90</span>, <span class="at">hjust =</span> <span class="dv">0</span>), </span>
<span id="cb279-17"><a href="#cb279-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb279-18"><a href="#cb279-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(), </span>
<span id="cb279-19"><a href="#cb279-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.spacing.x =</span> <span class="fu">unit</span>(<span class="fl">0.4</span>, <span class="st">"lines"</span>), </span>
<span id="cb279-20"><a href="#cb279-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.spacing.y =</span> <span class="fu">unit</span>(<span class="fl">0.25</span>, <span class="st">"lines"</span>), </span>
<span id="cb279-21"><a href="#cb279-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">strip.text.x.top =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">10</span>, </span>
<span id="cb279-22"><a href="#cb279-22" aria-hidden="true" tabindex="-1"></a>              <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">face =</span> <span class="st">"bold"</span>, <span class="at">angle =</span> <span class="dv">0</span>), </span>
<span id="cb279-23"><a href="#cb279-23" aria-hidden="true" tabindex="-1"></a>        <span class="at">strip.text.y.left =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">9</span>, <span class="at">color =</span> <span class="st">"black"</span>, </span>
<span id="cb279-24"><a href="#cb279-24" aria-hidden="true" tabindex="-1"></a>                                         <span class="at">face =</span> <span class="st">"bold"</span>, <span class="at">angle =</span> <span class="dv">0</span>), </span>
<span id="cb279-25"><a href="#cb279-25" aria-hidden="true" tabindex="-1"></a>        <span class="at">strip.background =</span> <span class="fu">element_rect</span>(<span class="at">color =</span> <span class="st">"darkgrey"</span>, </span>
<span id="cb279-26"><a href="#cb279-26" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">fill =</span> <span class="st">"whitesmoke"</span>, </span>
<span id="cb279-27"><a href="#cb279-27" aria-hidden="true" tabindex="-1"></a>                                        <span class="at">size =</span> <span class="fl">1.5</span>, <span class="at">linetype =</span> <span class="st">"solid"</span>)) <span class="sc">+</span></span>
<span id="cb279-28"><a href="#cb279-28" aria-hidden="true" tabindex="-1"></a>  custom_scale_color <span class="sc">+</span></span>
<span id="cb279-29"><a href="#cb279-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">""</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The `size` argument of `element_rect()` is deprecated as of ggplot2 3.4.0.
ℹ Please use the `linewidth` argument instead.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb281"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb281-1"><a href="#cb281-1" aria-hidden="true" tabindex="-1"></a>p_lr_prod_change</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="multinichenet1_files/figure-html/unnamed-chunk-122-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>